\begin{longtblr}
{
colspec={Xp{2cm}p{2.25cm}XXXXXXXXX},
hlines={solid, 1pt},
vlines={solid, 1pt},
cells={halign=c, valign=m},
width=\linewidth,
rowhead=1,
row{odd}={Grey},
row{1}={Brown, fg=white, halign=c, valign=m},
caption={\textbf{Table S25. Denisovan-specific coding mutations in the focal 72kb \textit{MUC19} region.} \newline Allelic and functional information for the 11 Denisovan-specific coding mutations found within the focal 72kb region. Each variant’s effect on protein structure is categorized using Grantham scores as follows: Grantham score $< 50$ = conservative, Grantham score 51–100 = moderately conservative, Grantham score 101–150 = moderately radical, and Grantham score $> 150$ = radical. The PhyloP score measures the conservation of the reference base-pair, with higher positive scores indicating stronger conservation.},
}
Chr12 Position (Hg19) & rsID & Mut. Type & Ref. Allele & Ref. Amino Acid & Denisovan Allele & Denisovan Amino Acid & Protein Position & Grantham Score & PhyloP Score \\
40808672 & rs4768252 & Missense & C & Ser & T & Leu & 275 & 145 & -1.221 \\
40808726 & rs4768253 & Missense & C & Thr & T & Met & 293 & 81 & -0.879 \\
40809983 & rs142268259 & Missense & C & Asp & A & Glu & 410 & 45 & -0.061 \\
40814107 & rs2114566 & Missense & A & Ile & G & Val & 511 & 29 & -0.189 \\
40814197 & rs17467164 & Missense & G & Val & A & Ile & 541 & 29 & -0.336 \\
40815060 & rs149221842 & Missense & C & His & T & Tyr & 572 & 83 & -0.569 \\
40820403 & rs11564249 & Synonymous & C & Asp & T & Asp & 711 & 0 & 0.818 \\
40821795 & rs61736852 & Missense & G & Glu & A & Lys & 766 & 56 & 0.581 \\
40821871 & rs17467284 & Missense & G & Arg & T & Leu & 791 & 102 & 5.15 \\
40826201 & rs11564125 & Missense & G & Val & A & Ile & 883 & 29 & -1.722 \\
40826535 & rs11564104 & Synonymous & T & Ser & C & Ser & 939 & 0 & -0.035 \\
\end{longtblr}
