\begin{longtblr}
{
colspec={XXXXXXXXXX},
hlines={solid, 1pt},
vlines={solid, 1pt},
cells={halign=c, valign=m},
width=\linewidth,
rowhead=1,
row{odd}={Grey},
row{1}={Brown, fg=white, halign=c, valign=m},
caption={\textbf{Table S64. Integrated Haplotype Scores (\textit{iHS}) for archaic SNPs in the focal 72kb \textit{MUC19} region among 1000 Genomes Project non-African populations.} \newline The normalized $\mid iHS \mid$ scores for archaic SNPs observed at the focal 72kb region for each non-African population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each population, the normalized \textit{iHS} scores were calculated for archaic SNPs with a minor allele frequency $> 5\%$. For each population, the table reports the total number of \textit{iHS} scores observed for archaic SNPs, the number of extreme \textit{iHS} scores for archaic SNPs (i.e., $\mid iHS \mid > 2$), the observed proportion of extreme \textit{iHS} scores for archaic SNPs (calculated as the number of extreme \textit{iHS} scores normalized by the total number of \textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \textit{iHS} scores for archaic SNPs determined from the genomic background distribution (i.e., non-overlapping 72kb windows with more than 10 archaic SNPs), the total number of 72kb windows in the genomic background distribution, the number of 72kb windows where the observed proportion of extreme \textit{iHS} scores for archaic SNPs is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \textit{iHS} scores for archaic SNPs (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \textit{iHS} scores). An observed proportion of extreme \textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection. Note that a "---" denotes that the population did not meet the minimum number of archaic SNPs threshold for the \textit{iHS} analysis.},
}
Super Population & Population & Focal 72kb Region (Total SNPs) & Focal 72kb Region (SNPs with $\mid iHS \mid > 2$) & Focal 72kb Region (Prop. of SNPs with $\mid iHS \mid > 2$) & 72kb Non-overlapping Windows ($99^{th}$ Percentile) & 72kb Non-overlapping Windows (Total SNPs $> 10$) & Focal 72kb Region $>$ 72kb Non-overlapping Windows & Focal 72kb Region (Percentile Rank) \\
AMR & MXL & 140 & 135 & 0.964 & 0.938 & 2699 & 2677 & 99.185 \\
AMR & PEL & 140 & 58 & 0.414 & 1 & 2332 & 2225 & 95.412 \\
AMR & CLM & 140 & 6 & 0.043 & 0.824 & 2630 & 2339 & 88.935 \\
AMR & PUR & 140 & 0 & 0 & 0.878 & 2344 & 0 & 0 \\
SAS & BEB & 140 & 0 & 0 & 0.732 & 3048 & 0 & 0 \\
SAS & STU & 140 & 0 & 0 & 0.733 & 2901 & 0 & 0 \\
SAS & ITU & 140 & 0 & 0 & 0.733 & 2900 & 0 & 0 \\
SAS & PJL & 139 & 0 & 0 & 0.802 & 2952 & 0 & 0 \\
SAS & GIH & 140 & 0 & 0 & 0.848 & 2883 & 0 & 0 \\
EAS & CHB & 0 & 0 & --- & --- & --- & --- & --- \\
EAS & KHV & 140 & 0 & 0 & 0.686 & 3248 & 0 & 0 \\
EAS & CHS & 0 & 0 & --- & --- & --- & --- & --- \\
EAS & JPT & 2 & 0 & 0 & --- & --- & --- & --- \\
EAS & CDX & 140 & 0 & 0 & 0.799 & 3116 & 0 & 0 \\
EUR & TSI & 139 & 1 & 0.007 & 0.766 & 2668 & 2352 & 88.156 \\
EUR & CEU & 0 & 0 & --- & --- & --- & --- & --- \\
EUR & IBS & 0 & 0 & --- & --- & --- & --- & --- \\
EUR & GBR & 0 & 0 & --- & --- & --- & --- & --- \\
EUR & FIN & 0 & 0 & --- & --- & --- & --- & --- \\
\end{longtblr}
