{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Missense Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "allel 1.3.5\n",
      "numpy 1.22.3\n",
      "pandas 1.4.2\n"
     ]
    }
   ],
   "source": [
    "# Import packages.\n",
    "import analysis_functions as af\n",
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "# Print version numbers.\n",
    "print('allel', allel.__version__)\n",
    "print('numpy', np.__version__)\n",
    "print('pandas', pd.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Intialize the pandas preferences.\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the missense mutation dataframe.\n",
    "den_mis_df = pd.read_csv('../meta_data/denisovan_specific_derived_missense_muts_info.csv')\n",
    "# Extract the positions of the missense mutations.\n",
    "den_mis_pos = den_mis_df['POS'].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TGP & SGDP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the meta data file for the tgp and sgdp.\n",
    "tgp_df = pd.read_csv(\n",
    "    '../meta_data/tgp_mod.txt', sep='\\t',\n",
    "    names=['IND', 'POP', 'SUPERPOP'],\n",
    ")\n",
    "pap_df = pd.read_csv(\n",
    "    '../meta_data/sgdp_pap.txt', sep='\\t',\n",
    "    names=['IDX', 'IND', 'POP', 'SUPERPOP'],\n",
    ")\n",
    "sgdp_amr_df = pd.read_csv('../meta_data/sgdp_amr.csv')\n",
    "# Intialize the tgp and sgdp pop list.\n",
    "tgp_pop_list = [\n",
    "    'LWK', 'GWD', 'MSL', 'ESN', 'YRI', # AFR.\n",
    "    'BEB', 'STU', 'ITU', 'PJL', 'GIH', # SAS.\n",
    "    'CHB', 'KHV', 'CHS', 'JPT', 'CDX', # EAS.    \n",
    "    'TSI', 'CEU', 'IBS', 'GBR', 'FIN', # EUR.\n",
    "    'PEL', 'MXL', 'CLM', 'PUR', # AMR.\n",
    "]\n",
    "sgdp_amr_list = [\n",
    "    'Mayan', 'Mixe','Mixtec',\n",
    "    'Pima', 'Zapotec', 'Chane',\n",
    "    'Karitiana', 'Surui', 'Piapoco',\n",
    "    'Quechua',\n",
    "]\n",
    "sgdp_amr_dicc = {\n",
    "    'Chipewyan': 'Canada',\n",
    "    'Cree':  'Canada',\n",
    "    'Mayan': 'Mexico',\n",
    "    'Mixe': 'Mexico',\n",
    "    'Mixtec': 'Mexico',\n",
    "    'Nahua': 'Mexico',\n",
    "    'Pima': 'Mexico',\n",
    "    'Zapotec': 'Mexico',\n",
    "    'Chane': 'Argentina',\n",
    "    'Karitiana': 'Brazil',\n",
    "    'Surui': 'Brazil',\n",
    "    'Piapoco': 'Colombia',\n",
    "    'Quechua': 'Peru',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the genotype matricies.\n",
    "tgp_72kb_gt, tgp_72kb_pos = af.load_hap_region('tgp_mod_arc_anc', 12, 40758000, 40830000)\n",
    "sgdp_72kb_gt, sgdp_72kb_pos = af.load_hap_region('sgdp_arc_anc', 12, 40758000, 40830000)\n",
    "# Create missense mutation masks.\n",
    "tgp_mis_mask = np.in1d(tgp_72kb_pos, den_mis_pos)\n",
    "sgdp_mis_mask = np.in1d(sgdp_72kb_pos, den_mis_pos)\n",
    "# Polarize the genotype matricies.\n",
    "tgp_mis_p_gt = af.polarize_gt(tgp_72kb_gt.compress(tgp_mis_mask, axis=0))\n",
    "sgdp_mis_p_gt = af.polarize_gt(sgdp_72kb_gt.compress(sgdp_mis_mask, axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Intialize a dictionary for the tgp samples.\n",
    "tgp_dicc = {\n",
    "    'Position (hg19)': den_mis_df['POS'].values,\n",
    "    'rsID': den_mis_df['ID'].values,\n",
    "    'Reference Allele': den_mis_df['REF'].values,\n",
    "    'Alternative Allele': den_mis_df['ALT'].values,\n",
    "    'Ancestral Allele': den_mis_df['ANC'].values,\n",
    "    'Derived Allele': den_mis_df['DER'].values,\n",
    "    r'Anc. AA $\\rightarrow$ Der. AA': den_mis_df['AA'].values,\n",
    "}\n",
    "# Add the sgdp information.\n",
    "tgp_dicc[r'Papuan ($n$ = 15)'] = af.calc_alt_freqs(sgdp_mis_p_gt.take(pap_df['IDX'].values, axis=1))\n",
    "tgp_dicc[r'SGDP AMR ($n$ = 22)'] = af.calc_alt_freqs(sgdp_mis_p_gt.take(sgdp_amr_df['IDX'].values, axis=1))\n",
    "# For every population...\n",
    "for pop in tgp_pop_list:\n",
    "    # Extract the sample indicies.\n",
    "    pop_idx = tgp_df[tgp_df['POP'] == pop].index.values\n",
    "    # Update the dataframe.\n",
    "    tgp_dicc[pop] = af.calc_alt_freqs(tgp_mis_p_gt.take(pop_idx, axis=1))\n",
    "# Convert the dataframe.\n",
    "tgp_den_mis_freq_df = pd.DataFrame(tgp_dicc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position (hg19)</th>\n",
       "      <th>rsID</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Alternative Allele</th>\n",
       "      <th>Ancestral Allele</th>\n",
       "      <th>Derived Allele</th>\n",
       "      <th>Anc. AA $\\rightarrow$ Der. AA</th>\n",
       "      <th>Papuan ($n$ = 15)</th>\n",
       "      <th>SGDP AMR ($n$ = 22)</th>\n",
       "      <th>LWK</th>\n",
       "      <th>GWD</th>\n",
       "      <th>MSL</th>\n",
       "      <th>ESN</th>\n",
       "      <th>YRI</th>\n",
       "      <th>BEB</th>\n",
       "      <th>STU</th>\n",
       "      <th>ITU</th>\n",
       "      <th>PJL</th>\n",
       "      <th>GIH</th>\n",
       "      <th>CHB</th>\n",
       "      <th>KHV</th>\n",
       "      <th>CHS</th>\n",
       "      <th>JPT</th>\n",
       "      <th>CDX</th>\n",
       "      <th>TSI</th>\n",
       "      <th>CEU</th>\n",
       "      <th>IBS</th>\n",
       "      <th>GBR</th>\n",
       "      <th>FIN</th>\n",
       "      <th>PEL</th>\n",
       "      <th>MXL</th>\n",
       "      <th>CLM</th>\n",
       "      <th>PUR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40808672</td>\n",
       "      <td>rs4768252</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Ser$\\rightarrow$Leu</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156977</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.03271</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.079787</td>\n",
       "      <td>0.086538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40808726</td>\n",
       "      <td>rs4768253</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Thr$\\rightarrow$Met</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.168605</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.101942</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.03271</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.086538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40815060</td>\n",
       "      <td>rs149221842</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>His$\\rightarrow$Tyr</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156977</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.107843</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.03271</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.086538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40821795</td>\n",
       "      <td>rs61736852</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Glu$\\rightarrow$Lys</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.102941</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.087379</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.03271</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.069149</td>\n",
       "      <td>0.086538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40821847</td>\n",
       "      <td>rs11564170</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Thr$\\rightarrow$Lys</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.102941</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.087379</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.03271</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.069149</td>\n",
       "      <td>0.086538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40821871</td>\n",
       "      <td>rs17467284</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Arg$\\rightarrow$Leu</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.102941</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.087379</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.03271</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.069149</td>\n",
       "      <td>0.086538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position (hg19)         rsID Reference Allele Alternative Allele  \\\n",
       "0         40808672    rs4768252                C                  T   \n",
       "1         40808726    rs4768253                C                  T   \n",
       "2         40815060  rs149221842                C                  T   \n",
       "3         40821795   rs61736852                G                  A   \n",
       "4         40821847   rs11564170                C                  A   \n",
       "5         40821871   rs17467284                G                  T   \n",
       "\n",
       "  Ancestral Allele Derived Allele Anc. AA $\\rightarrow$ Der. AA  \\\n",
       "0                C              T           Ser$\\rightarrow$Leu   \n",
       "1                C              T           Thr$\\rightarrow$Met   \n",
       "2                C              T           His$\\rightarrow$Tyr   \n",
       "3                G              A           Glu$\\rightarrow$Lys   \n",
       "4                C              A           Thr$\\rightarrow$Lys   \n",
       "5                G              T           Arg$\\rightarrow$Leu   \n",
       "\n",
       "   Papuan ($n$ = 15)  SGDP AMR ($n$ = 22)  LWK  GWD  MSL  ESN  YRI       BEB  \\\n",
       "0                0.1             0.363636  0.0  0.0  0.0  0.0  0.0  0.156977   \n",
       "1                0.1             0.363636  0.0  0.0  0.0  0.0  0.0  0.168605   \n",
       "2                0.1             0.363636  0.0  0.0  0.0  0.0  0.0  0.156977   \n",
       "3                0.1             0.363636  0.0  0.0  0.0  0.0  0.0  0.139535   \n",
       "4                0.1             0.363636  0.0  0.0  0.0  0.0  0.0  0.139535   \n",
       "5                0.1             0.363636  0.0  0.0  0.0  0.0  0.0  0.139535   \n",
       "\n",
       "        STU       ITU       PJL       GIH       CHB       KHV       CHS  \\\n",
       "0  0.112745  0.112745  0.052083  0.097087  0.033981  0.085859  0.038095   \n",
       "1  0.112745  0.112745  0.052083  0.101942  0.033981  0.085859  0.038095   \n",
       "2  0.112745  0.107843  0.052083  0.097087  0.033981  0.085859  0.038095   \n",
       "3  0.112745  0.102941  0.052083  0.087379  0.033981  0.085859  0.038095   \n",
       "4  0.112745  0.102941  0.052083  0.087379  0.033981  0.085859  0.038095   \n",
       "5  0.112745  0.102941  0.052083  0.087379  0.033981  0.085859  0.038095   \n",
       "\n",
       "        JPT       CDX       TSI       CEU      IBS       GBR       FIN  \\\n",
       "0  0.028846  0.123656  0.051402  0.015152  0.03271  0.005495  0.005051   \n",
       "1  0.028846  0.123656  0.051402  0.015152  0.03271  0.005495  0.005051   \n",
       "2  0.028846  0.123656  0.051402  0.015152  0.03271  0.005495  0.005051   \n",
       "3  0.028846  0.123656  0.051402  0.015152  0.03271  0.005495  0.005051   \n",
       "4  0.028846  0.123656  0.051402  0.015152  0.03271  0.005495  0.005051   \n",
       "5  0.028846  0.123656  0.051402  0.015152  0.03271  0.005495  0.005051   \n",
       "\n",
       "        PEL       MXL       CLM       PUR  \n",
       "0  0.217647  0.304688  0.079787  0.086538  \n",
       "1  0.217647  0.304688  0.074468  0.086538  \n",
       "2  0.217647  0.304688  0.074468  0.086538  \n",
       "3  0.217647  0.304688  0.069149  0.086538  \n",
       "4  0.217647  0.304688  0.069149  0.086538  \n",
       "5  0.217647  0.304688  0.069149  0.086538  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show the results for tgp + sgdp.\n",
    "tgp_den_mis_freq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Intialize a dictionary for the sgdp samples.\n",
    "sgdp_dicc = {\n",
    "    'Position (hg19)': den_mis_df['POS'].values,\n",
    "    'rsID': den_mis_df['ID'].values,\n",
    "    'Reference Allele': den_mis_df['REF'].values,\n",
    "    'Alternative Allele': den_mis_df['ALT'].values,\n",
    "    'Ancestral Allele': den_mis_df['ANC'].values,\n",
    "    'Derived Allele': den_mis_df['DER'].values,\n",
    "    r'Anc. AA $\\rightarrow$ Der. AA': den_mis_df['AA'].values,\n",
    "}\n",
    "# Add the papuan info.\n",
    "sgdp_dicc[r'Papuan ($n$ = 15)'] = af.calc_alt_freqs(sgdp_mis_p_gt.take(pap_df['IDX'].values, axis=1))\n",
    "# For every population...\n",
    "for pop in sgdp_amr_list:\n",
    "    # Subset the dataframe.\n",
    "    sub_df = sgdp_amr_df[sgdp_amr_df['POP'] == pop]\n",
    "    # Intialize a title.\n",
    "    title = pop+r' ($n$ = '+f'{sub_df.shape[0]})'\n",
    "    # Update the dataframe.\n",
    "    sgdp_dicc[title] = af.calc_alt_freqs(sgdp_mis_p_gt.take(sub_df['IDX'].values, axis=1))\n",
    "# Convert the dataframe.\n",
    "sgdp_den_mis_freq_df = pd.DataFrame(sgdp_dicc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position (hg19)</th>\n",
       "      <th>rsID</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Alternative Allele</th>\n",
       "      <th>Ancestral Allele</th>\n",
       "      <th>Derived Allele</th>\n",
       "      <th>Anc. AA $\\rightarrow$ Der. AA</th>\n",
       "      <th>Papuan ($n$ = 15)</th>\n",
       "      <th>Mayan ($n$ = 2)</th>\n",
       "      <th>Mixe ($n$ = 3)</th>\n",
       "      <th>Mixtec ($n$ = 2)</th>\n",
       "      <th>Pima ($n$ = 2)</th>\n",
       "      <th>Zapotec ($n$ = 2)</th>\n",
       "      <th>Chane ($n$ = 1)</th>\n",
       "      <th>Karitiana ($n$ = 3)</th>\n",
       "      <th>Surui ($n$ = 2)</th>\n",
       "      <th>Piapoco ($n$ = 2)</th>\n",
       "      <th>Quechua ($n$ = 3)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40808672</td>\n",
       "      <td>rs4768252</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Ser$\\rightarrow$Leu</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40808726</td>\n",
       "      <td>rs4768253</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Thr$\\rightarrow$Met</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40815060</td>\n",
       "      <td>rs149221842</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>His$\\rightarrow$Tyr</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40821795</td>\n",
       "      <td>rs61736852</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Glu$\\rightarrow$Lys</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40821847</td>\n",
       "      <td>rs11564170</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Thr$\\rightarrow$Lys</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40821871</td>\n",
       "      <td>rs17467284</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Arg$\\rightarrow$Leu</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position (hg19)         rsID Reference Allele Alternative Allele  \\\n",
       "0         40808672    rs4768252                C                  T   \n",
       "1         40808726    rs4768253                C                  T   \n",
       "2         40815060  rs149221842                C                  T   \n",
       "3         40821795   rs61736852                G                  A   \n",
       "4         40821847   rs11564170                C                  A   \n",
       "5         40821871   rs17467284                G                  T   \n",
       "\n",
       "  Ancestral Allele Derived Allele Anc. AA $\\rightarrow$ Der. AA  \\\n",
       "0                C              T           Ser$\\rightarrow$Leu   \n",
       "1                C              T           Thr$\\rightarrow$Met   \n",
       "2                C              T           His$\\rightarrow$Tyr   \n",
       "3                G              A           Glu$\\rightarrow$Lys   \n",
       "4                C              A           Thr$\\rightarrow$Lys   \n",
       "5                G              T           Arg$\\rightarrow$Leu   \n",
       "\n",
       "   Papuan ($n$ = 15)  Mayan ($n$ = 2)  Mixe ($n$ = 3)  Mixtec ($n$ = 2)  \\\n",
       "0                0.1              0.5        0.666667               0.0   \n",
       "1                0.1              0.5        0.666667               0.0   \n",
       "2                0.1              0.5        0.666667               0.0   \n",
       "3                0.1              0.5        0.666667               0.0   \n",
       "4                0.1              0.5        0.666667               0.0   \n",
       "5                0.1              0.5        0.666667               0.0   \n",
       "\n",
       "   Pima ($n$ = 2)  Zapotec ($n$ = 2)  Chane ($n$ = 1)  Karitiana ($n$ = 3)  \\\n",
       "0            0.25                0.5              0.0             0.166667   \n",
       "1            0.25                0.5              0.0             0.166667   \n",
       "2            0.25                0.5              0.0             0.166667   \n",
       "3            0.25                0.5              0.0             0.166667   \n",
       "4            0.25                0.5              0.0             0.166667   \n",
       "5            0.25                0.5              0.0             0.166667   \n",
       "\n",
       "   Surui ($n$ = 2)  Piapoco ($n$ = 2)  Quechua ($n$ = 3)  \n",
       "0              0.5                0.5           0.333333  \n",
       "1              0.5                0.5           0.333333  \n",
       "2              0.5                0.5           0.333333  \n",
       "3              0.5                0.5           0.333333  \n",
       "4              0.5                0.5           0.333333  \n",
       "5              0.5                0.5           0.333333  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show the results for the sgdp.\n",
    "sgdp_den_mis_freq_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ancient Americans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the poistions, ancestral, and derived alleles.\n",
    "missense_sites = den_mis_df['POS'].values\n",
    "anc_alleles = den_mis_df['ANC'].values\n",
    "der_alleles = den_mis_df['DER'].values\n",
    "# Load the ancient sample ids.\n",
    "anc_samps = np.array([\n",
    "    'USR1', 'CK-13', 'Anzick',\n",
    "    'CR-01', 'CT-01', 'NC_C',\n",
    "    'PS-06', 'SC-05', 'SM-02',\n",
    "    'SN-13', 'SN-17', 'SN-44',\n",
    "    '2417J', '2417Q', '333B',\n",
    "    'SRR8144644', 'SRR8144646',\n",
    "    'SRR8144647', 'SRR8144650',\n",
    "    'ERR2270782', 'ERR2270783',\n",
    "    'ERR2270784', 'ERR2270785',\n",
    "\n",
    "])\n",
    "# Intialize a dictionary to determine the allele from angsd.\n",
    "angsd_allele_dicc = {\n",
    "    0: 'A', 1: 'C',\n",
    "    2: 'G', 3: 'T',\n",
    "}\n",
    "# Intialize a dictionary for the ancient samples.\n",
    "anc_dicc = {\n",
    "    'Position (hg19)': den_mis_df['POS'].values,\n",
    "    'rsID': den_mis_df['ID'].values,\n",
    "    'Reference Allele': den_mis_df['REF'].values,\n",
    "    'Alternative Allele': den_mis_df['ALT'].values,\n",
    "    'Ancestral Allele': den_mis_df['ANC'].values,\n",
    "    'Derived Allele': den_mis_df['DER'].values,\n",
    "    r'Anc. AA $\\rightarrow$ Der. AA': den_mis_df['AA'].values,\n",
    "}\n",
    "# For all ancient samples...\n",
    "for samp in anc_samps:\n",
    "    # Intialize the individual.\n",
    "    anc_dicc[samp] = []\n",
    "    # Load the read data.\n",
    "    anc_df = pd.read_csv(f'../ancient_americans/{samp}_ac.txt.gz', sep='\\t')\n",
    "    # Extract the positions array.\n",
    "    anc_pos = anc_df['pos'].values\n",
    "    # For every Denisovan missense mutation...\n",
    "    for i in range(den_mis_df.shape[0]):\n",
    "        # Extract the missense info.\n",
    "        pos = missense_sites[i]\n",
    "        anc = anc_alleles[i]\n",
    "        der = der_alleles[i]\n",
    "        # Determine the index in the ancient sample.\n",
    "        pos_idx = np.where(anc_pos == pos)[0]\n",
    "        # If the site passed QC in the ancient sample...\n",
    "        if pos_idx.size > 0:\n",
    "            # Extract read counts.\n",
    "            read_counts = anc_df.iloc[pos_idx].to_numpy()[:, -4:].flatten()\n",
    "            # Determine which allele indicies have more than two reads.\n",
    "            allele_idx = np.where(read_counts >= 2)[0]\n",
    "            # If this site is not mono- or bi-allelic.\n",
    "            if allele_idx.size > 2:\n",
    "                # Append the dictionary.\n",
    "                anc_dicc[samp].append('.')\n",
    "                # Continue to the next site.\n",
    "                continue\n",
    "            # Else-if this is a mono-allelic site...\n",
    "            elif allele_idx.size == 1:\n",
    "                # Determine the allele identity.\n",
    "                anc_allele_call = angsd_allele_dicc[allele_idx[0]]\n",
    "                # If the ancient allele is the derived allele...\n",
    "                if anc_allele_call == der:\n",
    "                    # Append the dictionary.\n",
    "                    anc_dicc[samp].append(der+'/'+der)\n",
    "                # Else-if the ancient allele is the ancestral allele...\n",
    "                elif anc_allele_call == anc:\n",
    "                    # Append the dictionary.\n",
    "                    anc_dicc[samp].append(anc+'/'+anc)\n",
    "                # Else...\n",
    "                else:\n",
    "                    # Append the dictionary.\n",
    "                    anc_dicc[samp].append('.')\n",
    "                    # Continue to the next site.\n",
    "                    continue\n",
    "            # Else-if this is a bi-allelic site...\n",
    "            elif allele_idx.size == 2:\n",
    "                # Determine the allele identities.\n",
    "                anc_allele_calls = np.array([angsd_allele_dicc[j] for j in allele_idx])\n",
    "                # If the ancient alleles are the same as the hg19 alleles...\n",
    "                if np.array_equal(np.sort(anc_allele_calls), np.sort(np.array([anc, der]))):\n",
    "                    # Append the dictionary.\n",
    "                    anc_dicc[samp].append(anc+'/'+der)\n",
    "                # Else...\n",
    "                else:\n",
    "                    # Append the dictionary.\n",
    "                    anc_dicc[samp].append('.')\n",
    "                    # Continue to the next site.\n",
    "                    continue\n",
    "            # Else...\n",
    "            else:\n",
    "                # Append the dictionary.\n",
    "                anc_dicc[samp].append('.')\n",
    "                # Continue to the next site.\n",
    "                continue\n",
    "        # Else...\n",
    "        else:\n",
    "            # Append the dictionary.\n",
    "            anc_dicc[samp].append('.')\n",
    "            # Continue to the next site.\n",
    "            continue\n",
    "# Convert the dictionary to a dataframe.\n",
    "anc_genos_df = pd.DataFrame(data=anc_dicc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position (hg19)</th>\n",
       "      <th>rsID</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Alternative Allele</th>\n",
       "      <th>Ancestral Allele</th>\n",
       "      <th>Derived Allele</th>\n",
       "      <th>Anc. AA $\\rightarrow$ Der. AA</th>\n",
       "      <th>USR1</th>\n",
       "      <th>CK-13</th>\n",
       "      <th>Anzick</th>\n",
       "      <th>CR-01</th>\n",
       "      <th>CT-01</th>\n",
       "      <th>NC_C</th>\n",
       "      <th>PS-06</th>\n",
       "      <th>SC-05</th>\n",
       "      <th>SM-02</th>\n",
       "      <th>SN-13</th>\n",
       "      <th>SN-17</th>\n",
       "      <th>SN-44</th>\n",
       "      <th>2417J</th>\n",
       "      <th>2417Q</th>\n",
       "      <th>333B</th>\n",
       "      <th>SRR8144644</th>\n",
       "      <th>SRR8144646</th>\n",
       "      <th>SRR8144647</th>\n",
       "      <th>SRR8144650</th>\n",
       "      <th>ERR2270782</th>\n",
       "      <th>ERR2270783</th>\n",
       "      <th>ERR2270784</th>\n",
       "      <th>ERR2270785</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40808672</td>\n",
       "      <td>rs4768252</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Ser$\\rightarrow$Leu</td>\n",
       "      <td>C/T</td>\n",
       "      <td>.</td>\n",
       "      <td>C/T</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/C</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/T</td>\n",
       "      <td>T/T</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/T</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>T/T</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40808726</td>\n",
       "      <td>rs4768253</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Thr$\\rightarrow$Met</td>\n",
       "      <td>C/C</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/C</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/T</td>\n",
       "      <td>T/T</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40815060</td>\n",
       "      <td>rs149221842</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>His$\\rightarrow$Tyr</td>\n",
       "      <td>C/C</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/C</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>T/T</td>\n",
       "      <td>.</td>\n",
       "      <td>T/T</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>T/T</td>\n",
       "      <td>C/T</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40821795</td>\n",
       "      <td>rs61736852</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Glu$\\rightarrow$Lys</td>\n",
       "      <td>G/G</td>\n",
       "      <td>.</td>\n",
       "      <td>G/G</td>\n",
       "      <td>A/A</td>\n",
       "      <td>.</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A/A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G/G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "      <td>A/A</td>\n",
       "      <td>G/A</td>\n",
       "      <td>.</td>\n",
       "      <td>G/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40821847</td>\n",
       "      <td>rs11564170</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Thr$\\rightarrow$Lys</td>\n",
       "      <td>C/C</td>\n",
       "      <td>A/A</td>\n",
       "      <td>C/C</td>\n",
       "      <td>A/A</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/A</td>\n",
       "      <td>A/A</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A/A</td>\n",
       "      <td>.</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/C</td>\n",
       "      <td>A/A</td>\n",
       "      <td>C/A</td>\n",
       "      <td>C/C</td>\n",
       "      <td>C/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40821871</td>\n",
       "      <td>rs17467284</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Arg$\\rightarrow$Leu</td>\n",
       "      <td>G/G</td>\n",
       "      <td>.</td>\n",
       "      <td>G/G</td>\n",
       "      <td>T/T</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/T</td>\n",
       "      <td>T/T</td>\n",
       "      <td>.</td>\n",
       "      <td>G/G</td>\n",
       "      <td>.</td>\n",
       "      <td>T/T</td>\n",
       "      <td>.</td>\n",
       "      <td>G/T</td>\n",
       "      <td>G/T</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "      <td>T/T</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/G</td>\n",
       "      <td>G/T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position (hg19)         rsID Reference Allele Alternative Allele  \\\n",
       "0         40808672    rs4768252                C                  T   \n",
       "1         40808726    rs4768253                C                  T   \n",
       "2         40815060  rs149221842                C                  T   \n",
       "3         40821795   rs61736852                G                  A   \n",
       "4         40821847   rs11564170                C                  A   \n",
       "5         40821871   rs17467284                G                  T   \n",
       "\n",
       "  Ancestral Allele Derived Allele Anc. AA $\\rightarrow$ Der. AA USR1 CK-13  \\\n",
       "0                C              T           Ser$\\rightarrow$Leu  C/T     .   \n",
       "1                C              T           Thr$\\rightarrow$Met  C/C   T/T   \n",
       "2                C              T           His$\\rightarrow$Tyr  C/C     .   \n",
       "3                G              A           Glu$\\rightarrow$Lys  G/G     .   \n",
       "4                C              A           Thr$\\rightarrow$Lys  C/C   A/A   \n",
       "5                G              T           Arg$\\rightarrow$Leu  G/G     .   \n",
       "\n",
       "  Anzick CR-01 CT-01 NC_C PS-06 SC-05 SM-02 SN-13 SN-17 SN-44 2417J 2417Q  \\\n",
       "0    C/T   T/T   C/C    .   C/C   C/T   T/T     .   C/C   C/T     .   C/C   \n",
       "1    C/C   T/T   C/C  C/C   C/C   C/T   T/T   T/T   C/T   C/T     .     .   \n",
       "2    C/C   T/T   C/C    .   C/C   T/T     .   T/T   T/T   C/T     .     .   \n",
       "3    G/G   A/A     .  G/G   G/G   G/A     .     .   A/A     .     .   G/G   \n",
       "4    C/C   A/A   C/C  C/C   C/C   C/A   A/A     .   C/C   C/C     .     .   \n",
       "5    G/G   T/T   G/G  G/G   G/G   G/T   T/T     .   G/G     .   T/T     .   \n",
       "\n",
       "  333B SRR8144644 SRR8144646 SRR8144647 SRR8144650 ERR2270782 ERR2270783  \\\n",
       "0    .        C/C        C/C        C/C        C/C        T/T        T/T   \n",
       "1    .        C/C        C/C        C/C        C/C        T/T        C/T   \n",
       "2    .        C/C        C/C        C/C        C/C        T/T        C/T   \n",
       "3    .          .        G/G        G/G        G/G        A/A        G/A   \n",
       "4  A/A          .        C/C        C/C        C/C        A/A        C/A   \n",
       "5  G/T        G/T        G/G        G/G        G/G        T/T        G/G   \n",
       "\n",
       "  ERR2270784 ERR2270785  \n",
       "0        C/C        C/T  \n",
       "1        C/C        C/T  \n",
       "2        C/C        C/T  \n",
       "3          .        G/A  \n",
       "4        C/C        C/A  \n",
       "5        G/G        G/T  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show the results for the ancient american samples.\n",
    "anc_genos_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anc_der_intro_env",
   "language": "python",
   "name": "anc_der_intro_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
