{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "820a7de5",
   "metadata": {},
   "source": [
    "# Supplemental Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dade1e07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numpy 1.22.3\n",
      "pandas 1.4.2\n"
     ]
    }
   ],
   "source": [
    "# Load packages.\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "# Intialize the pandas preferences.\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', None)\n",
    "# Print version numbers.\n",
    "print('numpy', np.__version__)\n",
    "print('pandas', pd.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e841495f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to convert a number in scientific notation for LaTeX.\n",
    "def convert_sci_notation_to_latex(value):\n",
    "    # Convert to scientific notation with 3 decimal places.\n",
    "    sci_notation = '{:.3e}'.format(value)\n",
    "    # Split into mantissa and exponent.\n",
    "    mantissa, exponent = sci_notation.split('e')\n",
    "    # Convert the exponent to an integer to remove leading zeros,\n",
    "    exponent = int(exponent)\n",
    "    # Construct the LaTeX string.\n",
    "    latex_str = r'${}e^{{{}}}$'.format(mantissa, exponent)\n",
    "    return latex_str"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "103480b2",
   "metadata": {},
   "source": [
    "## S1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "eba608c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_muc19_introgressed_tract_frequency_per_super_population.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "cells={halign=c, valign=m},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S1. Frequency of introgressed tracts overlapping \\textit{MUC19} among 1000 Genomes Project super populations.} \\newline The frequency of introgressed tracts---i.e., the number of introgressed tracts normalized by the total number of chromosomes---and the mean tract length stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Introgressed tracts at \\textit{MUC19} are significantly enriched in AMR individuals (Fisher's Exact Test, Odds Ratio: 2.354, \\textit{P-value}: $2.144e^{-12}$), with AMR populations exhibiting a higher proportion of introgressed tracts compared to non-AMR populations, excluding AFR (Proportions \\textit{Z}-Test, \\textit{Z}-statistic: 7.441, \\textit{P-value}: $5.011e^{-14}$).  Note that a \"---\" denotes that there are no introgressed tracts overlapping \\textit{MUC19} for that group.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_muc19_introgressed_tract_frequency_per_super_population.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        f'{mat[i, 0]}'+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f14b8f15",
   "metadata": {},
   "source": [
    "\\textbf{Table S1. Frequency of introgressed tracts overlapping \\textit{MUC19} among 1000 Genomes Project super populations.} \\newline The frequency of introgressed tracts---i.e., the number of introgressed tracts normalized by the total number of chromosomes---and the mean tract length stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Introgressed tracts at \\textit{MUC19} are significantly enriched in AMR individuals (Fisher's Exact Test, Odds Ratio: 2.354, \\textit{P-value}: $2.144e^{-12}$), with AMR populations exhibiting a higher proportion of introgressed tracts compared to non-AMR populations, excluding AFR (Proportions \\textit{Z}-Test, \\textit{Z}-statistic: 7.441, \\textit{P-value}: $5.011e^{-14}$).  Note that a \"---\" denotes that there are no introgressed tracts overlapping \\textit{MUC19} for that group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f861d75c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Total Number of Chromosomes</th>\n",
       "      <th>Number of Introgressed Tracts</th>\n",
       "      <th>Introgressed Tract Frequency</th>\n",
       "      <th>Mean Tract Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>694</td>\n",
       "      <td>127</td>\n",
       "      <td>0.182997</td>\n",
       "      <td>609692.913386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAS</td>\n",
       "      <td>978</td>\n",
       "      <td>145</td>\n",
       "      <td>0.148262</td>\n",
       "      <td>266268.965517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EAS</td>\n",
       "      <td>1008</td>\n",
       "      <td>88</td>\n",
       "      <td>0.087302</td>\n",
       "      <td>349102.272727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EUR</td>\n",
       "      <td>1006</td>\n",
       "      <td>27</td>\n",
       "      <td>0.026839</td>\n",
       "      <td>467037.037037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFR</td>\n",
       "      <td>1008</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Super Population  Total Number of Chromosomes  \\\n",
       "0              AMR                          694   \n",
       "1              SAS                          978   \n",
       "2              EAS                         1008   \n",
       "3              EUR                         1006   \n",
       "4              AFR                         1008   \n",
       "\n",
       "   Number of Introgressed Tracts  Introgressed Tract Frequency  \\\n",
       "0                            127                      0.182997   \n",
       "1                            145                      0.148262   \n",
       "2                             88                      0.087302   \n",
       "3                             27                      0.026839   \n",
       "4                              0                      0.000000   \n",
       "\n",
       "   Mean Tract Length  \n",
       "0      609692.913386  \n",
       "1      266268.965517  \n",
       "2      349102.272727  \n",
       "3      467037.037037  \n",
       "4                NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "923f8399",
   "metadata": {},
   "source": [
    "## S2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3eac1d31",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_muc19_introgressed_tract_frequency_per_population.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S2. Frequency of introgressed tracts overlapping \\textit{MUC19} among 1000 Genomes Project populations.} \\newline The frequency of introgressed tracts---i.e., the number of introgressed tracts normalized by the total number of chromosomes---and the mean tract length for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Note that a \"---\" denotes that there are no introgressed tracts overlapping \\textit{MUC19} for that group.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_muc19_introgressed_tract_frequency_per_population.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(2, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83bef667",
   "metadata": {},
   "source": [
    "\\textbf{Table S2. Frequency of introgressed tracts overlapping \\textit{MUC19} among 1000 Genomes Project populations.} \\newline The frequency of introgressed tracts---i.e., the number of introgressed tracts normalized by the total number of chromosomes---and the mean tract length for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Note that a \"---\" denotes that there are no introgressed tracts overlapping \\textit{MUC19} for that group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3619a294",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Total Number of Chromosomes</th>\n",
       "      <th>Number of Introgressed Tracts</th>\n",
       "      <th>Introgressed Tract Frequency</th>\n",
       "      <th>Mean Tract Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>39</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>710666.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>170</td>\n",
       "      <td>42</td>\n",
       "      <td>0.247059</td>\n",
       "      <td>574714.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>188</td>\n",
       "      <td>23</td>\n",
       "      <td>0.122340</td>\n",
       "      <td>443304.347826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>208</td>\n",
       "      <td>23</td>\n",
       "      <td>0.110577</td>\n",
       "      <td>668739.130435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>172</td>\n",
       "      <td>34</td>\n",
       "      <td>0.197674</td>\n",
       "      <td>289941.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>204</td>\n",
       "      <td>33</td>\n",
       "      <td>0.161765</td>\n",
       "      <td>220878.787879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>204</td>\n",
       "      <td>35</td>\n",
       "      <td>0.171569</td>\n",
       "      <td>227800.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>192</td>\n",
       "      <td>18</td>\n",
       "      <td>0.093750</td>\n",
       "      <td>238277.777778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>206</td>\n",
       "      <td>25</td>\n",
       "      <td>0.121359</td>\n",
       "      <td>368000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>206</td>\n",
       "      <td>11</td>\n",
       "      <td>0.053398</td>\n",
       "      <td>413454.545455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>198</td>\n",
       "      <td>22</td>\n",
       "      <td>0.111111</td>\n",
       "      <td>357818.181818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>210</td>\n",
       "      <td>13</td>\n",
       "      <td>0.061905</td>\n",
       "      <td>324692.307692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>208</td>\n",
       "      <td>14</td>\n",
       "      <td>0.067308</td>\n",
       "      <td>245500.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>186</td>\n",
       "      <td>28</td>\n",
       "      <td>0.150538</td>\n",
       "      <td>380107.142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>214</td>\n",
       "      <td>13</td>\n",
       "      <td>0.060748</td>\n",
       "      <td>458769.230769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>198</td>\n",
       "      <td>3</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>612333.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>214</td>\n",
       "      <td>7</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>520000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>182</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>343000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>198</td>\n",
       "      <td>3</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>275333.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AFR</td>\n",
       "      <td>LWK</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AFR</td>\n",
       "      <td>GWD</td>\n",
       "      <td>226</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AFR</td>\n",
       "      <td>MSL</td>\n",
       "      <td>170</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AFR</td>\n",
       "      <td>ESN</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AFR</td>\n",
       "      <td>YRI</td>\n",
       "      <td>216</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Total Number of Chromosomes  \\\n",
       "0               AMR        MXL                          128   \n",
       "1               AMR        PEL                          170   \n",
       "2               AMR        CLM                          188   \n",
       "3               AMR        PUR                          208   \n",
       "4               SAS        BEB                          172   \n",
       "5               SAS        STU                          204   \n",
       "6               SAS        ITU                          204   \n",
       "7               SAS        PJL                          192   \n",
       "8               SAS        GIH                          206   \n",
       "9               EAS        CHB                          206   \n",
       "10              EAS        KHV                          198   \n",
       "11              EAS        CHS                          210   \n",
       "12              EAS        JPT                          208   \n",
       "13              EAS        CDX                          186   \n",
       "14              EUR        TSI                          214   \n",
       "15              EUR        CEU                          198   \n",
       "16              EUR        IBS                          214   \n",
       "17              EUR        GBR                          182   \n",
       "18              EUR        FIN                          198   \n",
       "19              AFR        LWK                          198   \n",
       "20              AFR        GWD                          226   \n",
       "21              AFR        MSL                          170   \n",
       "22              AFR        ESN                          198   \n",
       "23              AFR        YRI                          216   \n",
       "\n",
       "    Number of Introgressed Tracts  Introgressed Tract Frequency  \\\n",
       "0                              39                      0.304688   \n",
       "1                              42                      0.247059   \n",
       "2                              23                      0.122340   \n",
       "3                              23                      0.110577   \n",
       "4                              34                      0.197674   \n",
       "5                              33                      0.161765   \n",
       "6                              35                      0.171569   \n",
       "7                              18                      0.093750   \n",
       "8                              25                      0.121359   \n",
       "9                              11                      0.053398   \n",
       "10                             22                      0.111111   \n",
       "11                             13                      0.061905   \n",
       "12                             14                      0.067308   \n",
       "13                             28                      0.150538   \n",
       "14                             13                      0.060748   \n",
       "15                              3                      0.015152   \n",
       "16                              7                      0.032710   \n",
       "17                              1                      0.005495   \n",
       "18                              3                      0.015152   \n",
       "19                              0                      0.000000   \n",
       "20                              0                      0.000000   \n",
       "21                              0                      0.000000   \n",
       "22                              0                      0.000000   \n",
       "23                              0                      0.000000   \n",
       "\n",
       "    Mean Tract Length  \n",
       "0       710666.666667  \n",
       "1       574714.285714  \n",
       "2       443304.347826  \n",
       "3       668739.130435  \n",
       "4       289941.176471  \n",
       "5       220878.787879  \n",
       "6       227800.000000  \n",
       "7       238277.777778  \n",
       "8       368000.000000  \n",
       "9       413454.545455  \n",
       "10      357818.181818  \n",
       "11      324692.307692  \n",
       "12      245500.000000  \n",
       "13      380107.142857  \n",
       "14      458769.230769  \n",
       "15      612333.333333  \n",
       "16      520000.000000  \n",
       "17      343000.000000  \n",
       "18      275333.333333  \n",
       "19                NaN  \n",
       "20                NaN  \n",
       "21                NaN  \n",
       "22                NaN  \n",
       "23                NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fe60fa4",
   "metadata": {},
   "source": [
    "## S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "98d10dc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_denisovan_specific_snp_denisty_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S3. Number of Denisovan-specific SNPs within the focal 742kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The number of Denisovan-specific SNPs---i.e., SNPs rare or absent in African populations ($<1\\%$), present in the non-African population ($>1\\%$), and uniquely shared with the Denisovan---observed in the focal 742kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the Denisovan-specific SNP genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the number of Denisovan-specific SNPs is greater than or equal to what is observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_denisovan_specific_snp_denisty_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, 7))+r' & '\\\n",
    "        + (\n",
    "            r'$<3.164e^{-4}$' if mat[i, 7][0] == '<' else\n",
    "            r'$>0.99968$' if mat[i, 7][0] == '>' else\n",
    "            convert_sci_notation_to_latex(float(mat[i, 7])) if float(mat[i, 7]) < 0.001 else\n",
    "            '{:.3f}'.format(round(float(mat[i, 7]), 3)).rstrip('0').rstrip('.')\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c2983ca",
   "metadata": {},
   "source": [
    "\\textbf{Table S3. Number of Denisovan-specific SNPs within the focal 742kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The number of Denisovan-specific SNPs---i.e., SNPs rare or absent in African populations ($<1\\%$), present in the non-African population ($>1\\%$), and uniquely shared with the Denisovan---observed in the focal 742kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the Denisovan-specific SNP genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the number of Denisovan-specific SNPs is greater than or equal to what is observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8050acf5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Focal 742kb Region (Denisovan-specific SNPs)</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>135</td>\n",
       "      <td>1.750079</td>\n",
       "      <td>5.623103</td>\n",
       "      <td>0.100031</td>\n",
       "      <td>0.196131</td>\n",
       "      <td>&lt;3.164e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>135</td>\n",
       "      <td>1.663398</td>\n",
       "      <td>5.824271</td>\n",
       "      <td>0.103609</td>\n",
       "      <td>0.203148</td>\n",
       "      <td>&lt;3.164e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>135</td>\n",
       "      <td>2.001265</td>\n",
       "      <td>6.004427</td>\n",
       "      <td>0.106814</td>\n",
       "      <td>0.209432</td>\n",
       "      <td>&lt;3.164e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>135</td>\n",
       "      <td>1.608352</td>\n",
       "      <td>4.908564</td>\n",
       "      <td>0.087319</td>\n",
       "      <td>0.171209</td>\n",
       "      <td>&lt;3.164e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>135</td>\n",
       "      <td>4.686808</td>\n",
       "      <td>10.990951</td>\n",
       "      <td>0.195520</td>\n",
       "      <td>0.383359</td>\n",
       "      <td>0.00031635558367605187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>135</td>\n",
       "      <td>3.723505</td>\n",
       "      <td>9.674721</td>\n",
       "      <td>0.172106</td>\n",
       "      <td>0.337450</td>\n",
       "      <td>0.00031635558367605187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>135</td>\n",
       "      <td>3.767479</td>\n",
       "      <td>9.997154</td>\n",
       "      <td>0.177841</td>\n",
       "      <td>0.348696</td>\n",
       "      <td>0.00031635558367605187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>135</td>\n",
       "      <td>4.422335</td>\n",
       "      <td>10.531272</td>\n",
       "      <td>0.187343</td>\n",
       "      <td>0.367326</td>\n",
       "      <td>0.00031635558367605187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>135</td>\n",
       "      <td>3.390383</td>\n",
       "      <td>8.953539</td>\n",
       "      <td>0.159276</td>\n",
       "      <td>0.312295</td>\n",
       "      <td>0.00031635558367605187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>135</td>\n",
       "      <td>2.757039</td>\n",
       "      <td>9.942565</td>\n",
       "      <td>0.176870</td>\n",
       "      <td>0.346792</td>\n",
       "      <td>0.0006327111673521037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>135</td>\n",
       "      <td>3.296742</td>\n",
       "      <td>10.724169</td>\n",
       "      <td>0.190774</td>\n",
       "      <td>0.374054</td>\n",
       "      <td>0.0009490667510281556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>135</td>\n",
       "      <td>2.693451</td>\n",
       "      <td>10.166285</td>\n",
       "      <td>0.180850</td>\n",
       "      <td>0.354595</td>\n",
       "      <td>0.0009490667510281556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>135</td>\n",
       "      <td>2.481493</td>\n",
       "      <td>8.650754</td>\n",
       "      <td>0.153890</td>\n",
       "      <td>0.301734</td>\n",
       "      <td>0.00031635558367605187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>135</td>\n",
       "      <td>3.079405</td>\n",
       "      <td>10.265428</td>\n",
       "      <td>0.182614</td>\n",
       "      <td>0.358054</td>\n",
       "      <td>0.0009490667510281556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>135</td>\n",
       "      <td>1.321101</td>\n",
       "      <td>4.473235</td>\n",
       "      <td>0.079575</td>\n",
       "      <td>0.156024</td>\n",
       "      <td>&lt;3.164e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>135</td>\n",
       "      <td>1.633344</td>\n",
       "      <td>5.242371</td>\n",
       "      <td>0.093258</td>\n",
       "      <td>0.182852</td>\n",
       "      <td>&lt;3.164e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>135</td>\n",
       "      <td>1.367922</td>\n",
       "      <td>4.821493</td>\n",
       "      <td>0.085771</td>\n",
       "      <td>0.168172</td>\n",
       "      <td>&lt;3.164e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0</td>\n",
       "      <td>1.547612</td>\n",
       "      <td>5.149185</td>\n",
       "      <td>0.091600</td>\n",
       "      <td>0.179601</td>\n",
       "      <td>&gt;0.99968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>2</td>\n",
       "      <td>1.829801</td>\n",
       "      <td>5.442938</td>\n",
       "      <td>0.096826</td>\n",
       "      <td>0.189847</td>\n",
       "      <td>0.23821575450806706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Focal 742kb Region (Denisovan-specific SNPs)  \\\n",
       "0               AMR        MXL                                           135   \n",
       "1               AMR        PEL                                           135   \n",
       "2               AMR        CLM                                           135   \n",
       "3               AMR        PUR                                           135   \n",
       "4               SAS        BEB                                           135   \n",
       "5               SAS        STU                                           135   \n",
       "6               SAS        ITU                                           135   \n",
       "7               SAS        PJL                                           135   \n",
       "8               SAS        GIH                                           135   \n",
       "9               EAS        CHB                                           135   \n",
       "10              EAS        KHV                                           135   \n",
       "11              EAS        CHS                                           135   \n",
       "12              EAS        JPT                                           135   \n",
       "13              EAS        CDX                                           135   \n",
       "14              EUR        TSI                                           135   \n",
       "15              EUR        CEU                                           135   \n",
       "16              EUR        IBS                                           135   \n",
       "17              EUR        GBR                                             0   \n",
       "18              EUR        FIN                                             2   \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                            1.750079    \n",
       "1                                            1.663398    \n",
       "2                                            2.001265    \n",
       "3                                            1.608352    \n",
       "4                                            4.686808    \n",
       "5                                            3.723505    \n",
       "6                                            3.767479    \n",
       "7                                            4.422335    \n",
       "8                                            3.390383    \n",
       "9                                            2.757039    \n",
       "10                                           3.296742    \n",
       "11                                           2.693451    \n",
       "12                                           2.481493    \n",
       "13                                           3.079405    \n",
       "14                                           1.321101    \n",
       "15                                           1.633344    \n",
       "16                                           1.367922    \n",
       "17                                           1.547612    \n",
       "18                                           1.829801    \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                            5.623103       \n",
       "1                                            5.824271       \n",
       "2                                            6.004427       \n",
       "3                                            4.908564       \n",
       "4                                           10.990951       \n",
       "5                                            9.674721       \n",
       "6                                            9.997154       \n",
       "7                                           10.531272       \n",
       "8                                            8.953539       \n",
       "9                                            9.942565       \n",
       "10                                          10.724169       \n",
       "11                                          10.166285       \n",
       "12                                           8.650754       \n",
       "13                                          10.265428       \n",
       "14                                           4.473235       \n",
       "15                                           5.242371       \n",
       "16                                           4.821493       \n",
       "17                                           5.149185       \n",
       "18                                           5.442938       \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                            0.100031    \n",
       "1                                            0.103609    \n",
       "2                                            0.106814    \n",
       "3                                            0.087319    \n",
       "4                                            0.195520    \n",
       "5                                            0.172106    \n",
       "6                                            0.177841    \n",
       "7                                            0.187343    \n",
       "8                                            0.159276    \n",
       "9                                            0.176870    \n",
       "10                                           0.190774    \n",
       "11                                           0.180850    \n",
       "12                                           0.153890    \n",
       "13                                           0.182614    \n",
       "14                                           0.079575    \n",
       "15                                           0.093258    \n",
       "16                                           0.085771    \n",
       "17                                           0.091600    \n",
       "18                                           0.096826    \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                            0.196131              \n",
       "1                                            0.203148              \n",
       "2                                            0.209432              \n",
       "3                                            0.171209              \n",
       "4                                            0.383359              \n",
       "5                                            0.337450              \n",
       "6                                            0.348696              \n",
       "7                                            0.367326              \n",
       "8                                            0.312295              \n",
       "9                                            0.346792              \n",
       "10                                           0.374054              \n",
       "11                                           0.354595              \n",
       "12                                           0.301734              \n",
       "13                                           0.358054              \n",
       "14                                           0.156024              \n",
       "15                                           0.182852              \n",
       "16                                           0.168172              \n",
       "17                                           0.179601              \n",
       "18                                           0.189847              \n",
       "\n",
       "                 $P-value$  \n",
       "0               <3.164e-04  \n",
       "1               <3.164e-04  \n",
       "2               <3.164e-04  \n",
       "3               <3.164e-04  \n",
       "4   0.00031635558367605187  \n",
       "5   0.00031635558367605187  \n",
       "6   0.00031635558367605187  \n",
       "7   0.00031635558367605187  \n",
       "8   0.00031635558367605187  \n",
       "9    0.0006327111673521037  \n",
       "10   0.0009490667510281556  \n",
       "11   0.0009490667510281556  \n",
       "12  0.00031635558367605187  \n",
       "13   0.0009490667510281556  \n",
       "14              <3.164e-04  \n",
       "15              <3.164e-04  \n",
       "16              <3.164e-04  \n",
       "17                >0.99968  \n",
       "18     0.23821575450806706  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48b59fb5",
   "metadata": {},
   "source": [
    "## S4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "403814dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_denisovan_specific_snp_denisty_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S4. Number of Denisovan-specific SNPs within the focal 72kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The number of Denisovan-specific SNPs---i.e., SNPs rare or absent in African populations ($<1\\%$), present in the non-African population ($>1\\%$), and uniquely shared with the Denisovan---observed in the focal 72kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the Denisovan-specific SNP genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the number of Denisovan-specific SNPs is greater than or equal to what is observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_denisovan_specific_snp_denisty_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, 7))+r' & '\\\n",
    "        + (\n",
    "            r'$<3.389e^{-5}$' if mat[i, 7][0] == '<' else\n",
    "            r'$>0.9999966$' if mat[i, 7][0] == '>' else\n",
    "            convert_sci_notation_to_latex(float(mat[i, 7])) if float(mat[i, 7]) < 0.001 else\n",
    "            '{:.3f}'.format(round(float(mat[i, 7]), 3)).rstrip('0').rstrip('.')\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc887a2f",
   "metadata": {},
   "source": [
    "\\textbf{Table S4. Number of Denisovan-specific SNPs within the focal 72kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The number of Denisovan-specific SNPs---i.e., SNPs rare or absent in African populations ($<1\\%$), present in the non-African population ($>1\\%$), and uniquely shared with the Denisovan---observed in the focal 72kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the Denisovan-specific SNP genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the number of Denisovan-specific SNPs is greater than or equal to what is observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8857960a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Focal 72kb Region (Denisovan-specific SNPs)</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>135</td>\n",
       "      <td>0.178079</td>\n",
       "      <td>1.536302</td>\n",
       "      <td>0.008943</td>\n",
       "      <td>0.017528</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>135</td>\n",
       "      <td>0.167711</td>\n",
       "      <td>1.503239</td>\n",
       "      <td>0.008750</td>\n",
       "      <td>0.017151</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>135</td>\n",
       "      <td>0.201321</td>\n",
       "      <td>1.630700</td>\n",
       "      <td>0.009492</td>\n",
       "      <td>0.018605</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>135</td>\n",
       "      <td>0.161782</td>\n",
       "      <td>1.441084</td>\n",
       "      <td>0.008388</td>\n",
       "      <td>0.016441</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>135</td>\n",
       "      <td>0.467118</td>\n",
       "      <td>2.449108</td>\n",
       "      <td>0.014256</td>\n",
       "      <td>0.027942</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>135</td>\n",
       "      <td>0.371438</td>\n",
       "      <td>2.120086</td>\n",
       "      <td>0.012341</td>\n",
       "      <td>0.024188</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>135</td>\n",
       "      <td>0.368491</td>\n",
       "      <td>2.159334</td>\n",
       "      <td>0.012569</td>\n",
       "      <td>0.024636</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>135</td>\n",
       "      <td>0.432831</td>\n",
       "      <td>2.292055</td>\n",
       "      <td>0.013342</td>\n",
       "      <td>0.026150</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>135</td>\n",
       "      <td>0.334338</td>\n",
       "      <td>2.064443</td>\n",
       "      <td>0.012017</td>\n",
       "      <td>0.023553</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>135</td>\n",
       "      <td>0.280400</td>\n",
       "      <td>2.009982</td>\n",
       "      <td>0.011700</td>\n",
       "      <td>0.022932</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>135</td>\n",
       "      <td>0.335592</td>\n",
       "      <td>2.153180</td>\n",
       "      <td>0.012533</td>\n",
       "      <td>0.024566</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>135</td>\n",
       "      <td>0.275419</td>\n",
       "      <td>2.009061</td>\n",
       "      <td>0.011694</td>\n",
       "      <td>0.022922</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>135</td>\n",
       "      <td>0.256243</td>\n",
       "      <td>1.906994</td>\n",
       "      <td>0.011100</td>\n",
       "      <td>0.021757</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>135</td>\n",
       "      <td>0.311808</td>\n",
       "      <td>2.068040</td>\n",
       "      <td>0.012038</td>\n",
       "      <td>0.023595</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>135</td>\n",
       "      <td>0.128884</td>\n",
       "      <td>1.291494</td>\n",
       "      <td>0.007518</td>\n",
       "      <td>0.014735</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>135</td>\n",
       "      <td>0.157140</td>\n",
       "      <td>1.387747</td>\n",
       "      <td>0.008078</td>\n",
       "      <td>0.015833</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>135</td>\n",
       "      <td>0.134271</td>\n",
       "      <td>1.357556</td>\n",
       "      <td>0.007902</td>\n",
       "      <td>0.015489</td>\n",
       "      <td>&lt;3.389e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0</td>\n",
       "      <td>0.151482</td>\n",
       "      <td>1.283256</td>\n",
       "      <td>0.007470</td>\n",
       "      <td>0.014641</td>\n",
       "      <td>&gt;0.9999966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.182822</td>\n",
       "      <td>1.371097</td>\n",
       "      <td>0.007981</td>\n",
       "      <td>0.015643</td>\n",
       "      <td>0.024360494663730307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Focal 72kb Region (Denisovan-specific SNPs)  \\\n",
       "0               AMR        MXL                                          135   \n",
       "1               AMR        PEL                                          135   \n",
       "2               AMR        CLM                                          135   \n",
       "3               AMR        PUR                                          135   \n",
       "4               SAS        BEB                                          135   \n",
       "5               SAS        STU                                          135   \n",
       "6               SAS        ITU                                          135   \n",
       "7               SAS        PJL                                          135   \n",
       "8               SAS        GIH                                          135   \n",
       "9               EAS        CHB                                          135   \n",
       "10              EAS        KHV                                          135   \n",
       "11              EAS        CHS                                          135   \n",
       "12              EAS        JPT                                          135   \n",
       "13              EAS        CDX                                          135   \n",
       "14              EUR        TSI                                          135   \n",
       "15              EUR        CEU                                          135   \n",
       "16              EUR        IBS                                          135   \n",
       "17              EUR        GBR                                            0   \n",
       "18              EUR        FIN                                            2   \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                            0.178079   \n",
       "1                                            0.167711   \n",
       "2                                            0.201321   \n",
       "3                                            0.161782   \n",
       "4                                            0.467118   \n",
       "5                                            0.371438   \n",
       "6                                            0.368491   \n",
       "7                                            0.432831   \n",
       "8                                            0.334338   \n",
       "9                                            0.280400   \n",
       "10                                           0.335592   \n",
       "11                                           0.275419   \n",
       "12                                           0.256243   \n",
       "13                                           0.311808   \n",
       "14                                           0.128884   \n",
       "15                                           0.157140   \n",
       "16                                           0.134271   \n",
       "17                                           0.151482   \n",
       "18                                           0.182822   \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                            1.536302      \n",
       "1                                            1.503239      \n",
       "2                                            1.630700      \n",
       "3                                            1.441084      \n",
       "4                                            2.449108      \n",
       "5                                            2.120086      \n",
       "6                                            2.159334      \n",
       "7                                            2.292055      \n",
       "8                                            2.064443      \n",
       "9                                            2.009982      \n",
       "10                                           2.153180      \n",
       "11                                           2.009061      \n",
       "12                                           1.906994      \n",
       "13                                           2.068040      \n",
       "14                                           1.291494      \n",
       "15                                           1.387747      \n",
       "16                                           1.357556      \n",
       "17                                           1.283256      \n",
       "18                                           1.371097      \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                            0.008943   \n",
       "1                                            0.008750   \n",
       "2                                            0.009492   \n",
       "3                                            0.008388   \n",
       "4                                            0.014256   \n",
       "5                                            0.012341   \n",
       "6                                            0.012569   \n",
       "7                                            0.013342   \n",
       "8                                            0.012017   \n",
       "9                                            0.011700   \n",
       "10                                           0.012533   \n",
       "11                                           0.011694   \n",
       "12                                           0.011100   \n",
       "13                                           0.012038   \n",
       "14                                           0.007518   \n",
       "15                                           0.008078   \n",
       "16                                           0.007902   \n",
       "17                                           0.007470   \n",
       "18                                           0.007981   \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                            0.017528             \n",
       "1                                            0.017151             \n",
       "2                                            0.018605             \n",
       "3                                            0.016441             \n",
       "4                                            0.027942             \n",
       "5                                            0.024188             \n",
       "6                                            0.024636             \n",
       "7                                            0.026150             \n",
       "8                                            0.023553             \n",
       "9                                            0.022932             \n",
       "10                                           0.024566             \n",
       "11                                           0.022922             \n",
       "12                                           0.021757             \n",
       "13                                           0.023595             \n",
       "14                                           0.014735             \n",
       "15                                           0.015833             \n",
       "16                                           0.015489             \n",
       "17                                           0.014641             \n",
       "18                                           0.015643             \n",
       "\n",
       "               $P-value$  \n",
       "0             <3.389e-05  \n",
       "1             <3.389e-05  \n",
       "2             <3.389e-05  \n",
       "3             <3.389e-05  \n",
       "4             <3.389e-05  \n",
       "5             <3.389e-05  \n",
       "6             <3.389e-05  \n",
       "7             <3.389e-05  \n",
       "8             <3.389e-05  \n",
       "9             <3.389e-05  \n",
       "10            <3.389e-05  \n",
       "11            <3.389e-05  \n",
       "12            <3.389e-05  \n",
       "13            <3.389e-05  \n",
       "14            <3.389e-05  \n",
       "15            <3.389e-05  \n",
       "16            <3.389e-05  \n",
       "17            >0.9999966  \n",
       "18  0.024360494663730307  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "744c1448",
   "metadata": {},
   "source": [
    "## S5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f1c36524",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_neanderthal_specific_snp_denisty_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S5. Number of Neanderthal-specific SNPs within the focal 742kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The number of Neanderthal-specific SNPs---i.e., SNPs rare or absent in African populations ($<1\\%$), present in the non-African population ($>1\\%$), and uniquely shared with at least one of the three high-coverage Neanderthals---observed in the focal 742kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the Neanderthal-specific SNP genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the number of Neanderthal-specific SNPs is greater than or equal to what is observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_neanderthal_specific_snp_denisty_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7576729f",
   "metadata": {},
   "source": [
    "\\textbf{Table S5. Number of Neanderthal-specific SNPs within the focal 742kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The number of Neanderthal-specific SNPs---i.e., SNPs rare or absent in African populations ($<1\\%$), present in the non-African population ($>1\\%$), and uniquely shared with at least one of the three high-coverage Neanderthals---observed in the focal 742kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the Neanderthal-specific SNP genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the number of Neanderthal-specific SNPs is greater than or equal to what is observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "019afc76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Focal 742kb Region (Neanderthal-specific SNPs)</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>80</td>\n",
       "      <td>35.183170</td>\n",
       "      <td>45.209628</td>\n",
       "      <td>0.804243</td>\n",
       "      <td>1.576892</td>\n",
       "      <td>0.158811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>95</td>\n",
       "      <td>31.069915</td>\n",
       "      <td>42.421685</td>\n",
       "      <td>0.754648</td>\n",
       "      <td>1.479649</td>\n",
       "      <td>0.096805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>95</td>\n",
       "      <td>40.197722</td>\n",
       "      <td>48.246151</td>\n",
       "      <td>0.858261</td>\n",
       "      <td>1.682804</td>\n",
       "      <td>0.146156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>95</td>\n",
       "      <td>35.199937</td>\n",
       "      <td>45.629117</td>\n",
       "      <td>0.811706</td>\n",
       "      <td>1.591523</td>\n",
       "      <td>0.116735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>95</td>\n",
       "      <td>52.221765</td>\n",
       "      <td>56.294081</td>\n",
       "      <td>1.001427</td>\n",
       "      <td>1.963512</td>\n",
       "      <td>0.230623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>95</td>\n",
       "      <td>44.078773</td>\n",
       "      <td>52.299243</td>\n",
       "      <td>0.930362</td>\n",
       "      <td>1.824174</td>\n",
       "      <td>0.177475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>93</td>\n",
       "      <td>43.830117</td>\n",
       "      <td>52.051303</td>\n",
       "      <td>0.925951</td>\n",
       "      <td>1.815526</td>\n",
       "      <td>0.185384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>95</td>\n",
       "      <td>50.680165</td>\n",
       "      <td>55.329451</td>\n",
       "      <td>0.984267</td>\n",
       "      <td>1.929867</td>\n",
       "      <td>0.217336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>94</td>\n",
       "      <td>42.474217</td>\n",
       "      <td>51.579525</td>\n",
       "      <td>0.917559</td>\n",
       "      <td>1.799071</td>\n",
       "      <td>0.169567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>82</td>\n",
       "      <td>32.609301</td>\n",
       "      <td>47.829347</td>\n",
       "      <td>0.850846</td>\n",
       "      <td>1.668266</td>\n",
       "      <td>0.168301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>84</td>\n",
       "      <td>38.056944</td>\n",
       "      <td>51.288447</td>\n",
       "      <td>0.912381</td>\n",
       "      <td>1.788918</td>\n",
       "      <td>0.199937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>69</td>\n",
       "      <td>32.633028</td>\n",
       "      <td>47.634484</td>\n",
       "      <td>0.847379</td>\n",
       "      <td>1.661470</td>\n",
       "      <td>0.199937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>83</td>\n",
       "      <td>32.508700</td>\n",
       "      <td>47.423960</td>\n",
       "      <td>0.843634</td>\n",
       "      <td>1.654127</td>\n",
       "      <td>0.164189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>70</td>\n",
       "      <td>35.868080</td>\n",
       "      <td>49.539365</td>\n",
       "      <td>0.881266</td>\n",
       "      <td>1.727911</td>\n",
       "      <td>0.219551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>87</td>\n",
       "      <td>34.629548</td>\n",
       "      <td>45.714674</td>\n",
       "      <td>0.813228</td>\n",
       "      <td>1.594507</td>\n",
       "      <td>0.141727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>86</td>\n",
       "      <td>38.689022</td>\n",
       "      <td>48.180883</td>\n",
       "      <td>0.857099</td>\n",
       "      <td>1.680528</td>\n",
       "      <td>0.167985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>87</td>\n",
       "      <td>34.863651</td>\n",
       "      <td>46.256633</td>\n",
       "      <td>0.822869</td>\n",
       "      <td>1.613411</td>\n",
       "      <td>0.137931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>33</td>\n",
       "      <td>37.390066</td>\n",
       "      <td>47.164483</td>\n",
       "      <td>0.839019</td>\n",
       "      <td>1.645076</td>\n",
       "      <td>0.388168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>23</td>\n",
       "      <td>39.594116</td>\n",
       "      <td>47.670168</td>\n",
       "      <td>0.848014</td>\n",
       "      <td>1.662714</td>\n",
       "      <td>0.481810</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  \\\n",
       "0               AMR        MXL   \n",
       "1               AMR        PEL   \n",
       "2               AMR        CLM   \n",
       "3               AMR        PUR   \n",
       "4               SAS        BEB   \n",
       "5               SAS        STU   \n",
       "6               SAS        ITU   \n",
       "7               SAS        PJL   \n",
       "8               SAS        GIH   \n",
       "9               EAS        CHB   \n",
       "10              EAS        KHV   \n",
       "11              EAS        CHS   \n",
       "12              EAS        JPT   \n",
       "13              EAS        CDX   \n",
       "14              EUR        TSI   \n",
       "15              EUR        CEU   \n",
       "16              EUR        IBS   \n",
       "17              EUR        GBR   \n",
       "18              EUR        FIN   \n",
       "\n",
       "    Focal 742kb Region (Neanderthal-specific SNPs)  \\\n",
       "0                                               80   \n",
       "1                                               95   \n",
       "2                                               95   \n",
       "3                                               95   \n",
       "4                                               95   \n",
       "5                                               95   \n",
       "6                                               93   \n",
       "7                                               95   \n",
       "8                                               94   \n",
       "9                                               82   \n",
       "10                                              84   \n",
       "11                                              69   \n",
       "12                                              83   \n",
       "13                                              70   \n",
       "14                                              87   \n",
       "15                                              86   \n",
       "16                                              87   \n",
       "17                                              33   \n",
       "18                                              23   \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           35.183170    \n",
       "1                                           31.069915    \n",
       "2                                           40.197722    \n",
       "3                                           35.199937    \n",
       "4                                           52.221765    \n",
       "5                                           44.078773    \n",
       "6                                           43.830117    \n",
       "7                                           50.680165    \n",
       "8                                           42.474217    \n",
       "9                                           32.609301    \n",
       "10                                          38.056944    \n",
       "11                                          32.633028    \n",
       "12                                          32.508700    \n",
       "13                                          35.868080    \n",
       "14                                          34.629548    \n",
       "15                                          38.689022    \n",
       "16                                          34.863651    \n",
       "17                                          37.390066    \n",
       "18                                          39.594116    \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           45.209628       \n",
       "1                                           42.421685       \n",
       "2                                           48.246151       \n",
       "3                                           45.629117       \n",
       "4                                           56.294081       \n",
       "5                                           52.299243       \n",
       "6                                           52.051303       \n",
       "7                                           55.329451       \n",
       "8                                           51.579525       \n",
       "9                                           47.829347       \n",
       "10                                          51.288447       \n",
       "11                                          47.634484       \n",
       "12                                          47.423960       \n",
       "13                                          49.539365       \n",
       "14                                          45.714674       \n",
       "15                                          48.180883       \n",
       "16                                          46.256633       \n",
       "17                                          47.164483       \n",
       "18                                          47.670168       \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                            0.804243    \n",
       "1                                            0.754648    \n",
       "2                                            0.858261    \n",
       "3                                            0.811706    \n",
       "4                                            1.001427    \n",
       "5                                            0.930362    \n",
       "6                                            0.925951    \n",
       "7                                            0.984267    \n",
       "8                                            0.917559    \n",
       "9                                            0.850846    \n",
       "10                                           0.912381    \n",
       "11                                           0.847379    \n",
       "12                                           0.843634    \n",
       "13                                           0.881266    \n",
       "14                                           0.813228    \n",
       "15                                           0.857099    \n",
       "16                                           0.822869    \n",
       "17                                           0.839019    \n",
       "18                                           0.848014    \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                            1.576892              0.158811  \n",
       "1                                            1.479649              0.096805  \n",
       "2                                            1.682804              0.146156  \n",
       "3                                            1.591523              0.116735  \n",
       "4                                            1.963512              0.230623  \n",
       "5                                            1.824174              0.177475  \n",
       "6                                            1.815526              0.185384  \n",
       "7                                            1.929867              0.217336  \n",
       "8                                            1.799071              0.169567  \n",
       "9                                            1.668266              0.168301  \n",
       "10                                           1.788918              0.199937  \n",
       "11                                           1.661470              0.199937  \n",
       "12                                           1.654127              0.164189  \n",
       "13                                           1.727911              0.219551  \n",
       "14                                           1.594507              0.141727  \n",
       "15                                           1.680528              0.167985  \n",
       "16                                           1.613411              0.137931  \n",
       "17                                           1.645076              0.388168  \n",
       "18                                           1.662714              0.481810  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f13ea168",
   "metadata": {},
   "source": [
    "## S6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4c3758ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_neanderthal_specific_snp_denisty_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S6. Number of Neanderthal-specific SNPs within the focal 72kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The number of Neanderthal-specific SNPs---i.e., SNPs rare or absent in African populations ($<1\\%$), present in the non-African population ($>1\\%$), and uniquely shared with at least one of the three high-coverage Neanderthals---observed in the focal 72kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the Neanderthal-specific SNP genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the number of Neanderthal-specific SNPs is greater than or equal to what is observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_neanderthal_specific_snp_denisty_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, 7))+r' & '\\\n",
    "        + (\n",
    "            r'$<3.389e^{-5}$' if mat[i, 7][0] == '<' else\n",
    "            r'$>0.9999966$' if mat[i, 7][0] == '>' else\n",
    "            convert_sci_notation_to_latex(float(mat[i, 7])) if float(mat[i, 7]) < 0.001 else\n",
    "            '{:.3f}'.format(round(float(mat[i, 7]), 3)).rstrip('0').rstrip('.')\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb7641f1",
   "metadata": {},
   "source": [
    "\\textbf{Table S6. Number of Neanderthal-specific SNPs within the focal 72kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The number of Neanderthal-specific SNPs---i.e., SNPs rare or absent in African populations ($<1\\%$), present in the non-African population ($>1\\%$), and uniquely shared with at least one of the three high-coverage Neanderthals---observed in the focal 72kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the Neanderthal-specific SNP genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the number of Neanderthal-specific SNPs is greater than or equal to what is observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "200c80ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Focal 72kb Region (Neanderthal-specific SNPs)</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>4</td>\n",
       "      <td>3.427444</td>\n",
       "      <td>7.030794</td>\n",
       "      <td>0.040925</td>\n",
       "      <td>0.080215</td>\n",
       "      <td>0.26257834999152974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>4</td>\n",
       "      <td>3.024632</td>\n",
       "      <td>6.598628</td>\n",
       "      <td>0.038410</td>\n",
       "      <td>0.075284</td>\n",
       "      <td>0.23452481788920887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>4</td>\n",
       "      <td>3.913569</td>\n",
       "      <td>7.441163</td>\n",
       "      <td>0.043314</td>\n",
       "      <td>0.084897</td>\n",
       "      <td>0.2957818058614264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>4</td>\n",
       "      <td>3.426664</td>\n",
       "      <td>7.039147</td>\n",
       "      <td>0.040974</td>\n",
       "      <td>0.080310</td>\n",
       "      <td>0.2618668473657462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>4</td>\n",
       "      <td>5.110215</td>\n",
       "      <td>8.376788</td>\n",
       "      <td>0.048760</td>\n",
       "      <td>0.095572</td>\n",
       "      <td>0.3750974080975775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>4</td>\n",
       "      <td>4.306048</td>\n",
       "      <td>7.852042</td>\n",
       "      <td>0.045706</td>\n",
       "      <td>0.089585</td>\n",
       "      <td>0.31997289513806537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>4</td>\n",
       "      <td>4.292360</td>\n",
       "      <td>7.800725</td>\n",
       "      <td>0.045407</td>\n",
       "      <td>0.088999</td>\n",
       "      <td>0.31966796544130105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>4</td>\n",
       "      <td>4.964527</td>\n",
       "      <td>8.337270</td>\n",
       "      <td>0.048530</td>\n",
       "      <td>0.095121</td>\n",
       "      <td>0.3642554633237337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>4</td>\n",
       "      <td>4.158597</td>\n",
       "      <td>7.775598</td>\n",
       "      <td>0.045261</td>\n",
       "      <td>0.088713</td>\n",
       "      <td>0.3106555988480434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>4</td>\n",
       "      <td>3.178181</td>\n",
       "      <td>6.905960</td>\n",
       "      <td>0.040199</td>\n",
       "      <td>0.078791</td>\n",
       "      <td>0.2376079959342707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>4</td>\n",
       "      <td>3.709334</td>\n",
       "      <td>7.365245</td>\n",
       "      <td>0.042872</td>\n",
       "      <td>0.084031</td>\n",
       "      <td>0.2749788243266136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>4</td>\n",
       "      <td>3.176893</td>\n",
       "      <td>6.877790</td>\n",
       "      <td>0.040035</td>\n",
       "      <td>0.078469</td>\n",
       "      <td>0.23828561748263594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>4</td>\n",
       "      <td>3.160529</td>\n",
       "      <td>6.906671</td>\n",
       "      <td>0.040203</td>\n",
       "      <td>0.078799</td>\n",
       "      <td>0.23625275283754024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>4</td>\n",
       "      <td>3.509470</td>\n",
       "      <td>7.218321</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.082355</td>\n",
       "      <td>0.26003726918516007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>4</td>\n",
       "      <td>3.390073</td>\n",
       "      <td>7.037404</td>\n",
       "      <td>0.040964</td>\n",
       "      <td>0.080291</td>\n",
       "      <td>0.2566491614433339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>4</td>\n",
       "      <td>3.779129</td>\n",
       "      <td>7.358516</td>\n",
       "      <td>0.042833</td>\n",
       "      <td>0.083954</td>\n",
       "      <td>0.28372014230052517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>4</td>\n",
       "      <td>3.401321</td>\n",
       "      <td>7.123345</td>\n",
       "      <td>0.041464</td>\n",
       "      <td>0.081271</td>\n",
       "      <td>0.256276469591733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0</td>\n",
       "      <td>3.654311</td>\n",
       "      <td>7.271604</td>\n",
       "      <td>0.042327</td>\n",
       "      <td>0.082963</td>\n",
       "      <td>&gt;0.9999966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0</td>\n",
       "      <td>3.854379</td>\n",
       "      <td>7.407932</td>\n",
       "      <td>0.043120</td>\n",
       "      <td>0.084518</td>\n",
       "      <td>&gt;0.9999966</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Focal 72kb Region (Neanderthal-specific SNPs)  \\\n",
       "0               AMR        MXL                                              4   \n",
       "1               AMR        PEL                                              4   \n",
       "2               AMR        CLM                                              4   \n",
       "3               AMR        PUR                                              4   \n",
       "4               SAS        BEB                                              4   \n",
       "5               SAS        STU                                              4   \n",
       "6               SAS        ITU                                              4   \n",
       "7               SAS        PJL                                              4   \n",
       "8               SAS        GIH                                              4   \n",
       "9               EAS        CHB                                              4   \n",
       "10              EAS        KHV                                              4   \n",
       "11              EAS        CHS                                              4   \n",
       "12              EAS        JPT                                              4   \n",
       "13              EAS        CDX                                              4   \n",
       "14              EUR        TSI                                              4   \n",
       "15              EUR        CEU                                              4   \n",
       "16              EUR        IBS                                              4   \n",
       "17              EUR        GBR                                              0   \n",
       "18              EUR        FIN                                              0   \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                            3.427444   \n",
       "1                                            3.024632   \n",
       "2                                            3.913569   \n",
       "3                                            3.426664   \n",
       "4                                            5.110215   \n",
       "5                                            4.306048   \n",
       "6                                            4.292360   \n",
       "7                                            4.964527   \n",
       "8                                            4.158597   \n",
       "9                                            3.178181   \n",
       "10                                           3.709334   \n",
       "11                                           3.176893   \n",
       "12                                           3.160529   \n",
       "13                                           3.509470   \n",
       "14                                           3.390073   \n",
       "15                                           3.779129   \n",
       "16                                           3.401321   \n",
       "17                                           3.654311   \n",
       "18                                           3.854379   \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                            7.030794      \n",
       "1                                            6.598628      \n",
       "2                                            7.441163      \n",
       "3                                            7.039147      \n",
       "4                                            8.376788      \n",
       "5                                            7.852042      \n",
       "6                                            7.800725      \n",
       "7                                            8.337270      \n",
       "8                                            7.775598      \n",
       "9                                            6.905960      \n",
       "10                                           7.365245      \n",
       "11                                           6.877790      \n",
       "12                                           6.906671      \n",
       "13                                           7.218321      \n",
       "14                                           7.037404      \n",
       "15                                           7.358516      \n",
       "16                                           7.123345      \n",
       "17                                           7.271604      \n",
       "18                                           7.407932      \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                            0.040925   \n",
       "1                                            0.038410   \n",
       "2                                            0.043314   \n",
       "3                                            0.040974   \n",
       "4                                            0.048760   \n",
       "5                                            0.045706   \n",
       "6                                            0.045407   \n",
       "7                                            0.048530   \n",
       "8                                            0.045261   \n",
       "9                                            0.040199   \n",
       "10                                           0.042872   \n",
       "11                                           0.040035   \n",
       "12                                           0.040203   \n",
       "13                                           0.042017   \n",
       "14                                           0.040964   \n",
       "15                                           0.042833   \n",
       "16                                           0.041464   \n",
       "17                                           0.042327   \n",
       "18                                           0.043120   \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                            0.080215             \n",
       "1                                            0.075284             \n",
       "2                                            0.084897             \n",
       "3                                            0.080310             \n",
       "4                                            0.095572             \n",
       "5                                            0.089585             \n",
       "6                                            0.088999             \n",
       "7                                            0.095121             \n",
       "8                                            0.088713             \n",
       "9                                            0.078791             \n",
       "10                                           0.084031             \n",
       "11                                           0.078469             \n",
       "12                                           0.078799             \n",
       "13                                           0.082355             \n",
       "14                                           0.080291             \n",
       "15                                           0.083954             \n",
       "16                                           0.081271             \n",
       "17                                           0.082963             \n",
       "18                                           0.084518             \n",
       "\n",
       "              $P-value$  \n",
       "0   0.26257834999152974  \n",
       "1   0.23452481788920887  \n",
       "2    0.2957818058614264  \n",
       "3    0.2618668473657462  \n",
       "4    0.3750974080975775  \n",
       "5   0.31997289513806537  \n",
       "6   0.31966796544130105  \n",
       "7    0.3642554633237337  \n",
       "8    0.3106555988480434  \n",
       "9    0.2376079959342707  \n",
       "10   0.2749788243266136  \n",
       "11  0.23828561748263594  \n",
       "12  0.23625275283754024  \n",
       "13  0.26003726918516007  \n",
       "14   0.2566491614433339  \n",
       "15  0.28372014230052517  \n",
       "16    0.256276469591733  \n",
       "17           >0.9999966  \n",
       "18           >0.9999966  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9d2162",
   "metadata": {},
   "source": [
    "## S7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "76f1f7a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/u30_afr_b_den_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "cols[2] = r'Focal 742kb Region ($U_{AFR:B:DEN}$ $(1\\%, 30\\%, 100\\%)$)'\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXp{3cm}XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S7. $\\boldsymbol{U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)}$ values for the focal 742kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ values observed in the focal 742kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). The $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ statistic quantifies the number of sites where: 1) the Denisovan allele is found in the homozygous state, 2) the Denisovan allele is at low frequency ($<1\\%$) in the African super population, and 3) the Denisovan allele is at high frequency ($>30\\%$) in the non-African population (\\textit{B}). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ value is greater than or equal to that observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/u30_afr_b_den_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, 7))+r' & '\\\n",
    "        + (\n",
    "            r'$<3.139e^{-4}$' if mat[i, 7][0] == '<' else\n",
    "            r'$>0.999686$' if mat[i, 7][0] == '>' else\n",
    "            convert_sci_notation_to_latex(float(mat[i, 7])) if float(mat[i, 7]) < 0.001 else\n",
    "            '{:.3f}'.format(round(float(mat[i, 7]), 3)).rstrip('0').rstrip('.')\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e6272fb",
   "metadata": {},
   "source": [
    "\\textbf{Table S7. $\\boldsymbol{U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)}$ values for the focal 742kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ values observed in the focal 742kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). The $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ statistic quantifies the number of sites where: 1) the Denisovan allele is found in the homozygous state, 2) the Denisovan allele is at low frequency ($<1\\%$) in the African super population, and 3) the Denisovan allele is at high frequency ($>30\\%$) in the non-African population (\\textit{B}). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ value is greater than or equal to that observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "88ec6d83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population $\\left( B \\right)$</th>\n",
       "      <th>Focal 742kb Region ($U_{AFR:B:DEN}$ $(1\\%, 30\\%, 100\\%)$)</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>136</td>\n",
       "      <td>0.312304</td>\n",
       "      <td>2.778960</td>\n",
       "      <td>0.049241</td>\n",
       "      <td>0.096547</td>\n",
       "      <td>&lt;3.139e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>0</td>\n",
       "      <td>0.666981</td>\n",
       "      <td>4.357879</td>\n",
       "      <td>0.077218</td>\n",
       "      <td>0.151403</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>0</td>\n",
       "      <td>0.155995</td>\n",
       "      <td>1.829475</td>\n",
       "      <td>0.032417</td>\n",
       "      <td>0.063560</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>0</td>\n",
       "      <td>0.124922</td>\n",
       "      <td>1.640308</td>\n",
       "      <td>0.029065</td>\n",
       "      <td>0.056988</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.175455</td>\n",
       "      <td>1.829084</td>\n",
       "      <td>0.032410</td>\n",
       "      <td>0.063547</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>0</td>\n",
       "      <td>0.129944</td>\n",
       "      <td>1.299017</td>\n",
       "      <td>0.023018</td>\n",
       "      <td>0.045131</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>0</td>\n",
       "      <td>0.180791</td>\n",
       "      <td>1.895734</td>\n",
       "      <td>0.033591</td>\n",
       "      <td>0.065862</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>0</td>\n",
       "      <td>0.161645</td>\n",
       "      <td>1.771979</td>\n",
       "      <td>0.031398</td>\n",
       "      <td>0.061563</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>0</td>\n",
       "      <td>0.174513</td>\n",
       "      <td>1.843787</td>\n",
       "      <td>0.032671</td>\n",
       "      <td>0.064057</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.438481</td>\n",
       "      <td>2.281262</td>\n",
       "      <td>0.040422</td>\n",
       "      <td>0.079256</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>0</td>\n",
       "      <td>0.414940</td>\n",
       "      <td>2.307868</td>\n",
       "      <td>0.040894</td>\n",
       "      <td>0.080181</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>0</td>\n",
       "      <td>0.415568</td>\n",
       "      <td>2.257427</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>0.078428</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>0</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>2.482245</td>\n",
       "      <td>0.043984</td>\n",
       "      <td>0.086239</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>0</td>\n",
       "      <td>0.451977</td>\n",
       "      <td>2.405342</td>\n",
       "      <td>0.042621</td>\n",
       "      <td>0.083567</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0</td>\n",
       "      <td>0.215003</td>\n",
       "      <td>2.730160</td>\n",
       "      <td>0.048376</td>\n",
       "      <td>0.094852</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0</td>\n",
       "      <td>0.212806</td>\n",
       "      <td>2.190240</td>\n",
       "      <td>0.038809</td>\n",
       "      <td>0.076094</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0</td>\n",
       "      <td>0.210609</td>\n",
       "      <td>2.191240</td>\n",
       "      <td>0.038827</td>\n",
       "      <td>0.076129</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0</td>\n",
       "      <td>0.241368</td>\n",
       "      <td>2.416814</td>\n",
       "      <td>0.042824</td>\n",
       "      <td>0.083966</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.282800</td>\n",
       "      <td>2.965178</td>\n",
       "      <td>0.052541</td>\n",
       "      <td>0.103017</td>\n",
       "      <td>&gt;0.999686</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population $\\left( B \\right)$  \\\n",
       "0               AMR                           MXL   \n",
       "1               AMR                           PEL   \n",
       "2               AMR                           CLM   \n",
       "3               AMR                           PUR   \n",
       "4               SAS                           BEB   \n",
       "5               SAS                           STU   \n",
       "6               SAS                           ITU   \n",
       "7               SAS                           PJL   \n",
       "8               SAS                           GIH   \n",
       "9               EAS                           CHB   \n",
       "10              EAS                           KHV   \n",
       "11              EAS                           CHS   \n",
       "12              EAS                           JPT   \n",
       "13              EAS                           CDX   \n",
       "14              EUR                           TSI   \n",
       "15              EUR                           CEU   \n",
       "16              EUR                           IBS   \n",
       "17              EUR                           GBR   \n",
       "18              EUR                           FIN   \n",
       "\n",
       "    Focal 742kb Region ($U_{AFR:B:DEN}$ $(1\\%, 30\\%, 100\\%)$)  \\\n",
       "0                                                 136           \n",
       "1                                                   0           \n",
       "2                                                   0           \n",
       "3                                                   0           \n",
       "4                                                   0           \n",
       "5                                                   0           \n",
       "6                                                   0           \n",
       "7                                                   0           \n",
       "8                                                   0           \n",
       "9                                                   0           \n",
       "10                                                  0           \n",
       "11                                                  0           \n",
       "12                                                  0           \n",
       "13                                                  0           \n",
       "14                                                  0           \n",
       "15                                                  0           \n",
       "16                                                  0           \n",
       "17                                                  0           \n",
       "18                                                  0           \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                            0.312304    \n",
       "1                                            0.666981    \n",
       "2                                            0.155995    \n",
       "3                                            0.124922    \n",
       "4                                            0.175455    \n",
       "5                                            0.129944    \n",
       "6                                            0.180791    \n",
       "7                                            0.161645    \n",
       "8                                            0.174513    \n",
       "9                                            0.438481    \n",
       "10                                           0.414940    \n",
       "11                                           0.415568    \n",
       "12                                           0.444444    \n",
       "13                                           0.451977    \n",
       "14                                           0.215003    \n",
       "15                                           0.212806    \n",
       "16                                           0.210609    \n",
       "17                                           0.241368    \n",
       "18                                           0.282800    \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                            2.778960       \n",
       "1                                            4.357879       \n",
       "2                                            1.829475       \n",
       "3                                            1.640308       \n",
       "4                                            1.829084       \n",
       "5                                            1.299017       \n",
       "6                                            1.895734       \n",
       "7                                            1.771979       \n",
       "8                                            1.843787       \n",
       "9                                            2.281262       \n",
       "10                                           2.307868       \n",
       "11                                           2.257427       \n",
       "12                                           2.482245       \n",
       "13                                           2.405342       \n",
       "14                                           2.730160       \n",
       "15                                           2.190240       \n",
       "16                                           2.191240       \n",
       "17                                           2.416814       \n",
       "18                                           2.965178       \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                            0.049241    \n",
       "1                                            0.077218    \n",
       "2                                            0.032417    \n",
       "3                                            0.029065    \n",
       "4                                            0.032410    \n",
       "5                                            0.023018    \n",
       "6                                            0.033591    \n",
       "7                                            0.031398    \n",
       "8                                            0.032671    \n",
       "9                                            0.040422    \n",
       "10                                           0.040894    \n",
       "11                                           0.040000    \n",
       "12                                           0.043984    \n",
       "13                                           0.042621    \n",
       "14                                           0.048376    \n",
       "15                                           0.038809    \n",
       "16                                           0.038827    \n",
       "17                                           0.042824    \n",
       "18                                           0.052541    \n",
       "\n",
       "    742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$   $P-value$  \n",
       "0                                            0.096547             <3.139e-04  \n",
       "1                                            0.151403              >0.999686  \n",
       "2                                            0.063560              >0.999686  \n",
       "3                                            0.056988              >0.999686  \n",
       "4                                            0.063547              >0.999686  \n",
       "5                                            0.045131              >0.999686  \n",
       "6                                            0.065862              >0.999686  \n",
       "7                                            0.061563              >0.999686  \n",
       "8                                            0.064057              >0.999686  \n",
       "9                                            0.079256              >0.999686  \n",
       "10                                           0.080181              >0.999686  \n",
       "11                                           0.078428              >0.999686  \n",
       "12                                           0.086239              >0.999686  \n",
       "13                                           0.083567              >0.999686  \n",
       "14                                           0.094852              >0.999686  \n",
       "15                                           0.076094              >0.999686  \n",
       "16                                           0.076129              >0.999686  \n",
       "17                                           0.083966              >0.999686  \n",
       "18                                           0.103017              >0.999686  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9143c364",
   "metadata": {},
   "source": [
    "## S8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "398dc463",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/u30_afr_b_den_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "cols[2] = r'Focal 72kb Region ($U_{AFR:B:DEN}$ $(1\\%, 30\\%, 100\\%)$)'\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXp{3cm}XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S8. $\\boldsymbol{U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)}$ values for the focal 72kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ values observed in the focal 72kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). The $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ statistic quantifies the number of sites where: 1) the Denisovan allele is found in the homozygous state, 2) the Denisovan allele is at low frequency ($<1\\%$) in the African super population, and 3) the Denisovan allele is at high frequency ($>30\\%$) in the non-African population (\\textit{B}). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ value is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/u30_afr_b_den_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(mat[i, j]).rstrip('0').rstrip('.') for j in range(2, 7))+r' & '\\\n",
    "        + (\n",
    "            r'$<3.284e^{-5}$' if mat[i, 7][0] == '<' else\n",
    "            r'$>0.999967$' if mat[i, 7][0] == '>' else\n",
    "            convert_sci_notation_to_latex(float(mat[i, 7])) if float(mat[i, 7]) < 0.001 else\n",
    "            '{:.3f}'.format(round(float(mat[i, 7]), 3)).rstrip('0').rstrip('.')\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80ed90e8",
   "metadata": {},
   "source": [
    "\\textbf{Table S8. $\\boldsymbol{U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)}$ values for the focal 72kb \\textit{MUC19} region among non-African populations in the 1000 Genomes Project.} \\newline The $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ values observed in the focal 72kb region for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). The $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ statistic quantifies the number of sites where: 1) the Denisovan allele is found in the homozygous state, 2) the Denisovan allele is at low frequency ($<1\\%$) in the African super population, and 3) the Denisovan allele is at high frequency ($>30\\%$) in the non-African population (\\textit{B}). For each non-African population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the $U_{AFR:B:DEN}(1\\%, 30\\%, 100\\%)$ value is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c915da70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population $\\left( B \\right)$</th>\n",
       "      <th>Focal 72kb Region ($U_{AFR:B:DEN}$ $(1\\%, 30\\%, 100\\%)$)</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>136</td>\n",
       "      <td>0.030045</td>\n",
       "      <td>0.859689</td>\n",
       "      <td>0.004926</td>\n",
       "      <td>0.009656</td>\n",
       "      <td>&lt;3.284e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>0</td>\n",
       "      <td>0.066428</td>\n",
       "      <td>0.932168</td>\n",
       "      <td>0.005342</td>\n",
       "      <td>0.010470</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>0</td>\n",
       "      <td>0.015564</td>\n",
       "      <td>0.504027</td>\n",
       "      <td>0.002888</td>\n",
       "      <td>0.005661</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>0</td>\n",
       "      <td>0.012084</td>\n",
       "      <td>0.456902</td>\n",
       "      <td>0.002618</td>\n",
       "      <td>0.005132</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.017469</td>\n",
       "      <td>0.482460</td>\n",
       "      <td>0.002765</td>\n",
       "      <td>0.005419</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>0</td>\n",
       "      <td>0.012511</td>\n",
       "      <td>0.379597</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>0.004264</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>0</td>\n",
       "      <td>0.016451</td>\n",
       "      <td>0.472907</td>\n",
       "      <td>0.002710</td>\n",
       "      <td>0.005312</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>0</td>\n",
       "      <td>0.016221</td>\n",
       "      <td>0.476064</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>0.005347</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>0</td>\n",
       "      <td>0.017633</td>\n",
       "      <td>0.486757</td>\n",
       "      <td>0.002789</td>\n",
       "      <td>0.005467</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>0</td>\n",
       "      <td>0.040881</td>\n",
       "      <td>0.545617</td>\n",
       "      <td>0.003127</td>\n",
       "      <td>0.006128</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>0</td>\n",
       "      <td>0.041243</td>\n",
       "      <td>0.645778</td>\n",
       "      <td>0.003701</td>\n",
       "      <td>0.007253</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>0</td>\n",
       "      <td>0.039535</td>\n",
       "      <td>0.555645</td>\n",
       "      <td>0.003184</td>\n",
       "      <td>0.006241</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>0</td>\n",
       "      <td>0.042293</td>\n",
       "      <td>0.623828</td>\n",
       "      <td>0.003575</td>\n",
       "      <td>0.007007</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>0</td>\n",
       "      <td>0.043738</td>\n",
       "      <td>0.660298</td>\n",
       "      <td>0.003784</td>\n",
       "      <td>0.007416</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0</td>\n",
       "      <td>0.021541</td>\n",
       "      <td>0.595838</td>\n",
       "      <td>0.003414</td>\n",
       "      <td>0.006692</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0</td>\n",
       "      <td>0.020917</td>\n",
       "      <td>0.543796</td>\n",
       "      <td>0.003116</td>\n",
       "      <td>0.006108</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0</td>\n",
       "      <td>0.020063</td>\n",
       "      <td>0.551205</td>\n",
       "      <td>0.003159</td>\n",
       "      <td>0.006191</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0</td>\n",
       "      <td>0.022723</td>\n",
       "      <td>0.601335</td>\n",
       "      <td>0.003446</td>\n",
       "      <td>0.006754</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.028535</td>\n",
       "      <td>0.649479</td>\n",
       "      <td>0.003722</td>\n",
       "      <td>0.007295</td>\n",
       "      <td>&gt;0.999967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population $\\left( B \\right)$  \\\n",
       "0               AMR                           MXL   \n",
       "1               AMR                           PEL   \n",
       "2               AMR                           CLM   \n",
       "3               AMR                           PUR   \n",
       "4               SAS                           BEB   \n",
       "5               SAS                           STU   \n",
       "6               SAS                           ITU   \n",
       "7               SAS                           PJL   \n",
       "8               SAS                           GIH   \n",
       "9               EAS                           CHB   \n",
       "10              EAS                           KHV   \n",
       "11              EAS                           CHS   \n",
       "12              EAS                           JPT   \n",
       "13              EAS                           CDX   \n",
       "14              EUR                           TSI   \n",
       "15              EUR                           CEU   \n",
       "16              EUR                           IBS   \n",
       "17              EUR                           GBR   \n",
       "18              EUR                           FIN   \n",
       "\n",
       "    Focal 72kb Region ($U_{AFR:B:DEN}$ $(1\\%, 30\\%, 100\\%)$)  \\\n",
       "0                                                 136          \n",
       "1                                                   0          \n",
       "2                                                   0          \n",
       "3                                                   0          \n",
       "4                                                   0          \n",
       "5                                                   0          \n",
       "6                                                   0          \n",
       "7                                                   0          \n",
       "8                                                   0          \n",
       "9                                                   0          \n",
       "10                                                  0          \n",
       "11                                                  0          \n",
       "12                                                  0          \n",
       "13                                                  0          \n",
       "14                                                  0          \n",
       "15                                                  0          \n",
       "16                                                  0          \n",
       "17                                                  0          \n",
       "18                                                  0          \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                            0.030045   \n",
       "1                                            0.066428   \n",
       "2                                            0.015564   \n",
       "3                                            0.012084   \n",
       "4                                            0.017469   \n",
       "5                                            0.012511   \n",
       "6                                            0.016451   \n",
       "7                                            0.016221   \n",
       "8                                            0.017633   \n",
       "9                                            0.040881   \n",
       "10                                           0.041243   \n",
       "11                                           0.039535   \n",
       "12                                           0.042293   \n",
       "13                                           0.043738   \n",
       "14                                           0.021541   \n",
       "15                                           0.020917   \n",
       "16                                           0.020063   \n",
       "17                                           0.022723   \n",
       "18                                           0.028535   \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                            0.859689      \n",
       "1                                            0.932168      \n",
       "2                                            0.504027      \n",
       "3                                            0.456902      \n",
       "4                                            0.482460      \n",
       "5                                            0.379597      \n",
       "6                                            0.472907      \n",
       "7                                            0.476064      \n",
       "8                                            0.486757      \n",
       "9                                            0.545617      \n",
       "10                                           0.645778      \n",
       "11                                           0.555645      \n",
       "12                                           0.623828      \n",
       "13                                           0.660298      \n",
       "14                                           0.595838      \n",
       "15                                           0.543796      \n",
       "16                                           0.551205      \n",
       "17                                           0.601335      \n",
       "18                                           0.649479      \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                            0.004926   \n",
       "1                                            0.005342   \n",
       "2                                            0.002888   \n",
       "3                                            0.002618   \n",
       "4                                            0.002765   \n",
       "5                                            0.002175   \n",
       "6                                            0.002710   \n",
       "7                                            0.002728   \n",
       "8                                            0.002789   \n",
       "9                                            0.003127   \n",
       "10                                           0.003701   \n",
       "11                                           0.003184   \n",
       "12                                           0.003575   \n",
       "13                                           0.003784   \n",
       "14                                           0.003414   \n",
       "15                                           0.003116   \n",
       "16                                           0.003159   \n",
       "17                                           0.003446   \n",
       "18                                           0.003722   \n",
       "\n",
       "    72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$   $P-value$  \n",
       "0                                            0.009656            <3.284e-05  \n",
       "1                                            0.010470             >0.999967  \n",
       "2                                            0.005661             >0.999967  \n",
       "3                                            0.005132             >0.999967  \n",
       "4                                            0.005419             >0.999967  \n",
       "5                                            0.004264             >0.999967  \n",
       "6                                            0.005312             >0.999967  \n",
       "7                                            0.005347             >0.999967  \n",
       "8                                            0.005467             >0.999967  \n",
       "9                                            0.006128             >0.999967  \n",
       "10                                           0.007253             >0.999967  \n",
       "11                                           0.006241             >0.999967  \n",
       "12                                           0.007007             >0.999967  \n",
       "13                                           0.007416             >0.999967  \n",
       "14                                           0.006692             >0.999967  \n",
       "15                                           0.006108             >0.999967  \n",
       "16                                           0.006191             >0.999967  \n",
       "17                                           0.006754             >0.999967  \n",
       "18                                           0.007295             >0.999967  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5651cf97",
   "metadata": {},
   "source": [
    "## S9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "07b0dcbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/q95_u30_afr_b_den_72kb_and_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "cols[-2] = r'Focal 72kb Region ($Q95_{AFR:B:DEN}$ $(1\\%, 100\\%)$)'\n",
    "cols[-1] = r'Focal 742kb Region ($Q95_{AFR:B:DEN}$ $(1\\%, 100\\%)$)'\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S9. $\\boldsymbol{Q95_{AFR:B:DEN}(1\\%, 100\\%)}$ values for the focal 72kb and 742kb \\textit{MUC19} regions among non-African populations in the 1000 Genomes Project.} \\newline The $Q95_{AFR:B:DEN}(1\\%, 100\\%)$ values observed in the focal 72kb and 742kb regions for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). The $Q95_{AFR:B:DEN}(1\\%, 100\\%)$ statistic quantifies $95^{th}$ percentile of the Denisovan alleles in the non-African population (\\textit{B}) conditioned on: 1) the Denisovan allele is found in the homozygous state, and 2) the Denisovan allele is at low frequency ($<1\\%$) in the African super population.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/q95_u30_afr_b_den_72kb_and_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e767b5f4",
   "metadata": {},
   "source": [
    "\\textbf{Table S9. $\\boldsymbol{Q95_{AFR:B:DEN}(1\\%, 100\\%)}$ values for the focal 72kb and 742kb \\textit{MUC19} regions among non-African populations in the 1000 Genomes Project.} \\newline The $Q95_{AFR:B:DEN}(1\\%, 100\\%)$ values observed in the focal 72kb and 742kb regions for each non-African population, stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). The $Q95_{AFR:B:DEN}(1\\%, 100\\%)$ statistic quantifies $95^{th}$ percentile of the Denisovan alleles in the non-African population (\\textit{B}) conditioned on: 1) the Denisovan allele is found in the homozygous state, and 2) the Denisovan allele is at low frequency ($<1\\%$) in the African super population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4d338a1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population $\\left( B \\right)$</th>\n",
       "      <th>Focal 72kb Region ($Q95_{AFR:B:DEN}$ $(1\\%, 100\\%)$)</th>\n",
       "      <th>Focal 742kb Region ($Q95_{AFR:B:DEN}$ $(1\\%, 100\\%)$)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.074468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.086538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>0.156977</td>\n",
       "      <td>0.156977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>0.107843</td>\n",
       "      <td>0.107843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.097087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.033981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.090909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>0.040238</td>\n",
       "      <td>0.038095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population $\\left( B \\right)$  \\\n",
       "0               AMR                           MXL   \n",
       "1               AMR                           PEL   \n",
       "2               AMR                           CLM   \n",
       "3               AMR                           PUR   \n",
       "4               SAS                           BEB   \n",
       "5               SAS                           STU   \n",
       "6               SAS                           ITU   \n",
       "7               SAS                           PJL   \n",
       "8               SAS                           GIH   \n",
       "9               EAS                           CHB   \n",
       "10              EAS                           KHV   \n",
       "11              EAS                           CHS   \n",
       "12              EAS                           JPT   \n",
       "13              EAS                           CDX   \n",
       "14              EUR                           TSI   \n",
       "15              EUR                           CEU   \n",
       "16              EUR                           IBS   \n",
       "17              EUR                           GBR   \n",
       "18              EUR                           FIN   \n",
       "\n",
       "    Focal 72kb Region ($Q95_{AFR:B:DEN}$ $(1\\%, 100\\%)$)  \\\n",
       "0                                            0.304688      \n",
       "1                                            0.217647      \n",
       "2                                            0.074468      \n",
       "3                                            0.086538      \n",
       "4                                            0.156977      \n",
       "5                                            0.112745      \n",
       "6                                            0.107843      \n",
       "7                                            0.052083      \n",
       "8                                            0.097087      \n",
       "9                                            0.033981      \n",
       "10                                           0.090909      \n",
       "11                                           0.040238      \n",
       "12                                           0.028846      \n",
       "13                                           0.123656      \n",
       "14                                           0.051402      \n",
       "15                                           0.015152      \n",
       "16                                           0.032710      \n",
       "17                                           0.005495      \n",
       "18                                           0.005051      \n",
       "\n",
       "    Focal 742kb Region ($Q95_{AFR:B:DEN}$ $(1\\%, 100\\%)$)  \n",
       "0                                            0.304688      \n",
       "1                                            0.217647      \n",
       "2                                            0.074468      \n",
       "3                                            0.086538      \n",
       "4                                            0.156977      \n",
       "5                                            0.112745      \n",
       "6                                            0.107843      \n",
       "7                                            0.052083      \n",
       "8                                            0.097087      \n",
       "9                                            0.033981      \n",
       "10                                           0.090909      \n",
       "11                                           0.038095      \n",
       "12                                           0.028846      \n",
       "13                                           0.123656      \n",
       "14                                           0.051402      \n",
       "15                                           0.015152      \n",
       "16                                           0.032710      \n",
       "17                                           0.005495      \n",
       "18                                           0.005051      "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ea797fc",
   "metadata": {},
   "source": [
    "## S10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "da117736",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/mxl_chb_ceu_pbs_per_region_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "mat[:, 0] = np.array([r'$A =$ All MXL Inds.', r'$A =$ MXL Inds. $>50\\%$ IAA', r'$A =$ MXL Inds. $<50\\%$ IAA'])\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S10. $\\boldsymbol{PBS_{A:CHB:CEU}}$ values for the focal 742kb \\textit{MUC19} region among different Indigenous American ancestry (IAA) demes in MXL.} \\newline The $PBS_{A:CHB:CEU}$ values observed in the focal 742kb region for different IAA-partitioned demes in MXL: all individuals in the MXL population ($n = 64$; \\textit{A} = All MXL Individuals), MXL individuals with genome-wide IAA $> 50\\%$ ($n = 27$; \\textit{A} = MXL Individuals $> 50\\%$ IAA), and MXL individuals with genome-wide IAA $< 50\\%$ ($n = 37$; \\textit{A} = MXL Individuals $< 50\\%$ IAA). The $PBS_{A:CHB:CEU}$ statistic uses the logarithmic transformation of pairwise $F_{ST}$ estimates to measure the branch length in deme \\textit{A} since its divergence from the two control populations: CHB and CEU. For each IAA-partitioned deme, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{A:CHB:CEU}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the $PBS_{A:CHB:CEU}$ value is greater than or equal to that observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/mxl_chb_ceu_pbs_per_region_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adba4655",
   "metadata": {},
   "source": [
    "\\textbf{Table S10. $\\boldsymbol{PBS_{A:CHB:CEU}}$ values for the focal 742kb \\textit{MUC19} region among different Indigenous American ancestry (IAA) demes in MXL.} \\newline The $PBS_{A:CHB:CEU}$ values observed in the focal 742kb region for different IAA-partitioned demes in MXL: all individuals in the MXL population ($n = 64$; \\textit{A} = All MXL Individuals), MXL individuals with genome-wide IAA $> 50\\%$ ($n = 27$; \\textit{A} = MXL Individuals $> 50\\%$ IAA), and MXL individuals with genome-wide IAA $< 50\\%$ ($n = 37$; \\textit{A} = MXL Individuals $< 50\\%$ IAA). The $PBS_{A:CHB:CEU}$ statistic uses the logarithmic transformation of pairwise $F_{ST}$ estimates to measure the branch length in deme \\textit{A} since its divergence from the two control populations: CHB and CEU. For each IAA-partitioned deme, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{A:CHB:CEU}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the $PBS_{A:CHB:CEU}$ value is greater than or equal to that observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "23bb1b6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$PBS_{A:CHB:CEU}$</th>\n",
       "      <th>Focal 742kb Region $\\left( PBS \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>$A =$ All MXL Inds.</td>\n",
       "      <td>0.065842</td>\n",
       "      <td>0.005256</td>\n",
       "      <td>0.011149</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>0.003972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>$A =$ MXL Inds. &gt;50% IAA</td>\n",
       "      <td>0.133193</td>\n",
       "      <td>0.018499</td>\n",
       "      <td>0.023611</td>\n",
       "      <td>0.000413</td>\n",
       "      <td>0.000809</td>\n",
       "      <td>0.003361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>$A =$ MXL Inds. &lt;50% IAA</td>\n",
       "      <td>0.029065</td>\n",
       "      <td>0.002137</td>\n",
       "      <td>0.007044</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000241</td>\n",
       "      <td>0.009777</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          $PBS_{A:CHB:CEU}$  Focal 742kb Region $\\left( PBS \\right)$  \\\n",
       "0       $A =$ All MXL Inds.                                 0.065842   \n",
       "1  $A =$ MXL Inds. >50% IAA                                 0.133193   \n",
       "2  $A =$ MXL Inds. <50% IAA                                 0.029065   \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           0.005256    \n",
       "1                                           0.018499    \n",
       "2                                           0.002137    \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.011149       \n",
       "1                                           0.023611       \n",
       "2                                           0.007044       \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.000195    \n",
       "1                                           0.000413    \n",
       "2                                           0.000123    \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                           0.000382              0.003972  \n",
       "1                                           0.000809              0.003361  \n",
       "2                                           0.000241              0.009777  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81bb577d",
   "metadata": {},
   "source": [
    "## S11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "25d09437",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/mxl_chb_ceu_pbs_per_region_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "mat[:, 0] = np.array([r'$A =$ All MXL Inds.', r'$A =$ MXL Inds. $>50\\%$ IAA', r'$A =$ MXL Inds. $<50\\%$ IAA'])\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S11. $\\boldsymbol{PBS_{A:CHB:CEU}}$ values for the focal 72kb \\textit{MUC19} region among different Indigenous American ancestry (IAA) demes in MXL.} \\newline The $PBS_{A:CHB:CEU}$ values observed in the focal 72kb region for different IAA-partitioned demes in MXL: all individuals in the MXL population ($n = 64$; \\textit{A} = All MXL Individuals), MXL individuals with genome-wide IAA $> 50\\%$ ($n = 27$; \\textit{A} = MXL Individuals $> 50\\%$ IAA), and MXL individuals with genome-wide IAA $< 50\\%$ ($n = 37$; \\textit{A} = MXL Individuals $< 50\\%$ IAA). The $PBS_{A:CHB:CEU}$ statistic uses the logarithmic transformation of pairwise $F_{ST}$ estimates to measure the branch length in deme \\textit{A} since its divergence from the two control populations: CHB and CEU. For each IAA-partitioned deme, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{A:CHB:CEU}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the $PBS_{A:CHB:CEU}$ value is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/mxl_chb_ceu_pbs_per_region_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3956eb99",
   "metadata": {},
   "source": [
    "\\textbf{Table S11. $\\boldsymbol{PBS_{A:CHB:CEU}}$ values for the focal 72kb \\textit{MUC19} region among different Indigenous American ancestry (IAA) demes in MXL.} \\newline The $PBS_{A:CHB:CEU}$ values observed in the focal 72kb region for different IAA-partitioned demes in MXL: all individuals in the MXL population ($n = 64$; \\textit{A} = All MXL Individuals), MXL individuals with genome-wide IAA $> 50\\%$ ($n = 27$; \\textit{A} = MXL Individuals $> 50\\%$ IAA), and MXL individuals with genome-wide IAA $< 50\\%$ ($n = 37$; \\textit{A} = MXL Individuals $< 50\\%$ IAA). The $PBS_{A:CHB:CEU}$ statistic uses the logarithmic transformation of pairwise $F_{ST}$ estimates to measure the branch length in deme \\textit{A} since its divergence from the two control populations: CHB and CEU. For each IAA-partitioned deme, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{A:CHB:CEU}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the $PBS_{A:CHB:CEU}$ value is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "707c3f27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$PBS_{A:CHB:CEU}$</th>\n",
       "      <th>Focal 72kb Region $\\left( PBS \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>$A =$ All MXL Inds.</td>\n",
       "      <td>0.127226</td>\n",
       "      <td>0.008982</td>\n",
       "      <td>0.018510</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.002241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>$A =$ MXL Inds. &gt;50% IAA</td>\n",
       "      <td>0.217440</td>\n",
       "      <td>0.022913</td>\n",
       "      <td>0.036177</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.003378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>$A =$ MXL Inds. &lt;50% IAA</td>\n",
       "      <td>0.071933</td>\n",
       "      <td>0.005169</td>\n",
       "      <td>0.013232</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000146</td>\n",
       "      <td>0.006345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          $PBS_{A:CHB:CEU}$  Focal 72kb Region $\\left( PBS \\right)$  \\\n",
       "0       $A =$ All MXL Inds.                                0.127226   \n",
       "1  $A =$ MXL Inds. >50% IAA                                0.217440   \n",
       "2  $A =$ MXL Inds. <50% IAA                                0.071933   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           0.008982   \n",
       "1                                           0.022913   \n",
       "2                                           0.005169   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.018510      \n",
       "1                                           0.036177      \n",
       "2                                           0.013232      \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.000104   \n",
       "1                                           0.000203   \n",
       "2                                           0.000074   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                           0.000204             0.002241  \n",
       "1                                           0.000398             0.003378  \n",
       "2                                           0.000146             0.006345  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d5a2e6",
   "metadata": {},
   "source": [
    "## S12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2e7876b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_ihs_critical_scores_proportions_742kb.csv.gz')\n",
    "df = df[df['SNP Set'] == 'All SNPs']\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 3, 4, 5, 6, 7, 8, 9]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 3, 4, 5, 6, 7, 8, 9]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S12. Integrated Haplotype Scores (\\textit{iHS}) for the focal 742kb \\textit{MUC19} region among 1000 Genomes Project populations.} \\newline The normalized $\\mid iHS \\mid$ scores observed at the focal 742kb region for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). For each population, the normalized \\textit{iHS} scores were calculated for all SNPs with a minor allele frequency $> 5\\%$. For each population, the table reports the total number of \\textit{iHS} scores observed, the number of extreme \\textit{iHS} scores (i.e., $\\mid iHS \\mid > 2$), the observed proportion of extreme \\textit{iHS} scores (calculated as the number of extreme \\textit{iHS} scores normalized by the total number of \\textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \\textit{iHS} scores determined from the genomic background distribution (i.e., non-overlapping 742kb windows with more than 10 SNPs), the total number of 742kb windows in the genomic background distribution, the number of 742kb windows where the observed proportion of extreme \\textit{iHS} scores is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \\textit{iHS} scores (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \\textit{iHS} scores). An observed proportion of extreme \\textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_ihs_critical_scores_proportions_all_snps_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            '0' if mat[i, j] == 0 else\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8a493b0",
   "metadata": {},
   "source": [
    "\\textbf{Table S12. Integrated Haplotype Scores (\\textit{iHS}) for the focal 742kb \\textit{MUC19} region among 1000 Genomes Project populations.} \\newline The normalized $\\mid iHS \\mid$ scores observed at the focal 742kb region for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). For each population, the normalized \\textit{iHS} scores were calculated for all SNPs with a minor allele frequency $> 5\\%$. For each population, the table reports the total number of \\textit{iHS} scores observed, the number of extreme \\textit{iHS} scores (i.e., $\\mid iHS \\mid > 2$), the observed proportion of extreme \\textit{iHS} scores (calculated as the number of extreme \\textit{iHS} scores normalized by the total number of \\textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \\textit{iHS} scores determined from the genomic background distribution (i.e., non-overlapping 742kb windows with more than 10 SNPs), the total number of 742kb windows in the genomic background distribution, the number of 742kb windows where the observed proportion of extreme \\textit{iHS} scores is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \\textit{iHS} scores (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \\textit{iHS} scores). An observed proportion of extreme \\textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6fa25eed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Focal 742kb Region (Total SNPs)</th>\n",
       "      <th>Focal 742kb Region (SNPs with $\\mid iHS \\mid &gt; 2$)</th>\n",
       "      <th>Focal 742kb Region (Prop. of SNPs with $\\mid iHS \\mid &gt; 2$)</th>\n",
       "      <th>742kb Non-overlapping Windows ($99^{th}$ Percentile)</th>\n",
       "      <th>742kb Non-overlapping Windows (Total SNPs $&gt; 10$)</th>\n",
       "      <th>Focal 742kb Region $&gt;$ 742kb Non-overlapping Windows</th>\n",
       "      <th>Focal 742kb Region (Percentile Rank)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>2248</td>\n",
       "      <td>599</td>\n",
       "      <td>0.266459</td>\n",
       "      <td>0.210119</td>\n",
       "      <td>3595.0</td>\n",
       "      <td>3578.0</td>\n",
       "      <td>99.527121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>2158</td>\n",
       "      <td>162</td>\n",
       "      <td>0.075070</td>\n",
       "      <td>0.280090</td>\n",
       "      <td>3591.0</td>\n",
       "      <td>2983.0</td>\n",
       "      <td>83.068783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>2397</td>\n",
       "      <td>53</td>\n",
       "      <td>0.022111</td>\n",
       "      <td>0.237278</td>\n",
       "      <td>3596.0</td>\n",
       "      <td>1165.0</td>\n",
       "      <td>32.397108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>2407</td>\n",
       "      <td>68</td>\n",
       "      <td>0.028251</td>\n",
       "      <td>0.213893</td>\n",
       "      <td>3597.0</td>\n",
       "      <td>1503.0</td>\n",
       "      <td>41.784821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>2177</td>\n",
       "      <td>78</td>\n",
       "      <td>0.035829</td>\n",
       "      <td>0.273717</td>\n",
       "      <td>3595.0</td>\n",
       "      <td>2106.0</td>\n",
       "      <td>58.581363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>2197</td>\n",
       "      <td>65</td>\n",
       "      <td>0.029586</td>\n",
       "      <td>0.279665</td>\n",
       "      <td>3596.0</td>\n",
       "      <td>1781.0</td>\n",
       "      <td>49.527253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>1976</td>\n",
       "      <td>23</td>\n",
       "      <td>0.011640</td>\n",
       "      <td>0.259856</td>\n",
       "      <td>3596.0</td>\n",
       "      <td>668.0</td>\n",
       "      <td>18.576196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>2064</td>\n",
       "      <td>32</td>\n",
       "      <td>0.015504</td>\n",
       "      <td>0.272384</td>\n",
       "      <td>3596.0</td>\n",
       "      <td>945.0</td>\n",
       "      <td>26.279199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>2195</td>\n",
       "      <td>24</td>\n",
       "      <td>0.010934</td>\n",
       "      <td>0.262338</td>\n",
       "      <td>3596.0</td>\n",
       "      <td>596.0</td>\n",
       "      <td>16.573971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>1591</td>\n",
       "      <td>48</td>\n",
       "      <td>0.030170</td>\n",
       "      <td>0.233153</td>\n",
       "      <td>3594.0</td>\n",
       "      <td>1795.0</td>\n",
       "      <td>49.944352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>2010</td>\n",
       "      <td>59</td>\n",
       "      <td>0.029353</td>\n",
       "      <td>0.224836</td>\n",
       "      <td>3594.0</td>\n",
       "      <td>1719.0</td>\n",
       "      <td>47.829716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>1586</td>\n",
       "      <td>40</td>\n",
       "      <td>0.025221</td>\n",
       "      <td>0.249063</td>\n",
       "      <td>3595.0</td>\n",
       "      <td>1529.0</td>\n",
       "      <td>42.531293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>1618</td>\n",
       "      <td>78</td>\n",
       "      <td>0.048208</td>\n",
       "      <td>0.239478</td>\n",
       "      <td>3594.0</td>\n",
       "      <td>2481.0</td>\n",
       "      <td>69.031720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>1955</td>\n",
       "      <td>76</td>\n",
       "      <td>0.038875</td>\n",
       "      <td>0.222919</td>\n",
       "      <td>3594.0</td>\n",
       "      <td>2158.0</td>\n",
       "      <td>60.044519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>2323</td>\n",
       "      <td>83</td>\n",
       "      <td>0.035730</td>\n",
       "      <td>0.225909</td>\n",
       "      <td>3594.0</td>\n",
       "      <td>1995.0</td>\n",
       "      <td>55.509182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>2090</td>\n",
       "      <td>66</td>\n",
       "      <td>0.031579</td>\n",
       "      <td>0.239886</td>\n",
       "      <td>3595.0</td>\n",
       "      <td>1824.0</td>\n",
       "      <td>50.737135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>1944</td>\n",
       "      <td>40</td>\n",
       "      <td>0.020576</td>\n",
       "      <td>0.228981</td>\n",
       "      <td>3595.0</td>\n",
       "      <td>1178.0</td>\n",
       "      <td>32.767733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>1986</td>\n",
       "      <td>58</td>\n",
       "      <td>0.029204</td>\n",
       "      <td>0.214151</td>\n",
       "      <td>3596.0</td>\n",
       "      <td>1766.0</td>\n",
       "      <td>49.110122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>2007</td>\n",
       "      <td>10</td>\n",
       "      <td>0.004983</td>\n",
       "      <td>0.230146</td>\n",
       "      <td>3594.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>6.037841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>AFR</td>\n",
       "      <td>LWK</td>\n",
       "      <td>2218</td>\n",
       "      <td>4</td>\n",
       "      <td>0.001803</td>\n",
       "      <td>0.192903</td>\n",
       "      <td>3598.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.333519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>AFR</td>\n",
       "      <td>GWD</td>\n",
       "      <td>2324</td>\n",
       "      <td>19</td>\n",
       "      <td>0.008176</td>\n",
       "      <td>0.204919</td>\n",
       "      <td>3597.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>2.613289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>AFR</td>\n",
       "      <td>MSL</td>\n",
       "      <td>2152</td>\n",
       "      <td>45</td>\n",
       "      <td>0.020911</td>\n",
       "      <td>0.207785</td>\n",
       "      <td>3598.0</td>\n",
       "      <td>634.0</td>\n",
       "      <td>17.620901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>AFR</td>\n",
       "      <td>ESN</td>\n",
       "      <td>2321</td>\n",
       "      <td>28</td>\n",
       "      <td>0.012064</td>\n",
       "      <td>0.194454</td>\n",
       "      <td>3597.0</td>\n",
       "      <td>280.0</td>\n",
       "      <td>7.784265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>AFR</td>\n",
       "      <td>YRI</td>\n",
       "      <td>2277</td>\n",
       "      <td>34</td>\n",
       "      <td>0.014932</td>\n",
       "      <td>0.220867</td>\n",
       "      <td>3598.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>10.061145</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Focal 742kb Region (Total SNPs)  \\\n",
       "0               AMR        MXL                             2248   \n",
       "4               AMR        PEL                             2158   \n",
       "8               AMR        CLM                             2397   \n",
       "12              AMR        PUR                             2407   \n",
       "16              SAS        BEB                             2177   \n",
       "20              SAS        STU                             2197   \n",
       "24              SAS        ITU                             1976   \n",
       "28              SAS        PJL                             2064   \n",
       "32              SAS        GIH                             2195   \n",
       "36              EAS        CHB                             1591   \n",
       "40              EAS        KHV                             2010   \n",
       "44              EAS        CHS                             1586   \n",
       "48              EAS        JPT                             1618   \n",
       "52              EAS        CDX                             1955   \n",
       "56              EUR        TSI                             2323   \n",
       "60              EUR        CEU                             2090   \n",
       "64              EUR        IBS                             1944   \n",
       "68              EUR        GBR                             1986   \n",
       "72              EUR        FIN                             2007   \n",
       "76              AFR        LWK                             2218   \n",
       "80              AFR        GWD                             2324   \n",
       "84              AFR        MSL                             2152   \n",
       "88              AFR        ESN                             2321   \n",
       "92              AFR        YRI                             2277   \n",
       "\n",
       "    Focal 742kb Region (SNPs with $\\mid iHS \\mid > 2$)  \\\n",
       "0                                                 599    \n",
       "4                                                 162    \n",
       "8                                                  53    \n",
       "12                                                 68    \n",
       "16                                                 78    \n",
       "20                                                 65    \n",
       "24                                                 23    \n",
       "28                                                 32    \n",
       "32                                                 24    \n",
       "36                                                 48    \n",
       "40                                                 59    \n",
       "44                                                 40    \n",
       "48                                                 78    \n",
       "52                                                 76    \n",
       "56                                                 83    \n",
       "60                                                 66    \n",
       "64                                                 40    \n",
       "68                                                 58    \n",
       "72                                                 10    \n",
       "76                                                  4    \n",
       "80                                                 19    \n",
       "84                                                 45    \n",
       "88                                                 28    \n",
       "92                                                 34    \n",
       "\n",
       "    Focal 742kb Region (Prop. of SNPs with $\\mid iHS \\mid > 2$)  \\\n",
       "0                                            0.266459             \n",
       "4                                            0.075070             \n",
       "8                                            0.022111             \n",
       "12                                           0.028251             \n",
       "16                                           0.035829             \n",
       "20                                           0.029586             \n",
       "24                                           0.011640             \n",
       "28                                           0.015504             \n",
       "32                                           0.010934             \n",
       "36                                           0.030170             \n",
       "40                                           0.029353             \n",
       "44                                           0.025221             \n",
       "48                                           0.048208             \n",
       "52                                           0.038875             \n",
       "56                                           0.035730             \n",
       "60                                           0.031579             \n",
       "64                                           0.020576             \n",
       "68                                           0.029204             \n",
       "72                                           0.004983             \n",
       "76                                           0.001803             \n",
       "80                                           0.008176             \n",
       "84                                           0.020911             \n",
       "88                                           0.012064             \n",
       "92                                           0.014932             \n",
       "\n",
       "    742kb Non-overlapping Windows ($99^{th}$ Percentile)  \\\n",
       "0                                            0.210119      \n",
       "4                                            0.280090      \n",
       "8                                            0.237278      \n",
       "12                                           0.213893      \n",
       "16                                           0.273717      \n",
       "20                                           0.279665      \n",
       "24                                           0.259856      \n",
       "28                                           0.272384      \n",
       "32                                           0.262338      \n",
       "36                                           0.233153      \n",
       "40                                           0.224836      \n",
       "44                                           0.249063      \n",
       "48                                           0.239478      \n",
       "52                                           0.222919      \n",
       "56                                           0.225909      \n",
       "60                                           0.239886      \n",
       "64                                           0.228981      \n",
       "68                                           0.214151      \n",
       "72                                           0.230146      \n",
       "76                                           0.192903      \n",
       "80                                           0.204919      \n",
       "84                                           0.207785      \n",
       "88                                           0.194454      \n",
       "92                                           0.220867      \n",
       "\n",
       "    742kb Non-overlapping Windows (Total SNPs $> 10$)  \\\n",
       "0                                              3595.0   \n",
       "4                                              3591.0   \n",
       "8                                              3596.0   \n",
       "12                                             3597.0   \n",
       "16                                             3595.0   \n",
       "20                                             3596.0   \n",
       "24                                             3596.0   \n",
       "28                                             3596.0   \n",
       "32                                             3596.0   \n",
       "36                                             3594.0   \n",
       "40                                             3594.0   \n",
       "44                                             3595.0   \n",
       "48                                             3594.0   \n",
       "52                                             3594.0   \n",
       "56                                             3594.0   \n",
       "60                                             3595.0   \n",
       "64                                             3595.0   \n",
       "68                                             3596.0   \n",
       "72                                             3594.0   \n",
       "76                                             3598.0   \n",
       "80                                             3597.0   \n",
       "84                                             3598.0   \n",
       "88                                             3597.0   \n",
       "92                                             3598.0   \n",
       "\n",
       "    Focal 742kb Region $>$ 742kb Non-overlapping Windows  \\\n",
       "0                                              3578.0      \n",
       "4                                              2983.0      \n",
       "8                                              1165.0      \n",
       "12                                             1503.0      \n",
       "16                                             2106.0      \n",
       "20                                             1781.0      \n",
       "24                                              668.0      \n",
       "28                                              945.0      \n",
       "32                                              596.0      \n",
       "36                                             1795.0      \n",
       "40                                             1719.0      \n",
       "44                                             1529.0      \n",
       "48                                             2481.0      \n",
       "52                                             2158.0      \n",
       "56                                             1995.0      \n",
       "60                                             1824.0      \n",
       "64                                             1178.0      \n",
       "68                                             1766.0      \n",
       "72                                              217.0      \n",
       "76                                               12.0      \n",
       "80                                               94.0      \n",
       "84                                              634.0      \n",
       "88                                              280.0      \n",
       "92                                              362.0      \n",
       "\n",
       "    Focal 742kb Region (Percentile Rank)  \n",
       "0                              99.527121  \n",
       "4                              83.068783  \n",
       "8                              32.397108  \n",
       "12                             41.784821  \n",
       "16                             58.581363  \n",
       "20                             49.527253  \n",
       "24                             18.576196  \n",
       "28                             26.279199  \n",
       "32                             16.573971  \n",
       "36                             49.944352  \n",
       "40                             47.829716  \n",
       "44                             42.531293  \n",
       "48                             69.031720  \n",
       "52                             60.044519  \n",
       "56                             55.509182  \n",
       "60                             50.737135  \n",
       "64                             32.767733  \n",
       "68                             49.110122  \n",
       "72                              6.037841  \n",
       "76                              0.333519  \n",
       "80                              2.613289  \n",
       "84                             17.620901  \n",
       "88                              7.784265  \n",
       "92                             10.061145  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 3, 4, 5, 6, 7, 8, 9]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "234b534f",
   "metadata": {},
   "source": [
    "## S13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d8ea6d49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_ihs_critical_scores_proportions_72kb.csv.gz')\n",
    "df = df[df['SNP Set'] == 'All SNPs']\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 3, 4, 5, 6, 7, 8, 9]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 3, 4, 5, 6, 7, 8, 9]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S13. Integrated Haplotype Scores (\\textit{iHS}) for the focal 72kb \\textit{MUC19} region among 1000 Genomes Project populations.} \\newline The normalized $\\mid iHS \\mid$ scores observed at the focal 742kb region for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). For each population, the normalized \\textit{iHS} scores were calculated for all SNPs with a minor allele frequency $> 5\\%$. For each population, the table reports the total number of \\textit{iHS} scores observed, the number of extreme \\textit{iHS} scores (i.e., $\\mid iHS \\mid > 2$), the observed proportion of extreme \\textit{iHS} scores (calculated as the number of extreme \\textit{iHS} scores normalized by the total number of \\textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \\textit{iHS} scores determined from the genomic background distribution (i.e., non-overlapping 72kb windows with more than 10 SNPs), the total number of 72kb windows in the genomic background distribution, the number of 72kb windows where the observed proportion of extreme \\textit{iHS} scores is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \\textit{iHS} scores (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \\textit{iHS} scores). An observed proportion of extreme \\textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_ihs_critical_scores_proportions_all_snps_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            '0' if mat[i, j] == 0 else\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c031fa54",
   "metadata": {},
   "source": [
    "\\textbf{Table S13. Integrated Haplotype Scores (\\textit{iHS}) for the focal 72kb \\textit{MUC19} region among 1000 Genomes Project populations.} \\newline The normalized $\\mid iHS \\mid$ scores observed at the focal 742kb region for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). For each population, the normalized \\textit{iHS} scores were calculated for all SNPs with a minor allele frequency $> 5\\%$. For each population, the table reports the total number of \\textit{iHS} scores observed, the number of extreme \\textit{iHS} scores (i.e., $\\mid iHS \\mid > 2$), the observed proportion of extreme \\textit{iHS} scores (calculated as the number of extreme \\textit{iHS} scores normalized by the total number of \\textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \\textit{iHS} scores determined from the genomic background distribution (i.e., non-overlapping 72kb windows with more than 10 SNPs), the total number of 72kb windows in the genomic background distribution, the number of 72kb windows where the observed proportion of extreme \\textit{iHS} scores is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \\textit{iHS} scores (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \\textit{iHS} scores). An observed proportion of extreme \\textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "5afea0d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Focal 72kb Region (Total SNPs)</th>\n",
       "      <th>Focal 72kb Region (SNPs with $\\mid iHS \\mid &gt; 2$)</th>\n",
       "      <th>Focal 72kb Region (Prop. of SNPs with $\\mid iHS \\mid &gt; 2$)</th>\n",
       "      <th>72kb Non-overlapping Windows ($99^{th}$ Percentile)</th>\n",
       "      <th>72kb Non-overlapping Windows (Total SNPs $&gt; 10$)</th>\n",
       "      <th>Focal 72kb Region $&gt;$ 72kb Non-overlapping Windows</th>\n",
       "      <th>Focal 72kb Region (Percentile Rank)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>425</td>\n",
       "      <td>229</td>\n",
       "      <td>0.538824</td>\n",
       "      <td>0.462339</td>\n",
       "      <td>36411.0</td>\n",
       "      <td>36202.0</td>\n",
       "      <td>99.425998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>403</td>\n",
       "      <td>86</td>\n",
       "      <td>0.213400</td>\n",
       "      <td>0.491600</td>\n",
       "      <td>36293.0</td>\n",
       "      <td>34299.0</td>\n",
       "      <td>94.505828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>431</td>\n",
       "      <td>7</td>\n",
       "      <td>0.016241</td>\n",
       "      <td>0.473293</td>\n",
       "      <td>36459.0</td>\n",
       "      <td>19972.0</td>\n",
       "      <td>54.779341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>441</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002268</td>\n",
       "      <td>0.447304</td>\n",
       "      <td>36474.0</td>\n",
       "      <td>11772.0</td>\n",
       "      <td>32.275045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>420</td>\n",
       "      <td>6</td>\n",
       "      <td>0.014286</td>\n",
       "      <td>0.525785</td>\n",
       "      <td>36388.0</td>\n",
       "      <td>21536.0</td>\n",
       "      <td>59.184346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>413</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.523772</td>\n",
       "      <td>36376.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>403</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.520320</td>\n",
       "      <td>36383.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>413</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.516129</td>\n",
       "      <td>36409.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>415</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>36354.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>168</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005952</td>\n",
       "      <td>0.489533</td>\n",
       "      <td>36193.0</td>\n",
       "      <td>17256.0</td>\n",
       "      <td>47.677728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>394</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002538</td>\n",
       "      <td>0.494527</td>\n",
       "      <td>36283.0</td>\n",
       "      <td>15128.0</td>\n",
       "      <td>41.694457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>169</td>\n",
       "      <td>3</td>\n",
       "      <td>0.017751</td>\n",
       "      <td>0.495570</td>\n",
       "      <td>36208.0</td>\n",
       "      <td>22884.0</td>\n",
       "      <td>63.201502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.489517</td>\n",
       "      <td>36192.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>393</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.479004</td>\n",
       "      <td>36219.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>452</td>\n",
       "      <td>13</td>\n",
       "      <td>0.028761</td>\n",
       "      <td>0.478261</td>\n",
       "      <td>36375.0</td>\n",
       "      <td>24780.0</td>\n",
       "      <td>68.123711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>232</td>\n",
       "      <td>6</td>\n",
       "      <td>0.025862</td>\n",
       "      <td>0.482557</td>\n",
       "      <td>36393.0</td>\n",
       "      <td>24602.0</td>\n",
       "      <td>67.600912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>217</td>\n",
       "      <td>4</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>0.486590</td>\n",
       "      <td>36372.0</td>\n",
       "      <td>22140.0</td>\n",
       "      <td>60.871000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>226</td>\n",
       "      <td>2</td>\n",
       "      <td>0.008850</td>\n",
       "      <td>0.487308</td>\n",
       "      <td>36348.0</td>\n",
       "      <td>18741.0</td>\n",
       "      <td>51.559921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>234</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004274</td>\n",
       "      <td>0.479831</td>\n",
       "      <td>36374.0</td>\n",
       "      <td>14941.0</td>\n",
       "      <td>41.076043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>AFR</td>\n",
       "      <td>LWK</td>\n",
       "      <td>217</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.401442</td>\n",
       "      <td>36570.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>AFR</td>\n",
       "      <td>GWD</td>\n",
       "      <td>219</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004566</td>\n",
       "      <td>0.430478</td>\n",
       "      <td>36562.0</td>\n",
       "      <td>9341.0</td>\n",
       "      <td>25.548384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>AFR</td>\n",
       "      <td>MSL</td>\n",
       "      <td>208</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004808</td>\n",
       "      <td>0.424061</td>\n",
       "      <td>36576.0</td>\n",
       "      <td>9130.0</td>\n",
       "      <td>24.961724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>AFR</td>\n",
       "      <td>ESN</td>\n",
       "      <td>223</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004484</td>\n",
       "      <td>0.431935</td>\n",
       "      <td>36565.0</td>\n",
       "      <td>9545.0</td>\n",
       "      <td>26.104198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>AFR</td>\n",
       "      <td>YRI</td>\n",
       "      <td>211</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.448686</td>\n",
       "      <td>36554.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Focal 72kb Region (Total SNPs)  \\\n",
       "0               AMR        MXL                             425   \n",
       "4               AMR        PEL                             403   \n",
       "8               AMR        CLM                             431   \n",
       "12              AMR        PUR                             441   \n",
       "16              SAS        BEB                             420   \n",
       "20              SAS        STU                             413   \n",
       "24              SAS        ITU                             403   \n",
       "28              SAS        PJL                             413   \n",
       "32              SAS        GIH                             415   \n",
       "36              EAS        CHB                             168   \n",
       "40              EAS        KHV                             394   \n",
       "44              EAS        CHS                             169   \n",
       "48              EAS        JPT                             180   \n",
       "52              EAS        CDX                             393   \n",
       "56              EUR        TSI                             452   \n",
       "60              EUR        CEU                             232   \n",
       "64              EUR        IBS                             217   \n",
       "68              EUR        GBR                             226   \n",
       "72              EUR        FIN                             234   \n",
       "76              AFR        LWK                             217   \n",
       "80              AFR        GWD                             219   \n",
       "84              AFR        MSL                             208   \n",
       "88              AFR        ESN                             223   \n",
       "92              AFR        YRI                             211   \n",
       "\n",
       "    Focal 72kb Region (SNPs with $\\mid iHS \\mid > 2$)  \\\n",
       "0                                                 229   \n",
       "4                                                  86   \n",
       "8                                                   7   \n",
       "12                                                  1   \n",
       "16                                                  6   \n",
       "20                                                  0   \n",
       "24                                                  0   \n",
       "28                                                  0   \n",
       "32                                                  0   \n",
       "36                                                  1   \n",
       "40                                                  1   \n",
       "44                                                  3   \n",
       "48                                                  0   \n",
       "52                                                  0   \n",
       "56                                                 13   \n",
       "60                                                  6   \n",
       "64                                                  4   \n",
       "68                                                  2   \n",
       "72                                                  1   \n",
       "76                                                  0   \n",
       "80                                                  1   \n",
       "84                                                  1   \n",
       "88                                                  1   \n",
       "92                                                  0   \n",
       "\n",
       "    Focal 72kb Region (Prop. of SNPs with $\\mid iHS \\mid > 2$)  \\\n",
       "0                                            0.538824            \n",
       "4                                            0.213400            \n",
       "8                                            0.016241            \n",
       "12                                           0.002268            \n",
       "16                                           0.014286            \n",
       "20                                           0.000000            \n",
       "24                                           0.000000            \n",
       "28                                           0.000000            \n",
       "32                                           0.000000            \n",
       "36                                           0.005952            \n",
       "40                                           0.002538            \n",
       "44                                           0.017751            \n",
       "48                                           0.000000            \n",
       "52                                           0.000000            \n",
       "56                                           0.028761            \n",
       "60                                           0.025862            \n",
       "64                                           0.018433            \n",
       "68                                           0.008850            \n",
       "72                                           0.004274            \n",
       "76                                           0.000000            \n",
       "80                                           0.004566            \n",
       "84                                           0.004808            \n",
       "88                                           0.004484            \n",
       "92                                           0.000000            \n",
       "\n",
       "    72kb Non-overlapping Windows ($99^{th}$ Percentile)  \\\n",
       "0                                            0.462339     \n",
       "4                                            0.491600     \n",
       "8                                            0.473293     \n",
       "12                                           0.447304     \n",
       "16                                           0.525785     \n",
       "20                                           0.523772     \n",
       "24                                           0.520320     \n",
       "28                                           0.516129     \n",
       "32                                           0.500000     \n",
       "36                                           0.489533     \n",
       "40                                           0.494527     \n",
       "44                                           0.495570     \n",
       "48                                           0.489517     \n",
       "52                                           0.479004     \n",
       "56                                           0.478261     \n",
       "60                                           0.482557     \n",
       "64                                           0.486590     \n",
       "68                                           0.487308     \n",
       "72                                           0.479831     \n",
       "76                                           0.401442     \n",
       "80                                           0.430478     \n",
       "84                                           0.424061     \n",
       "88                                           0.431935     \n",
       "92                                           0.448686     \n",
       "\n",
       "    72kb Non-overlapping Windows (Total SNPs $> 10$)  \\\n",
       "0                                            36411.0   \n",
       "4                                            36293.0   \n",
       "8                                            36459.0   \n",
       "12                                           36474.0   \n",
       "16                                           36388.0   \n",
       "20                                           36376.0   \n",
       "24                                           36383.0   \n",
       "28                                           36409.0   \n",
       "32                                           36354.0   \n",
       "36                                           36193.0   \n",
       "40                                           36283.0   \n",
       "44                                           36208.0   \n",
       "48                                           36192.0   \n",
       "52                                           36219.0   \n",
       "56                                           36375.0   \n",
       "60                                           36393.0   \n",
       "64                                           36372.0   \n",
       "68                                           36348.0   \n",
       "72                                           36374.0   \n",
       "76                                           36570.0   \n",
       "80                                           36562.0   \n",
       "84                                           36576.0   \n",
       "88                                           36565.0   \n",
       "92                                           36554.0   \n",
       "\n",
       "    Focal 72kb Region $>$ 72kb Non-overlapping Windows  \\\n",
       "0                                             36202.0    \n",
       "4                                             34299.0    \n",
       "8                                             19972.0    \n",
       "12                                            11772.0    \n",
       "16                                            21536.0    \n",
       "20                                                0.0    \n",
       "24                                                0.0    \n",
       "28                                                0.0    \n",
       "32                                                0.0    \n",
       "36                                            17256.0    \n",
       "40                                            15128.0    \n",
       "44                                            22884.0    \n",
       "48                                                0.0    \n",
       "52                                                0.0    \n",
       "56                                            24780.0    \n",
       "60                                            24602.0    \n",
       "64                                            22140.0    \n",
       "68                                            18741.0    \n",
       "72                                            14941.0    \n",
       "76                                                0.0    \n",
       "80                                             9341.0    \n",
       "84                                             9130.0    \n",
       "88                                             9545.0    \n",
       "92                                                0.0    \n",
       "\n",
       "    Focal 72kb Region (Percentile Rank)  \n",
       "0                             99.425998  \n",
       "4                             94.505828  \n",
       "8                             54.779341  \n",
       "12                            32.275045  \n",
       "16                            59.184346  \n",
       "20                             0.000000  \n",
       "24                             0.000000  \n",
       "28                             0.000000  \n",
       "32                             0.000000  \n",
       "36                            47.677728  \n",
       "40                            41.694457  \n",
       "44                            63.201502  \n",
       "48                             0.000000  \n",
       "52                             0.000000  \n",
       "56                            68.123711  \n",
       "60                            67.600912  \n",
       "64                            60.871000  \n",
       "68                            51.559921  \n",
       "72                            41.076043  \n",
       "76                             0.000000  \n",
       "80                            25.548384  \n",
       "84                            24.961724  \n",
       "88                            26.104198  \n",
       "92                             0.000000  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 3, 4, 5, 6, 7, 8, 9]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b9ad4d7",
   "metadata": {},
   "source": [
    "## S14"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "31092c1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_repeat_region_introgressed_tract_frequency_per_super_population.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "cells={halign=c, valign=m},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S14. Frequency of introgressed tracts overlapping \\textit{MUC19} short-read repeat region among 1000 Genomes Project super populations.} \\newline The frequency of introgressed tracts---i.e., the number of introgressed tracts normalized by the total number of chromosomes---and the mean tract length stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Introgressed tracts at the \\textit{MUC19} short-read repeat region (hg19, Chr12:40876395-40885001) are significantly enriched in AMR individuals (Fisher's Exact Test, Odds Ratio: 5.940, \\textit{P-value}: $1.953e^{-31}$), with AMR populations exhibiting a higher proportion of introgressed tracts compared to non-AMR populations, excluding AFR (Proportions \\textit{Z}-Test, \\textit{Z}-statistic: 13.269, \\textit{P-value}: $1.742e^{-40}$). Note that a \"---\" denotes that there are no introgressed tracts overlapping short-read repeat region for that group.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_repeat_region_introgressed_tract_frequency_per_super_population.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        f'{mat[i, 0]}'+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40e5dd50",
   "metadata": {},
   "source": [
    "\\textbf{Table S14. Frequency of introgressed tracts overlapping \\textit{MUC19} short-read repeat region among 1000 Genomes Project super populations.} \\newline The frequency of introgressed tracts---i.e., the number of introgressed tracts normalized by the total number of chromosomes---stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Introgressed tracts at the \\textit{MUC19} short-read repeat region (hg19, Chr12:40875941-40885367) are significantly enriched in AMR individuals (Fisher's Exact Test, Odds Ratio: 5.940, \\textit{P-value}: $1.953e^{-31}$), with AMR populations exhibiting a higher proportion of introgressed tracts compared to non-AMR populations, excluding AFR (Proportions \\textit{Z}-Test, \\textit{Z}-statistic: 13.269, \\textit{P-value}: $1.742e^{-40}$)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0dc16890",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Total Number of Chromosomes</th>\n",
       "      <th>Number of Introgressed Tracts</th>\n",
       "      <th>Introgressed Tract Frequency</th>\n",
       "      <th>Mean Tract Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>694</td>\n",
       "      <td>109</td>\n",
       "      <td>0.157061</td>\n",
       "      <td>547954.128440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAS</td>\n",
       "      <td>978</td>\n",
       "      <td>27</td>\n",
       "      <td>0.027607</td>\n",
       "      <td>283666.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EAS</td>\n",
       "      <td>1008</td>\n",
       "      <td>58</td>\n",
       "      <td>0.057540</td>\n",
       "      <td>335327.586207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EUR</td>\n",
       "      <td>1006</td>\n",
       "      <td>6</td>\n",
       "      <td>0.005964</td>\n",
       "      <td>460500.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFR</td>\n",
       "      <td>1008</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Super Population  Total Number of Chromosomes  \\\n",
       "0              AMR                          694   \n",
       "1              SAS                          978   \n",
       "2              EAS                         1008   \n",
       "3              EUR                         1006   \n",
       "4              AFR                         1008   \n",
       "\n",
       "   Number of Introgressed Tracts  Introgressed Tract Frequency  \\\n",
       "0                            109                      0.157061   \n",
       "1                             27                      0.027607   \n",
       "2                             58                      0.057540   \n",
       "3                              6                      0.005964   \n",
       "4                              0                      0.000000   \n",
       "\n",
       "   Mean Tract Length  \n",
       "0      547954.128440  \n",
       "1      283666.666667  \n",
       "2      335327.586207  \n",
       "3      460500.000000  \n",
       "4                NaN  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d825dbd8",
   "metadata": {},
   "source": [
    "## S15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "76e0f1e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_repeat_region_introgressed_tract_frequency_per_population.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S15. Frequency of introgressed tracts overlapping \\textit{MUC19} short-read repeat region among 1000 Genomes Project populations.} \\newline The frequency of introgressed tracts---i.e., the number of introgressed tracts normalized by the total number of chromosomes---and the mean tract length for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Note that a \"---\" denotes that there are no introgressed tracts overlapping short-read repeat region for that group.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_repeat_region_introgressed_tract_frequency_per_population.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(2, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77c1fcc8",
   "metadata": {},
   "source": [
    "\\textbf{Table S15. Frequency of introgressed tracts overlapping \\textit{MUC19} short-read repeat region among 1000 Genomes Project populations.} \\newline The frequency of introgressed tracts---i.e., the number of introgressed tracts normalized by the total number of chromosomes---and the mean tract length for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Note that a \"---\" denotes that there are no introgressed tracts overlapping short-read repeat region for that group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "598aca7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Total Number of Chromosomes</th>\n",
       "      <th>Number of Introgressed Tracts</th>\n",
       "      <th>Introgressed Tract Frequency</th>\n",
       "      <th>Mean Tract Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>36</td>\n",
       "      <td>0.281250</td>\n",
       "      <td>653250.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>170</td>\n",
       "      <td>38</td>\n",
       "      <td>0.223529</td>\n",
       "      <td>511368.421053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>188</td>\n",
       "      <td>18</td>\n",
       "      <td>0.095745</td>\n",
       "      <td>381166.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>208</td>\n",
       "      <td>17</td>\n",
       "      <td>0.081731</td>\n",
       "      <td>583352.941176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>172</td>\n",
       "      <td>4</td>\n",
       "      <td>0.023256</td>\n",
       "      <td>427750.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>204</td>\n",
       "      <td>3</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>136333.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>204</td>\n",
       "      <td>8</td>\n",
       "      <td>0.039216</td>\n",
       "      <td>271250.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>192</td>\n",
       "      <td>7</td>\n",
       "      <td>0.036458</td>\n",
       "      <td>223142.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>206</td>\n",
       "      <td>5</td>\n",
       "      <td>0.024272</td>\n",
       "      <td>361400.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>206</td>\n",
       "      <td>9</td>\n",
       "      <td>0.043689</td>\n",
       "      <td>363666.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>198</td>\n",
       "      <td>14</td>\n",
       "      <td>0.070707</td>\n",
       "      <td>306214.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>210</td>\n",
       "      <td>10</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>283000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>208</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>326750.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>186</td>\n",
       "      <td>17</td>\n",
       "      <td>0.091398</td>\n",
       "      <td>379117.647059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>214</td>\n",
       "      <td>3</td>\n",
       "      <td>0.014019</td>\n",
       "      <td>334666.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>198</td>\n",
       "      <td>2</td>\n",
       "      <td>0.010101</td>\n",
       "      <td>592500.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>214</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>574000.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>182</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AFR</td>\n",
       "      <td>LWK</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AFR</td>\n",
       "      <td>GWD</td>\n",
       "      <td>226</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AFR</td>\n",
       "      <td>MSL</td>\n",
       "      <td>170</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AFR</td>\n",
       "      <td>ESN</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AFR</td>\n",
       "      <td>YRI</td>\n",
       "      <td>216</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Total Number of Chromosomes  \\\n",
       "0               AMR        MXL                          128   \n",
       "1               AMR        PEL                          170   \n",
       "2               AMR        CLM                          188   \n",
       "3               AMR        PUR                          208   \n",
       "4               SAS        BEB                          172   \n",
       "5               SAS        STU                          204   \n",
       "6               SAS        ITU                          204   \n",
       "7               SAS        PJL                          192   \n",
       "8               SAS        GIH                          206   \n",
       "9               EAS        CHB                          206   \n",
       "10              EAS        KHV                          198   \n",
       "11              EAS        CHS                          210   \n",
       "12              EAS        JPT                          208   \n",
       "13              EAS        CDX                          186   \n",
       "14              EUR        TSI                          214   \n",
       "15              EUR        CEU                          198   \n",
       "16              EUR        IBS                          214   \n",
       "17              EUR        GBR                          182   \n",
       "18              EUR        FIN                          198   \n",
       "19              AFR        LWK                          198   \n",
       "20              AFR        GWD                          226   \n",
       "21              AFR        MSL                          170   \n",
       "22              AFR        ESN                          198   \n",
       "23              AFR        YRI                          216   \n",
       "\n",
       "    Number of Introgressed Tracts  Introgressed Tract Frequency  \\\n",
       "0                              36                      0.281250   \n",
       "1                              38                      0.223529   \n",
       "2                              18                      0.095745   \n",
       "3                              17                      0.081731   \n",
       "4                               4                      0.023256   \n",
       "5                               3                      0.014706   \n",
       "6                               8                      0.039216   \n",
       "7                               7                      0.036458   \n",
       "8                               5                      0.024272   \n",
       "9                               9                      0.043689   \n",
       "10                             14                      0.070707   \n",
       "11                             10                      0.047619   \n",
       "12                              8                      0.038462   \n",
       "13                             17                      0.091398   \n",
       "14                              3                      0.014019   \n",
       "15                              2                      0.010101   \n",
       "16                              1                      0.004673   \n",
       "17                              0                      0.000000   \n",
       "18                              0                      0.000000   \n",
       "19                              0                      0.000000   \n",
       "20                              0                      0.000000   \n",
       "21                              0                      0.000000   \n",
       "22                              0                      0.000000   \n",
       "23                              0                      0.000000   \n",
       "\n",
       "    Mean Tract Length  \n",
       "0       653250.000000  \n",
       "1       511368.421053  \n",
       "2       381166.666667  \n",
       "3       583352.941176  \n",
       "4       427750.000000  \n",
       "5       136333.333333  \n",
       "6       271250.000000  \n",
       "7       223142.857143  \n",
       "8       361400.000000  \n",
       "9       363666.666667  \n",
       "10      306214.285714  \n",
       "11      283000.000000  \n",
       "12      326750.000000  \n",
       "13      379117.647059  \n",
       "14      334666.666667  \n",
       "15      592500.000000  \n",
       "16      574000.000000  \n",
       "17                NaN  \n",
       "18                NaN  \n",
       "19                NaN  \n",
       "20                NaN  \n",
       "21                NaN  \n",
       "22                NaN  \n",
       "23                NaN  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7edd3d6a",
   "metadata": {},
   "source": [
    "## S16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f9873a8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_population_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S16. Estimated repeat copies of the \\textit{MUC19} 30bp variable number tandem repeat from short-read sequencing data among 1000 Genomes Project populations.} \\newline Summary of the distribution of repeat copy numbers for \\textit{MUC19} per population stratified by super populations: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR).  For each population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the repeat copy distribution are reported. We also report the total number of individuals, the number of individuals with elevated repeat copies (i.e., those with $> 487$ repeat copies), and the proportion of individuals with elevated repeat copies per population. Populations are sorted by their mean repeat copy number.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_population_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "277483bf",
   "metadata": {},
   "source": [
    "\\textbf{Table S16. Estimated repeat copies of the \\textit{MUC19} 30bp variable number tandem repeat from short-read sequencing data among 1000 Genomes Project populations.} \\newline Summary of the distribution of repeat copy numbers for \\textit{MUC19} per population stratified by super populations: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR).  For each population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the repeat copy distribution are reported. We also report the total number of individuals, the number of individuals with elevated repeat copies (i.e., those with $> 487$ repeat copies), and the proportion of individuals with elevated repeat copies per population.  Populations are sorted by their mean repeat copy number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f54527b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Number of Inds.</th>\n",
       "      <th>Repeat Copies $\\left( \\mu \\right)$</th>\n",
       "      <th>Repeat Copies $\\left( \\sigma \\right)$</th>\n",
       "      <th>Repeat Copies $\\left( SEM \\right)$</th>\n",
       "      <th>Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>Number of Inds. with Elevated Repeats</th>\n",
       "      <th>Proportion of Inds. with Elevated Repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>64</td>\n",
       "      <td>492.843750</td>\n",
       "      <td>147.127731</td>\n",
       "      <td>18.536352</td>\n",
       "      <td>37.041943</td>\n",
       "      <td>32</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>85</td>\n",
       "      <td>409.941176</td>\n",
       "      <td>140.779456</td>\n",
       "      <td>15.360298</td>\n",
       "      <td>30.545637</td>\n",
       "      <td>24</td>\n",
       "      <td>0.282353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>94</td>\n",
       "      <td>398.882979</td>\n",
       "      <td>101.087265</td>\n",
       "      <td>10.482261</td>\n",
       "      <td>20.815693</td>\n",
       "      <td>20</td>\n",
       "      <td>0.212766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>104</td>\n",
       "      <td>391.346154</td>\n",
       "      <td>95.725546</td>\n",
       "      <td>9.432118</td>\n",
       "      <td>18.706382</td>\n",
       "      <td>15</td>\n",
       "      <td>0.144231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>93</td>\n",
       "      <td>365.032258</td>\n",
       "      <td>61.999385</td>\n",
       "      <td>6.463883</td>\n",
       "      <td>12.837829</td>\n",
       "      <td>6</td>\n",
       "      <td>0.064516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>99</td>\n",
       "      <td>361.262626</td>\n",
       "      <td>50.016883</td>\n",
       "      <td>5.052468</td>\n",
       "      <td>10.026459</td>\n",
       "      <td>2</td>\n",
       "      <td>0.020202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>107</td>\n",
       "      <td>358.803738</td>\n",
       "      <td>44.555843</td>\n",
       "      <td>4.327646</td>\n",
       "      <td>8.579979</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>99</td>\n",
       "      <td>357.050505</td>\n",
       "      <td>48.049926</td>\n",
       "      <td>4.853776</td>\n",
       "      <td>9.632160</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>103</td>\n",
       "      <td>356.902913</td>\n",
       "      <td>51.543462</td>\n",
       "      <td>5.103563</td>\n",
       "      <td>10.122893</td>\n",
       "      <td>3</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>104</td>\n",
       "      <td>354.788462</td>\n",
       "      <td>38.661916</td>\n",
       "      <td>3.809472</td>\n",
       "      <td>7.555189</td>\n",
       "      <td>1</td>\n",
       "      <td>0.009615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>96</td>\n",
       "      <td>353.614583</td>\n",
       "      <td>47.445225</td>\n",
       "      <td>4.867777</td>\n",
       "      <td>9.663760</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>105</td>\n",
       "      <td>352.076190</td>\n",
       "      <td>47.589454</td>\n",
       "      <td>4.666530</td>\n",
       "      <td>9.253904</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>107</td>\n",
       "      <td>351.925234</td>\n",
       "      <td>39.528459</td>\n",
       "      <td>3.839343</td>\n",
       "      <td>7.611872</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AFR</td>\n",
       "      <td>YRI</td>\n",
       "      <td>108</td>\n",
       "      <td>350.361111</td>\n",
       "      <td>38.850487</td>\n",
       "      <td>3.755818</td>\n",
       "      <td>7.445472</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>102</td>\n",
       "      <td>350.235294</td>\n",
       "      <td>47.460236</td>\n",
       "      <td>4.722470</td>\n",
       "      <td>9.368110</td>\n",
       "      <td>3</td>\n",
       "      <td>0.029412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>102</td>\n",
       "      <td>349.823529</td>\n",
       "      <td>44.253151</td>\n",
       "      <td>4.403353</td>\n",
       "      <td>8.735068</td>\n",
       "      <td>2</td>\n",
       "      <td>0.019608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AFR</td>\n",
       "      <td>ESN</td>\n",
       "      <td>99</td>\n",
       "      <td>348.272727</td>\n",
       "      <td>38.940467</td>\n",
       "      <td>3.933581</td>\n",
       "      <td>7.806064</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>91</td>\n",
       "      <td>348.186813</td>\n",
       "      <td>45.543582</td>\n",
       "      <td>4.800715</td>\n",
       "      <td>9.537458</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>86</td>\n",
       "      <td>347.372093</td>\n",
       "      <td>39.937465</td>\n",
       "      <td>4.331826</td>\n",
       "      <td>8.612831</td>\n",
       "      <td>1</td>\n",
       "      <td>0.011628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>103</td>\n",
       "      <td>343.961165</td>\n",
       "      <td>45.212105</td>\n",
       "      <td>4.476665</td>\n",
       "      <td>8.879445</td>\n",
       "      <td>2</td>\n",
       "      <td>0.019417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AFR</td>\n",
       "      <td>MSL</td>\n",
       "      <td>85</td>\n",
       "      <td>343.529412</td>\n",
       "      <td>34.436844</td>\n",
       "      <td>3.757368</td>\n",
       "      <td>7.471938</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>99</td>\n",
       "      <td>341.898990</td>\n",
       "      <td>35.976281</td>\n",
       "      <td>3.634153</td>\n",
       "      <td>7.211859</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AFR</td>\n",
       "      <td>GWD</td>\n",
       "      <td>113</td>\n",
       "      <td>341.725664</td>\n",
       "      <td>40.330767</td>\n",
       "      <td>3.810899</td>\n",
       "      <td>7.550808</td>\n",
       "      <td>2</td>\n",
       "      <td>0.017699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AFR</td>\n",
       "      <td>LWK</td>\n",
       "      <td>99</td>\n",
       "      <td>340.707071</td>\n",
       "      <td>34.856489</td>\n",
       "      <td>3.521037</td>\n",
       "      <td>6.987383</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Number of Inds.  \\\n",
       "0               AMR        MXL               64   \n",
       "1               AMR        PEL               85   \n",
       "2               AMR        CLM               94   \n",
       "3               AMR        PUR              104   \n",
       "4               EAS        CDX               93   \n",
       "5               EAS        KHV               99   \n",
       "6               EUR        TSI              107   \n",
       "7               EUR        CEU               99   \n",
       "8               SAS        GIH              103   \n",
       "9               EAS        JPT              104   \n",
       "10              SAS        PJL               96   \n",
       "11              EAS        CHS              105   \n",
       "12              EUR        IBS              107   \n",
       "13              AFR        YRI              108   \n",
       "14              SAS        ITU              102   \n",
       "15              SAS        STU              102   \n",
       "16              AFR        ESN               99   \n",
       "17              EUR        GBR               91   \n",
       "18              SAS        BEB               86   \n",
       "19              EAS        CHB              103   \n",
       "20              AFR        MSL               85   \n",
       "21              EUR        FIN               99   \n",
       "22              AFR        GWD              113   \n",
       "23              AFR        LWK               99   \n",
       "\n",
       "    Repeat Copies $\\left( \\mu \\right)$  Repeat Copies $\\left( \\sigma \\right)$  \\\n",
       "0                           492.843750                             147.127731   \n",
       "1                           409.941176                             140.779456   \n",
       "2                           398.882979                             101.087265   \n",
       "3                           391.346154                              95.725546   \n",
       "4                           365.032258                              61.999385   \n",
       "5                           361.262626                              50.016883   \n",
       "6                           358.803738                              44.555843   \n",
       "7                           357.050505                              48.049926   \n",
       "8                           356.902913                              51.543462   \n",
       "9                           354.788462                              38.661916   \n",
       "10                          353.614583                              47.445225   \n",
       "11                          352.076190                              47.589454   \n",
       "12                          351.925234                              39.528459   \n",
       "13                          350.361111                              38.850487   \n",
       "14                          350.235294                              47.460236   \n",
       "15                          349.823529                              44.253151   \n",
       "16                          348.272727                              38.940467   \n",
       "17                          348.186813                              45.543582   \n",
       "18                          347.372093                              39.937465   \n",
       "19                          343.961165                              45.212105   \n",
       "20                          343.529412                              34.436844   \n",
       "21                          341.898990                              35.976281   \n",
       "22                          341.725664                              40.330767   \n",
       "23                          340.707071                              34.856489   \n",
       "\n",
       "    Repeat Copies $\\left( SEM \\right)$  \\\n",
       "0                            18.536352   \n",
       "1                            15.360298   \n",
       "2                            10.482261   \n",
       "3                             9.432118   \n",
       "4                             6.463883   \n",
       "5                             5.052468   \n",
       "6                             4.327646   \n",
       "7                             4.853776   \n",
       "8                             5.103563   \n",
       "9                             3.809472   \n",
       "10                            4.867777   \n",
       "11                            4.666530   \n",
       "12                            3.839343   \n",
       "13                            3.755818   \n",
       "14                            4.722470   \n",
       "15                            4.403353   \n",
       "16                            3.933581   \n",
       "17                            4.800715   \n",
       "18                            4.331826   \n",
       "19                            4.476665   \n",
       "20                            3.757368   \n",
       "21                            3.634153   \n",
       "22                            3.810899   \n",
       "23                            3.521037   \n",
       "\n",
       "    Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                      37.041943   \n",
       "1                                      30.545637   \n",
       "2                                      20.815693   \n",
       "3                                      18.706382   \n",
       "4                                      12.837829   \n",
       "5                                      10.026459   \n",
       "6                                       8.579979   \n",
       "7                                       9.632160   \n",
       "8                                      10.122893   \n",
       "9                                       7.555189   \n",
       "10                                      9.663760   \n",
       "11                                      9.253904   \n",
       "12                                      7.611872   \n",
       "13                                      7.445472   \n",
       "14                                      9.368110   \n",
       "15                                      8.735068   \n",
       "16                                      7.806064   \n",
       "17                                      9.537458   \n",
       "18                                      8.612831   \n",
       "19                                      8.879445   \n",
       "20                                      7.471938   \n",
       "21                                      7.211859   \n",
       "22                                      7.550808   \n",
       "23                                      6.987383   \n",
       "\n",
       "    Number of Inds. with Elevated Repeats  \\\n",
       "0                                      32   \n",
       "1                                      24   \n",
       "2                                      20   \n",
       "3                                      15   \n",
       "4                                       6   \n",
       "5                                       2   \n",
       "6                                       1   \n",
       "7                                       1   \n",
       "8                                       3   \n",
       "9                                       1   \n",
       "10                                      1   \n",
       "11                                      0   \n",
       "12                                      0   \n",
       "13                                      0   \n",
       "14                                      3   \n",
       "15                                      2   \n",
       "16                                      1   \n",
       "17                                      1   \n",
       "18                                      1   \n",
       "19                                      2   \n",
       "20                                      0   \n",
       "21                                      0   \n",
       "22                                      2   \n",
       "23                                      0   \n",
       "\n",
       "    Proportion of Inds. with Elevated Repeats  \n",
       "0                                    0.500000  \n",
       "1                                    0.282353  \n",
       "2                                    0.212766  \n",
       "3                                    0.144231  \n",
       "4                                    0.064516  \n",
       "5                                    0.020202  \n",
       "6                                    0.009346  \n",
       "7                                    0.010101  \n",
       "8                                    0.029126  \n",
       "9                                    0.009615  \n",
       "10                                   0.010417  \n",
       "11                                   0.000000  \n",
       "12                                   0.000000  \n",
       "13                                   0.000000  \n",
       "14                                   0.029412  \n",
       "15                                   0.019608  \n",
       "16                                   0.010101  \n",
       "17                                   0.010989  \n",
       "18                                   0.011628  \n",
       "19                                   0.019417  \n",
       "20                                   0.000000  \n",
       "21                                   0.000000  \n",
       "22                                   0.017699  \n",
       "23                                   0.000000  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0c328da",
   "metadata": {},
   "source": [
    "## S17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "edc08c28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_superpopulation_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S17. Estimated repeat copies of the \\textit{MUC19} 30bp variable number tandem repeat from short-read sequencing data among 1000 Genomes Project super populations.} \\newline Summary of the distribution of repeat copy numbers for \\textit{MUC19} per super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR).  For each super population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the repeat copy distribution are reported. We also report the total number of individuals, the number of individuals with elevated repeat copies (i.e., those with $> 487$ repeat copies), and the proportion of individuals with elevated repeat copies per super population. Super populations are sorted by their mean repeat copy number.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_superpopulation_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54cdfd3e",
   "metadata": {},
   "source": [
    "\\textbf{Table S18. Estimated repeat copies of the \\textit{MUC19} 30bp variable number tandem repeat from short-read sequencing data among 1000 Genomes Project super populations.} \\newline Summary of the distribution of repeat copy numbers for \\textit{MUC19} per super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR).  For each super population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the repeat copy distribution are reported. We also report the total number of individuals, the number of individuals with elevated repeat copies (i.e., those with $> 487$ repeat copies), and the proportion of individuals with elevated repeat copies per super population. Super populations are sorted by their mean repeat copy number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8ac7c2d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Number of Inds.</th>\n",
       "      <th>Repeat Copies $\\left( \\mu \\right)$</th>\n",
       "      <th>Repeat Copies $\\left( \\sigma \\right)$</th>\n",
       "      <th>Repeat Copies $\\left( SEM \\right)$</th>\n",
       "      <th>Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>Number of Inds. with Elevated Repeats</th>\n",
       "      <th>Proportion of Inds. with Elevated Repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>347</td>\n",
       "      <td>416.662824</td>\n",
       "      <td>125.383165</td>\n",
       "      <td>6.740641</td>\n",
       "      <td>13.257788</td>\n",
       "      <td>91</td>\n",
       "      <td>0.262248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EAS</td>\n",
       "      <td>504</td>\n",
       "      <td>355.172619</td>\n",
       "      <td>49.514147</td>\n",
       "      <td>2.207727</td>\n",
       "      <td>4.337502</td>\n",
       "      <td>11</td>\n",
       "      <td>0.021825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EUR</td>\n",
       "      <td>503</td>\n",
       "      <td>351.747515</td>\n",
       "      <td>43.319905</td>\n",
       "      <td>1.933462</td>\n",
       "      <td>3.798674</td>\n",
       "      <td>3</td>\n",
       "      <td>0.005964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAS</td>\n",
       "      <td>489</td>\n",
       "      <td>351.713701</td>\n",
       "      <td>46.596776</td>\n",
       "      <td>2.109337</td>\n",
       "      <td>4.144503</td>\n",
       "      <td>10</td>\n",
       "      <td>0.020450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFR</td>\n",
       "      <td>504</td>\n",
       "      <td>344.966270</td>\n",
       "      <td>37.937930</td>\n",
       "      <td>1.691569</td>\n",
       "      <td>3.323411</td>\n",
       "      <td>3</td>\n",
       "      <td>0.005952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Super Population  Number of Inds.  Repeat Copies $\\left( \\mu \\right)$  \\\n",
       "0              AMR              347                          416.662824   \n",
       "1              EAS              504                          355.172619   \n",
       "2              EUR              503                          351.747515   \n",
       "3              SAS              489                          351.713701   \n",
       "4              AFR              504                          344.966270   \n",
       "\n",
       "   Repeat Copies $\\left( \\sigma \\right)$  Repeat Copies $\\left( SEM \\right)$  \\\n",
       "0                             125.383165                            6.740641   \n",
       "1                              49.514147                            2.207727   \n",
       "2                              43.319905                            1.933462   \n",
       "3                              46.596776                            2.109337   \n",
       "4                              37.937930                            1.691569   \n",
       "\n",
       "   Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                     13.257788   \n",
       "1                                      4.337502   \n",
       "2                                      3.798674   \n",
       "3                                      4.144503   \n",
       "4                                      3.323411   \n",
       "\n",
       "   Number of Inds. with Elevated Repeats  \\\n",
       "0                                     91   \n",
       "1                                     11   \n",
       "2                                      3   \n",
       "3                                     10   \n",
       "4                                      3   \n",
       "\n",
       "   Proportion of Inds. with Elevated Repeats  \n",
       "0                                   0.262248  \n",
       "1                                   0.021825  \n",
       "2                                   0.005964  \n",
       "3                                   0.020450  \n",
       "4                                   0.005952  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6423a34",
   "metadata": {},
   "source": [
    "## S18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f12ea7d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_group_comparisons_all_individuals.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 8, 9, 10, 11, 12, 13, 14, 15]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 8, 9, 10, 11, 12, 13, 14, 15]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S18. Distribution of repeat copies for different focal groups among all individuals in the 1000 Genome Project.} \\newline Summary of the distribution of repeat copy numbers for individuals with at least one introgressed tract overlapping the repeat region vs individuals with no introgressed tracts overlapping the repeat region and admixed American vs non-admixed American individuals. For each focal group, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the repeat copy distribution are reported. },\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_group_comparisons_all_individuals_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca692e5d",
   "metadata": {},
   "source": [
    "\\textbf{Table S17. Distribution of repeat copies for different focal groups among all individuals in the 1000 Genome Project.} \\newline Summary of the distribution of repeat copy numbers for individuals with at least one introgressed tract overlapping the repeat region vs individuals with no introgressed tracts overlapping the repeat region and admixed American vs non-admixed American individuals. For each focal group, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the repeat copy distribution are reported. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "33b0bc32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group 1</th>\n",
       "      <th>Group 2</th>\n",
       "      <th>Group 1 Repeat Copies $\\left( \\mu \\right)$</th>\n",
       "      <th>Group 2 Repeat Copies $\\left( \\mu \\right)$</th>\n",
       "      <th>Group 1 Repeat Copies $\\left( \\sigma \\right)$</th>\n",
       "      <th>Group 2 Repeat Copies $\\left( \\sigma \\right)$</th>\n",
       "      <th>Group 1 Repeat Copies $\\left( SEM \\right)$</th>\n",
       "      <th>Group 2 Repeat Copies $\\left( SEM \\right)$</th>\n",
       "      <th>Group 1 Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>Group 2 Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All Inds.: $\\geq 1$ Overlapping Tract</td>\n",
       "      <td>All Inds.: No Overlapping Tract</td>\n",
       "      <td>517.886486</td>\n",
       "      <td>347.160037</td>\n",
       "      <td>114.083294</td>\n",
       "      <td>39.384456</td>\n",
       "      <td>8.410332</td>\n",
       "      <td>0.847222</td>\n",
       "      <td>16.593085</td>\n",
       "      <td>1.661456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>All AMR Inds.</td>\n",
       "      <td>All non-AMR Inds.</td>\n",
       "      <td>416.662824</td>\n",
       "      <td>350.893500</td>\n",
       "      <td>125.383165</td>\n",
       "      <td>44.689743</td>\n",
       "      <td>6.740641</td>\n",
       "      <td>0.999543</td>\n",
       "      <td>13.257788</td>\n",
       "      <td>1.960255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Group 1                          Group 2  \\\n",
       "0  All Inds.: $\\geq 1$ Overlapping Tract  All Inds.: No Overlapping Tract   \n",
       "1                          All AMR Inds.                All non-AMR Inds.   \n",
       "\n",
       "   Group 1 Repeat Copies $\\left( \\mu \\right)$  \\\n",
       "0                                  517.886486   \n",
       "1                                  416.662824   \n",
       "\n",
       "   Group 2 Repeat Copies $\\left( \\mu \\right)$  \\\n",
       "0                                  347.160037   \n",
       "1                                  350.893500   \n",
       "\n",
       "   Group 1 Repeat Copies $\\left( \\sigma \\right)$  \\\n",
       "0                                     114.083294   \n",
       "1                                     125.383165   \n",
       "\n",
       "   Group 2 Repeat Copies $\\left( \\sigma \\right)$  \\\n",
       "0                                      39.384456   \n",
       "1                                      44.689743   \n",
       "\n",
       "   Group 1 Repeat Copies $\\left( SEM \\right)$  \\\n",
       "0                                    8.410332   \n",
       "1                                    6.740641   \n",
       "\n",
       "   Group 2 Repeat Copies $\\left( SEM \\right)$  \\\n",
       "0                                    0.847222   \n",
       "1                                    0.999543   \n",
       "\n",
       "   Group 1 Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                          16.593085      \n",
       "1                                          13.257788      \n",
       "\n",
       "   Group 2 Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$  \n",
       "0                                           1.661456     \n",
       "1                                           1.960255     "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 8, 9, 10, 11, 12, 13, 14, 15]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "244b8fcd",
   "metadata": {},
   "source": [
    "## S19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c07d285f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_group_comparisons_all_individuals.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 3, 4, 5, 6, 7, 16, 17, 18, 19]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 3, 4, 5, 6, 7, 16, 17, 18, 19]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXp{1.55cm}XXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S19. Comparisons of the distributions of repeat copies between different focal groups among all individuals in the 1000 Genome Project.} \\newline Comparing the distributions of repeat copy numbers between focal groups: individuals with at least one introgressed tract overlapping the repeat region vs individuals with no introgressed tracts overlapping the repeat region and admixed American vs non-admixed American individuals. For each comparison, the total number of individuals, the number of individuals with elevated repeat copies (i.e., those with $> 487$ repeat copies), the proportion of individuals with elevated repeat copies per population, and the results of the Proportions \\textit{Z}-Test, and Mann-Whitney \\textit{U}-Test is reported. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_group_comparisons_all_individuals_comparison.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(2, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f0e2c30",
   "metadata": {},
   "source": [
    "\\textbf{Table S18. Comparisons of the distributions of repeat copies between different focal groups among all individuals in the 1000 Genome Project.} \\newline Comparing the distributions of repeat copy numbers between focal groups: individuals with at least one introgressed tract overlapping the repeat region vs individuals with no introgressed tracts overlapping the repeat region and admixed American vs non-admixed American individuals. For each comparison, the total number of individuals, the number of individuals with elevated repeat copies (i.e., those with $> 487$ repeat copies), the proportion of individuals with elevated repeat copies per population, and the results of the Proportions \\textit{Z}-Test, and Mann-Whitney \\textit{U}-Test is reported. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "7fb05034",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group 1</th>\n",
       "      <th>Group 2</th>\n",
       "      <th>Number of Inds. (Group 1)</th>\n",
       "      <th>Number of Inds. (Group 2)</th>\n",
       "      <th>Number of Inds. with Elevated Repeats (Group 1)</th>\n",
       "      <th>Number of Inds. with Elevated Repeats (Group 2)</th>\n",
       "      <th>Prop. of Inds. with Elevated Repeats (Group 1)</th>\n",
       "      <th>Prop. of Inds. with Elevated Repeats (Group 2)</th>\n",
       "      <th>$Z-Statistic$</th>\n",
       "      <th>$P-value$ (Prop. $Z$-Test)</th>\n",
       "      <th>$U-Statistic$</th>\n",
       "      <th>$P-value$ (Mann-Whitney $U$-Test)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All Inds.: $\\geq 1$ Overlapping Tract</td>\n",
       "      <td>All Inds.: No Overlapping Tract</td>\n",
       "      <td>185</td>\n",
       "      <td>2162</td>\n",
       "      <td>102</td>\n",
       "      <td>16</td>\n",
       "      <td>0.551351</td>\n",
       "      <td>0.007401</td>\n",
       "      <td>32.496294</td>\n",
       "      <td>6.014519e-232</td>\n",
       "      <td>375108.5</td>\n",
       "      <td>1.596665e-87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>All AMR Inds.</td>\n",
       "      <td>All non-AMR Inds.</td>\n",
       "      <td>347</td>\n",
       "      <td>2000</td>\n",
       "      <td>91</td>\n",
       "      <td>27</td>\n",
       "      <td>0.262248</td>\n",
       "      <td>0.013500</td>\n",
       "      <td>19.574900</td>\n",
       "      <td>1.265659e-85</td>\n",
       "      <td>445495.0</td>\n",
       "      <td>1.424411e-17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Group 1                          Group 2  \\\n",
       "0  All Inds.: $\\geq 1$ Overlapping Tract  All Inds.: No Overlapping Tract   \n",
       "1                          All AMR Inds.                All non-AMR Inds.   \n",
       "\n",
       "   Number of Inds. (Group 1)  Number of Inds. (Group 2)  \\\n",
       "0                        185                       2162   \n",
       "1                        347                       2000   \n",
       "\n",
       "   Number of Inds. with Elevated Repeats (Group 1)  \\\n",
       "0                                              102   \n",
       "1                                               91   \n",
       "\n",
       "   Number of Inds. with Elevated Repeats (Group 2)  \\\n",
       "0                                               16   \n",
       "1                                               27   \n",
       "\n",
       "   Prop. of Inds. with Elevated Repeats (Group 1)  \\\n",
       "0                                        0.551351   \n",
       "1                                        0.262248   \n",
       "\n",
       "   Prop. of Inds. with Elevated Repeats (Group 2)  $Z-Statistic$  \\\n",
       "0                                        0.007401      32.496294   \n",
       "1                                        0.013500      19.574900   \n",
       "\n",
       "   $P-value$ (Prop. $Z$-Test)  $U-Statistic$  \\\n",
       "0               6.014519e-232       375108.5   \n",
       "1                1.265659e-85       445495.0   \n",
       "\n",
       "   $P-value$ (Mann-Whitney $U$-Test)  \n",
       "0                       1.596665e-87  \n",
       "1                       1.424411e-17  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 2, 3, 4, 5, 6, 7, 16, 17, 18, 19]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c42afdf",
   "metadata": {},
   "source": [
    "## S20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6ca23bbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_group_comparisons_outlier_individuals.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 7, 8, 9, 10, 11, 12, 13, 14]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 7, 8, 9, 10, 11, 12, 13, 14]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S20. Distribution of repeat copies for different focal groups among individuals with an elevated number of repeat copies in the 1000 Genome Project.} \\newline Summary of the distribution of repeat copy numbers for different focal groups of outlier individuals: individuals with at least one introgressed tract overlapping the repeat region vs individuals with no introgressed tracts overlapping the repeat region and admixed American vs non-admixed American individuals. For each focal group, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the repeat copy distribution are reported.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_group_comparisons_outlier_individuals_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75697c0f",
   "metadata": {},
   "source": [
    "\\textbf{Table S19. Distribution of repeat copies for different focal groups among individuals with an elevated number of repeat copies in the 1000 Genome Project.} \\newline Summary of the distribution of repeat copy numbers for different focal groups of outlier individuals: individuals with at least one introgressed tract overlapping the repeat region vs individuals with no introgressed tracts overlapping the repeat region and admixed American vs non-admixed American individuals. For each focal group, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the repeat copy distribution are reported. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b471ead2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group 1</th>\n",
       "      <th>Group 2</th>\n",
       "      <th>Group 1 Repeat Copies $\\left( \\mu \\right)$</th>\n",
       "      <th>Group 2 Repeat Copies $\\left( \\mu \\right)$</th>\n",
       "      <th>Group 1 Repeat Copies $\\left( \\sigma \\right)$</th>\n",
       "      <th>Group 2 Repeat Copies $\\left( \\sigma \\right)$</th>\n",
       "      <th>Group 1 Repeat Copies $\\left( SEM \\right)$</th>\n",
       "      <th>Group 2 Repeat Copies $\\left( SEM \\right)$</th>\n",
       "      <th>Group 1 Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>Group 2 Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Outlier Inds.: $\\geq 1$ Overlapping Tract</td>\n",
       "      <td>Outlier Inds.: No Overlapping Tract</td>\n",
       "      <td>594.480392</td>\n",
       "      <td>510.6875</td>\n",
       "      <td>92.117011</td>\n",
       "      <td>16.115671</td>\n",
       "      <td>9.165985</td>\n",
       "      <td>4.161048</td>\n",
       "      <td>18.182849</td>\n",
       "      <td>8.869064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Outlier AMR Inds.</td>\n",
       "      <td>Outlier non-AMR Inds.</td>\n",
       "      <td>600.956044</td>\n",
       "      <td>523.0000</td>\n",
       "      <td>94.367869</td>\n",
       "      <td>33.259919</td>\n",
       "      <td>9.947247</td>\n",
       "      <td>6.522807</td>\n",
       "      <td>19.761942</td>\n",
       "      <td>13.407821</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     Group 1  \\\n",
       "0  Outlier Inds.: $\\geq 1$ Overlapping Tract   \n",
       "1                          Outlier AMR Inds.   \n",
       "\n",
       "                               Group 2  \\\n",
       "0  Outlier Inds.: No Overlapping Tract   \n",
       "1                Outlier non-AMR Inds.   \n",
       "\n",
       "   Group 1 Repeat Copies $\\left( \\mu \\right)$  \\\n",
       "0                                  594.480392   \n",
       "1                                  600.956044   \n",
       "\n",
       "   Group 2 Repeat Copies $\\left( \\mu \\right)$  \\\n",
       "0                                    510.6875   \n",
       "1                                    523.0000   \n",
       "\n",
       "   Group 1 Repeat Copies $\\left( \\sigma \\right)$  \\\n",
       "0                                      92.117011   \n",
       "1                                      94.367869   \n",
       "\n",
       "   Group 2 Repeat Copies $\\left( \\sigma \\right)$  \\\n",
       "0                                      16.115671   \n",
       "1                                      33.259919   \n",
       "\n",
       "   Group 1 Repeat Copies $\\left( SEM \\right)$  \\\n",
       "0                                    9.165985   \n",
       "1                                    9.947247   \n",
       "\n",
       "   Group 2 Repeat Copies $\\left( SEM \\right)$  \\\n",
       "0                                    4.161048   \n",
       "1                                    6.522807   \n",
       "\n",
       "   Group 1 Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                          18.182849      \n",
       "1                                          19.761942      \n",
       "\n",
       "   Group 2 Repeat Copies $\\left( \\pm CI_{95\\%} \\right)$  \n",
       "0                                           8.869064     \n",
       "1                                          13.407821     "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 7, 8, 9, 10, 11, 12, 13, 14]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4233dbc2",
   "metadata": {},
   "source": [
    "## S21"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "05667d58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_group_comparisons_outlier_individuals.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 3, 4, 5, 6, 15, 16, 17, 18]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 3, 4, 5, 6, 15, 16, 17, 18]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S21. Comparisons of the distributions of repeat copies between different focal groups among individuals with an elevated number of repeat copies in the 1000 Genome Project.} \\newline Comparing the distributions of repeat copy numbers between focal groups of outlier individuals: individuals with at least one introgressed tract overlapping the repeat region vs individuals with no introgressed tracts overlapping the repeat region and admixed American vs non-admixed American individuals. For each comparison, the total number of individuals, the number of individuals with elevated repeat copies (i.e., those with $> 487$ repeat copies), the proportion of individuals with elevated repeat copies per population, and the results of the Proportions \\textit{Z}-Test, and Mann-Whitney \\textit{U}-Test is reported. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_group_comparisons_outlier_individuals_comparison.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(2, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afec793d",
   "metadata": {},
   "source": [
    "\\textbf{Table S20. Comparisons of the distributions of repeat copies between different focal groups among individuals with an elevated number of repeat copies in the 1000 Genome Project.} \\newline Comparing the distributions of repeat copy numbers between focal groups of outlier individuals: individuals with at least one introgressed tract overlapping the repeat region vs individuals with no introgressed tracts overlapping the repeat region and admixed American vs non-admixed American individuals. For each comparison, the total number of individuals, the number of individuals with elevated repeat copies (i.e., those with $> 487$ repeat copies), the proportion of individuals with elevated repeat copies per population, and the results of the Proportions \\textit{Z}-Test, and Mann-Whitney \\textit{U}-Test is reported. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "19705d22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group 1</th>\n",
       "      <th>Group 2</th>\n",
       "      <th>Total Number of Inds. with Elevated Repeats</th>\n",
       "      <th>Number of Inds. (Group 1)</th>\n",
       "      <th>Number of Inds. (Group 2)</th>\n",
       "      <th>Prop. of Inds. (Group 1)</th>\n",
       "      <th>Prop. of Inds. (Group 2)</th>\n",
       "      <th>$Z-Statistic$</th>\n",
       "      <th>$P-value$ (Prop. $Z$-Test)</th>\n",
       "      <th>$U-Statistic$</th>\n",
       "      <th>$P-value$ (Mann-Whitney $U$-Test)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Outlier Inds.: $\\geq 1$ Overlapping Tract</td>\n",
       "      <td>Outlier Inds.: No Overlapping Tract</td>\n",
       "      <td>118</td>\n",
       "      <td>102</td>\n",
       "      <td>16</td>\n",
       "      <td>0.864407</td>\n",
       "      <td>0.135593</td>\n",
       "      <td>11.196246</td>\n",
       "      <td>2.126889e-29</td>\n",
       "      <td>1412.5</td>\n",
       "      <td>1.398450e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Outlier AMR Inds.</td>\n",
       "      <td>Outlier non-AMR Inds.</td>\n",
       "      <td>118</td>\n",
       "      <td>91</td>\n",
       "      <td>27</td>\n",
       "      <td>0.771186</td>\n",
       "      <td>0.228814</td>\n",
       "      <td>8.332090</td>\n",
       "      <td>3.971366e-17</td>\n",
       "      <td>1988.0</td>\n",
       "      <td>5.788634e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     Group 1  \\\n",
       "0  Outlier Inds.: $\\geq 1$ Overlapping Tract   \n",
       "1                          Outlier AMR Inds.   \n",
       "\n",
       "                               Group 2  \\\n",
       "0  Outlier Inds.: No Overlapping Tract   \n",
       "1                Outlier non-AMR Inds.   \n",
       "\n",
       "   Total Number of Inds. with Elevated Repeats  Number of Inds. (Group 1)  \\\n",
       "0                                          118                        102   \n",
       "1                                          118                         91   \n",
       "\n",
       "   Number of Inds. (Group 2)  Prop. of Inds. (Group 1)  \\\n",
       "0                         16                  0.864407   \n",
       "1                         27                  0.771186   \n",
       "\n",
       "   Prop. of Inds. (Group 2)  $Z-Statistic$  $P-value$ (Prop. $Z$-Test)  \\\n",
       "0                  0.135593      11.196246                2.126889e-29   \n",
       "1                  0.228814       8.332090                3.971366e-17   \n",
       "\n",
       "   $U-Statistic$  $P-value$ (Mann-Whitney $U$-Test)  \n",
       "0         1412.5                       1.398450e-06  \n",
       "1         1988.0                       5.788634e-07  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 2, 3, 4, 5, 6, 15, 16, 17, 18]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ec15529",
   "metadata": {},
   "source": [
    "## S22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "034f6299",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_introgressed_tract_correlations.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S22. Correlation between the introgressed ancestry and repeat copies in the 1000 Genome Project.} \\newline Spearman's correlation coefificients and associated \\textit{P-values} between the number of introgressed tracts overlapping the repeat region and the number of repeat copies for all individuals in the 1000 Genome Project, individuals stratified by super population, and individuals stratified by population. A \\textit{P-value} less than 0.05 is considered statistically significant, and a \"---\" denotes that there are no introgressed tracts overlapping the repeat region for that group.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_introgressed_tract_correlations.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c19c4338",
   "metadata": {},
   "source": [
    "\\textbf{Table S21. Correlation between the introgressed ancestry and repeat copies in the 1000 Genome Project.} \\newline Spearman's correlation coefificients and associated \\textit{P-values} between the number of introgressed tracts overlapping the repeat region and the number of repeat copies for all individuals in the 1000 Genome Project, individuals stratified by super population, and individuals stratified by population. A \\textit{P-value} less than 0.05 is considered statistically significant, and a \"---\" denotes that there are no introgressed tracts overlapping the repeat region for that group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "09af07db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Spearman's $\\rho$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TGP</td>\n",
       "      <td>0.409165</td>\n",
       "      <td>2.022428e-95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>0.736302</td>\n",
       "      <td>1.741238e-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAS</td>\n",
       "      <td>0.215730</td>\n",
       "      <td>1.471162e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EAS</td>\n",
       "      <td>0.531768</td>\n",
       "      <td>3.914002e-38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EUR</td>\n",
       "      <td>0.104177</td>\n",
       "      <td>1.943935e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LWK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GWD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MSL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ESN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MXL</td>\n",
       "      <td>0.885166</td>\n",
       "      <td>2.838516e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PEL</td>\n",
       "      <td>0.812446</td>\n",
       "      <td>3.814654e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CLM</td>\n",
       "      <td>0.631744</td>\n",
       "      <td>8.639555e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>PUR</td>\n",
       "      <td>0.528468</td>\n",
       "      <td>8.118912e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BEB</td>\n",
       "      <td>0.360387</td>\n",
       "      <td>6.532452e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>STU</td>\n",
       "      <td>0.255253</td>\n",
       "      <td>9.619854e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ITU</td>\n",
       "      <td>0.205621</td>\n",
       "      <td>3.814592e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PJL</td>\n",
       "      <td>0.123630</td>\n",
       "      <td>2.301144e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GIH</td>\n",
       "      <td>0.205908</td>\n",
       "      <td>3.692016e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CHB</td>\n",
       "      <td>0.461462</td>\n",
       "      <td>9.297354e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>KHV</td>\n",
       "      <td>0.585871</td>\n",
       "      <td>1.886625e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CHS</td>\n",
       "      <td>0.499451</td>\n",
       "      <td>5.838589e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>JPT</td>\n",
       "      <td>0.419380</td>\n",
       "      <td>9.369610e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CDX</td>\n",
       "      <td>0.653281</td>\n",
       "      <td>1.270084e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TSI</td>\n",
       "      <td>0.043081</td>\n",
       "      <td>6.595008e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CEU</td>\n",
       "      <td>0.243717</td>\n",
       "      <td>1.506000e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>IBS</td>\n",
       "      <td>0.116364</td>\n",
       "      <td>2.326325e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GBR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>FIN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Group  Spearman's $\\rho$     $P-value$\n",
       "0    TGP           0.409165  2.022428e-95\n",
       "1    AFR                NaN           NaN\n",
       "2    AMR           0.736302  1.741238e-60\n",
       "3    SAS           0.215730  1.471162e-06\n",
       "4    EAS           0.531768  3.914002e-38\n",
       "5    EUR           0.104177  1.943935e-02\n",
       "6    LWK                NaN           NaN\n",
       "7    GWD                NaN           NaN\n",
       "8    MSL                NaN           NaN\n",
       "9    ESN                NaN           NaN\n",
       "10   YRI                NaN           NaN\n",
       "11   MXL           0.885166  2.838516e-22\n",
       "12   PEL           0.812446  3.814654e-21\n",
       "13   CLM           0.631744  8.639555e-12\n",
       "14   PUR           0.528468  8.118912e-09\n",
       "15   BEB           0.360387  6.532452e-04\n",
       "16   STU           0.255253  9.619854e-03\n",
       "17   ITU           0.205621  3.814592e-02\n",
       "18   PJL           0.123630  2.301144e-01\n",
       "19   GIH           0.205908  3.692016e-02\n",
       "20   CHB           0.461462  9.297354e-07\n",
       "21   KHV           0.585871  1.886625e-10\n",
       "22   CHS           0.499451  5.838589e-08\n",
       "23   JPT           0.419380  9.369610e-06\n",
       "24   CDX           0.653281  1.270084e-12\n",
       "25   TSI           0.043081  6.595008e-01\n",
       "26   CEU           0.243717  1.506000e-02\n",
       "27   IBS           0.116364  2.326325e-01\n",
       "28   GBR                NaN           NaN\n",
       "29   FIN                NaN           NaN"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b95121",
   "metadata": {},
   "source": [
    "## S23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "04250709",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_amr_ancestry_proportions_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S23. Ancestry component distribution for the \\textit{MUC19} short-read repeat region among admixed American populations in 1000 Genomes Project.} \\newline Summary of the per-individual ancestry proportions at the \\textit{MUC19} short-read repeat region (hg19, Chr12:40876395-40885001) for each of the admixed American populations in the 1000 Genomes project. For each ancestry component per population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the per-individual ancestry proportion distribution are reported.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_amr_ancestry_proportions_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12292d30",
   "metadata": {},
   "source": [
    "\\textbf{Table S22. Ancestry component distribution for the \\textit{MUC19} short-read repeat region among admixed American populations in 1000 Genomes Project.} \\newline Summary of the per-individual ancestry proportions at the \\textit{MUC19} short-read repeat region (hg19, Chr12:40875941-40885367) for each of the admixed American populations in the 1000 Genomes project. For each ancestry component per population, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the per-individual ancestry proportion distribution are reported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a3e88dfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Ancestry Componenet</th>\n",
       "      <th>Ancestry Proportion $\\left( \\mu \\right)$</th>\n",
       "      <th>Ancestry Proportion $\\left( \\sigma \\right)$</th>\n",
       "      <th>Ancestry Proportion $\\left( SEM \\right)$</th>\n",
       "      <th>Ancestry Proportion $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MXL</td>\n",
       "      <td>Indigenous American</td>\n",
       "      <td>0.601562</td>\n",
       "      <td>0.344371</td>\n",
       "      <td>0.043387</td>\n",
       "      <td>0.086701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MXL</td>\n",
       "      <td>European</td>\n",
       "      <td>0.335938</td>\n",
       "      <td>0.331364</td>\n",
       "      <td>0.041748</td>\n",
       "      <td>0.083427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MXL</td>\n",
       "      <td>African</td>\n",
       "      <td>0.054688</td>\n",
       "      <td>0.156055</td>\n",
       "      <td>0.019661</td>\n",
       "      <td>0.039289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PEL</td>\n",
       "      <td>Indigenous American</td>\n",
       "      <td>0.694118</td>\n",
       "      <td>0.343803</td>\n",
       "      <td>0.037512</td>\n",
       "      <td>0.074597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PEL</td>\n",
       "      <td>European</td>\n",
       "      <td>0.270588</td>\n",
       "      <td>0.339142</td>\n",
       "      <td>0.037003</td>\n",
       "      <td>0.073585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PEL</td>\n",
       "      <td>African</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>0.012836</td>\n",
       "      <td>0.025526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CLM</td>\n",
       "      <td>Indigenous American</td>\n",
       "      <td>0.281915</td>\n",
       "      <td>0.305608</td>\n",
       "      <td>0.031690</td>\n",
       "      <td>0.062930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CLM</td>\n",
       "      <td>European</td>\n",
       "      <td>0.648936</td>\n",
       "      <td>0.340758</td>\n",
       "      <td>0.035335</td>\n",
       "      <td>0.070168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CLM</td>\n",
       "      <td>African</td>\n",
       "      <td>0.063830</td>\n",
       "      <td>0.182098</td>\n",
       "      <td>0.018883</td>\n",
       "      <td>0.037497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PUR</td>\n",
       "      <td>Indigenous American</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.250739</td>\n",
       "      <td>0.024706</td>\n",
       "      <td>0.048999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PUR</td>\n",
       "      <td>European</td>\n",
       "      <td>0.682692</td>\n",
       "      <td>0.332948</td>\n",
       "      <td>0.032806</td>\n",
       "      <td>0.065064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PUR</td>\n",
       "      <td>African</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.260149</td>\n",
       "      <td>0.025633</td>\n",
       "      <td>0.050837</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Population  Ancestry Componenet  Ancestry Proportion $\\left( \\mu \\right)$  \\\n",
       "0         MXL  Indigenous American                                  0.601562   \n",
       "1         MXL             European                                  0.335938   \n",
       "2         MXL              African                                  0.054688   \n",
       "3         PEL  Indigenous American                                  0.694118   \n",
       "4         PEL             European                                  0.270588   \n",
       "5         PEL              African                                  0.029412   \n",
       "6         CLM  Indigenous American                                  0.281915   \n",
       "7         CLM             European                                  0.648936   \n",
       "8         CLM              African                                  0.063830   \n",
       "9         PUR  Indigenous American                                  0.153846   \n",
       "10        PUR             European                                  0.682692   \n",
       "11        PUR              African                                  0.153846   \n",
       "\n",
       "    Ancestry Proportion $\\left( \\sigma \\right)$  \\\n",
       "0                                      0.344371   \n",
       "1                                      0.331364   \n",
       "2                                      0.156055   \n",
       "3                                      0.343803   \n",
       "4                                      0.339142   \n",
       "5                                      0.117647   \n",
       "6                                      0.305608   \n",
       "7                                      0.340758   \n",
       "8                                      0.182098   \n",
       "9                                      0.250739   \n",
       "10                                     0.332948   \n",
       "11                                     0.260149   \n",
       "\n",
       "    Ancestry Proportion $\\left( SEM \\right)$  \\\n",
       "0                                   0.043387   \n",
       "1                                   0.041748   \n",
       "2                                   0.019661   \n",
       "3                                   0.037512   \n",
       "4                                   0.037003   \n",
       "5                                   0.012836   \n",
       "6                                   0.031690   \n",
       "7                                   0.035335   \n",
       "8                                   0.018883   \n",
       "9                                   0.024706   \n",
       "10                                  0.032806   \n",
       "11                                  0.025633   \n",
       "\n",
       "    Ancestry Proportion $\\left( \\pm CI_{95\\%} \\right)$  \n",
       "0                                            0.086701   \n",
       "1                                            0.083427   \n",
       "2                                            0.039289   \n",
       "3                                            0.074597   \n",
       "4                                            0.073585   \n",
       "5                                            0.025526   \n",
       "6                                            0.062930   \n",
       "7                                            0.070168   \n",
       "8                                            0.037497   \n",
       "9                                            0.048999   \n",
       "10                                           0.065064   \n",
       "11                                           0.050837   "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14bdcf28",
   "metadata": {},
   "source": [
    "## S24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "575eb604",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_short_read_vntr_amr_ancestry_proportions_correlations.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S24. Correlation between ancestry components and repeat copies among Admixed American populations in the 1000 Genome Project.} \\newline Spearman's correlation coefificients and associated \\textit{P-values} between the ancestry proportion (per ancestry component) and the number of repeat copies for all admixed American individuals in the 1000 Genome Project stratified by population. After correcting for three multiple comparisons per admixed American population---i.e., one per ancestry component---a \\textit{P-value} less than 0.0167 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_short_read_vntr_amr_ancestry_proportions_correlations.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(2, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39afdeef",
   "metadata": {},
   "source": [
    "\\textbf{Table S23.Correlation between ancestry components and repeat copies among Admixed American populations in the 1000 Genome Project.} \\newline Spearman's correlation coefificients and associated \\textit{P-values} between the ancestry proportion (per ancestry component) and the number of repeat copies for all admixed American individuals in the 1000 Genome Project stratified by population. After correcting for three multiple comparisons per admixed American population---i.e., one per ancestry component---a \\textit{P-value} less than 0.0167 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "cdd74b93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Ancestry Componenet</th>\n",
       "      <th>Spearman's $\\rho$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MXL</td>\n",
       "      <td>Indigenous American</td>\n",
       "      <td>0.438100</td>\n",
       "      <td>2.939568e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MXL</td>\n",
       "      <td>European</td>\n",
       "      <td>-0.353268</td>\n",
       "      <td>4.191332e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MXL</td>\n",
       "      <td>African</td>\n",
       "      <td>-0.288633</td>\n",
       "      <td>2.072134e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PEL</td>\n",
       "      <td>Indigenous American</td>\n",
       "      <td>0.390762</td>\n",
       "      <td>2.177656e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PEL</td>\n",
       "      <td>European</td>\n",
       "      <td>-0.313211</td>\n",
       "      <td>3.514602e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PEL</td>\n",
       "      <td>African</td>\n",
       "      <td>-0.199727</td>\n",
       "      <td>6.685667e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CLM</td>\n",
       "      <td>Indigenous American</td>\n",
       "      <td>0.553194</td>\n",
       "      <td>7.377577e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CLM</td>\n",
       "      <td>European</td>\n",
       "      <td>-0.494947</td>\n",
       "      <td>3.956818e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CLM</td>\n",
       "      <td>African</td>\n",
       "      <td>-0.047328</td>\n",
       "      <td>6.505690e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PUR</td>\n",
       "      <td>Indigenous American</td>\n",
       "      <td>0.413539</td>\n",
       "      <td>1.281210e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PUR</td>\n",
       "      <td>European</td>\n",
       "      <td>-0.216058</td>\n",
       "      <td>2.760688e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PUR</td>\n",
       "      <td>African</td>\n",
       "      <td>-0.127257</td>\n",
       "      <td>1.979811e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Population  Ancestry Componenet  Spearman's $\\rho$     $P-value$\n",
       "0         MXL  Indigenous American           0.438100  2.939568e-04\n",
       "1         MXL             European          -0.353268  4.191332e-03\n",
       "2         MXL              African          -0.288633  2.072134e-02\n",
       "3         PEL  Indigenous American           0.390762  2.177656e-04\n",
       "4         PEL             European          -0.313211  3.514602e-03\n",
       "5         PEL              African          -0.199727  6.685667e-02\n",
       "6         CLM  Indigenous American           0.553194  7.377577e-09\n",
       "7         CLM             European          -0.494947  3.956818e-07\n",
       "8         CLM              African          -0.047328  6.505690e-01\n",
       "9         PUR  Indigenous American           0.413539  1.281210e-05\n",
       "10        PUR             European          -0.216058  2.760688e-02\n",
       "11        PUR              African          -0.127257  1.979811e-01"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6b4b662",
   "metadata": {},
   "source": [
    "## S25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "acc35a6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('../meta_data/muc19_72kb_denisovan_coding_muts_info.csv')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "mat[:, -1] = np.round(mat[:, -1].astype(float), 3)\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={Xp{2cm}p{2.25cm}XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S25. Denisovan-specific coding mutations in the focal 72kb \\textit{MUC19} region.} \\newline Allelic and functional information for the 11 Denisovan-specific coding mutations found within the focal 72kb region. Each variant’s effect on protein structure is categorized using Grantham scores as follows: Grantham score $< 50$ = conservative, Grantham score 51–100 = moderately conservative, Grantham score 101–150 = moderately radical, and Grantham score $> 150$ = radical. The PhyloP score measures the conservation of the reference base-pair, with higher positive scores indicating stronger conservation.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/muc19_72kb_denisovan_coding_muts_info.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(mat.shape[1]))\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42995136",
   "metadata": {},
   "source": [
    "\\textbf{Table S24. Denisovan-specific coding mutations in the focal 72kb \\textit{MUC19} region.} \\newline Allelic and functional information for the 11 Denisovan-specific coding mutations found within the focal 72kb region. Each variant’s effect on protein structure is categorized using Grantham scores as follows: Grantham score $< 50$ = conservative, Grantham score 51–100 = moderately conservative, Grantham score 101–150 = moderately radical, and Grantham score $> 150$ = radical. The PhyloP score measures the conservation of the mutation, with higher positive scores indicating stronger conservation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "da56936e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr12 Position (Hg19)</th>\n",
       "      <th>rsID</th>\n",
       "      <th>Mut. Type</th>\n",
       "      <th>Ref. Allele</th>\n",
       "      <th>Ref. Amino Acid</th>\n",
       "      <th>Denisovan Allele</th>\n",
       "      <th>Denisovan Amino Acid</th>\n",
       "      <th>Protein Position</th>\n",
       "      <th>Grantham Score</th>\n",
       "      <th>PhyloP Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40808672</td>\n",
       "      <td>rs4768252</td>\n",
       "      <td>Missense</td>\n",
       "      <td>C</td>\n",
       "      <td>Ser</td>\n",
       "      <td>T</td>\n",
       "      <td>Leu</td>\n",
       "      <td>275</td>\n",
       "      <td>145</td>\n",
       "      <td>-1.220990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40808726</td>\n",
       "      <td>rs4768253</td>\n",
       "      <td>Missense</td>\n",
       "      <td>C</td>\n",
       "      <td>Thr</td>\n",
       "      <td>T</td>\n",
       "      <td>Met</td>\n",
       "      <td>293</td>\n",
       "      <td>81</td>\n",
       "      <td>-0.879299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40809983</td>\n",
       "      <td>rs142268259</td>\n",
       "      <td>Missense</td>\n",
       "      <td>C</td>\n",
       "      <td>Asp</td>\n",
       "      <td>A</td>\n",
       "      <td>Glu</td>\n",
       "      <td>410</td>\n",
       "      <td>45</td>\n",
       "      <td>-0.060835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40814107</td>\n",
       "      <td>rs2114566</td>\n",
       "      <td>Missense</td>\n",
       "      <td>A</td>\n",
       "      <td>Ile</td>\n",
       "      <td>G</td>\n",
       "      <td>Val</td>\n",
       "      <td>511</td>\n",
       "      <td>29</td>\n",
       "      <td>-0.188913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40814197</td>\n",
       "      <td>rs17467164</td>\n",
       "      <td>Missense</td>\n",
       "      <td>G</td>\n",
       "      <td>Val</td>\n",
       "      <td>A</td>\n",
       "      <td>Ile</td>\n",
       "      <td>541</td>\n",
       "      <td>29</td>\n",
       "      <td>-0.336425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40815060</td>\n",
       "      <td>rs149221842</td>\n",
       "      <td>Missense</td>\n",
       "      <td>C</td>\n",
       "      <td>His</td>\n",
       "      <td>T</td>\n",
       "      <td>Tyr</td>\n",
       "      <td>572</td>\n",
       "      <td>83</td>\n",
       "      <td>-0.569071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>40820403</td>\n",
       "      <td>rs11564249</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>C</td>\n",
       "      <td>Asp</td>\n",
       "      <td>T</td>\n",
       "      <td>Asp</td>\n",
       "      <td>711</td>\n",
       "      <td>0</td>\n",
       "      <td>0.818386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>40821795</td>\n",
       "      <td>rs61736852</td>\n",
       "      <td>Missense</td>\n",
       "      <td>G</td>\n",
       "      <td>Glu</td>\n",
       "      <td>A</td>\n",
       "      <td>Lys</td>\n",
       "      <td>766</td>\n",
       "      <td>56</td>\n",
       "      <td>0.581063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>40821871</td>\n",
       "      <td>rs17467284</td>\n",
       "      <td>Missense</td>\n",
       "      <td>G</td>\n",
       "      <td>Arg</td>\n",
       "      <td>T</td>\n",
       "      <td>Leu</td>\n",
       "      <td>791</td>\n",
       "      <td>102</td>\n",
       "      <td>5.150000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>40826201</td>\n",
       "      <td>rs11564125</td>\n",
       "      <td>Missense</td>\n",
       "      <td>G</td>\n",
       "      <td>Val</td>\n",
       "      <td>A</td>\n",
       "      <td>Ile</td>\n",
       "      <td>883</td>\n",
       "      <td>29</td>\n",
       "      <td>-1.722080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>40826535</td>\n",
       "      <td>rs11564104</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>T</td>\n",
       "      <td>Ser</td>\n",
       "      <td>C</td>\n",
       "      <td>Ser</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.035008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chr12 Position (Hg19)         rsID   Mut. Type Ref. Allele  \\\n",
       "0                40808672    rs4768252    Missense           C   \n",
       "1                40808726    rs4768253    Missense           C   \n",
       "2                40809983  rs142268259    Missense           C   \n",
       "3                40814107    rs2114566    Missense           A   \n",
       "4                40814197   rs17467164    Missense           G   \n",
       "5                40815060  rs149221842    Missense           C   \n",
       "6                40820403   rs11564249  Synonymous           C   \n",
       "7                40821795   rs61736852    Missense           G   \n",
       "8                40821871   rs17467284    Missense           G   \n",
       "9                40826201   rs11564125    Missense           G   \n",
       "10               40826535   rs11564104  Synonymous           T   \n",
       "\n",
       "   Ref. Amino Acid Denisovan Allele Denisovan Amino Acid  Protein Position  \\\n",
       "0              Ser                T                  Leu               275   \n",
       "1              Thr                T                  Met               293   \n",
       "2              Asp                A                  Glu               410   \n",
       "3              Ile                G                  Val               511   \n",
       "4              Val                A                  Ile               541   \n",
       "5              His                T                  Tyr               572   \n",
       "6              Asp                T                  Asp               711   \n",
       "7              Glu                A                  Lys               766   \n",
       "8              Arg                T                  Leu               791   \n",
       "9              Val                A                  Ile               883   \n",
       "10             Ser                C                  Ser               939   \n",
       "\n",
       "    Grantham Score  PhyloP Score  \n",
       "0              145     -1.220990  \n",
       "1               81     -0.879299  \n",
       "2               45     -0.060835  \n",
       "3               29     -0.188913  \n",
       "4               29     -0.336425  \n",
       "5               83     -0.569071  \n",
       "6                0      0.818386  \n",
       "7               56      0.581063  \n",
       "8              102      5.150000  \n",
       "9               29     -1.722080  \n",
       "10               0     -0.035008  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10c9d01e",
   "metadata": {},
   "source": [
    "## S26"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "eb5c9b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('../meta_data/ancient_sample_info.csv')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S26. Metadata for the ancient Indigenous American genomes.} \\newline The location, sample ID, age, genomic coverage, and reference for the 23 ancient Indigenous American individuals that predate European colonization and the African slave trade used in this study.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/ancient_sample_info.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(mat.shape[1]))\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4fefca",
   "metadata": {},
   "source": [
    "\\textbf{Table S25. Metadata for the ancient Indigenous American genomes.} \\newline The location, sample ID, age, genomic coverage, and reference for the 23 ancient Indigenous American individuals that predate European colonization and the African slave trade used in this study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9724dc4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Location</th>\n",
       "      <th>ID</th>\n",
       "      <th>Age (years)</th>\n",
       "      <th>Coverage (x)</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alaska</td>\n",
       "      <td>USR1</td>\n",
       "      <td>11,500</td>\n",
       "      <td>17.00</td>\n",
       "      <td>Moreno-Mayar et al. 2018 Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>California</td>\n",
       "      <td>CR-01</td>\n",
       "      <td>1,111</td>\n",
       "      <td>8.36</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>California</td>\n",
       "      <td>CT-01</td>\n",
       "      <td>387</td>\n",
       "      <td>6.68</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>California</td>\n",
       "      <td>NC\\_C</td>\n",
       "      <td>1,450</td>\n",
       "      <td>1.12</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>California</td>\n",
       "      <td>PS-06</td>\n",
       "      <td>1,570</td>\n",
       "      <td>4.12</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>California</td>\n",
       "      <td>SC-05</td>\n",
       "      <td>1,101</td>\n",
       "      <td>13.67</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>California</td>\n",
       "      <td>SM-02</td>\n",
       "      <td>826</td>\n",
       "      <td>7.38</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>California</td>\n",
       "      <td>SN-13</td>\n",
       "      <td>881</td>\n",
       "      <td>1.74</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>California</td>\n",
       "      <td>SN-17</td>\n",
       "      <td>4,517</td>\n",
       "      <td>7.46</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>California</td>\n",
       "      <td>SN-44</td>\n",
       "      <td>4,646</td>\n",
       "      <td>9.36</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Montana</td>\n",
       "      <td>Clovis</td>\n",
       "      <td>12,500</td>\n",
       "      <td>14.40</td>\n",
       "      <td>Rasmussen et al. 2014 Nature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Ontario</td>\n",
       "      <td>CK-13</td>\n",
       "      <td>4,267</td>\n",
       "      <td>3.02</td>\n",
       "      <td>Scheib et al. 2018 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Patagonia</td>\n",
       "      <td>ERR2270782</td>\n",
       "      <td>1,000</td>\n",
       "      <td>7.80</td>\n",
       "      <td>de la Fuente et al. 2018 PNAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Patagonia</td>\n",
       "      <td>ERR2270783</td>\n",
       "      <td>1,000</td>\n",
       "      <td>3.50</td>\n",
       "      <td>de la Fuente et al. 2018 PNAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Patagonia</td>\n",
       "      <td>ERR2270784</td>\n",
       "      <td>1,000</td>\n",
       "      <td>1.70</td>\n",
       "      <td>de la Fuente et al. 2018 PNAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Patagonia</td>\n",
       "      <td>ERR2270785</td>\n",
       "      <td>1,000</td>\n",
       "      <td>9.10</td>\n",
       "      <td>de la Fuente et al. 2018 PNAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Peru</td>\n",
       "      <td>SRR8144644</td>\n",
       "      <td>1,800</td>\n",
       "      <td>3.34</td>\n",
       "      <td>Lindo et al. 2018 Science Advances</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Peru</td>\n",
       "      <td>SRR8144646</td>\n",
       "      <td>1,800</td>\n",
       "      <td>5.89</td>\n",
       "      <td>Lindo et al. 2018 Science Advances</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Peru</td>\n",
       "      <td>SRR8144647</td>\n",
       "      <td>1,800</td>\n",
       "      <td>6.51</td>\n",
       "      <td>Lindo et al. 2018 Science Advances</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Peru</td>\n",
       "      <td>SRR8144650</td>\n",
       "      <td>1,800</td>\n",
       "      <td>5.62</td>\n",
       "      <td>Lindo et al. 2018 Science Advances</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Quer\\'{e}taro</td>\n",
       "      <td>2417J</td>\n",
       "      <td>1,161</td>\n",
       "      <td>1.39</td>\n",
       "      <td>Villa-Islas et al. 2023 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Quer\\'{e}taro</td>\n",
       "      <td>2417Q</td>\n",
       "      <td>1,257</td>\n",
       "      <td>4.74</td>\n",
       "      <td>Villa-Islas et al. 2023 Science</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Quer\\'{e}taro</td>\n",
       "      <td>333B</td>\n",
       "      <td>1,351</td>\n",
       "      <td>2.28</td>\n",
       "      <td>Villa-Islas et al. 2023 Science</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Location          ID Age (years)  Coverage (x)  \\\n",
       "0          Alaska        USR1      11,500         17.00   \n",
       "1      California       CR-01       1,111          8.36   \n",
       "2      California       CT-01         387          6.68   \n",
       "3      California       NC\\_C       1,450          1.12   \n",
       "4      California       PS-06       1,570          4.12   \n",
       "5      California       SC-05       1,101         13.67   \n",
       "6      California       SM-02         826          7.38   \n",
       "7      California       SN-13         881          1.74   \n",
       "8      California       SN-17       4,517          7.46   \n",
       "9      California       SN-44       4,646          9.36   \n",
       "10        Montana      Clovis      12,500         14.40   \n",
       "11        Ontario       CK-13       4,267          3.02   \n",
       "12      Patagonia  ERR2270782       1,000          7.80   \n",
       "13      Patagonia  ERR2270783       1,000          3.50   \n",
       "14      Patagonia  ERR2270784       1,000          1.70   \n",
       "15      Patagonia  ERR2270785       1,000          9.10   \n",
       "16           Peru  SRR8144644       1,800          3.34   \n",
       "17           Peru  SRR8144646       1,800          5.89   \n",
       "18           Peru  SRR8144647       1,800          6.51   \n",
       "19           Peru  SRR8144650       1,800          5.62   \n",
       "20  Quer\\'{e}taro       2417J       1,161          1.39   \n",
       "21  Quer\\'{e}taro       2417Q       1,257          4.74   \n",
       "22  Quer\\'{e}taro        333B       1,351          2.28   \n",
       "\n",
       "                             Reference  \n",
       "0      Moreno-Mayar et al. 2018 Nature  \n",
       "1           Scheib et al. 2018 Science  \n",
       "2           Scheib et al. 2018 Science  \n",
       "3           Scheib et al. 2018 Science  \n",
       "4           Scheib et al. 2018 Science  \n",
       "5           Scheib et al. 2018 Science  \n",
       "6           Scheib et al. 2018 Science  \n",
       "7           Scheib et al. 2018 Science  \n",
       "8           Scheib et al. 2018 Science  \n",
       "9           Scheib et al. 2018 Science  \n",
       "10        Rasmussen et al. 2014 Nature  \n",
       "11          Scheib et al. 2018 Science  \n",
       "12       de la Fuente et al. 2018 PNAS  \n",
       "13       de la Fuente et al. 2018 PNAS  \n",
       "14       de la Fuente et al. 2018 PNAS  \n",
       "15       de la Fuente et al. 2018 PNAS  \n",
       "16  Lindo et al. 2018 Science Advances  \n",
       "17  Lindo et al. 2018 Science Advances  \n",
       "18  Lindo et al. 2018 Science Advances  \n",
       "19  Lindo et al. 2018 Science Advances  \n",
       "20     Villa-Islas et al. 2023 Science  \n",
       "21     Villa-Islas et al. 2023 Science  \n",
       "22     Villa-Islas et al. 2023 Science  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a0bdc84",
   "metadata": {},
   "source": [
    "## S27"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "8516dbb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/amr_v_non_amr_denisovan_specific_missense_mutation_proportions_ztest_fet.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 2, 3, 5, 6, 7, 8, 9, 10]]\n",
    "# Intialize the columns.\n",
    "cols = np.where(\n",
    "    df.columns.values[[0, 2, 3, 5, 6, 7, 8, 9, 10]] == 'Denisovan-specific Mis. Mut. Pos.',\n",
    "    'Position', df.columns.values[[0, 2, 3, 5, 6, 7, 8, 9, 10]],\n",
    ")\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S27. Comparison of the per-individual frequency distributions for each Denisovan-specific missense mutation in the focal 72kb \\textit{MUC19} region between admixed American and non-admixed American individuals in 1000 Genomes Project} \\newline Comparisons of the per-individual frequency distribution for each of the nine Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region between the 347 admixed American (AMR) individuals ($n = 694$ chromosomes) and 1,496 non-admixed American (Non-AMR) individuals ($n = 2,992$ chromosomes), excluding African populations, in the 1000 Genomes Project. For each comparison, the number of chromosomes harboring the respective Denisovan-specific missense mutation, the frequency of the Denisovan-specific missense mutation, and the results of the Proportions \\textit{Z}-Test and Fisher's Exact Test are reported. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/amr_v_non_amr_denisovan_specific_missense_mutation_proportions_ztest_fet.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join((\n",
    "            f'{int(mat[i, j])}' if mat[i, j] > 100 else\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a57c088b",
   "metadata": {},
   "source": [
    "\\textbf{Table S26. Comparison of the per-individual frequency distributions for each Denisovan-specific missense mutation in the focal 72kb \\textit{MUC19} region between admixed American and non-admixed American individuals in 1000 Genomes Project} \\newline Comparisons of the per-individual frequency distribution for each of the nine Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region between the 347 admixed American (AMR) individuals ($n = 694$ chromosomes) and 1,496 non-admixed American (Non-AMR) individuals ($n = 2,992$ chromosomes), excluding African populations, in the 1000 Genomes Project. For each comparison, the number of chromosomes harboring the respective Denisovan-specific missense mutation, the frequency of the Denisovan-specific missense mutation, and the results of the Proportions \\textit{Z}-Test and Fisher's Exact Test are reported. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "bd54edf2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Denisovan-specific Mis. Mut. Pos.</th>\n",
       "      <th>Denisovan-specific Mis. Mut. Count (AMR)</th>\n",
       "      <th>Denisovan-specific Mis. Freq. (AMR)</th>\n",
       "      <th>Denisovan-specific Mis. Mut. Count (Non-AMR)</th>\n",
       "      <th>Denisovan-specific Mis. Freq. (Non-AMR)</th>\n",
       "      <th>$Z-Statistic$</th>\n",
       "      <th>$P-value$ (Prop. $Z$-Test)</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>$P-value$ (Fisher's Exact Test)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40808672</td>\n",
       "      <td>109</td>\n",
       "      <td>0.157061</td>\n",
       "      <td>187</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>8.258522</td>\n",
       "      <td>7.374357e-17</td>\n",
       "      <td>2.794872</td>\n",
       "      <td>1.939194e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40808726</td>\n",
       "      <td>108</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>190</td>\n",
       "      <td>0.063503</td>\n",
       "      <td>8.020432</td>\n",
       "      <td>5.268677e-16</td>\n",
       "      <td>2.717945</td>\n",
       "      <td>9.213214e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40809983</td>\n",
       "      <td>108</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>186</td>\n",
       "      <td>0.062166</td>\n",
       "      <td>8.187167</td>\n",
       "      <td>1.337230e-16</td>\n",
       "      <td>2.780359</td>\n",
       "      <td>3.175081e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40814107</td>\n",
       "      <td>108</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>186</td>\n",
       "      <td>0.062166</td>\n",
       "      <td>8.187167</td>\n",
       "      <td>1.337230e-16</td>\n",
       "      <td>2.780359</td>\n",
       "      <td>3.175081e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40814197</td>\n",
       "      <td>108</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>188</td>\n",
       "      <td>0.062834</td>\n",
       "      <td>8.103488</td>\n",
       "      <td>2.670271e-16</td>\n",
       "      <td>2.748820</td>\n",
       "      <td>5.430066e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40815060</td>\n",
       "      <td>108</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>186</td>\n",
       "      <td>0.062166</td>\n",
       "      <td>8.187167</td>\n",
       "      <td>1.337230e-16</td>\n",
       "      <td>2.780359</td>\n",
       "      <td>3.175081e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>40821795</td>\n",
       "      <td>107</td>\n",
       "      <td>0.154179</td>\n",
       "      <td>180</td>\n",
       "      <td>0.060160</td>\n",
       "      <td>8.327868</td>\n",
       "      <td>4.115575e-17</td>\n",
       "      <td>2.847662</td>\n",
       "      <td>1.319832e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>40821871</td>\n",
       "      <td>107</td>\n",
       "      <td>0.154179</td>\n",
       "      <td>180</td>\n",
       "      <td>0.060160</td>\n",
       "      <td>8.327868</td>\n",
       "      <td>4.115575e-17</td>\n",
       "      <td>2.847662</td>\n",
       "      <td>1.319832e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>40826201</td>\n",
       "      <td>108</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>181</td>\n",
       "      <td>0.060495</td>\n",
       "      <td>8.399161</td>\n",
       "      <td>2.248396e-17</td>\n",
       "      <td>2.862256</td>\n",
       "      <td>8.010657e-15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Denisovan-specific Mis. Mut. Pos.  \\\n",
       "0                           40808672   \n",
       "1                           40808726   \n",
       "2                           40809983   \n",
       "3                           40814107   \n",
       "4                           40814197   \n",
       "5                           40815060   \n",
       "6                           40821795   \n",
       "7                           40821871   \n",
       "8                           40826201   \n",
       "\n",
       "   Denisovan-specific Mis. Mut. Count (AMR)  \\\n",
       "0                                       109   \n",
       "1                                       108   \n",
       "2                                       108   \n",
       "3                                       108   \n",
       "4                                       108   \n",
       "5                                       108   \n",
       "6                                       107   \n",
       "7                                       107   \n",
       "8                                       108   \n",
       "\n",
       "   Denisovan-specific Mis. Freq. (AMR)  \\\n",
       "0                             0.157061   \n",
       "1                             0.155620   \n",
       "2                             0.155620   \n",
       "3                             0.155620   \n",
       "4                             0.155620   \n",
       "5                             0.155620   \n",
       "6                             0.154179   \n",
       "7                             0.154179   \n",
       "8                             0.155620   \n",
       "\n",
       "   Denisovan-specific Mis. Mut. Count (Non-AMR)  \\\n",
       "0                                           187   \n",
       "1                                           190   \n",
       "2                                           186   \n",
       "3                                           186   \n",
       "4                                           188   \n",
       "5                                           186   \n",
       "6                                           180   \n",
       "7                                           180   \n",
       "8                                           181   \n",
       "\n",
       "   Denisovan-specific Mis. Freq. (Non-AMR)  $Z-Statistic$  \\\n",
       "0                                 0.062500       8.258522   \n",
       "1                                 0.063503       8.020432   \n",
       "2                                 0.062166       8.187167   \n",
       "3                                 0.062166       8.187167   \n",
       "4                                 0.062834       8.103488   \n",
       "5                                 0.062166       8.187167   \n",
       "6                                 0.060160       8.327868   \n",
       "7                                 0.060160       8.327868   \n",
       "8                                 0.060495       8.399161   \n",
       "\n",
       "   $P-value$ (Prop. $Z$-Test)  Odds Ratio  $P-value$ (Fisher's Exact Test)  \n",
       "0                7.374357e-17    2.794872                     1.939194e-14  \n",
       "1                5.268677e-16    2.717945                     9.213214e-14  \n",
       "2                1.337230e-16    2.780359                     3.175081e-14  \n",
       "3                1.337230e-16    2.780359                     3.175081e-14  \n",
       "4                2.670271e-16    2.748820                     5.430066e-14  \n",
       "5                1.337230e-16    2.780359                     3.175081e-14  \n",
       "6                4.115575e-17    2.847662                     1.319832e-14  \n",
       "7                4.115575e-17    2.847662                     1.319832e-14  \n",
       "8                2.248396e-17    2.862256                     8.010657e-15  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 2, 3, 5, 6, 7, 8, 9, 10]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f8a3cb",
   "metadata": {},
   "source": [
    "## S28"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "45ec44c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_denisovan_specific_missense_mutation_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [10, 11, 12, 13, 14]].T\n",
    "idx = df.iloc[:, [10, 11, 12, 13, 14]].T.index.values\n",
    "# Intialize the columns.\n",
    "cols = ['Super Population']\n",
    "cols.extend(f'{pos}' for pos in df['Chr12 Position (Hg19)'].values)\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S28. Frequency of the Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region among super populations in the 1000 Genomes Project.} \\newline Frequencies of the nine Denisovan-specific missense mutations found within the focal 72kb \\textit{MUC19} region per super population in the 1000 Genomes Project.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_denisovan_specific_missense_mutation_summary_per_spop.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        f'{idx[i]} & '\\\n",
    "        + r' & '.join(f'{round(mat[i, j], 3)}' for j in range(mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4854fe5a",
   "metadata": {},
   "source": [
    "\\textbf{Table S27. Frequency of the Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region among super populations in the 1000 Genomes Project.} \\newline Frequencies of the nine Denisovan-specific missense mutations found within the focal 72kb \\textit{MUC19} region per super population in the 1000 Genomes Project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "88aaa452",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AFR</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMR</th>\n",
       "      <td>0.157061</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>0.155620</td>\n",
       "      <td>0.154179</td>\n",
       "      <td>0.154179</td>\n",
       "      <td>0.155620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAS</th>\n",
       "      <td>0.105317</td>\n",
       "      <td>0.108384</td>\n",
       "      <td>0.104294</td>\n",
       "      <td>0.104294</td>\n",
       "      <td>0.105317</td>\n",
       "      <td>0.104294</td>\n",
       "      <td>0.098160</td>\n",
       "      <td>0.098160</td>\n",
       "      <td>0.098160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EAS</th>\n",
       "      <td>0.060516</td>\n",
       "      <td>0.060516</td>\n",
       "      <td>0.060516</td>\n",
       "      <td>0.060516</td>\n",
       "      <td>0.061508</td>\n",
       "      <td>0.060516</td>\n",
       "      <td>0.060516</td>\n",
       "      <td>0.060516</td>\n",
       "      <td>0.061508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EUR</th>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.022863</td>\n",
       "      <td>0.022863</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0         1         2         3         4         5         6  \\\n",
       "AFR  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "AMR  0.157061  0.155620  0.155620  0.155620  0.155620  0.155620  0.154179   \n",
       "SAS  0.105317  0.108384  0.104294  0.104294  0.105317  0.104294  0.098160   \n",
       "EAS  0.060516  0.060516  0.060516  0.060516  0.061508  0.060516  0.060516   \n",
       "EUR  0.022863  0.022863  0.022863  0.022863  0.022863  0.022863  0.022863   \n",
       "\n",
       "            7         8  \n",
       "AFR  0.000000  0.000000  \n",
       "AMR  0.154179  0.155620  \n",
       "SAS  0.098160  0.098160  \n",
       "EAS  0.060516  0.061508  \n",
       "EUR  0.022863  0.022863  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [10, 11, 12, 13, 14]].T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf067549",
   "metadata": {},
   "source": [
    "## S29"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "05565404",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_denisovan_specific_missense_mutation_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, 15:].T\n",
    "idx = df.iloc[:, 15:].T.index.values\n",
    "# Intialize the columns.\n",
    "cols = ['Population']\n",
    "cols.extend(f'{pos}' for pos in df['Chr12 Position (Hg19)'].values)\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S29. Frequency of the Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region among populations in the 1000 Genomes Project.} \\newline Frequencies of the nine Denisovan-specific missense mutations found within the focal 72kb \\textit{MUC19} region per population in the 1000 Genomes Project.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_denisovan_specific_missense_mutation_summary_per_pop.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        f'{idx[i]} & '\\\n",
    "        + r' & '.join(f'{round(mat[i, j], 3)}' for j in range(mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ef42a2a",
   "metadata": {},
   "source": [
    "\\textbf{Table S28. Frequency of the Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region among populations in the 1000 Genomes Project.} \\newline Frequencies of the nine Denisovan-specific missense mutations found within the focal 72kb \\textit{MUC19} region per population in the 1000 Genomes Project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f5e06556",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GWD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ESN</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MSL</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YRI</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LWK</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PUR</th>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.086538</td>\n",
       "      <td>0.091346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLM</th>\n",
       "      <td>0.079787</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.069149</td>\n",
       "      <td>0.069149</td>\n",
       "      <td>0.069149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PEL</th>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "      <td>0.217647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MXL</th>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.304688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PJL</th>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "      <td>0.052083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BEB</th>\n",
       "      <td>0.156977</td>\n",
       "      <td>0.168605</td>\n",
       "      <td>0.156977</td>\n",
       "      <td>0.156977</td>\n",
       "      <td>0.156977</td>\n",
       "      <td>0.156977</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>0.139535</td>\n",
       "      <td>0.139535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STU</th>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITU</th>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.107843</td>\n",
       "      <td>0.107843</td>\n",
       "      <td>0.112745</td>\n",
       "      <td>0.107843</td>\n",
       "      <td>0.102941</td>\n",
       "      <td>0.102941</td>\n",
       "      <td>0.102941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GIH</th>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.101942</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.097087</td>\n",
       "      <td>0.087379</td>\n",
       "      <td>0.087379</td>\n",
       "      <td>0.087379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHS</th>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.038095</td>\n",
       "      <td>0.038095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDX</th>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "      <td>0.123656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KHV</th>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.085859</td>\n",
       "      <td>0.090909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHB</th>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.038835</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.033981</td>\n",
       "      <td>0.033981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JPT</th>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>0.028846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBR</th>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "      <td>0.005495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIN</th>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "      <td>0.005051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IBS</th>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "      <td>0.032710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEU</th>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.015152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSI</th>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "      <td>0.051402</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0         1         2         3         4         5         6  \\\n",
       "GWD  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "ESN  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "MSL  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "YRI  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "LWK  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "PUR  0.086538  0.086538  0.086538  0.086538  0.086538  0.086538  0.086538   \n",
       "CLM  0.079787  0.074468  0.074468  0.074468  0.074468  0.074468  0.069149   \n",
       "PEL  0.217647  0.217647  0.217647  0.217647  0.217647  0.217647  0.217647   \n",
       "MXL  0.304688  0.304688  0.304688  0.304688  0.304688  0.304688  0.304688   \n",
       "PJL  0.052083  0.052083  0.052083  0.052083  0.052083  0.052083  0.052083   \n",
       "BEB  0.156977  0.168605  0.156977  0.156977  0.156977  0.156977  0.139535   \n",
       "STU  0.112745  0.112745  0.112745  0.112745  0.112745  0.112745  0.112745   \n",
       "ITU  0.112745  0.112745  0.107843  0.107843  0.112745  0.107843  0.102941   \n",
       "GIH  0.097087  0.101942  0.097087  0.097087  0.097087  0.097087  0.087379   \n",
       "CHS  0.038095  0.038095  0.038095  0.038095  0.038095  0.038095  0.038095   \n",
       "CDX  0.123656  0.123656  0.123656  0.123656  0.123656  0.123656  0.123656   \n",
       "KHV  0.085859  0.085859  0.085859  0.085859  0.085859  0.085859  0.085859   \n",
       "CHB  0.033981  0.033981  0.033981  0.033981  0.038835  0.033981  0.033981   \n",
       "JPT  0.028846  0.028846  0.028846  0.028846  0.028846  0.028846  0.028846   \n",
       "GBR  0.005495  0.005495  0.005495  0.005495  0.005495  0.005495  0.005495   \n",
       "FIN  0.005051  0.005051  0.005051  0.005051  0.005051  0.005051  0.005051   \n",
       "IBS  0.032710  0.032710  0.032710  0.032710  0.032710  0.032710  0.032710   \n",
       "CEU  0.015152  0.015152  0.015152  0.015152  0.015152  0.015152  0.015152   \n",
       "TSI  0.051402  0.051402  0.051402  0.051402  0.051402  0.051402  0.051402   \n",
       "\n",
       "            7         8  \n",
       "GWD  0.000000  0.000000  \n",
       "ESN  0.000000  0.000000  \n",
       "MSL  0.000000  0.000000  \n",
       "YRI  0.000000  0.000000  \n",
       "LWK  0.000000  0.000000  \n",
       "PUR  0.086538  0.091346  \n",
       "CLM  0.069149  0.069149  \n",
       "PEL  0.217647  0.217647  \n",
       "MXL  0.304688  0.304688  \n",
       "PJL  0.052083  0.052083  \n",
       "BEB  0.139535  0.139535  \n",
       "STU  0.112745  0.112745  \n",
       "ITU  0.102941  0.102941  \n",
       "GIH  0.087379  0.087379  \n",
       "CHS  0.038095  0.038095  \n",
       "CDX  0.123656  0.123656  \n",
       "KHV  0.085859  0.090909  \n",
       "CHB  0.033981  0.033981  \n",
       "JPT  0.028846  0.028846  \n",
       "GBR  0.005495  0.005495  \n",
       "FIN  0.005051  0.005051  \n",
       "IBS  0.032710  0.032710  \n",
       "CEU  0.015152  0.015152  \n",
       "TSI  0.051402  0.051402  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, 15:].T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fe492d5",
   "metadata": {},
   "source": [
    "## S29"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "77a0e45a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/ancient_americans_denisovan_specific_missense_mutation_aac.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, 1:].T\n",
    "idx = np.where(df.columns.values[1:] == 'NC_C', 'NC\\_C', df.columns.values[1:])\n",
    "# Intialize the columns.\n",
    "cols = ['ID']\n",
    "cols.extend(f'{int(pos)}' for pos in df['POS'].values)\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={p{2.25cm}XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S30. Denisovan-specific missense mutation allele counts in the focal 72kb \\textit{MUC19} region among 23 ancient Indigenous American individuals.} \\newline Allele counts (i.e., 0, 1, or 2) of the nine Denisovan-specific missense mutations found within the focal 72kb \\textit{MUC19} region for each of the 23 ancient Indigenous American individuals that predate European colonization and the African slave trade. A \"---\"  denotes that the site did not pass quality control.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/ancient_americans_denisovan_specific_missense_mutation_aac.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        f'{idx[i]} & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            f'{int(mat[i, j])}'\n",
    "            for j in range(mat.shape[1])\n",
    "        ))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f732ca78",
   "metadata": {},
   "source": [
    "\\textbf{Table S29. Denisovan-specific missense mutation allele counts in the focal 72kb \\textit{MUC19} region among 23 ancient Indigenous American individuals.} \\newline Allele counts (i.e., 0, 1, or 2) of the nine Denisovan-specific missense mutations found within the focal 72kb \\textit{MUC19} region for each of the 23 ancient Indigenous American individuals that predate European colonization and the African slave trade. A \"---\"  denotes that the site did not pass quality control."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "b0137794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>USR1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CK-13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Anzick</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CR-01</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CT-01</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NC_C</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PS-06</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SC-05</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SM-02</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SN-13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SN-17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SN-44</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2417J</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2417Q</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333B</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8144644</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8144646</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8144647</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8144650</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2270782</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2270783</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2270784</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR2270785</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0    1    2    3    4    5    6    7    8\n",
       "USR1        1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "CK-13       NaN  2.0  0.0  NaN  0.0  NaN  NaN  NaN  NaN\n",
       "Anzick      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "CR-01       2.0  2.0  2.0  2.0  NaN  2.0  2.0  2.0  2.0\n",
       "CT-01       0.0  0.0  NaN  0.0  0.0  0.0  NaN  0.0  0.0\n",
       "NC_C        NaN  0.0  NaN  NaN  NaN  NaN  0.0  0.0  NaN\n",
       "PS-06       0.0  0.0  0.0  NaN  NaN  0.0  0.0  0.0  NaN\n",
       "SC-05       1.0  1.0  1.0  1.0  1.0  2.0  1.0  1.0  1.0\n",
       "SM-02       2.0  2.0  2.0  2.0  2.0  NaN  NaN  2.0  2.0\n",
       "SN-13       NaN  2.0  NaN  2.0  NaN  2.0  NaN  NaN  NaN\n",
       "SN-17       0.0  1.0  1.0  0.0  2.0  2.0  2.0  0.0  1.0\n",
       "SN-44       1.0  1.0  1.0  1.0  1.0  1.0  NaN  NaN  1.0\n",
       "2417J       NaN  NaN  NaN  NaN  NaN  NaN  NaN  2.0  NaN\n",
       "2417Q       0.0  NaN  NaN  0.0  NaN  NaN  0.0  NaN  NaN\n",
       "333B        NaN  NaN  NaN  0.0  2.0  NaN  NaN  1.0  NaN\n",
       "SRR8144644  0.0  0.0  2.0  0.0  2.0  0.0  NaN  1.0  0.0\n",
       "SRR8144646  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "SRR8144647  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "SRR8144650  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0\n",
       "ERR2270782  2.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0  2.0\n",
       "ERR2270783  2.0  1.0  NaN  0.0  0.0  1.0  1.0  0.0  1.0\n",
       "ERR2270784  0.0  0.0  0.0  NaN  NaN  0.0  NaN  0.0  NaN\n",
       "ERR2270785  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, 1:].T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84009467",
   "metadata": {},
   "source": [
    "## S31"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "4c5e11eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/mxl_sgdp_amr_anc_amr_denisovan_specific_missense_mutation_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = np.where(df.columns.values == 'Denisovan-specific Mis. Mut. Pos.', 'Position', df.columns.values)\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S31. Per-individual frequency distribution for each Denisovan-specific missense mutation in the focal 72kb \\textit{MUC19} region among Mexicans in 1000 Genomes Project, 23 ancient Indigenous American individuals, and Indigenous American individuals in the Simons Genome Diversity Project.} \\newline Summary of the per-individual frequency (i.e., 0, 0.5, or 1) distribution for each of the nine Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region per group: 64 Mexican individuals in the 1000 Genomes project ($n = 128$ chromosomes), 23 ancient Indigenous American individuals ($n = 46$ chromosomes), and 22 Indigenous American individuals in the Simons Genome Diversity Project ($n = 44$ chromosomes). For each group, the total number of chromosomes that passed quality control, the mean (i.e., frequency), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the per-individual Denisovan-specific missense mutation frequency distribution are reported.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/mxl_sgdp_amr_anc_amr_denisovan_specific_missense_mutation_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            f'{int(mat[i, j])}' if mat[i, j] > 1 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(2, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c371507b",
   "metadata": {},
   "source": [
    "\\textbf{Table S30. Per-individual frequency distribution for each Denisovan-specific missense mutation in the focal 72kb \\textit{MUC19} region among Mexicans in 1000 Genomes Project, 23 ancient Indigenous American individuals, and Indigenous American individuals in the Simons Genome Diversity Project.} \\newline Summary of the per-individual frequency (i.e., 0, 0.5, or 1) distribution for each of the nine Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region per group: 64 Mexican individuals in the 1000 Genomes project ($n = 128$ chromosomes), 23 ancient Indigenous American individuals ($n = 46$ chromosomes), and 22 Indigenous American individuals in the Simons Genome Diversity Project ($n = 44$ chromosomes). For each group, the total number of chromosomes that passed quality control, the mean (i.e., frequency), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the per-individual Denisovan-specific missense mutation frequency distribution are reported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "6cab8787",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Denisovan-specific Mis. Mut. Pos.</th>\n",
       "      <th>Group</th>\n",
       "      <th>Number of Chroms.</th>\n",
       "      <th>Denisovan-specific Mis. Mut. Freq.</th>\n",
       "      <th>Denisovan-specific Mis. Mut. $\\left( \\sigma \\right)$</th>\n",
       "      <th>Denisovan-specific Mis. Mut. $\\left( SEM \\right)$</th>\n",
       "      <th>Denisovan-specific Mis. Mut. $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40808672</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40808672</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40808672</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>36</td>\n",
       "      <td>0.361111</td>\n",
       "      <td>0.401579</td>\n",
       "      <td>0.097397</td>\n",
       "      <td>0.205490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40808726</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40808726</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40808726</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>40</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.414578</td>\n",
       "      <td>0.095111</td>\n",
       "      <td>0.199069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>40809983</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>40809983</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>40809983</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>32</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.414578</td>\n",
       "      <td>0.107044</td>\n",
       "      <td>0.228158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>40814107</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>40814107</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>40814107</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>36</td>\n",
       "      <td>0.305556</td>\n",
       "      <td>0.412946</td>\n",
       "      <td>0.100154</td>\n",
       "      <td>0.211307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>40814197</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>40814197</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>40814197</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>32</td>\n",
       "      <td>0.406250</td>\n",
       "      <td>0.440835</td>\n",
       "      <td>0.113823</td>\n",
       "      <td>0.242608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>40815060</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>40815060</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>40815060</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>34</td>\n",
       "      <td>0.382353</td>\n",
       "      <td>0.438225</td>\n",
       "      <td>0.109556</td>\n",
       "      <td>0.232249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>40821795</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>40821795</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>40821795</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>28</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>0.405636</td>\n",
       "      <td>0.112503</td>\n",
       "      <td>0.243049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>40821871</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>40821871</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>40821871</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>38</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.404271</td>\n",
       "      <td>0.095288</td>\n",
       "      <td>0.200192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>40826201</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>0.304688</td>\n",
       "      <td>0.301263</td>\n",
       "      <td>0.037956</td>\n",
       "      <td>0.075848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>40826201</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>44</td>\n",
       "      <td>0.363636</td>\n",
       "      <td>0.308288</td>\n",
       "      <td>0.067274</td>\n",
       "      <td>0.139904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>40826201</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>30</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>0.385861</td>\n",
       "      <td>0.103126</td>\n",
       "      <td>0.221183</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Denisovan-specific Mis. Mut. Pos.              Group  Number of Chroms.  \\\n",
       "0                            40808672                MXL                128   \n",
       "1                            40808672           SDGP AMR                 44   \n",
       "2                            40808672  Ancient Americans                 36   \n",
       "3                            40808726                MXL                128   \n",
       "4                            40808726           SDGP AMR                 44   \n",
       "5                            40808726  Ancient Americans                 40   \n",
       "6                            40809983                MXL                128   \n",
       "7                            40809983           SDGP AMR                 44   \n",
       "8                            40809983  Ancient Americans                 32   \n",
       "9                            40814107                MXL                128   \n",
       "10                           40814107           SDGP AMR                 44   \n",
       "11                           40814107  Ancient Americans                 36   \n",
       "12                           40814197                MXL                128   \n",
       "13                           40814197           SDGP AMR                 44   \n",
       "14                           40814197  Ancient Americans                 32   \n",
       "15                           40815060                MXL                128   \n",
       "16                           40815060           SDGP AMR                 44   \n",
       "17                           40815060  Ancient Americans                 34   \n",
       "18                           40821795                MXL                128   \n",
       "19                           40821795           SDGP AMR                 44   \n",
       "20                           40821795  Ancient Americans                 28   \n",
       "21                           40821871                MXL                128   \n",
       "22                           40821871           SDGP AMR                 44   \n",
       "23                           40821871  Ancient Americans                 38   \n",
       "24                           40826201                MXL                128   \n",
       "25                           40826201           SDGP AMR                 44   \n",
       "26                           40826201  Ancient Americans                 30   \n",
       "\n",
       "    Denisovan-specific Mis. Mut. Freq.  \\\n",
       "0                             0.304688   \n",
       "1                             0.363636   \n",
       "2                             0.361111   \n",
       "3                             0.304688   \n",
       "4                             0.363636   \n",
       "5                             0.375000   \n",
       "6                             0.304688   \n",
       "7                             0.363636   \n",
       "8                             0.375000   \n",
       "9                             0.304688   \n",
       "10                            0.363636   \n",
       "11                            0.305556   \n",
       "12                            0.304688   \n",
       "13                            0.363636   \n",
       "14                            0.406250   \n",
       "15                            0.304688   \n",
       "16                            0.363636   \n",
       "17                            0.382353   \n",
       "18                            0.304688   \n",
       "19                            0.363636   \n",
       "20                            0.321429   \n",
       "21                            0.304688   \n",
       "22                            0.363636   \n",
       "23                            0.315789   \n",
       "24                            0.304688   \n",
       "25                            0.363636   \n",
       "26                            0.366667   \n",
       "\n",
       "    Denisovan-specific Mis. Mut. $\\left( \\sigma \\right)$  \\\n",
       "0                                            0.301263      \n",
       "1                                            0.308288      \n",
       "2                                            0.401579      \n",
       "3                                            0.301263      \n",
       "4                                            0.308288      \n",
       "5                                            0.414578      \n",
       "6                                            0.301263      \n",
       "7                                            0.308288      \n",
       "8                                            0.414578      \n",
       "9                                            0.301263      \n",
       "10                                           0.308288      \n",
       "11                                           0.412946      \n",
       "12                                           0.301263      \n",
       "13                                           0.308288      \n",
       "14                                           0.440835      \n",
       "15                                           0.301263      \n",
       "16                                           0.308288      \n",
       "17                                           0.438225      \n",
       "18                                           0.301263      \n",
       "19                                           0.308288      \n",
       "20                                           0.405636      \n",
       "21                                           0.301263      \n",
       "22                                           0.308288      \n",
       "23                                           0.404271      \n",
       "24                                           0.301263      \n",
       "25                                           0.308288      \n",
       "26                                           0.385861      \n",
       "\n",
       "    Denisovan-specific Mis. Mut. $\\left( SEM \\right)$  \\\n",
       "0                                            0.037956   \n",
       "1                                            0.067274   \n",
       "2                                            0.097397   \n",
       "3                                            0.037956   \n",
       "4                                            0.067274   \n",
       "5                                            0.095111   \n",
       "6                                            0.037956   \n",
       "7                                            0.067274   \n",
       "8                                            0.107044   \n",
       "9                                            0.037956   \n",
       "10                                           0.067274   \n",
       "11                                           0.100154   \n",
       "12                                           0.037956   \n",
       "13                                           0.067274   \n",
       "14                                           0.113823   \n",
       "15                                           0.037956   \n",
       "16                                           0.067274   \n",
       "17                                           0.109556   \n",
       "18                                           0.037956   \n",
       "19                                           0.067274   \n",
       "20                                           0.112503   \n",
       "21                                           0.037956   \n",
       "22                                           0.067274   \n",
       "23                                           0.095288   \n",
       "24                                           0.037956   \n",
       "25                                           0.067274   \n",
       "26                                           0.103126   \n",
       "\n",
       "    Denisovan-specific Mis. Mut. $\\left( \\pm CI_{95\\%} \\right)$  \n",
       "0                                            0.075848            \n",
       "1                                            0.139904            \n",
       "2                                            0.205490            \n",
       "3                                            0.075848            \n",
       "4                                            0.139904            \n",
       "5                                            0.199069            \n",
       "6                                            0.075848            \n",
       "7                                            0.139904            \n",
       "8                                            0.228158            \n",
       "9                                            0.075848            \n",
       "10                                           0.139904            \n",
       "11                                           0.211307            \n",
       "12                                           0.075848            \n",
       "13                                           0.139904            \n",
       "14                                           0.242608            \n",
       "15                                           0.075848            \n",
       "16                                           0.139904            \n",
       "17                                           0.232249            \n",
       "18                                           0.075848            \n",
       "19                                           0.139904            \n",
       "20                                           0.243049            \n",
       "21                                           0.075848            \n",
       "22                                           0.139904            \n",
       "23                                           0.200192            \n",
       "24                                           0.075848            \n",
       "25                                           0.139904            \n",
       "26                                           0.221183            "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ba2f968",
   "metadata": {},
   "source": [
    "## S32"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "82f85999",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/mxl_sgdp_amr_anc_amr_denisovan_specific_missense_mutation_proportions_ztest_fet.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 3, 4, 6, 7, 9, 10, 11, 12]]\n",
    "# Intialize the columns.\n",
    "cols = np.where(\n",
    "    df.columns.values[[0, 1, 2, 3, 4, 6, 7, 9, 10, 11, 12]] == 'Denisovan-specific Mis. Mut. Pos.',\n",
    "    'Position', df.columns.values[[0, 1, 2, 3, 4, 6, 7, 9, 10, 11, 12]]\n",
    ")\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S32. Comparison of the per-individual frequency distributions for each Denisovan-specific missense mutation in the focal 72kb \\textit{MUC19} region between Mexicans in 1000 Genomes Project, 23 ancient Indigenous American individuals, and Indigenous American individuals in the Simons Genome Diversity Project.} \\newline Comparisons of the per-individual frequency distribution for each of the nine Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region between groups: 64 Mexican individuals in the 1000 Genomes project ($n = 128$ chromosomes), 23 ancient Indigenous American individuals ($n = 46$ chromosomes), and 22 Indigenous American individuals in the Simons Genome Diversity Project ($n = 44$ chromosomes). For each comparison, the total number of chromosomes that passed quality control, the number of chromosomes harboring the respective Denisovan-specific missense mutation, and the results of the Proportions \\textit{Z}-Test and Fisher's Exact Test are reported. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/mxl_sgdp_amr_anc_amr_denisovan_specific_missense_mutation_proportions_ztest_fet.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            f'{int(mat[i, j])}' if mat[i, j] > 1 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87816dfc",
   "metadata": {},
   "source": [
    "\\textbf{Table S31. Comparison of the per-individual frequency distributions for each Denisovan-specific missense mutation in the focal 72kb \\textit{MUC19} region between Mexicans in 1000 Genomes Project, 23 ancient Indigenous American individuals, and Indigenous American individuals in the Simons Genome Diversity Project.} \\newline Comparisons of the per-individual frequency distribution for each of the nine Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region between groups: 64 Mexican individuals in the 1000 Genomes project ($n = 128$ chromosomes), 23 ancient Indigenous American individuals ($n = 46$ chromosomes), and 22 Indigenous American individuals in the Simons Genome Diversity Project ($n = 44$ chromosomes). For each comparison, the total number of chromosomes that passed quality control, the number of chromosomes harboring the respective Denisovan-specific missense mutation, and the results of the Proportions \\textit{Z}-Test and Fisher's Exact Test are reported. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "a64ab0bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Denisovan-specific Mis. Mut. Pos.</th>\n",
       "      <th>Group 1</th>\n",
       "      <th>Group 2</th>\n",
       "      <th>Number of Chroms. (Group 1)</th>\n",
       "      <th>Denisovan-specific Mis. Mut. Count (Group 1)</th>\n",
       "      <th>Number of Chroms. (Group 2)</th>\n",
       "      <th>Denisovan-specific Mis. Mut. Count (Group 2)</th>\n",
       "      <th>$Z-Statistic$</th>\n",
       "      <th>$P-value$ (Prop. $Z$-Test)</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>$P-value$ (Fisher's Exact Test)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40808672</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>36</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-0.642730</td>\n",
       "      <td>0.520400</td>\n",
       "      <td>0.775281</td>\n",
       "      <td>0.546776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40808726</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>40</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-0.831137</td>\n",
       "      <td>0.405896</td>\n",
       "      <td>0.730337</td>\n",
       "      <td>0.440712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40809983</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>32</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-0.763439</td>\n",
       "      <td>0.445202</td>\n",
       "      <td>0.730337</td>\n",
       "      <td>0.525291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40814107</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>36</td>\n",
       "      <td>11.0</td>\n",
       "      <td>-0.009995</td>\n",
       "      <td>0.992025</td>\n",
       "      <td>0.995914</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40814197</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>32</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-1.097134</td>\n",
       "      <td>0.272583</td>\n",
       "      <td>0.640449</td>\n",
       "      <td>0.295742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>40815060</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>34</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-0.862247</td>\n",
       "      <td>0.388552</td>\n",
       "      <td>0.707865</td>\n",
       "      <td>0.412878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>40821795</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>28</td>\n",
       "      <td>9.0</td>\n",
       "      <td>-0.173859</td>\n",
       "      <td>0.861976</td>\n",
       "      <td>0.925094</td>\n",
       "      <td>0.825686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>40821871</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>38</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-0.130262</td>\n",
       "      <td>0.896359</td>\n",
       "      <td>0.949438</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>40826201</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>30</td>\n",
       "      <td>11.0</td>\n",
       "      <td>-0.656966</td>\n",
       "      <td>0.511203</td>\n",
       "      <td>0.756895</td>\n",
       "      <td>0.519337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>40808672</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>40808726</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>40809983</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>40814107</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>40814197</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>40815060</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>40821795</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>40821871</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>40826201</td>\n",
       "      <td>MXL</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>128</td>\n",
       "      <td>39.0</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>-0.723264</td>\n",
       "      <td>0.469518</td>\n",
       "      <td>0.766854</td>\n",
       "      <td>0.461491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>40808672</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>36</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.023375</td>\n",
       "      <td>0.981352</td>\n",
       "      <td>1.010989</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>40808726</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>40</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-0.107794</td>\n",
       "      <td>0.914159</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>40809983</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>32</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-0.101397</td>\n",
       "      <td>0.919235</td>\n",
       "      <td>0.952381</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>40814107</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>36</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.546557</td>\n",
       "      <td>0.584683</td>\n",
       "      <td>1.298701</td>\n",
       "      <td>0.640061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>40814197</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>32</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-0.377580</td>\n",
       "      <td>0.705743</td>\n",
       "      <td>0.835165</td>\n",
       "      <td>0.812051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>40815060</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>34</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-0.169607</td>\n",
       "      <td>0.865320</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>40821795</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>28</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.366729</td>\n",
       "      <td>0.713821</td>\n",
       "      <td>1.206349</td>\n",
       "      <td>0.802342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>40821871</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>38</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.455620</td>\n",
       "      <td>0.648663</td>\n",
       "      <td>1.238095</td>\n",
       "      <td>0.815674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>40826201</td>\n",
       "      <td>SDGP AMR</td>\n",
       "      <td>Ancient Americans</td>\n",
       "      <td>44</td>\n",
       "      <td>16.0</td>\n",
       "      <td>30</td>\n",
       "      <td>11.0</td>\n",
       "      <td>-0.026586</td>\n",
       "      <td>0.978790</td>\n",
       "      <td>0.987013</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Denisovan-specific Mis. Mut. Pos.   Group 1            Group 2  \\\n",
       "0                            40808672       MXL  Ancient Americans   \n",
       "1                            40808726       MXL  Ancient Americans   \n",
       "2                            40809983       MXL  Ancient Americans   \n",
       "3                            40814107       MXL  Ancient Americans   \n",
       "4                            40814197       MXL  Ancient Americans   \n",
       "5                            40815060       MXL  Ancient Americans   \n",
       "6                            40821795       MXL  Ancient Americans   \n",
       "7                            40821871       MXL  Ancient Americans   \n",
       "8                            40826201       MXL  Ancient Americans   \n",
       "9                            40808672       MXL           SDGP AMR   \n",
       "10                           40808726       MXL           SDGP AMR   \n",
       "11                           40809983       MXL           SDGP AMR   \n",
       "12                           40814107       MXL           SDGP AMR   \n",
       "13                           40814197       MXL           SDGP AMR   \n",
       "14                           40815060       MXL           SDGP AMR   \n",
       "15                           40821795       MXL           SDGP AMR   \n",
       "16                           40821871       MXL           SDGP AMR   \n",
       "17                           40826201       MXL           SDGP AMR   \n",
       "18                           40808672  SDGP AMR  Ancient Americans   \n",
       "19                           40808726  SDGP AMR  Ancient Americans   \n",
       "20                           40809983  SDGP AMR  Ancient Americans   \n",
       "21                           40814107  SDGP AMR  Ancient Americans   \n",
       "22                           40814197  SDGP AMR  Ancient Americans   \n",
       "23                           40815060  SDGP AMR  Ancient Americans   \n",
       "24                           40821795  SDGP AMR  Ancient Americans   \n",
       "25                           40821871  SDGP AMR  Ancient Americans   \n",
       "26                           40826201  SDGP AMR  Ancient Americans   \n",
       "\n",
       "    Number of Chroms. (Group 1)  Denisovan-specific Mis. Mut. Count (Group 1)  \\\n",
       "0                           128                                          39.0   \n",
       "1                           128                                          39.0   \n",
       "2                           128                                          39.0   \n",
       "3                           128                                          39.0   \n",
       "4                           128                                          39.0   \n",
       "5                           128                                          39.0   \n",
       "6                           128                                          39.0   \n",
       "7                           128                                          39.0   \n",
       "8                           128                                          39.0   \n",
       "9                           128                                          39.0   \n",
       "10                          128                                          39.0   \n",
       "11                          128                                          39.0   \n",
       "12                          128                                          39.0   \n",
       "13                          128                                          39.0   \n",
       "14                          128                                          39.0   \n",
       "15                          128                                          39.0   \n",
       "16                          128                                          39.0   \n",
       "17                          128                                          39.0   \n",
       "18                           44                                          16.0   \n",
       "19                           44                                          16.0   \n",
       "20                           44                                          16.0   \n",
       "21                           44                                          16.0   \n",
       "22                           44                                          16.0   \n",
       "23                           44                                          16.0   \n",
       "24                           44                                          16.0   \n",
       "25                           44                                          16.0   \n",
       "26                           44                                          16.0   \n",
       "\n",
       "    Number of Chroms. (Group 2)  Denisovan-specific Mis. Mut. Count (Group 2)  \\\n",
       "0                            36                                          13.0   \n",
       "1                            40                                          15.0   \n",
       "2                            32                                          12.0   \n",
       "3                            36                                          11.0   \n",
       "4                            32                                          13.0   \n",
       "5                            34                                          13.0   \n",
       "6                            28                                           9.0   \n",
       "7                            38                                          12.0   \n",
       "8                            30                                          11.0   \n",
       "9                            44                                          16.0   \n",
       "10                           44                                          16.0   \n",
       "11                           44                                          16.0   \n",
       "12                           44                                          16.0   \n",
       "13                           44                                          16.0   \n",
       "14                           44                                          16.0   \n",
       "15                           44                                          16.0   \n",
       "16                           44                                          16.0   \n",
       "17                           44                                          16.0   \n",
       "18                           36                                          13.0   \n",
       "19                           40                                          15.0   \n",
       "20                           32                                          12.0   \n",
       "21                           36                                          11.0   \n",
       "22                           32                                          13.0   \n",
       "23                           34                                          13.0   \n",
       "24                           28                                           9.0   \n",
       "25                           38                                          12.0   \n",
       "26                           30                                          11.0   \n",
       "\n",
       "    $Z-Statistic$  $P-value$ (Prop. $Z$-Test)  Odds Ratio  \\\n",
       "0       -0.642730                    0.520400    0.775281   \n",
       "1       -0.831137                    0.405896    0.730337   \n",
       "2       -0.763439                    0.445202    0.730337   \n",
       "3       -0.009995                    0.992025    0.995914   \n",
       "4       -1.097134                    0.272583    0.640449   \n",
       "5       -0.862247                    0.388552    0.707865   \n",
       "6       -0.173859                    0.861976    0.925094   \n",
       "7       -0.130262                    0.896359    0.949438   \n",
       "8       -0.656966                    0.511203    0.756895   \n",
       "9       -0.723264                    0.469518    0.766854   \n",
       "10      -0.723264                    0.469518    0.766854   \n",
       "11      -0.723264                    0.469518    0.766854   \n",
       "12      -0.723264                    0.469518    0.766854   \n",
       "13      -0.723264                    0.469518    0.766854   \n",
       "14      -0.723264                    0.469518    0.766854   \n",
       "15      -0.723264                    0.469518    0.766854   \n",
       "16      -0.723264                    0.469518    0.766854   \n",
       "17      -0.723264                    0.469518    0.766854   \n",
       "18       0.023375                    0.981352    1.010989   \n",
       "19      -0.107794                    0.914159    0.952381   \n",
       "20      -0.101397                    0.919235    0.952381   \n",
       "21       0.546557                    0.584683    1.298701   \n",
       "22      -0.377580                    0.705743    0.835165   \n",
       "23      -0.169607                    0.865320    0.923077   \n",
       "24       0.366729                    0.713821    1.206349   \n",
       "25       0.455620                    0.648663    1.238095   \n",
       "26      -0.026586                    0.978790    0.987013   \n",
       "\n",
       "    $P-value$ (Fisher's Exact Test)  \n",
       "0                          0.546776  \n",
       "1                          0.440712  \n",
       "2                          0.525291  \n",
       "3                          1.000000  \n",
       "4                          0.295742  \n",
       "5                          0.412878  \n",
       "6                          0.825686  \n",
       "7                          1.000000  \n",
       "8                          0.519337  \n",
       "9                          0.461491  \n",
       "10                         0.461491  \n",
       "11                         0.461491  \n",
       "12                         0.461491  \n",
       "13                         0.461491  \n",
       "14                         0.461491  \n",
       "15                         0.461491  \n",
       "16                         0.461491  \n",
       "17                         0.461491  \n",
       "18                         1.000000  \n",
       "19                         1.000000  \n",
       "20                         1.000000  \n",
       "21                         0.640061  \n",
       "22                         0.812051  \n",
       "23                         1.000000  \n",
       "24                         0.802342  \n",
       "25                         0.815674  \n",
       "26                         1.000000  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 2, 3, 4, 6, 7, 9, 10, 11, 12]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fc256ae",
   "metadata": {},
   "source": [
    "## S34"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "95909e46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/sgdp_denisovan_specific_missense_mutation_summary.csv.gz')\n",
    "# Intialize the african columns.\n",
    "afr_cols = [\n",
    "    'AFR', 'Khomani_San', 'Ju_hoan_North', 'Dinka', 'Mandenka',\n",
    "    'Mbuti', 'Yoruba', 'BantuHerero', 'BantuKenya', 'BantuTswana', 'Biaka', 'Esan',\n",
    "    'Gambian', 'Luhya', 'Luo', 'Masai', 'Mende', 'Mozabite', 'Saharawi', 'Somali'\n",
    "]\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df[afr_cols].to_numpy().T\n",
    "idx = [\n",
    "    'AFR', 'Khomani San', r\"Ju$|$'hoan North\", 'Dinka', 'Mandenka',\n",
    "    'Mbuti', 'Yoruba', 'Bantu Herero', 'Bantu Kenya', 'Bantu Tswana', 'Biaka', 'Esan',\n",
    "    'Gambian', 'Luhya', 'Luo', 'Masai', 'Mende', 'Mozabite', 'Saharawi', 'Somali'\n",
    "]\n",
    "# Intialize the columns.\n",
    "cols = ['Group']\n",
    "cols.extend(f'{pos}' for pos in df['Chr12 Position (Hg19)'].values)\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S33. Frequency of the Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region among the African super population and African populations in the Simons Genome Diversity Project.} \\newline Frequencies of the nine Denisovan-specific missense mutations found within the focal 72kb \\textit{MUC19} region for the entire African super population and stratified by African population in the Simons Genome Diversity Project.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/sgdp_denisovan_specific_missense_mutation_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        f'{idx[i]} & '\\\n",
    "        + r' & '.join(f'{round(mat[i, j], 3)}' for j in range(mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a65989a",
   "metadata": {},
   "source": [
    "\\textbf{Table S32. Frequency of the Denisovan-specific missense mutations in the focal 72kb \\textit{MUC19} region among the African super population and African populations in the Simons Genome Diversity Project.} \\newline Frequencies of the nine Denisovan-specific missense mutations found within the focal 72kb \\textit{MUC19} region for the entire African super population and stratified by African population in the Simons Genome Diversity Project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "c2fa6cde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AFR</th>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Khomani_San</th>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ju_hoan_North</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dinka</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mandenka</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mbuti</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Yoruba</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BantuHerero</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BantuKenya</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BantuTswana</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Biaka</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Esan</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gambian</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Luhya</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Luo</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Masai</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mende</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mozabite</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Saharawi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Somali</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      0         1         2         3         4         5  \\\n",
       "AFR            0.011364  0.011364  0.011364  0.011364  0.011364  0.011364   \n",
       "Khomani_San    0.250000  0.250000  0.250000  0.250000  0.250000  0.250000   \n",
       "Ju_hoan_North  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Dinka          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Mandenka       0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Mbuti          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Yoruba         0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "BantuHerero    0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "BantuKenya     0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "BantuTswana    0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Biaka          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Esan           0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Gambian        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Luhya          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Luo            0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Masai          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Mende          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Mozabite       0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Saharawi       0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Somali         0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "                      6         7         8  \n",
       "AFR            0.011364  0.011364  0.011364  \n",
       "Khomani_San    0.250000  0.250000  0.250000  \n",
       "Ju_hoan_North  0.000000  0.000000  0.000000  \n",
       "Dinka          0.000000  0.000000  0.000000  \n",
       "Mandenka       0.000000  0.000000  0.000000  \n",
       "Mbuti          0.000000  0.000000  0.000000  \n",
       "Yoruba         0.000000  0.000000  0.000000  \n",
       "BantuHerero    0.000000  0.000000  0.000000  \n",
       "BantuKenya     0.000000  0.000000  0.000000  \n",
       "BantuTswana    0.000000  0.000000  0.000000  \n",
       "Biaka          0.000000  0.000000  0.000000  \n",
       "Esan           0.000000  0.000000  0.000000  \n",
       "Gambian        0.000000  0.000000  0.000000  \n",
       "Luhya          0.000000  0.000000  0.000000  \n",
       "Luo            0.000000  0.000000  0.000000  \n",
       "Masai          0.000000  0.000000  0.000000  \n",
       "Mende          0.000000  0.000000  0.000000  \n",
       "Mozabite       0.000000  0.000000  0.000000  \n",
       "Saharawi       0.000000  0.000000  0.000000  \n",
       "Somali         0.000000  0.000000  0.000000  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[afr_cols].T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2134574",
   "metadata": {},
   "source": [
    "## S34"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "7574b7e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_muc19_introgressed_hap_frequency_per_super_population_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXX},\n",
    "cells={halign=c, valign=m},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S34. Frequency of introgressed haplotypes at the focal 72kb \\textit{MUC19} region among 1000 Genomes Project super populations.} \\newline The frequency of introgressed haplotypes---i.e., the number of introgressed haplotypes normalized by the total number of chromosomes---stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Introgressed haplotypes at the focal 72kb \\textit{MUC19} region are significantly enriched in AMR individuals (Fisher's Exact Test, Odds Ratio: 2.848, \\textit{P-value}: $1.320e^{-14}$), with AMR populations exhibiting a higher proportion of introgressed tracts compared to non-AMR populations, excluding AFR (Proportions \\textit{Z}-Test, \\textit{Z}-statistic: 8.328, \\textit{P-value}: $4.116e^{-17}$).},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_muc19_introgressed_hap_frequency_per_super_population_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        f'{mat[i, 0]}'+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc495380",
   "metadata": {},
   "source": [
    "\\textbf{Table S33. Frequency of introgressed haplotypes at the focal 72kb \\textit{MUC19} region among 1000 Genomes Project super populations.} \\newline The frequency of introgressed haplotypes---i.e., the number of introgressed haplotypes normalized by the total number of chromosomes---stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Introgressed haplotypes at the focal 72kb \\textit{MUC19} region are significantly enriched in AMR individuals (Fisher's Exact Test, Odds Ratio: 2.848, \\textit{P-value}: $1.320e^{-14}$), with AMR populations exhibiting a higher proportion of introgressed tracts compared to non-AMR populations, excluding AFR (Proportions \\textit{Z}-Test, \\textit{Z}-statistic: 8.328, \\textit{P-value}: $4.116e^{-17}$)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "49a27278",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Total Number of Chromosomes</th>\n",
       "      <th>Number of Introgressed Haps.</th>\n",
       "      <th>Introgressed Hap. Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>694</td>\n",
       "      <td>107</td>\n",
       "      <td>0.154179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAS</td>\n",
       "      <td>978</td>\n",
       "      <td>96</td>\n",
       "      <td>0.098160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EAS</td>\n",
       "      <td>1008</td>\n",
       "      <td>61</td>\n",
       "      <td>0.060516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EUR</td>\n",
       "      <td>1006</td>\n",
       "      <td>23</td>\n",
       "      <td>0.022863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFR</td>\n",
       "      <td>1008</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Super Population  Total Number of Chromosomes  Number of Introgressed Haps.  \\\n",
       "0              AMR                          694                           107   \n",
       "1              SAS                          978                            96   \n",
       "2              EAS                         1008                            61   \n",
       "3              EUR                         1006                            23   \n",
       "4              AFR                         1008                             0   \n",
       "\n",
       "   Introgressed Hap. Frequency  \n",
       "0                     0.154179  \n",
       "1                     0.098160  \n",
       "2                     0.060516  \n",
       "3                     0.022863  \n",
       "4                     0.000000  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9da8365c",
   "metadata": {},
   "source": [
    "## S35"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "b20aa0ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_muc19_introgressed_hap_frequency_per_population_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S35. Frequency of introgressed haplotypes at the focal 72kb \\textit{MUC19} region among 1000 Genomes Project populations.} \\newline The frequency of introgressed haplotypes---i.e., the number of introgressed haplotypes normalized by the total number of chromosomes---for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR).},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_muc19_introgressed_hap_frequency_per_population_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1602aae3",
   "metadata": {},
   "source": [
    "\\textbf{Table S35. Frequency of introgressed haplotypes at the focal 72kb \\textit{MUC19} region among 1000 Genomes Project populations.} \\newline The frequency of introgressed haplotypes---i.e., the number of introgressed haplotypes normalized by the total number of chromosomes---for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "75311518",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Total Number of Chromosomes</th>\n",
       "      <th>Number of Introgressed Haps.</th>\n",
       "      <th>Introgressed Hap. Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>39</td>\n",
       "      <td>0.304688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>170</td>\n",
       "      <td>37</td>\n",
       "      <td>0.217647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>188</td>\n",
       "      <td>13</td>\n",
       "      <td>0.069149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>208</td>\n",
       "      <td>18</td>\n",
       "      <td>0.086538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>172</td>\n",
       "      <td>24</td>\n",
       "      <td>0.139535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>204</td>\n",
       "      <td>23</td>\n",
       "      <td>0.112745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>204</td>\n",
       "      <td>21</td>\n",
       "      <td>0.102941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>192</td>\n",
       "      <td>10</td>\n",
       "      <td>0.052083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>206</td>\n",
       "      <td>18</td>\n",
       "      <td>0.087379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>206</td>\n",
       "      <td>7</td>\n",
       "      <td>0.033981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>198</td>\n",
       "      <td>17</td>\n",
       "      <td>0.085859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>210</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>208</td>\n",
       "      <td>6</td>\n",
       "      <td>0.028846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>186</td>\n",
       "      <td>23</td>\n",
       "      <td>0.123656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>214</td>\n",
       "      <td>11</td>\n",
       "      <td>0.051402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>198</td>\n",
       "      <td>3</td>\n",
       "      <td>0.015152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>214</td>\n",
       "      <td>7</td>\n",
       "      <td>0.032710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>182</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>198</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AFR</td>\n",
       "      <td>LWK</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AFR</td>\n",
       "      <td>GWD</td>\n",
       "      <td>226</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AFR</td>\n",
       "      <td>MSL</td>\n",
       "      <td>170</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AFR</td>\n",
       "      <td>ESN</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AFR</td>\n",
       "      <td>YRI</td>\n",
       "      <td>216</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Total Number of Chromosomes  \\\n",
       "0               AMR        MXL                          128   \n",
       "1               AMR        PEL                          170   \n",
       "2               AMR        CLM                          188   \n",
       "3               AMR        PUR                          208   \n",
       "4               SAS        BEB                          172   \n",
       "5               SAS        STU                          204   \n",
       "6               SAS        ITU                          204   \n",
       "7               SAS        PJL                          192   \n",
       "8               SAS        GIH                          206   \n",
       "9               EAS        CHB                          206   \n",
       "10              EAS        KHV                          198   \n",
       "11              EAS        CHS                          210   \n",
       "12              EAS        JPT                          208   \n",
       "13              EAS        CDX                          186   \n",
       "14              EUR        TSI                          214   \n",
       "15              EUR        CEU                          198   \n",
       "16              EUR        IBS                          214   \n",
       "17              EUR        GBR                          182   \n",
       "18              EUR        FIN                          198   \n",
       "19              AFR        LWK                          198   \n",
       "20              AFR        GWD                          226   \n",
       "21              AFR        MSL                          170   \n",
       "22              AFR        ESN                          198   \n",
       "23              AFR        YRI                          216   \n",
       "\n",
       "    Number of Introgressed Haps.  Introgressed Hap. Frequency  \n",
       "0                             39                     0.304688  \n",
       "1                             37                     0.217647  \n",
       "2                             13                     0.069149  \n",
       "3                             18                     0.086538  \n",
       "4                             24                     0.139535  \n",
       "5                             23                     0.112745  \n",
       "6                             21                     0.102941  \n",
       "7                             10                     0.052083  \n",
       "8                             18                     0.087379  \n",
       "9                              7                     0.033981  \n",
       "10                            17                     0.085859  \n",
       "11                             8                     0.038095  \n",
       "12                             6                     0.028846  \n",
       "13                            23                     0.123656  \n",
       "14                            11                     0.051402  \n",
       "15                             3                     0.015152  \n",
       "16                             7                     0.032710  \n",
       "17                             1                     0.005495  \n",
       "18                             1                     0.005051  \n",
       "19                             0                     0.000000  \n",
       "20                             0                     0.000000  \n",
       "21                             0                     0.000000  \n",
       "22                             0                     0.000000  \n",
       "23                             0                     0.000000  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22cfc929",
   "metadata": {},
   "source": [
    "## S36"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "90a6253b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/altai_nean_afr_v_denisovan_seq_div_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S36. Sequence divergence from the Denisovan for the Altai Neanderthal and all African Individuals in the 1000 Genomes Project (1KG) at the focal 72kb \\textit{MUC19} region.} \\newline The observed sequence divergence from the Densisovan for the Altai Neanderthal (i.e., the number of pairwise differences between the Altai Neanderthal's and Denisovan's unphased genotypes normalized by the effective sequence length) and all Africans in the 1KG (i.e., the average number of pairwise differences between all African haplotypes in the 1KG and the Denisovan's unphased genotypes normalized by the effective sequence length). For each comparison to the Denisovan, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the sequence divergence genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the sequence divergence is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant. Note that at the focal 72kb region, the effective sequence length with respect to the Denisovan is 46136bp and 48435bp for the Altai Neanderthal and African comparison, respectively.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/altai_nean_afr_v_denisovan_seq_div_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(1, 2))+r' & '\\\n",
    "        + r' & '.join('{:.6f}'.format(round(mat[i, j], 6)).rstrip('0').rstrip('.') for j in range(2, 3))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a81e145",
   "metadata": {},
   "source": [
    "\\textbf{Table S36. Sequence divergence from the Denisovan for the Altai Neanderthal and all African Individuals in the 1000 Genomes Project (1KG) at the focal 72kb \\textit{MUC19} region.} \\newline The observed sequence divergence from the Densisovan for the Altai Neanderthal (i.e., the number of pairwise differences between the Altai Neanderthal's and Denisovan's unphased genotypes normalized by the effective sequence length) and all Africans in the 1KG (i.e., the average number of pairwise differences between all African haplotypes in the 1KG and the Denisovan's unphased genotypes normalized by the effective sequence length). For each comparison to the Denisovan, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the sequence divergence genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the sequence divergence is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant. Note that at the focal 72kb region, the effective sequence length with respect to the Denisovan is 46136bp and 48435bp for the Altai Neanderthal and African comparison, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "bdd5836f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Comparison</th>\n",
       "      <th>Focal 72kb Region (Pairwise Diffs.)</th>\n",
       "      <th>Focal 72kb Region (Seq. Div.)</th>\n",
       "      <th>72kb Nonoverlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Nonoverlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Nonoverlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Nonoverlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>174.500000</td>\n",
       "      <td>0.003782</td>\n",
       "      <td>0.001076</td>\n",
       "      <td>0.000601</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.002305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFR</td>\n",
       "      <td>194.884921</td>\n",
       "      <td>0.004024</td>\n",
       "      <td>0.001416</td>\n",
       "      <td>0.000553</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.001905</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Comparison  Focal 72kb Region (Pairwise Diffs.)  \\\n",
       "0  Altai Nean.                           174.500000   \n",
       "1          AFR                           194.884921   \n",
       "\n",
       "   Focal 72kb Region (Seq. Div.)  \\\n",
       "0                       0.003782   \n",
       "1                       0.004024   \n",
       "\n",
       "   72kb Nonoverlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                          0.001076   \n",
       "1                                          0.001416   \n",
       "\n",
       "   72kb Nonoverlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.000601     \n",
       "1                                           0.000553     \n",
       "\n",
       "   72kb Nonoverlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                          0.000004   \n",
       "1                                          0.000003   \n",
       "\n",
       "   72kb Nonoverlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                           0.000007            0.002305  \n",
       "1                                           0.000006            0.001905  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45ae892f",
   "metadata": {},
   "source": [
    "## S37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "5f0bf484",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/focal_mxl_sequence_divergence_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={p{2cm}XXXXp{2cm}p{2cm}p{2cm}p{2cm}p{1cm}p{1cm}X},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S37. Sequence divergence between the \\textit{Denisovan-like} haplotype in MXL and unphased archaic individuals at the focal 72kb \\textit{MUC19} region.} \\newline The observed sequence divergence---i.e., the number of pairwise differences between a modern human haplotype and an archaic genotype normalized by the effective sequence length---between the \\textit{Denisovan-like} haplotype in the focal MXL individual (i.e., NA19664), who harbors two \\textit{Denisovan-like} haplotypes and has no heterozygous sites at the focal 72kb region, and the four unphased high-coverage archaic individuals. For each archaic individual, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the sequence divergence genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the sequence divergence is less than or equal to that observed at the focal 72kb region. After correcting for two multiple comparisons---i.e., one per haplotype---a \\textit{P-value} less than 0.025 is considered statistically significant. Note that at the focal 72kb region, the effective sequence length with respect to the NA19664 individual is 48435bp, 48654bp, 48121bp, and 48447bp for the Denisovan, Altai Neanderthal, Chagyrskaya Neanderthal, and Vindija Neanderthal, respectively.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/focal_mxl_sequence_divergence_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join('{:.6f}'.format(round(mat[i, j], 6)).rstrip('0').rstrip('.') for j in range(1, 5))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(5, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f66da78",
   "metadata": {},
   "source": [
    "\\textbf{Table S37. Sequence divergence between the \\textit{Denisovan-like} haplotype in MXL and unphased archaic individuals at the focal 72kb \\textit{MUC19} region.} \\newline The observed sequence divergence---i.e., the number of pairwise differences between a modern human haplotype and an archaic genotype normalized by the effective sequence length---between the \\textit{Denisovan-like} haplotype in the focal MXL individual (i.e., NA19664), who harbors two \\textit{Denisovan-like} haplotypes and has no heterozygous sites at the focal 72kb region, and the four unphased high-coverage archaic individuals. For each archaic individual, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the sequence divergence genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the sequence divergence is less than or equal to that observed at the focal 72kb region. After correcting for two multiple comparisons---i.e., one per haplotype---a \\textit{P-value} less than 0.025 is considered statistically significant. Note that at the focal 72kb region, the effective sequence length with respect to the NA19664 individual is 48435bp, 48654bp, 48121bp, and 48447bp for the Denisovan, Altai Neanderthal, Chagyrskaya Neanderthal, and Vindija Neanderthal, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "47da8d8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Individual</th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Archaic</th>\n",
       "      <th>Focal 72kb Region (Pairwise Diffs. Hap. 1)</th>\n",
       "      <th>Focal 72kb Region (Pairwise Diffs. Hap. 2)</th>\n",
       "      <th>Focal 72kb Region (Seq. Div. Hap. 1)</th>\n",
       "      <th>Focal 72kb Region (Seq. Div. Hap. 2)</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$ (Hap. 1)</th>\n",
       "      <th>$P-value$ (Hap. 2)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA19664</td>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Denisovan</td>\n",
       "      <td>47.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>0.000970</td>\n",
       "      <td>0.000970</td>\n",
       "      <td>0.001413</td>\n",
       "      <td>0.000613</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.236521</td>\n",
       "      <td>0.236521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA19664</td>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>177.5</td>\n",
       "      <td>177.5</td>\n",
       "      <td>0.003648</td>\n",
       "      <td>0.003648</td>\n",
       "      <td>0.001357</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.994895</td>\n",
       "      <td>0.994895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA19664</td>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>87.5</td>\n",
       "      <td>87.5</td>\n",
       "      <td>0.001818</td>\n",
       "      <td>0.001818</td>\n",
       "      <td>0.001371</td>\n",
       "      <td>0.000607</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.810828</td>\n",
       "      <td>0.810828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA19664</td>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>88.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>0.001816</td>\n",
       "      <td>0.001816</td>\n",
       "      <td>0.001381</td>\n",
       "      <td>0.000604</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.806098</td>\n",
       "      <td>0.806098</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Individual Super Population Population            Archaic  \\\n",
       "0    NA19664              AMR        MXL          Denisovan   \n",
       "1    NA19664              AMR        MXL        Altai Nean.   \n",
       "2    NA19664              AMR        MXL  Chagyrskaya Nean.   \n",
       "3    NA19664              AMR        MXL      Vindija Nean.   \n",
       "\n",
       "   Focal 72kb Region (Pairwise Diffs. Hap. 1)  \\\n",
       "0                                        47.0   \n",
       "1                                       177.5   \n",
       "2                                        87.5   \n",
       "3                                        88.0   \n",
       "\n",
       "   Focal 72kb Region (Pairwise Diffs. Hap. 2)  \\\n",
       "0                                        47.0   \n",
       "1                                       177.5   \n",
       "2                                        87.5   \n",
       "3                                        88.0   \n",
       "\n",
       "   Focal 72kb Region (Seq. Div. Hap. 1)  Focal 72kb Region (Seq. Div. Hap. 2)  \\\n",
       "0                              0.000970                              0.000970   \n",
       "1                              0.003648                              0.003648   \n",
       "2                              0.001818                              0.001818   \n",
       "3                              0.001816                              0.001816   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           0.001413   \n",
       "1                                           0.001357   \n",
       "2                                           0.001371   \n",
       "3                                           0.001381   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.000613      \n",
       "1                                           0.000603      \n",
       "2                                           0.000607      \n",
       "3                                           0.000604      \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.000002   \n",
       "1                                           0.000002   \n",
       "2                                           0.000002   \n",
       "3                                           0.000002   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                           0.000005             \n",
       "1                                           0.000005             \n",
       "2                                           0.000005             \n",
       "3                                           0.000005             \n",
       "\n",
       "   $P-value$ (Hap. 1)  $P-value$ (Hap. 2)  \n",
       "0            0.236521            0.236521  \n",
       "1            0.994895            0.994895  \n",
       "2            0.810828            0.810828  \n",
       "3            0.806098            0.806098  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "809a6b8b",
   "metadata": {},
   "source": [
    "## S38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "87f24f4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/dplus_yri_mxl_arcs_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 9, 10, 11, 12, 13, 14]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 9, 10, 11, 12, 13, 14]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S38. \\textit{D+} tests for introgression at the focal 72kb \\textit{MUC19}.} \\newline \\textit{D+} results for the focal 72kb \\textit{MUC19} region using the Yoruba in Ibadan, Nigeria population (YRI) as \\textit{P1}, the introgressed haplotype in the focal MXL individual (i.e., NA19664) as \\textit{P2}, the four high-coverage archaics as \\textit{P3}, and the EPO ancestral sequence to polarize ancestral states. For each archaic(\\textit{P3}), the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the \\textit{D+} genomic background distribution used to compute the \\textit{P-value} are also reported. \\textit{P-values} were computed by building a \\textit{Z}-distribution of \\textit{D+} values from the non-overlapping 72kb windows of comparable effective sequence length. A \\textit{P-value} less than 0.05 is considered statistically significant and suggests that introgressed haplotype in MXL (\\textit{P2}) shares more derived and ancestral alleles with the given archaic individual (\\textit{P3}) than expected under a model of no gene flow.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/dplus_yri_mxl_arcs_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a25ccf3",
   "metadata": {},
   "source": [
    "\\textbf{Table S38. \\textit{D+} tests for introgression at the focal 72kb \\textit{MUC19}.} \\newline \\textit{D+} results for the focal 72kb \\textit{MUC19} region using the Yoruba in Ibadan, Nigeria population (YRI) as \\textit{P1}, the introgressed haplotype in the focal MXL individual (i.e., NA19664) as \\textit{P2}, the four high-coverage archaics as \\textit{P3}, and the EPO ancestral sequence to polarize ancestral states. For each archaic(\\textit{P3}), the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the \\textit{D+} genomic background distribution used to compute the \\textit{P-value} are also reported. \\textit{P-values} were computed by building a \\textit{Z}-distribution of \\textit{D+} values from the non-overlapping 72kb windows of comparable effective sequence length. A \\textit{P-value} less than 0.05 is considered statistically significant and suggests that introgressed haplotype in MXL (\\textit{P2}) shares more derived and ancestral alleles with the given archaic individual (\\textit{P3}) than expected under a model of no gene flow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "a13f49d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$P1$</th>\n",
       "      <th>$P2$</th>\n",
       "      <th>$P3$</th>\n",
       "      <th>Focal 72kb Region $\\left( D+ \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NA19664</td>\n",
       "      <td>Denisovan</td>\n",
       "      <td>0.742784</td>\n",
       "      <td>0.006973</td>\n",
       "      <td>0.177212</td>\n",
       "      <td>0.001030</td>\n",
       "      <td>0.291487</td>\n",
       "      <td>0.000014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NA19664</td>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>-0.621697</td>\n",
       "      <td>0.019682</td>\n",
       "      <td>0.191165</td>\n",
       "      <td>0.001111</td>\n",
       "      <td>0.314438</td>\n",
       "      <td>0.999427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NA19664</td>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>0.174662</td>\n",
       "      <td>0.020329</td>\n",
       "      <td>0.193022</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.317493</td>\n",
       "      <td>0.182764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NA19664</td>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>0.181960</td>\n",
       "      <td>0.021133</td>\n",
       "      <td>0.193560</td>\n",
       "      <td>0.001125</td>\n",
       "      <td>0.318379</td>\n",
       "      <td>0.173591</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  $P1$     $P2$               $P3$  Focal 72kb Region $\\left( D+ \\right)$  \\\n",
       "0  YRI  NA19664          Denisovan                               0.742784   \n",
       "1  YRI  NA19664        Altai Nean.                              -0.621697   \n",
       "2  YRI  NA19664  Chagyrskaya Nean.                               0.174662   \n",
       "3  YRI  NA19664      Vindija Nean.                               0.181960   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           0.006973   \n",
       "1                                           0.019682   \n",
       "2                                           0.020329   \n",
       "3                                           0.021133   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.177212      \n",
       "1                                           0.191165      \n",
       "2                                           0.193022      \n",
       "3                                           0.193560      \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.001030   \n",
       "1                                           0.001111   \n",
       "2                                           0.001124   \n",
       "3                                           0.001125   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                           0.291487             0.000014  \n",
       "1                                           0.314438             0.999427  \n",
       "2                                           0.317493             0.182764  \n",
       "3                                           0.318379             0.173591  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 2, 9, 10, 11, 12, 13, 14]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ea56ab2",
   "metadata": {},
   "source": [
    "## S39"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "aa222ad4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/focal_mxl_sequence_divergence_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={p{2cm}XXXXp{2cm}p{2cm}p{2cm}p{2cm}p{1cm}p{1cm}X},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S39. Sequence divergence between the longest introgressed tract in MXL and unphased archaic individuals at the focal 742kb \\textit{MUC19} region.} \\newline The observed sequence divergence---i.e., the number of pairwise differences between a modern human haplotype and an archaic genotype normalized by the effective sequence length---between the MXL individual (i.e., NA19725) harboring the longest introgressed tract in MXL (i.e., 742kb) and the four unphased high-coverage archaic individuals. For each archaic individual, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the sequence divergence genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the sequence divergence is less than or equal to that observed at the focal 742kb region. After correcting for two multiple comparisons---i.e., one per haplotype---a \\textit{P-value} less than 0.025 is considered statistically significant. Note that the 742kb longest introgressed tract in MXL corresponds to NA19725's second haplotype (i.e., Hap. 2). Note that at the focal 742kb region, the effective sequence length with respect to the NA19725 individual is 495788bp, 497642bp, 494606bp, and 499368bp for the Denisovan, Altai Neanderthal, Chagyrskaya Neanderthal, and Vindija Neanderthal, respectively.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/focal_mxl_sequence_divergence_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join('{:.6f}'.format(round(mat[i, j], 6)).rstrip('0').rstrip('.') for j in range(1, 5))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(5, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b18095b",
   "metadata": {},
   "source": [
    "\\textbf{Table S39. Sequence divergence between the longest introgressed tract in MXL and unphased archaic individuals at the focal 742kb \\textit{MUC19} region.} \\newline The observed sequence divergence---i.e., the number of pairwise differences between a modern human haplotype and an archaic genotype normalized by the effective sequence length---between the MXL individual (i.e., NA19725) harboring the longest introgressed tract in MXL (i.e., 742kb) and the four unphased high-coverage archaic individuals. For each archaic individual, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the sequence divergence genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the sequence divergence is less than or equal to that observed at the focal 742kb region. After correcting for two multiple comparisons---i.e., one per haplotype---a \\textit{P-value} less than 0.025 is considered statistically significant. Note that the 742kb longest introgressed tract in MXL corresponds to NA19725's second haplotype (i.e., Hap. 2). Note that at the focal 742kb region, the effective sequence length with respect to the NA19725 individual is 495788bp, 497642bp, 494606bp, and 499368bp for the Denisovan, Altai Neanderthal, Chagyrskaya Neanderthal, and Vindija Neanderthal, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "25f51104",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Archaic</th>\n",
       "      <th>Focal 742kb Region (Pairwise Diffs. Hap. 1)</th>\n",
       "      <th>Focal 742kb Region (Pairwise Diffs. Hap. 2)</th>\n",
       "      <th>Focal 742kb Region (Seq. Div. Hap. 1)</th>\n",
       "      <th>Focal 742kb Region (Seq. Div. Hap. 2)</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$ (Hap. 1)</th>\n",
       "      <th>$P-value$ (Hap. 2)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>792.5</td>\n",
       "      <td>399.5</td>\n",
       "      <td>0.001598</td>\n",
       "      <td>0.000806</td>\n",
       "      <td>0.001407</td>\n",
       "      <td>0.000346</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.755493</td>\n",
       "      <td>0.018832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>580.5</td>\n",
       "      <td>450.5</td>\n",
       "      <td>0.001167</td>\n",
       "      <td>0.000905</td>\n",
       "      <td>0.001352</td>\n",
       "      <td>0.000338</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.294210</td>\n",
       "      <td>0.064632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>662.0</td>\n",
       "      <td>327.0</td>\n",
       "      <td>0.001338</td>\n",
       "      <td>0.000661</td>\n",
       "      <td>0.001365</td>\n",
       "      <td>0.000340</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.511927</td>\n",
       "      <td>0.006434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>658.5</td>\n",
       "      <td>327.5</td>\n",
       "      <td>0.001319</td>\n",
       "      <td>0.000656</td>\n",
       "      <td>0.001374</td>\n",
       "      <td>0.000340</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.469669</td>\n",
       "      <td>0.006879</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Archaic  Focal 742kb Region (Pairwise Diffs. Hap. 1)  \\\n",
       "0          Denisovan                                        792.5   \n",
       "1        Altai Nean.                                        580.5   \n",
       "2  Chagyrskaya Nean.                                        662.0   \n",
       "3      Vindija Nean.                                        658.5   \n",
       "\n",
       "   Focal 742kb Region (Pairwise Diffs. Hap. 2)  \\\n",
       "0                                        399.5   \n",
       "1                                        450.5   \n",
       "2                                        327.0   \n",
       "3                                        327.5   \n",
       "\n",
       "   Focal 742kb Region (Seq. Div. Hap. 1)  \\\n",
       "0                               0.001598   \n",
       "1                               0.001167   \n",
       "2                               0.001338   \n",
       "3                               0.001319   \n",
       "\n",
       "   Focal 742kb Region (Seq. Div. Hap. 2)  \\\n",
       "0                               0.000806   \n",
       "1                               0.000905   \n",
       "2                               0.000661   \n",
       "3                               0.000656   \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           0.001407    \n",
       "1                                           0.001352    \n",
       "2                                           0.001365    \n",
       "3                                           0.001374    \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.000346       \n",
       "1                                           0.000338       \n",
       "2                                           0.000340       \n",
       "3                                           0.000340       \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.000004    \n",
       "1                                           0.000004    \n",
       "2                                           0.000004    \n",
       "3                                           0.000004    \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                                           0.000008              \n",
       "1                                           0.000008              \n",
       "2                                           0.000008              \n",
       "3                                           0.000008              \n",
       "\n",
       "   $P-value$ (Hap. 1)  $P-value$ (Hap. 2)  \n",
       "0            0.755493            0.018832  \n",
       "1            0.294210            0.064632  \n",
       "2            0.511927            0.006434  \n",
       "3            0.469669            0.006879  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddc3c705",
   "metadata": {},
   "source": [
    "## S40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "367dfc65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_muc19_introgressed_hap_frequency_per_super_population_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXX},\n",
    "cells={halign=c, valign=m},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S40. Frequency of introgressed haplotypes at the focal 742kb \\textit{MUC19} region among 1000 Genomes Project super populations.} \\newline The frequency of introgressed haplotypes---i.e., the number of introgressed haplotypes normalized by the total number of chromosomes---stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Introgressed haplotypes at the focal 742kb \\textit{MUC19} region are significantly enriched in AMR individuals (Fisher's Exact Test, Odds Ratio: 10.033, \\textit{P-value}: $1.601e^{-37}$), with AMR populations exhibiting a higher proportion of introgressed tracts compared to non-AMR populations, excluding AFR (Proportions \\textit{Z}-Test, \\textit{Z}-statistic: 14.909, \\textit{P-value}: $1.441e^{-50}$).},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_muc19_introgressed_hap_frequency_per_super_population_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        f'{mat[i, 0]}'+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f9d766",
   "metadata": {},
   "source": [
    "\\textbf{Table S40. Frequency of introgressed haplotypes at the focal 742kb \\textit{MUC19} region among 1000 Genomes Project super populations.} \\newline The frequency of introgressed haplotypes---i.e., the number of introgressed haplotypes normalized by the total number of chromosomes---stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR). Introgressed haplotypes at the focal 742kb \\textit{MUC19} region are significantly enriched in AMR individuals (Fisher's Exact Test, Odds Ratio: 10.033, \\textit{P-value}: $1.601e^{-37}$), with AMR populations exhibiting a higher proportion of introgressed tracts compared to non-AMR populations, excluding AFR (Proportions \\textit{Z}-Test, \\textit{Z}-statistic: 14.909, \\textit{P-value}: $1.441e^{-50}$)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "8ff85e74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Total Number of Chromosomes</th>\n",
       "      <th>Number of Introgressed Haps.</th>\n",
       "      <th>Introgressed Hap. Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>694</td>\n",
       "      <td>94</td>\n",
       "      <td>0.135447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAS</td>\n",
       "      <td>978</td>\n",
       "      <td>21</td>\n",
       "      <td>0.021472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EAS</td>\n",
       "      <td>1008</td>\n",
       "      <td>23</td>\n",
       "      <td>0.022817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EUR</td>\n",
       "      <td>1006</td>\n",
       "      <td>2</td>\n",
       "      <td>0.001988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFR</td>\n",
       "      <td>1008</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Super Population  Total Number of Chromosomes  Number of Introgressed Haps.  \\\n",
       "0              AMR                          694                            94   \n",
       "1              SAS                          978                            21   \n",
       "2              EAS                         1008                            23   \n",
       "3              EUR                         1006                             2   \n",
       "4              AFR                         1008                             0   \n",
       "\n",
       "   Introgressed Hap. Frequency  \n",
       "0                     0.135447  \n",
       "1                     0.021472  \n",
       "2                     0.022817  \n",
       "3                     0.001988  \n",
       "4                     0.000000  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a409c357",
   "metadata": {},
   "source": [
    "## S41"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "df8d036a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_muc19_introgressed_hap_frequency_per_population_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S41. Frequency of introgressed haplotypes at the focal 742kb \\textit{MUC19} region among 1000 Genomes Project populations.} \\newline The frequency of introgressed haplotypes---i.e., the number of introgressed haplotypes normalized by the total number of chromosomes--- for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR).},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_muc19_introgressed_hap_frequency_per_population_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fe68164",
   "metadata": {},
   "source": [
    "\\textbf{Table S41. Frequency of introgressed haplotypes at the focal 742kb \\textit{MUC19} region among 1000 Genomes Project populations.} \\newline The frequency of introgressed haplotypes---i.e., the number of introgressed haplotypes normalized by the total number of chromosomes--- for each population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), Europeans (EUR), and Africans (AFR)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "4cb950a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Total Number of Chromosomes</th>\n",
       "      <th>Number of Introgressed Haps.</th>\n",
       "      <th>Introgressed Hap. Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>128</td>\n",
       "      <td>40</td>\n",
       "      <td>0.312500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>170</td>\n",
       "      <td>30</td>\n",
       "      <td>0.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>188</td>\n",
       "      <td>10</td>\n",
       "      <td>0.053191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>208</td>\n",
       "      <td>14</td>\n",
       "      <td>0.067308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>172</td>\n",
       "      <td>6</td>\n",
       "      <td>0.034884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>204</td>\n",
       "      <td>5</td>\n",
       "      <td>0.024510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>204</td>\n",
       "      <td>3</td>\n",
       "      <td>0.014706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>192</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>206</td>\n",
       "      <td>6</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>206</td>\n",
       "      <td>3</td>\n",
       "      <td>0.014563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>198</td>\n",
       "      <td>6</td>\n",
       "      <td>0.030303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>210</td>\n",
       "      <td>2</td>\n",
       "      <td>0.009524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>208</td>\n",
       "      <td>4</td>\n",
       "      <td>0.019231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>186</td>\n",
       "      <td>8</td>\n",
       "      <td>0.043011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>214</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>198</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>214</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>182</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AFR</td>\n",
       "      <td>LWK</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AFR</td>\n",
       "      <td>GWD</td>\n",
       "      <td>226</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AFR</td>\n",
       "      <td>MSL</td>\n",
       "      <td>170</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AFR</td>\n",
       "      <td>ESN</td>\n",
       "      <td>198</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AFR</td>\n",
       "      <td>YRI</td>\n",
       "      <td>216</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Total Number of Chromosomes  \\\n",
       "0               AMR        MXL                          128   \n",
       "1               AMR        PEL                          170   \n",
       "2               AMR        CLM                          188   \n",
       "3               AMR        PUR                          208   \n",
       "4               SAS        BEB                          172   \n",
       "5               SAS        STU                          204   \n",
       "6               SAS        ITU                          204   \n",
       "7               SAS        PJL                          192   \n",
       "8               SAS        GIH                          206   \n",
       "9               EAS        CHB                          206   \n",
       "10              EAS        KHV                          198   \n",
       "11              EAS        CHS                          210   \n",
       "12              EAS        JPT                          208   \n",
       "13              EAS        CDX                          186   \n",
       "14              EUR        TSI                          214   \n",
       "15              EUR        CEU                          198   \n",
       "16              EUR        IBS                          214   \n",
       "17              EUR        GBR                          182   \n",
       "18              EUR        FIN                          198   \n",
       "19              AFR        LWK                          198   \n",
       "20              AFR        GWD                          226   \n",
       "21              AFR        MSL                          170   \n",
       "22              AFR        ESN                          198   \n",
       "23              AFR        YRI                          216   \n",
       "\n",
       "    Number of Introgressed Haps.  Introgressed Hap. Frequency  \n",
       "0                             40                     0.312500  \n",
       "1                             30                     0.176471  \n",
       "2                             10                     0.053191  \n",
       "3                             14                     0.067308  \n",
       "4                              6                     0.034884  \n",
       "5                              5                     0.024510  \n",
       "6                              3                     0.014706  \n",
       "7                              1                     0.005208  \n",
       "8                              6                     0.029126  \n",
       "9                              3                     0.014563  \n",
       "10                             6                     0.030303  \n",
       "11                             2                     0.009524  \n",
       "12                             4                     0.019231  \n",
       "13                             8                     0.043011  \n",
       "14                             1                     0.004673  \n",
       "15                             1                     0.005051  \n",
       "16                             0                     0.000000  \n",
       "17                             0                     0.000000  \n",
       "18                             0                     0.000000  \n",
       "19                             0                     0.000000  \n",
       "20                             0                     0.000000  \n",
       "21                             0                     0.000000  \n",
       "22                             0                     0.000000  \n",
       "23                             0                     0.000000  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2060cc43",
   "metadata": {},
   "source": [
    "## S42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "d10533ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/dplus_yri_mxl_arcs_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 9, 10, 11, 12, 13, 14]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 9, 10, 11, 12, 13, 14]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S42. \\textit{D+} tests for introgression at the focal 742kb \\textit{MUC19}.} \\newline \\textit{D+} results for the focal 742kb \\textit{MUC19} region using the Yoruba in Ibadan, Nigeria population (YRI) as \\textit{P1}, the introgressed haplotype in the focal MXL individual (i.e., NA19664) as \\textit{P2}, the four high-coverage archaics as \\textit{P3}, and the EPO ancestral sequence to polarize ancestral states. For each archaic(\\textit{P3}), the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the \\textit{D+} genomic background distribution used to compute the \\textit{P-value} are also reported. \\textit{P-values} were computed by building a \\textit{Z}-distribution of \\textit{D+} values from the non-overlapping 742kb windows of comparable effective sequence length. A \\textit{P-value} less than 0.05 is considered statistically significant and suggests that introgressed haplotype in MXL (\\textit{P2}) shares more derived and ancestral alleles with the given archaic individual (\\textit{P3}) than expected under a model of no gene flow.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/dplus_yri_mxl_arcs_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "765c424a",
   "metadata": {},
   "source": [
    "\\textbf{Table S42. \\textit{D+} tests for introgression at the focal 742kb \\textit{MUC19}.} \\newline \\textit{D+} results for the focal 742kb \\textit{MUC19} region using the Yoruba in Ibadan, Nigeria population (YRI) as \\textit{P1}, the introgressed haplotype in the focal MXL individual (i.e., NA19664) as \\textit{P2}, the four high-coverage archaics as \\textit{P3}, and the EPO ancestral sequence to polarize ancestral states. For each archaic(\\textit{P3}), the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the \\textit{D+} genomic background distribution used to compute the \\textit{P-value} are also reported. \\textit{P-values} were computed by building a \\textit{Z}-distribution of \\textit{D+} values from the non-overlapping 742kb windows of comparable effective sequence length. A \\textit{P-value} less than 0.05 is considered statistically significant and suggests that introgressed haplotype in MXL (\\textit{P2}) shares more derived and ancestral alleles with the given archaic individual (\\textit{P3}) than expected under a model of no gene flow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "b90a1d16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$P1$</th>\n",
       "      <th>$P2$</th>\n",
       "      <th>$P3$</th>\n",
       "      <th>Focal 742kb Region $\\left( D+ \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NA19664</td>\n",
       "      <td>Denisovan</td>\n",
       "      <td>0.376505</td>\n",
       "      <td>0.002145</td>\n",
       "      <td>0.072385</td>\n",
       "      <td>0.001291</td>\n",
       "      <td>0.119063</td>\n",
       "      <td>9.888840e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NA19664</td>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>0.090846</td>\n",
       "      <td>0.016794</td>\n",
       "      <td>0.085560</td>\n",
       "      <td>0.001526</td>\n",
       "      <td>0.140733</td>\n",
       "      <td>1.441673e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NA19664</td>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>0.381192</td>\n",
       "      <td>0.017834</td>\n",
       "      <td>0.087986</td>\n",
       "      <td>0.001572</td>\n",
       "      <td>0.144725</td>\n",
       "      <td>7.374818e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YRI</td>\n",
       "      <td>NA19664</td>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>0.382582</td>\n",
       "      <td>0.018550</td>\n",
       "      <td>0.088386</td>\n",
       "      <td>0.001575</td>\n",
       "      <td>0.145382</td>\n",
       "      <td>7.505230e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  $P1$     $P2$               $P3$  Focal 742kb Region $\\left( D+ \\right)$  \\\n",
       "0  YRI  NA19664          Denisovan                                0.376505   \n",
       "1  YRI  NA19664        Altai Nean.                                0.090846   \n",
       "2  YRI  NA19664  Chagyrskaya Nean.                                0.381192   \n",
       "3  YRI  NA19664      Vindija Nean.                                0.382582   \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           0.002145    \n",
       "1                                           0.016794    \n",
       "2                                           0.017834    \n",
       "3                                           0.018550    \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.072385       \n",
       "1                                           0.085560       \n",
       "2                                           0.087986       \n",
       "3                                           0.088386       \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.001291    \n",
       "1                                           0.001526    \n",
       "2                                           0.001572    \n",
       "3                                           0.001575    \n",
       "\n",
       "   742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$     $P-value$  \n",
       "0                                           0.119063             9.888840e-08  \n",
       "1                                           0.140733             1.441673e-01  \n",
       "2                                           0.144725             7.374818e-06  \n",
       "3                                           0.145382             7.505230e-06  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 2, 9, 10, 11, 12, 13, 14]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d49dcc4f",
   "metadata": {},
   "source": [
    "## S43"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "767d5fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/archaic_heterozygous_sites_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S43. The number of heterozygous sites in the focal 72kb \\textit{MUC19} region among archaic individuals.} \\newline The number of heterozygous observed in the focal 72kb \\textit{MUC19} region for each of the high-coverage archaic individuals. For each archaic individual, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the heterozygous sites genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the number of heterozygous sites is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/archaic_heterozygous_sites_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e052ec",
   "metadata": {},
   "source": [
    "\\textbf{Table S43. The number of heterozygous sites in the focal 72kb \\textit{MUC19} region among archaic individuals.} \\newline The number of heterozygous observed in the focal 72kb \\textit{MUC19} region for each of the high-coverage archaic individuals. For each archaic individual, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the heterozygous sites genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the number of heterozygous sites is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "f237689d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Archaic</th>\n",
       "      <th>Focal 72kb Region (Het. Sites)</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>6</td>\n",
       "      <td>8.746692</td>\n",
       "      <td>14.168528</td>\n",
       "      <td>0.081193</td>\n",
       "      <td>0.159141</td>\n",
       "      <td>0.454668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>1</td>\n",
       "      <td>7.861556</td>\n",
       "      <td>15.261282</td>\n",
       "      <td>0.087445</td>\n",
       "      <td>0.171395</td>\n",
       "      <td>0.678825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>168</td>\n",
       "      <td>6.534416</td>\n",
       "      <td>13.241708</td>\n",
       "      <td>0.076011</td>\n",
       "      <td>0.148986</td>\n",
       "      <td>0.000231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>171</td>\n",
       "      <td>7.837157</td>\n",
       "      <td>13.597947</td>\n",
       "      <td>0.077908</td>\n",
       "      <td>0.152702</td>\n",
       "      <td>0.000328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Archaic  Focal 72kb Region (Het. Sites)  \\\n",
       "0          Denisovan                               6   \n",
       "1        Altai Nean.                               1   \n",
       "2  Chagyrskaya Nean.                             168   \n",
       "3      Vindija Nean.                             171   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                      8.746692   \n",
       "1                                      7.861556   \n",
       "2                                      6.534416   \n",
       "3                                      7.837157   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                        14.168528   \n",
       "1                                        15.261282   \n",
       "2                                        13.241708   \n",
       "3                                        13.597947   \n",
       "\n",
       "   Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                      0.081193   \n",
       "1                                      0.087445   \n",
       "2                                      0.076011   \n",
       "3                                      0.077908   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                           0.159141        0.454668  \n",
       "1                                           0.171395        0.678825  \n",
       "2                                           0.148986        0.000231  \n",
       "3                                           0.152702        0.000328  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3fd4378",
   "metadata": {},
   "source": [
    "## S44"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "a0c46ff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/afr_het_hom_heterozygous_sites_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S44. The average number of heterozygous sites in the focal 72kb \\textit{MUC19} region among African individuals and individuals carrying a \\textit{Denisovan-like} haplotype in the 1000 Genomes Project.} \\newline The average number of heterozygous observed in the focal 72kb \\textit{MUC19} region for each focal group in the 1000 Genomes Project: all African individuals ($n = 504$), individuals harboring one \\textit{Denisovan-like} haplotype ($n = 255$), and individuals harboring two \\textit{Denisovan-like} haplotypes ($n = 16$). For each group, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the average number of heterozygous sites genomic background distribution used to compute the \\textit{P-value} are also reported. For the African individuals and individuals harboring one \\textit{Denisovan-like} haplotype groups, the \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the average number of heterozygous sites is greater than or equal to that observed at the focal 72kb region. For the individuals harboring two \\textit{Denisovan-like} haplotypes group, the \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the average number of heterozygous sites is less than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/afr_het_hom_heterozygous_sites_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb0cfb12",
   "metadata": {},
   "source": [
    "\\textbf{Table S44. The average number of heterozygous sites in the focal 72kb \\textit{MUC19} region among African individuals and individuals carrying a \\textit{Denisovan-like} haplotype in the 1000 Genomes Project.} \\newline The average number of heterozygous observed in the focal 72kb \\textit{MUC19} region for each focal group in the 1000 Genomes Project: all African individuals ($n = 504$), individuals harboring one \\textit{Denisovan-like} haplotype ($n = 255$), and individuals harboring two \\textit{Denisovan-like} haplotypes ($n = 16$). For each group, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the average number of heterozygous sites genomic background distribution used to compute the \\textit{P-value} are also reported. For the African individuals and individuals harboring one \\textit{Denisovan-like} haplotype groups, the \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the average number of heterozygous sites is greater than or equal to that observed at the focal 72kb region. For the individuals harboring two \\textit{Denisovan-like} haplotypes group, the \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the average number of heterozygous sites is less than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "35021a32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Focal 72kb Region (Average Number of Het. Sites)</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>African Inds.</td>\n",
       "      <td>75.315476</td>\n",
       "      <td>73.318893</td>\n",
       "      <td>25.707996</td>\n",
       "      <td>0.144443</td>\n",
       "      <td>0.283114</td>\n",
       "      <td>0.423606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Inds. with One $Denisovan-like$ Hap.</td>\n",
       "      <td>287.341176</td>\n",
       "      <td>54.639081</td>\n",
       "      <td>23.127764</td>\n",
       "      <td>0.129946</td>\n",
       "      <td>0.254698</td>\n",
       "      <td>0.000032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Inds. with Two $Denisovan-like$ Haps.</td>\n",
       "      <td>4.187500</td>\n",
       "      <td>52.579116</td>\n",
       "      <td>24.890720</td>\n",
       "      <td>0.139851</td>\n",
       "      <td>0.274113</td>\n",
       "      <td>0.000694</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Group  \\\n",
       "0                          African Inds.   \n",
       "1   Inds. with One $Denisovan-like$ Hap.   \n",
       "2  Inds. with Two $Denisovan-like$ Haps.   \n",
       "\n",
       "   Focal 72kb Region (Average Number of Het. Sites)  \\\n",
       "0                                         75.315476   \n",
       "1                                        287.341176   \n",
       "2                                          4.187500   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                     73.318893   \n",
       "1                                     54.639081   \n",
       "2                                     52.579116   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                        25.707996   \n",
       "1                                        23.127764   \n",
       "2                                        24.890720   \n",
       "\n",
       "   Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                      0.144443   \n",
       "1                                      0.129946   \n",
       "2                                      0.139851   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                           0.283114        0.423606  \n",
       "1                                           0.254698        0.000032  \n",
       "2                                           0.274113        0.000694  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f512aec2",
   "metadata": {},
   "source": [
    "## S45"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "15894dbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/afr_archaic_heterozygosity_genome_wide.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S45. Genome-wide heterozygosity among archaic individuals and African individuals in the 1000 Genomes Project.} \\newline The genome-wide heterozygosity (i.e., the number of heterozygous sites normalized by the effective sequence length) for each of the high coverage archaic individuals and the mean genome-wide heterozygosity among all African individuals in the 1000 Genomes Project.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/afr_archaic_heterozygosity_genome_wide.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            '{:.6f}'.format(round(mat[i, j], 6)).rstrip('0').rstrip('.')\n",
    "        for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "197cc6fa",
   "metadata": {},
   "source": [
    "\\textbf{Table S45. Genome-wide heterozygosity among archaic individuals and African individuals in the 1000 Genomes Project.} \\newline The genome-wide heterozygosity (i.e., the number of heterozygous sites normalized by the effective sequence length) for each of the high coverage archaic individuals and the mean genome-wide heterozygosity among all African individuals in the 1000 Genomes Project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "7047fb9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group</th>\n",
       "      <th>Genome-Wide Heterozygosity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFR</td>\n",
       "      <td>0.000985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>0.000191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>0.000171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>0.000144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>0.000170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Group  Genome-Wide Heterozygosity\n",
       "0                AFR                    0.000985\n",
       "1          Denisovan                    0.000191\n",
       "2        Altai Nean.                    0.000171\n",
       "3  Chagyrskaya Nean.                    0.000144\n",
       "4      Vindija Nean.                    0.000170"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "591ddba7",
   "metadata": {},
   "source": [
    "## S46"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "eb33f896",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/dplus_altai_neanderthal_p2_denisovan_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 9, 10, 11, 12, 13, 14]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 9, 10, 11, 12, 13, 14]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S46. \\textit{D+} tests for Denisovan introgression at the focal 72kb \\textit{MUC19} region among the late Neanderthal individuals.} \\newline \\textit{D+} results for the focal 72kb \\textit{MUC19} region using Altai Neanderthal as \\textit{P1}, each of the late Neanderthals as \\textit{P2}, the Denisovan as \\textit{P3}, and the EPO ancestral sequence to polarize ancestral states. For each late Neanderthal (\\textit{P2}), the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the \\textit{D+} genomic background distribution used to compute the \\textit{P-value} are also reported. \\textit{P-values} were computed by building a \\textit{Z}-distribution of \\textit{D+} values from the non-overlapping 72kb windows of comparable effective sequence length. A \\textit{P-value} less than 0.05 is considered statistically significant and suggests that the late Neanderthal (\\textit{P2}) shares more derived and ancestral alleles with the Denisovan (\\textit{P3}) than expected under a model of no gene flow.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/dplus_altai_neanderthal_p2_denisovan_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e561c5c9",
   "metadata": {},
   "source": [
    "\\textbf{Table S46. \\textit{D+} tests for Denisovan introgression at the focal 72kb \\textit{MUC19} region among the late Neanderthal individuals.} \\newline \\textit{D+} results for the focal 72kb \\textit{MUC19} region using Altai Neanderthal as \\textit{P1}, each of the late Neanderthals as \\textit{P2}, the Denisovan as \\textit{P3}, and the EPO ancestral sequence to polarize ancestral states. For each late Neanderthal (\\textit{P2}), the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the \\textit{D+} genomic background distribution used to compute the \\textit{P-value} are also reported. \\textit{P-values} were computed by building a \\textit{Z}-distribution of \\textit{D+} values from the non-overlapping 72kb windows of comparable effective sequence length. A \\textit{P-value} less than 0.05 is considered statistically significant and suggests that the late Neanderthal (\\textit{P2}) shares more derived and ancestral alleles with the Denisovan (\\textit{P3}) than expected under a model of no gene flow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "52b01d7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$P1$</th>\n",
       "      <th>$P2$</th>\n",
       "      <th>$P3$</th>\n",
       "      <th>Focal 72kb Region $\\left( D+ \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>Denisovan</td>\n",
       "      <td>0.783439</td>\n",
       "      <td>-0.113348</td>\n",
       "      <td>0.414336</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>0.681522</td>\n",
       "      <td>0.029323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>Denisovan</td>\n",
       "      <td>0.818750</td>\n",
       "      <td>-0.169137</td>\n",
       "      <td>0.391438</td>\n",
       "      <td>0.002314</td>\n",
       "      <td>0.643858</td>\n",
       "      <td>0.018235</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          $P1$               $P2$       $P3$  \\\n",
       "0  Altai Nean.  Chagyrskaya Nean.  Denisovan   \n",
       "1  Altai Nean.      Vindija Nean.  Denisovan   \n",
       "\n",
       "   Focal 72kb Region $\\left( D+ \\right)$  \\\n",
       "0                               0.783439   \n",
       "1                               0.818750   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                          -0.113348   \n",
       "1                                          -0.169137   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.414336      \n",
       "1                                           0.391438      \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.002454   \n",
       "1                                           0.002314   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                           0.681522             0.029323  \n",
       "1                                           0.643858             0.018235  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 2, 9, 10, 11, 12, 13, 14]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18a78789",
   "metadata": {},
   "source": [
    "## S47"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "eb339475",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/phased_late_neanderthals_archaic_pairwise_diffs_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S47. The number of pairwise differences at the focal 72kb \\textit{MUC19} region among the Denisovan, Altai Neanderthal, and the phased late Neanderthal individuals.} \\newline The number of pairwise differences observed at the focal 72kb \\textit{MUC19} region between all pairwise possibilities of the unphased Denisovan genotypes, unphased Altai Neanderthal genotypes, phased Chagyrskaya Neanderthal haplotypes, and phased Vindija Neanderthal haplotypes.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/phased_late_neanderthals_archaic_pairwise_diffs_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            '{:.6f}'.format(round(mat[i, j], 1)).rstrip('0').rstrip('.')\n",
    "        for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eecea921",
   "metadata": {},
   "source": [
    "\\textbf{Table S47. The number of pairwise differences at the focal 72kb \\textit{MUC19} region among the Denisovan, Altai Neanderthal, and the phased late Neanderthal individuals.} \\newline The number of pairwise differences observed at the focal 72kb \\textit{MUC19} region between all pairwise possibilities of the unphased Denisovan genotypes, unphased Altai Neanderthal genotypes, phased Chagyrskaya Neanderthal haplotypes, and phased Vindija Neanderthal haplotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "360dedb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Archaic 1</th>\n",
       "      <th>Archaic 2</th>\n",
       "      <th>Focal 72kb Region (Pairwise Diffs.)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>173.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>Chagyrskaya Nean. Hap. 1</td>\n",
       "      <td>164.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>Chagyrskaya Nean. Hap. 2</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>Vindija Nean. Hap. 1</td>\n",
       "      <td>170.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>Vindija Nean. Hap. 2</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>Chagyrskaya Nean. Hap. 1</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>Chagyrskaya Nean. Hap. 2</td>\n",
       "      <td>159.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>Vindija Nean. Hap. 1</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>Vindija Nean. Hap. 2</td>\n",
       "      <td>161.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Chagyrskaya Nean. Hap. 1</td>\n",
       "      <td>Chagyrskaya Nean. Hap. 2</td>\n",
       "      <td>165.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Chagyrskaya Nean. Hap. 1</td>\n",
       "      <td>Vindija Nean. Hap. 1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Chagyrskaya Nean. Hap. 1</td>\n",
       "      <td>Vindija Nean. Hap. 2</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Chagyrskaya Nean. Hap. 2</td>\n",
       "      <td>Vindija Nean. Hap. 1</td>\n",
       "      <td>153.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Chagyrskaya Nean. Hap. 2</td>\n",
       "      <td>Vindija Nean. Hap. 2</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Vindija Nean. Hap. 1</td>\n",
       "      <td>Vindija Nean. Hap. 2</td>\n",
       "      <td>167.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Archaic 1                 Archaic 2  \\\n",
       "0                  Denisovan               Altai Nean.   \n",
       "1                  Denisovan  Chagyrskaya Nean. Hap. 1   \n",
       "2                  Denisovan  Chagyrskaya Nean. Hap. 2   \n",
       "3                  Denisovan      Vindija Nean. Hap. 1   \n",
       "4                  Denisovan      Vindija Nean. Hap. 2   \n",
       "5                Altai Nean.  Chagyrskaya Nean. Hap. 1   \n",
       "6                Altai Nean.  Chagyrskaya Nean. Hap. 2   \n",
       "7                Altai Nean.      Vindija Nean. Hap. 1   \n",
       "8                Altai Nean.      Vindija Nean. Hap. 2   \n",
       "9   Chagyrskaya Nean. Hap. 1  Chagyrskaya Nean. Hap. 2   \n",
       "10  Chagyrskaya Nean. Hap. 1      Vindija Nean. Hap. 1   \n",
       "11  Chagyrskaya Nean. Hap. 1      Vindija Nean. Hap. 2   \n",
       "12  Chagyrskaya Nean. Hap. 2      Vindija Nean. Hap. 1   \n",
       "13  Chagyrskaya Nean. Hap. 2      Vindija Nean. Hap. 2   \n",
       "14      Vindija Nean. Hap. 1      Vindija Nean. Hap. 2   \n",
       "\n",
       "    Focal 72kb Region (Pairwise Diffs.)  \n",
       "0                                 173.5  \n",
       "1                                 164.0  \n",
       "2                                  43.0  \n",
       "3                                 170.0  \n",
       "4                                  41.0  \n",
       "5                                   3.5  \n",
       "6                                 159.5  \n",
       "7                                   4.0  \n",
       "8                                 161.0  \n",
       "9                                 165.0  \n",
       "10                                  1.0  \n",
       "11                                154.0  \n",
       "12                                153.0  \n",
       "13                                  0.0  \n",
       "14                                167.0  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21d8dda3",
   "metadata": {},
   "source": [
    "## S48"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "b2705be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/focal_mxl_haplotype_divergence_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={p{2cm}XXXXp{2cm}p{2cm}p{2cm}p{2cm}p{1cm}p{1cm}X},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S48. Sequence divergence between the \\textit{Denisovan-like} haplotype in MXL and phased late Neanderthal haplotypes at the focal 72kb \\textit{MUC19} region.} \\newline The observed sequence divergence---i.e., the number of pairwise differences between a modern human haplotype and a late Neanderthal haplotype normalized by the effective sequence length---between the \\textit{Denisovan-like} haplotype in the focal MXL individual (i.e., NA19664), who harbors two \\textit{Denisovan-like} haplotypes and has no heterozygous sites at the focal 72kb region, and the phased haplotypes for both of the late Neanderthals. To assess the significance, we built a distribution of sequence divergence between each of the modern human's two chromosomes and the late Neanderthal's pseudo-haplotype, which was generated by randomly sampling one allele at each position from the unphased late Neanderthal's genotypes. For each late Neanderthal, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the pseudo-haplotype divergence genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the sequence divergence is less than or equal to that observed at the focal 72kb region. After correcting for four multiple comparisons---i.e., two per each modern human haplotype---a \\textit{P-value} less than 0.0125 is considered statistically significant. Note that the \\textit{Denisovan-like} in each of the late Neanderthal's corresponds to their second haplotype (i.e., Chagyrskaya Nean. Hap. 2 and Vindija Nean. Hap. 2). Also, note that at the phased 72kb region, the effective sequence length with respect to the NA19664 individual is 48119bp and 48444bp for the  Chagyrskaya and Vindija Neanderthal, respectively.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/focal_mxl_haplotype_divergence_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join('{:.6f}'.format(round(mat[i, j], 6)).rstrip('0').rstrip('.') for j in range(1, 5))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(5, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7be4bd9a",
   "metadata": {},
   "source": [
    "\\textbf{Table S48. Sequence divergence between the \\textit{Denisovan-like} haplotype in MXL and phased late Neanderthal haplotypes at the focal 72kb \\textit{MUC19} region.} \\newline The observed sequence divergence---i.e., the number of pairwise differences between a modern human haplotype and a late Neanderthal haplotype normalized by the effective sequence length---between the \\textit{Denisovan-like} haplotype in the focal MXL individual (i.e., NA19664), who harbors two \\textit{Denisovan-like} haplotypes and has no heterozygous sites at the focal 72kb region, and the phased haplotypes for both of the late Neanderthals. To assess the significance, we built a distribution of sequence divergence between each of the modern human's two chromosomes and the late Neanderthal's pseudo-haplotype, which was generated by randomly sampling one allele at each position from the unphased late Neanderthal's genotypes. For each late Neanderthal, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the pseudo-haplotype divergence genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the sequence divergence is less than or equal to that observed at the focal 72kb region. After correcting for four multiple comparisons---i.e., two per each modern human haplotype---a \\textit{P-value} less than 0.0125 is considered statistically significant. Note that the \\textit{Denisovan-like} in each of the late Neanderthal's corresponds to their second haplotype (i.e., Chagyrskaya Nean. Hap. 2 and Vindija Nean. Hap. 2). Also, note that at the phased 72kb region, the effective sequence length with respect to the NA19664 individual is 48119bp and 48444bp for the  Chagyrskaya and Vindija Neanderthal, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "0afcc726",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Individual</th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Archaic Hap.</th>\n",
       "      <th>Focal 72kb Region (Pairwise Diffs. Hap. 1)</th>\n",
       "      <th>Focal 72kb Region (Pairwise Diffs. Hap. 2)</th>\n",
       "      <th>Focal 72kb Region (Seq. Div. Hap. 1)</th>\n",
       "      <th>Focal 72kb Region (Seq. Div. Hap. 2)</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Nonoverlapping Windows $\\left( \\pm CI_{{95\\%}} \\right)$</th>\n",
       "      <th>$P-value$ (Hap. 1)</th>\n",
       "      <th>$P-value$ (Hap. 2)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA19664</td>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Chagyrskaya Nean. Hap. 1</td>\n",
       "      <td>168.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>0.003491</td>\n",
       "      <td>0.003491</td>\n",
       "      <td>0.001371</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.993163</td>\n",
       "      <td>0.993163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA19664</td>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Chagyrskaya Nean. Hap. 2</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.001371</td>\n",
       "      <td>0.000608</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.002636</td>\n",
       "      <td>0.002636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA19664</td>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Vindija Nean. Hap. 1</td>\n",
       "      <td>169.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>0.003489</td>\n",
       "      <td>0.003489</td>\n",
       "      <td>0.001381</td>\n",
       "      <td>0.000605</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.993190</td>\n",
       "      <td>0.993190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA19664</td>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Vindija Nean. Hap. 2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.001381</td>\n",
       "      <td>0.000605</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.001723</td>\n",
       "      <td>0.001723</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Individual Super Population Population              Archaic Hap.  \\\n",
       "0    NA19664              AMR        MXL  Chagyrskaya Nean. Hap. 1   \n",
       "1    NA19664              AMR        MXL  Chagyrskaya Nean. Hap. 2   \n",
       "2    NA19664              AMR        MXL      Vindija Nean. Hap. 1   \n",
       "3    NA19664              AMR        MXL      Vindija Nean. Hap. 2   \n",
       "\n",
       "   Focal 72kb Region (Pairwise Diffs. Hap. 1)  \\\n",
       "0                                       168.0   \n",
       "1                                         5.0   \n",
       "2                                       169.0   \n",
       "3                                         4.0   \n",
       "\n",
       "   Focal 72kb Region (Pairwise Diffs. Hap. 2)  \\\n",
       "0                                       168.0   \n",
       "1                                         5.0   \n",
       "2                                       169.0   \n",
       "3                                         4.0   \n",
       "\n",
       "   Focal 72kb Region (Seq. Div. Hap. 1)  Focal 72kb Region (Seq. Div. Hap. 2)  \\\n",
       "0                              0.003491                              0.003491   \n",
       "1                              0.000104                              0.000104   \n",
       "2                              0.003489                              0.003489   \n",
       "3                              0.000083                              0.000083   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           0.001371   \n",
       "1                                           0.001371   \n",
       "2                                           0.001381   \n",
       "3                                           0.001381   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.000608      \n",
       "1                                           0.000608      \n",
       "2                                           0.000605      \n",
       "3                                           0.000605      \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.000002   \n",
       "1                                           0.000002   \n",
       "2                                           0.000002   \n",
       "3                                           0.000002   \n",
       "\n",
       "   72kb Nonoverlapping Windows $\\left( \\pm CI_{{95\\%}} \\right)$  \\\n",
       "0                                           0.000005              \n",
       "1                                           0.000005              \n",
       "2                                           0.000005              \n",
       "3                                           0.000005              \n",
       "\n",
       "   $P-value$ (Hap. 1)  $P-value$ (Hap. 2)  \n",
       "0            0.993163            0.993163  \n",
       "1            0.002636            0.002636  \n",
       "2            0.993190            0.993190  \n",
       "3            0.001723            0.001723  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbc57e04",
   "metadata": {},
   "source": [
    "## S49"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "0c350650",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/pseudo_ancestry_painting_scores_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S49. Pseudo-Ancestry Painting scores for the focal 72kb \\textit{MUC19} region.} \\newline The Pseudo-Ancestry Painting (\\textit{PAP}) scores observed in the focal 72kb \\textit{MUC19} region. The \\textit{PAP} score quantifies the number of heterozygous sites in the target individual (i.e., \\textit{Target}) that can be explained by the two source individuals (i.e., $Source^{1}$ and $Source^{2}$) being fixed different allelic states and then normalized by the total number of heterozygous sites in the target individual. \\textit{PAP} scores were first computed for the late Neanderthals as target individuals using the Denisovan and Altai Neanderthal as sources as well as a pairing of an MXL individual (i.e., NA19664) who is homozygous for the \\textit{Denisovan-like} haplotype at the 72kb region and a YRI individual (i.e., NA19190) who is homozygous for the \\textit{Human-like} haplotype. Additionally, as a negative control, we computed additional \\textit{PAP} configurations where the Denisovan and Altai Neanderthal are the target individuals and the focal MXL (i.e., NA19664) and YRI (i.e., NA19190) individuals are sources. For each \\textit{PAP} configuration, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the \\textit{PAP} score genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the \\textit{PAP} score is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/pseudo_ancestry_painting_scores_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            r'$>0.999967$' if mat[i, j] == 1 else\n",
    "            '0' if mat[i, j] == 0 else\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1504b9c",
   "metadata": {},
   "source": [
    "\\textbf{Table S49. Pseudo-Ancestry Painting scores for the focal 72kb \\textit{MUC19} region.} \\newline The Pseudo-Ancestry Painting (\\textit{PAP}) scores observed in the focal 72kb \\textit{MUC19} region. The \\textit{PAP} score quantifies the number of heterozygous sites in the target individual (i.e., \\textit{Target}) that can be explained by the two source individuals (i.e., $Source^{1}$ and $Source^{2}$) being fixed different allelic states and then normalized by the total number of heterozygous sites in the target individual. \\textit{PAP} scores were first computed for the late Neanderthals as target individuals using the Denisovan and Altai Neanderthal as sources as well as a pairing of an MXL individual (i.e., NA19664) who is homozygous for the \\textit{Denisovan-like} haplotype at the 72kb region and a YRI individual (i.e., NA19190) who is homozygous for the \\textit{Human-like} haplotype. Additionally, as a negative control, we computed additional \\textit{PAP} configurations where the Denisovan and Altai Neanderthal are the target individuals and the focal MXL (i.e., NA19664) and YRI (i.e., NA19190) individuals are sources. For each \\textit{PAP} configuration, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the \\textit{PAP} score genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the \\textit{PAP} score is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "12ffb4f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$Source^{1}$</th>\n",
       "      <th>$Target$</th>\n",
       "      <th>$Source^{2}$</th>\n",
       "      <th>Focal 72kb Region $\\left( PAP \\text{ Score} \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>0.791667</td>\n",
       "      <td>0.104511</td>\n",
       "      <td>0.173347</td>\n",
       "      <td>0.001255</td>\n",
       "      <td>0.002460</td>\n",
       "      <td>0.005398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>0.801170</td>\n",
       "      <td>0.084911</td>\n",
       "      <td>0.153893</td>\n",
       "      <td>0.001011</td>\n",
       "      <td>0.001982</td>\n",
       "      <td>0.002980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MXL (NA19664)</td>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>YRI (NA19190)</td>\n",
       "      <td>0.940120</td>\n",
       "      <td>0.007686</td>\n",
       "      <td>0.046153</td>\n",
       "      <td>0.000335</td>\n",
       "      <td>0.000656</td>\n",
       "      <td>0.000368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MXL (NA19664)</td>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>YRI (NA19190)</td>\n",
       "      <td>0.929412</td>\n",
       "      <td>0.006178</td>\n",
       "      <td>0.038118</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>0.000492</td>\n",
       "      <td>0.000087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MXL (NA19664)</td>\n",
       "      <td>Denisovan</td>\n",
       "      <td>YRI (NA19190)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004781</td>\n",
       "      <td>0.031838</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.000392</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MXL (NA19664)</td>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>YRI (NA19190)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008424</td>\n",
       "      <td>0.043898</td>\n",
       "      <td>0.000311</td>\n",
       "      <td>0.000610</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    $Source^{1}$           $Target$   $Source^{2}$  \\\n",
       "0      Denisovan  Chagyrskaya Nean.    Altai Nean.   \n",
       "1      Denisovan      Vindija Nean.    Altai Nean.   \n",
       "2  MXL (NA19664)  Chagyrskaya Nean.  YRI (NA19190)   \n",
       "3  MXL (NA19664)      Vindija Nean.  YRI (NA19190)   \n",
       "4  MXL (NA19664)          Denisovan  YRI (NA19190)   \n",
       "5  MXL (NA19664)        Altai Nean.  YRI (NA19190)   \n",
       "\n",
       "   Focal 72kb Region $\\left( PAP \\text{ Score} \\right)$  \\\n",
       "0                                           0.791667      \n",
       "1                                           0.801170      \n",
       "2                                           0.940120      \n",
       "3                                           0.929412      \n",
       "4                                           0.000000      \n",
       "5                                           0.000000      \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                           0.104511   \n",
       "1                                           0.084911   \n",
       "2                                           0.007686   \n",
       "3                                           0.006178   \n",
       "4                                           0.004781   \n",
       "5                                           0.008424   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                           0.173347      \n",
       "1                                           0.153893      \n",
       "2                                           0.046153      \n",
       "3                                           0.038118      \n",
       "4                                           0.031838      \n",
       "5                                           0.043898      \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                           0.001255   \n",
       "1                                           0.001011   \n",
       "2                                           0.000335   \n",
       "3                                           0.000251   \n",
       "4                                           0.000200   \n",
       "5                                           0.000311   \n",
       "\n",
       "   72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                           0.002460             0.005398  \n",
       "1                                           0.001982             0.002980  \n",
       "2                                           0.000656             0.000368  \n",
       "3                                           0.000492             0.000087  \n",
       "4                                           0.000392             1.000000  \n",
       "5                                           0.000610             1.000000  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf5d0247",
   "metadata": {},
   "source": [
    "## S50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "33700f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/sprime_mxl_chb_ceu_pbs_per_region_742kb_and_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S50. $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values for the focal 742kb and 72kb \\textit{MUC19} using SPrime sites.} \\newline The $PBS_{MXL:CHB:CEU}$ values observed in the focal 742kb and 72kb region for putatively introgressed sites in MXL that is a match with either the Altai Denisovan or the Altai Neanderthal as defined by the SPrime introgression map. For each focal region, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{MXL:CHB:CEU}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping windows of comparable effective sequence length where the $PBS_{MXL:CHB:CEU}$ value, when only considering putatively introgressed SPrime sites, is greater than or equal to that observed at the focal respective focal region. A \\textit{P-value} less than 0.05 is considered statistically significant.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/sprime_mxl_chb_ceu_pbs_per_region_742kb_and_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "948f7ef6",
   "metadata": {},
   "source": [
    "\\textbf{Table S50. $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values for the focal 742kb and 72kb \\textit{MUC19} using SPrime sites.} \\newline The $PBS_{MXL:CHB:CEU}$ values observed in the focal 742kb and 72kb region for putatively introgressed sites in MXL that is a match with either the Altai Denisovan or the Altai Neanderthal as defined by the SPrime introgression map. For each focal region, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{MXL:CHB:CEU}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping windows of comparable effective sequence length where the $PBS_{MXL:CHB:CEU}$ value, when only considering putatively introgressed SPrime sites, is greater than or equal to that observed at the focal respective focal region. A \\textit{P-value} less than 0.05 is considered statistically significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "8eca4dac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Focal Region</th>\n",
       "      <th>$PBS_{MXL:CHB:CEU}$</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>Non-overlapping Windows $\\left( \\pm CI_{{95\\%}} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>742kb</td>\n",
       "      <td>0.291486</td>\n",
       "      <td>0.009563</td>\n",
       "      <td>0.027237</td>\n",
       "      <td>0.000769</td>\n",
       "      <td>0.001508</td>\n",
       "      <td>0.000796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>72kb</td>\n",
       "      <td>0.279663</td>\n",
       "      <td>0.010036</td>\n",
       "      <td>0.030212</td>\n",
       "      <td>0.000416</td>\n",
       "      <td>0.000815</td>\n",
       "      <td>0.002083</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Focal Region  $PBS_{MXL:CHB:CEU}$  \\\n",
       "0        742kb             0.291486   \n",
       "1         72kb             0.279663   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                      0.009563   \n",
       "1                                      0.010036   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                         0.027237   \n",
       "1                                         0.030212   \n",
       "\n",
       "   Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                      0.000769   \n",
       "1                                      0.000416   \n",
       "\n",
       "   Non-overlapping Windows $\\left( \\pm CI_{{95\\%}} \\right)$  $P-value$  \n",
       "0                                           0.001508          0.000796  \n",
       "1                                           0.000815          0.002083  "
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f3b6855",
   "metadata": {},
   "source": [
    "## S51"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "42deaea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/amr_asn_eur_pbs_per_region_742kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S51. $\\boldsymbol{PBS_{AMR:EAS/SAS:EUR}}$ values for the focal 742kb \\textit{MUC19} region.} \\newline The $PBS_{A:B:C}$ values observed in the focal 742kb region for all unique combinations of: AMR populations as the target population (\\textit{A}), EAS and SAS populations as one of the control populations (\\textit{B}), and EUR populations as the second control population (\\textit{C}). For each unique $PBS_{A:B:C}$ configuration, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{A:B:C}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the $PBS_{A:B:C}$ value is greater than or equal to that observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant. Note that when the target population is MXL, $PBS_{MXL:B:C}$ is always significant, and is never significant any other AMR population is the target population.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/amr_asn_eur_pbs_per_region_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            ('0' if mat[i, j] == 0 else\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, 8))+r' & '\\\n",
    "        + (\n",
    "            r'$>0.99969$' if mat[i, 8] == 1 else\n",
    "            convert_sci_notation_to_latex(float(mat[i, 8])) if float(mat[i, 8]) < 0.001 else\n",
    "            '{:.3f}'.format(round(float(mat[i, 8]), 3)).rstrip('0').rstrip('.')\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa2c9e56",
   "metadata": {},
   "source": [
    "\\textbf{Table S51. $\\boldsymbol{PBS_{AMR:EAS/SAS:EUR}}$ values for the focal 742kb \\textit{MUC19} region.} \\newline The $PBS_{A:B:C}$ values observed in the focal 742kb region for all unique combinations of: AMR populations as the target population (\\textit{A}), EAS and SAS populations as one of the control populations (\\textit{B}), and EUR populations as the second control population (\\textit{C}). For each unique $PBS_{A:B:C}$ configuration, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{A:B:C}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 742kb windows of comparable effective sequence length where the $PBS_{A:B:C}$ value is greater than or equal to that observed at the focal 742kb region. A \\textit{P-value} less than 0.05 is considered statistically significant. Note that when the target population is MXL, $PBS_{MXL:B:C}$ is always significant, and is never significant any other AMR population is the target population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "4ba250e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$A$ Pop.</th>\n",
       "      <th>$B$ Pop.</th>\n",
       "      <th>$C$ Pop.</th>\n",
       "      <th>Focal 742kb Region $\\left( PBS_{A:B:C} \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.006010</td>\n",
       "      <td>0.004268</td>\n",
       "      <td>0.006232</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000214</td>\n",
       "      <td>0.257562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005621</td>\n",
       "      <td>0.007057</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.007634</td>\n",
       "      <td>0.004288</td>\n",
       "      <td>0.006229</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000214</td>\n",
       "      <td>0.201344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.002293</td>\n",
       "      <td>0.003760</td>\n",
       "      <td>0.005823</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.425298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.005953</td>\n",
       "      <td>0.004534</td>\n",
       "      <td>0.006249</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000214</td>\n",
       "      <td>0.289337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001491</td>\n",
       "      <td>0.004624</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003462</td>\n",
       "      <td>0.007061</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001566</td>\n",
       "      <td>0.004953</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001345</td>\n",
       "      <td>0.004350</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001596</td>\n",
       "      <td>0.004769</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001363</td>\n",
       "      <td>0.004576</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003355</td>\n",
       "      <td>0.007149</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.000245</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000288</td>\n",
       "      <td>0.001480</td>\n",
       "      <td>0.004982</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.000171</td>\n",
       "      <td>0.185151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.003791</td>\n",
       "      <td>0.001231</td>\n",
       "      <td>0.004295</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>0.095325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.002545</td>\n",
       "      <td>0.001445</td>\n",
       "      <td>0.004695</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>0.137489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001409</td>\n",
       "      <td>0.004641</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003380</td>\n",
       "      <td>0.007179</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001521</td>\n",
       "      <td>0.005029</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.000172</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.003006</td>\n",
       "      <td>0.001269</td>\n",
       "      <td>0.004327</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0.111213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.001817</td>\n",
       "      <td>0.001494</td>\n",
       "      <td>0.004737</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.160403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.010258</td>\n",
       "      <td>0.006041</td>\n",
       "      <td>0.007342</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000252</td>\n",
       "      <td>0.192789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000995</td>\n",
       "      <td>0.006906</td>\n",
       "      <td>0.007777</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.751299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.011692</td>\n",
       "      <td>0.006017</td>\n",
       "      <td>0.007329</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>0.159487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.003962</td>\n",
       "      <td>0.005306</td>\n",
       "      <td>0.006877</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.000236</td>\n",
       "      <td>0.457684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>0.006361</td>\n",
       "      <td>0.007354</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.000252</td>\n",
       "      <td>0.294531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.012877</td>\n",
       "      <td>0.005126</td>\n",
       "      <td>0.006790</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.106936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006118</td>\n",
       "      <td>0.007474</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.012636</td>\n",
       "      <td>0.005112</td>\n",
       "      <td>0.006870</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.000235</td>\n",
       "      <td>0.116407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.004221</td>\n",
       "      <td>0.004632</td>\n",
       "      <td>0.006478</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.384662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.009036</td>\n",
       "      <td>0.005634</td>\n",
       "      <td>0.007010</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>0.227009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001334</td>\n",
       "      <td>0.004532</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000155</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003355</td>\n",
       "      <td>0.007255</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001430</td>\n",
       "      <td>0.004936</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.002176</td>\n",
       "      <td>0.001210</td>\n",
       "      <td>0.004353</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.116712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.001386</td>\n",
       "      <td>0.001402</td>\n",
       "      <td>0.004672</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0.158265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001487</td>\n",
       "      <td>0.004641</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003445</td>\n",
       "      <td>0.007003</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001593</td>\n",
       "      <td>0.004960</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.000170</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.003092</td>\n",
       "      <td>0.001363</td>\n",
       "      <td>0.004377</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000150</td>\n",
       "      <td>0.122823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.001809</td>\n",
       "      <td>0.001583</td>\n",
       "      <td>0.004716</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.169569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.014897</td>\n",
       "      <td>0.006338</td>\n",
       "      <td>0.007156</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.000245</td>\n",
       "      <td>0.094714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.002713</td>\n",
       "      <td>0.007069</td>\n",
       "      <td>0.007561</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.672166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.015271</td>\n",
       "      <td>0.006343</td>\n",
       "      <td>0.007248</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>0.097159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.006846</td>\n",
       "      <td>0.005573</td>\n",
       "      <td>0.006799</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.308585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.008949</td>\n",
       "      <td>0.006643</td>\n",
       "      <td>0.007232</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>0.278032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.008643</td>\n",
       "      <td>0.004942</td>\n",
       "      <td>0.006804</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.199817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005970</td>\n",
       "      <td>0.007465</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.009792</td>\n",
       "      <td>0.004923</td>\n",
       "      <td>0.006836</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.000234</td>\n",
       "      <td>0.167430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.003544</td>\n",
       "      <td>0.004477</td>\n",
       "      <td>0.006479</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.402994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.007058</td>\n",
       "      <td>0.005440</td>\n",
       "      <td>0.007002</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>0.300336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.073922</td>\n",
       "      <td>0.017006</td>\n",
       "      <td>0.014576</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.007333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.070940</td>\n",
       "      <td>0.016964</td>\n",
       "      <td>0.014552</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.007333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.078372</td>\n",
       "      <td>0.016957</td>\n",
       "      <td>0.014594</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.004888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.066875</td>\n",
       "      <td>0.016979</td>\n",
       "      <td>0.014481</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>0.009777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.068855</td>\n",
       "      <td>0.017933</td>\n",
       "      <td>0.014607</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.000501</td>\n",
       "      <td>0.010694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.055320</td>\n",
       "      <td>0.005987</td>\n",
       "      <td>0.011866</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>0.009777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.052291</td>\n",
       "      <td>0.007654</td>\n",
       "      <td>0.013190</td>\n",
       "      <td>0.000231</td>\n",
       "      <td>0.000452</td>\n",
       "      <td>0.014971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.059716</td>\n",
       "      <td>0.006011</td>\n",
       "      <td>0.012065</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0.000414</td>\n",
       "      <td>0.007333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.053224</td>\n",
       "      <td>0.006143</td>\n",
       "      <td>0.012144</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>0.000416</td>\n",
       "      <td>0.010082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.055809</td>\n",
       "      <td>0.006463</td>\n",
       "      <td>0.012273</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.009777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.065842</td>\n",
       "      <td>0.005256</td>\n",
       "      <td>0.011149</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>0.003972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.061304</td>\n",
       "      <td>0.006950</td>\n",
       "      <td>0.012709</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.000436</td>\n",
       "      <td>0.006416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.071925</td>\n",
       "      <td>0.005322</td>\n",
       "      <td>0.011410</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.000391</td>\n",
       "      <td>0.003972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.069272</td>\n",
       "      <td>0.005388</td>\n",
       "      <td>0.011437</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.000392</td>\n",
       "      <td>0.003972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.066346</td>\n",
       "      <td>0.005684</td>\n",
       "      <td>0.011618</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.000398</td>\n",
       "      <td>0.004583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.063583</td>\n",
       "      <td>0.005435</td>\n",
       "      <td>0.011446</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.000392</td>\n",
       "      <td>0.004583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.060042</td>\n",
       "      <td>0.007101</td>\n",
       "      <td>0.012974</td>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>0.008555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.069558</td>\n",
       "      <td>0.005510</td>\n",
       "      <td>0.011677</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.000400</td>\n",
       "      <td>0.004583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.068915</td>\n",
       "      <td>0.005584</td>\n",
       "      <td>0.011703</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.000401</td>\n",
       "      <td>0.004277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.066046</td>\n",
       "      <td>0.005877</td>\n",
       "      <td>0.011834</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000406</td>\n",
       "      <td>0.005500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.087203</td>\n",
       "      <td>0.020919</td>\n",
       "      <td>0.015859</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.004277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.084596</td>\n",
       "      <td>0.020081</td>\n",
       "      <td>0.015681</td>\n",
       "      <td>0.000274</td>\n",
       "      <td>0.000538</td>\n",
       "      <td>0.005805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.091462</td>\n",
       "      <td>0.020812</td>\n",
       "      <td>0.015919</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.000546</td>\n",
       "      <td>0.004277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.077576</td>\n",
       "      <td>0.020713</td>\n",
       "      <td>0.015691</td>\n",
       "      <td>0.000274</td>\n",
       "      <td>0.000538</td>\n",
       "      <td>0.008249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.080197</td>\n",
       "      <td>0.021908</td>\n",
       "      <td>0.015823</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>0.000542</td>\n",
       "      <td>0.007333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.084820</td>\n",
       "      <td>0.019040</td>\n",
       "      <td>0.015499</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>0.000531</td>\n",
       "      <td>0.006111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.077461</td>\n",
       "      <td>0.018468</td>\n",
       "      <td>0.015360</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>0.007638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.087405</td>\n",
       "      <td>0.018940</td>\n",
       "      <td>0.015562</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.000533</td>\n",
       "      <td>0.005194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.072834</td>\n",
       "      <td>0.019112</td>\n",
       "      <td>0.015403</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.000528</td>\n",
       "      <td>0.010082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.075968</td>\n",
       "      <td>0.020306</td>\n",
       "      <td>0.015586</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>0.007638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.063423</td>\n",
       "      <td>0.005068</td>\n",
       "      <td>0.011145</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>0.005194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.058515</td>\n",
       "      <td>0.006821</td>\n",
       "      <td>0.012868</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.000441</td>\n",
       "      <td>0.008555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.068761</td>\n",
       "      <td>0.005153</td>\n",
       "      <td>0.011522</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.000395</td>\n",
       "      <td>0.003972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.068404</td>\n",
       "      <td>0.005252</td>\n",
       "      <td>0.011504</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.000394</td>\n",
       "      <td>0.003361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.065934</td>\n",
       "      <td>0.005478</td>\n",
       "      <td>0.011590</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.000397</td>\n",
       "      <td>0.003972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.062689</td>\n",
       "      <td>0.005937</td>\n",
       "      <td>0.011682</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.000400</td>\n",
       "      <td>0.006111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.058970</td>\n",
       "      <td>0.007577</td>\n",
       "      <td>0.013040</td>\n",
       "      <td>0.000228</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.009777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.067149</td>\n",
       "      <td>0.005976</td>\n",
       "      <td>0.011859</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>0.005194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.065048</td>\n",
       "      <td>0.006099</td>\n",
       "      <td>0.011958</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.000410</td>\n",
       "      <td>0.005805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.062084</td>\n",
       "      <td>0.006397</td>\n",
       "      <td>0.012092</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0.000414</td>\n",
       "      <td>0.006722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.095860</td>\n",
       "      <td>0.021647</td>\n",
       "      <td>0.015604</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>0.002750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.090333</td>\n",
       "      <td>0.020649</td>\n",
       "      <td>0.015376</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>0.003972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.099060</td>\n",
       "      <td>0.021605</td>\n",
       "      <td>0.015691</td>\n",
       "      <td>0.000274</td>\n",
       "      <td>0.000538</td>\n",
       "      <td>0.002139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.084479</td>\n",
       "      <td>0.021463</td>\n",
       "      <td>0.015463</td>\n",
       "      <td>0.000270</td>\n",
       "      <td>0.000530</td>\n",
       "      <td>0.005500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.084902</td>\n",
       "      <td>0.022613</td>\n",
       "      <td>0.015564</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>0.005805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.079544</td>\n",
       "      <td>0.018787</td>\n",
       "      <td>0.015486</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>0.000531</td>\n",
       "      <td>0.007333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.075992</td>\n",
       "      <td>0.018249</td>\n",
       "      <td>0.015372</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.000527</td>\n",
       "      <td>0.008249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.083518</td>\n",
       "      <td>0.018657</td>\n",
       "      <td>0.015496</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>0.000531</td>\n",
       "      <td>0.006111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.071114</td>\n",
       "      <td>0.018839</td>\n",
       "      <td>0.015394</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.000528</td>\n",
       "      <td>0.009777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.072948</td>\n",
       "      <td>0.019992</td>\n",
       "      <td>0.015591</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.000534</td>\n",
       "      <td>0.010694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.039316</td>\n",
       "      <td>0.060214</td>\n",
       "      <td>0.034610</td>\n",
       "      <td>0.000605</td>\n",
       "      <td>0.001186</td>\n",
       "      <td>0.722884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.039025</td>\n",
       "      <td>0.058218</td>\n",
       "      <td>0.034190</td>\n",
       "      <td>0.000598</td>\n",
       "      <td>0.001172</td>\n",
       "      <td>0.695692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.042458</td>\n",
       "      <td>0.060169</td>\n",
       "      <td>0.034641</td>\n",
       "      <td>0.000606</td>\n",
       "      <td>0.001187</td>\n",
       "      <td>0.675527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.032087</td>\n",
       "      <td>0.060870</td>\n",
       "      <td>0.034536</td>\n",
       "      <td>0.000604</td>\n",
       "      <td>0.001184</td>\n",
       "      <td>0.826459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.035839</td>\n",
       "      <td>0.061861</td>\n",
       "      <td>0.034684</td>\n",
       "      <td>0.000606</td>\n",
       "      <td>0.001189</td>\n",
       "      <td>0.788268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.019981</td>\n",
       "      <td>0.034397</td>\n",
       "      <td>0.031146</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.001068</td>\n",
       "      <td>0.626948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.019643</td>\n",
       "      <td>0.035683</td>\n",
       "      <td>0.031506</td>\n",
       "      <td>0.000551</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.658723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.023069</td>\n",
       "      <td>0.034424</td>\n",
       "      <td>0.031349</td>\n",
       "      <td>0.000548</td>\n",
       "      <td>0.001075</td>\n",
       "      <td>0.585396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.017703</td>\n",
       "      <td>0.035318</td>\n",
       "      <td>0.031359</td>\n",
       "      <td>0.000548</td>\n",
       "      <td>0.001075</td>\n",
       "      <td>0.687137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.022059</td>\n",
       "      <td>0.035712</td>\n",
       "      <td>0.031498</td>\n",
       "      <td>0.000551</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.623587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.022649</td>\n",
       "      <td>0.032606</td>\n",
       "      <td>0.030795</td>\n",
       "      <td>0.000538</td>\n",
       "      <td>0.001056</td>\n",
       "      <td>0.555454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.020801</td>\n",
       "      <td>0.034113</td>\n",
       "      <td>0.031287</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.001072</td>\n",
       "      <td>0.609838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.027423</td>\n",
       "      <td>0.032666</td>\n",
       "      <td>0.030993</td>\n",
       "      <td>0.000542</td>\n",
       "      <td>0.001062</td>\n",
       "      <td>0.481515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.025897</td>\n",
       "      <td>0.033438</td>\n",
       "      <td>0.030974</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>0.001062</td>\n",
       "      <td>0.519707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.024742</td>\n",
       "      <td>0.033863</td>\n",
       "      <td>0.031103</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.001066</td>\n",
       "      <td>0.541399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.019627</td>\n",
       "      <td>0.032833</td>\n",
       "      <td>0.031045</td>\n",
       "      <td>0.000543</td>\n",
       "      <td>0.001064</td>\n",
       "      <td>0.607088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.018777</td>\n",
       "      <td>0.034258</td>\n",
       "      <td>0.031521</td>\n",
       "      <td>0.000551</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.645585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.024294</td>\n",
       "      <td>0.032904</td>\n",
       "      <td>0.031231</td>\n",
       "      <td>0.000546</td>\n",
       "      <td>0.001071</td>\n",
       "      <td>0.534067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.024776</td>\n",
       "      <td>0.033715</td>\n",
       "      <td>0.031196</td>\n",
       "      <td>0.000545</td>\n",
       "      <td>0.001069</td>\n",
       "      <td>0.537427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.023679</td>\n",
       "      <td>0.034115</td>\n",
       "      <td>0.031309</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.001073</td>\n",
       "      <td>0.558509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.052839</td>\n",
       "      <td>0.066481</td>\n",
       "      <td>0.036563</td>\n",
       "      <td>0.000639</td>\n",
       "      <td>0.001253</td>\n",
       "      <td>0.590284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.052923</td>\n",
       "      <td>0.063654</td>\n",
       "      <td>0.036016</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.001235</td>\n",
       "      <td>0.556370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.055790</td>\n",
       "      <td>0.066401</td>\n",
       "      <td>0.036601</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>0.001255</td>\n",
       "      <td>0.553315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.043030</td>\n",
       "      <td>0.066962</td>\n",
       "      <td>0.036440</td>\n",
       "      <td>0.000637</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>0.742744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.047423</td>\n",
       "      <td>0.068183</td>\n",
       "      <td>0.036660</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.001257</td>\n",
       "      <td>0.698136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.048203</td>\n",
       "      <td>0.063681</td>\n",
       "      <td>0.035872</td>\n",
       "      <td>0.000627</td>\n",
       "      <td>0.001230</td>\n",
       "      <td>0.623587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.043535</td>\n",
       "      <td>0.061132</td>\n",
       "      <td>0.035384</td>\n",
       "      <td>0.000619</td>\n",
       "      <td>0.001213</td>\n",
       "      <td>0.661167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.049480</td>\n",
       "      <td>0.063592</td>\n",
       "      <td>0.035911</td>\n",
       "      <td>0.000628</td>\n",
       "      <td>0.001231</td>\n",
       "      <td>0.603727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.036035</td>\n",
       "      <td>0.064445</td>\n",
       "      <td>0.035836</td>\n",
       "      <td>0.000626</td>\n",
       "      <td>0.001228</td>\n",
       "      <td>0.805988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.040941</td>\n",
       "      <td>0.065663</td>\n",
       "      <td>0.036047</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.001236</td>\n",
       "      <td>0.750687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.016766</td>\n",
       "      <td>0.031939</td>\n",
       "      <td>0.030753</td>\n",
       "      <td>0.000538</td>\n",
       "      <td>0.001054</td>\n",
       "      <td>0.639474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.014548</td>\n",
       "      <td>0.033598</td>\n",
       "      <td>0.031336</td>\n",
       "      <td>0.000548</td>\n",
       "      <td>0.001074</td>\n",
       "      <td>0.705469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.020795</td>\n",
       "      <td>0.032005</td>\n",
       "      <td>0.031062</td>\n",
       "      <td>0.000543</td>\n",
       "      <td>0.001065</td>\n",
       "      <td>0.575619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.021564</td>\n",
       "      <td>0.032852</td>\n",
       "      <td>0.031044</td>\n",
       "      <td>0.000543</td>\n",
       "      <td>0.001064</td>\n",
       "      <td>0.581118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.020866</td>\n",
       "      <td>0.033189</td>\n",
       "      <td>0.031105</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.001066</td>\n",
       "      <td>0.597006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.021261</td>\n",
       "      <td>0.034631</td>\n",
       "      <td>0.031097</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.001066</td>\n",
       "      <td>0.614421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.020233</td>\n",
       "      <td>0.035823</td>\n",
       "      <td>0.031528</td>\n",
       "      <td>0.000551</td>\n",
       "      <td>0.001081</td>\n",
       "      <td>0.652612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.024413</td>\n",
       "      <td>0.034660</td>\n",
       "      <td>0.031266</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.001072</td>\n",
       "      <td>0.562481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.023438</td>\n",
       "      <td>0.035508</td>\n",
       "      <td>0.031300</td>\n",
       "      <td>0.000547</td>\n",
       "      <td>0.001073</td>\n",
       "      <td>0.593645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.022245</td>\n",
       "      <td>0.035920</td>\n",
       "      <td>0.031401</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0.001076</td>\n",
       "      <td>0.619310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.055930</td>\n",
       "      <td>0.067601</td>\n",
       "      <td>0.036007</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.001234</td>\n",
       "      <td>0.570425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.053094</td>\n",
       "      <td>0.064611</td>\n",
       "      <td>0.035405</td>\n",
       "      <td>0.000619</td>\n",
       "      <td>0.001214</td>\n",
       "      <td>0.573480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.057822</td>\n",
       "      <td>0.067569</td>\n",
       "      <td>0.036100</td>\n",
       "      <td>0.000631</td>\n",
       "      <td>0.001237</td>\n",
       "      <td>0.541094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.044367</td>\n",
       "      <td>0.068086</td>\n",
       "      <td>0.035949</td>\n",
       "      <td>0.000628</td>\n",
       "      <td>0.001232</td>\n",
       "      <td>0.738772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.046561</td>\n",
       "      <td>0.069252</td>\n",
       "      <td>0.036176</td>\n",
       "      <td>0.000632</td>\n",
       "      <td>0.001240</td>\n",
       "      <td>0.724412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.044166</td>\n",
       "      <td>0.063216</td>\n",
       "      <td>0.035986</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.001233</td>\n",
       "      <td>0.681943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.043305</td>\n",
       "      <td>0.060725</td>\n",
       "      <td>0.035485</td>\n",
       "      <td>0.000620</td>\n",
       "      <td>0.001216</td>\n",
       "      <td>0.661167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.046833</td>\n",
       "      <td>0.063113</td>\n",
       "      <td>0.035958</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>0.001233</td>\n",
       "      <td>0.641002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.035554</td>\n",
       "      <td>0.063961</td>\n",
       "      <td>0.035951</td>\n",
       "      <td>0.000628</td>\n",
       "      <td>0.001232</td>\n",
       "      <td>0.809044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.039160</td>\n",
       "      <td>0.065149</td>\n",
       "      <td>0.036187</td>\n",
       "      <td>0.000633</td>\n",
       "      <td>0.001240</td>\n",
       "      <td>0.770241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.001231</td>\n",
       "      <td>0.002673</td>\n",
       "      <td>0.004243</td>\n",
       "      <td>0.000074</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>0.449435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004606</td>\n",
       "      <td>0.005818</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.002596</td>\n",
       "      <td>0.002692</td>\n",
       "      <td>0.004392</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>0.338527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001883</td>\n",
       "      <td>0.003477</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.002104</td>\n",
       "      <td>0.002504</td>\n",
       "      <td>0.004071</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>0.359609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001362</td>\n",
       "      <td>0.004149</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004050</td>\n",
       "      <td>0.007233</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001431</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000952</td>\n",
       "      <td>0.003372</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001193</td>\n",
       "      <td>0.003797</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001321</td>\n",
       "      <td>0.004036</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004164</td>\n",
       "      <td>0.007313</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001428</td>\n",
       "      <td>0.004470</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.001048</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>0.003314</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.139627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.001071</td>\n",
       "      <td>0.001166</td>\n",
       "      <td>0.003775</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.166208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001349</td>\n",
       "      <td>0.004107</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004111</td>\n",
       "      <td>0.007310</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001455</td>\n",
       "      <td>0.004532</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000155</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000967</td>\n",
       "      <td>0.003374</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001189</td>\n",
       "      <td>0.003844</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.006419</td>\n",
       "      <td>0.003748</td>\n",
       "      <td>0.004978</td>\n",
       "      <td>0.000087</td>\n",
       "      <td>0.000171</td>\n",
       "      <td>0.213566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.002998</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>0.006134</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.000210</td>\n",
       "      <td>0.546593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.007593</td>\n",
       "      <td>0.003738</td>\n",
       "      <td>0.005156</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.000177</td>\n",
       "      <td>0.170180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002739</td>\n",
       "      <td>0.004188</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.005353</td>\n",
       "      <td>0.003607</td>\n",
       "      <td>0.004792</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.000164</td>\n",
       "      <td>0.255118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.007360</td>\n",
       "      <td>0.003084</td>\n",
       "      <td>0.004571</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.139322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004681</td>\n",
       "      <td>0.005966</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.006859</td>\n",
       "      <td>0.003071</td>\n",
       "      <td>0.004765</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.157959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002298</td>\n",
       "      <td>0.003891</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.004447</td>\n",
       "      <td>0.003098</td>\n",
       "      <td>0.004575</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.270394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001300</td>\n",
       "      <td>0.004043</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004187</td>\n",
       "      <td>0.007409</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001406</td>\n",
       "      <td>0.004428</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000152</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000931</td>\n",
       "      <td>0.003321</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001128</td>\n",
       "      <td>0.003755</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001350</td>\n",
       "      <td>0.004061</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004004</td>\n",
       "      <td>0.007114</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001445</td>\n",
       "      <td>0.004450</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000954</td>\n",
       "      <td>0.003381</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001180</td>\n",
       "      <td>0.003735</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.009540</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>0.004820</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>0.106936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.003198</td>\n",
       "      <td>0.005299</td>\n",
       "      <td>0.005786</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>0.000198</td>\n",
       "      <td>0.554537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.009655</td>\n",
       "      <td>0.003908</td>\n",
       "      <td>0.005014</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>0.000172</td>\n",
       "      <td>0.108463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.001149</td>\n",
       "      <td>0.002850</td>\n",
       "      <td>0.004095</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>0.509624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.004521</td>\n",
       "      <td>0.003707</td>\n",
       "      <td>0.004736</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.323862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.004478</td>\n",
       "      <td>0.002975</td>\n",
       "      <td>0.004586</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.250535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>0.004580</td>\n",
       "      <td>0.005986</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.672777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.005368</td>\n",
       "      <td>0.002945</td>\n",
       "      <td>0.004717</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.200428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002169</td>\n",
       "      <td>0.003818</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.003822</td>\n",
       "      <td>0.002952</td>\n",
       "      <td>0.004528</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000155</td>\n",
       "      <td>0.289948</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    $A$ Pop. $B$ Pop. $C$ Pop.  \\\n",
       "0        CLM      BEB      CEU   \n",
       "1        CLM      BEB      FIN   \n",
       "2        CLM      BEB      GBR   \n",
       "3        CLM      BEB      IBS   \n",
       "4        CLM      BEB      TSI   \n",
       "5        CLM      CDX      CEU   \n",
       "6        CLM      CDX      FIN   \n",
       "7        CLM      CDX      GBR   \n",
       "8        CLM      CDX      IBS   \n",
       "9        CLM      CDX      TSI   \n",
       "10       CLM      CHB      CEU   \n",
       "11       CLM      CHB      FIN   \n",
       "12       CLM      CHB      GBR   \n",
       "13       CLM      CHB      IBS   \n",
       "14       CLM      CHB      TSI   \n",
       "15       CLM      CHS      CEU   \n",
       "16       CLM      CHS      FIN   \n",
       "17       CLM      CHS      GBR   \n",
       "18       CLM      CHS      IBS   \n",
       "19       CLM      CHS      TSI   \n",
       "20       CLM      GIH      CEU   \n",
       "21       CLM      GIH      FIN   \n",
       "22       CLM      GIH      GBR   \n",
       "23       CLM      GIH      IBS   \n",
       "24       CLM      GIH      TSI   \n",
       "25       CLM      ITU      CEU   \n",
       "26       CLM      ITU      FIN   \n",
       "27       CLM      ITU      GBR   \n",
       "28       CLM      ITU      IBS   \n",
       "29       CLM      ITU      TSI   \n",
       "30       CLM      JPT      CEU   \n",
       "31       CLM      JPT      FIN   \n",
       "32       CLM      JPT      GBR   \n",
       "33       CLM      JPT      IBS   \n",
       "34       CLM      JPT      TSI   \n",
       "35       CLM      KHV      CEU   \n",
       "36       CLM      KHV      FIN   \n",
       "37       CLM      KHV      GBR   \n",
       "38       CLM      KHV      IBS   \n",
       "39       CLM      KHV      TSI   \n",
       "40       CLM      PJL      CEU   \n",
       "41       CLM      PJL      FIN   \n",
       "42       CLM      PJL      GBR   \n",
       "43       CLM      PJL      IBS   \n",
       "44       CLM      PJL      TSI   \n",
       "45       CLM      STU      CEU   \n",
       "46       CLM      STU      FIN   \n",
       "47       CLM      STU      GBR   \n",
       "48       CLM      STU      IBS   \n",
       "49       CLM      STU      TSI   \n",
       "50       MXL      BEB      CEU   \n",
       "51       MXL      BEB      FIN   \n",
       "52       MXL      BEB      GBR   \n",
       "53       MXL      BEB      IBS   \n",
       "54       MXL      BEB      TSI   \n",
       "55       MXL      CDX      CEU   \n",
       "56       MXL      CDX      FIN   \n",
       "57       MXL      CDX      GBR   \n",
       "58       MXL      CDX      IBS   \n",
       "59       MXL      CDX      TSI   \n",
       "60       MXL      CHB      CEU   \n",
       "61       MXL      CHB      FIN   \n",
       "62       MXL      CHB      GBR   \n",
       "63       MXL      CHB      IBS   \n",
       "64       MXL      CHB      TSI   \n",
       "65       MXL      CHS      CEU   \n",
       "66       MXL      CHS      FIN   \n",
       "67       MXL      CHS      GBR   \n",
       "68       MXL      CHS      IBS   \n",
       "69       MXL      CHS      TSI   \n",
       "70       MXL      GIH      CEU   \n",
       "71       MXL      GIH      FIN   \n",
       "72       MXL      GIH      GBR   \n",
       "73       MXL      GIH      IBS   \n",
       "74       MXL      GIH      TSI   \n",
       "75       MXL      ITU      CEU   \n",
       "76       MXL      ITU      FIN   \n",
       "77       MXL      ITU      GBR   \n",
       "78       MXL      ITU      IBS   \n",
       "79       MXL      ITU      TSI   \n",
       "80       MXL      JPT      CEU   \n",
       "81       MXL      JPT      FIN   \n",
       "82       MXL      JPT      GBR   \n",
       "83       MXL      JPT      IBS   \n",
       "84       MXL      JPT      TSI   \n",
       "85       MXL      KHV      CEU   \n",
       "86       MXL      KHV      FIN   \n",
       "87       MXL      KHV      GBR   \n",
       "88       MXL      KHV      IBS   \n",
       "89       MXL      KHV      TSI   \n",
       "90       MXL      PJL      CEU   \n",
       "91       MXL      PJL      FIN   \n",
       "92       MXL      PJL      GBR   \n",
       "93       MXL      PJL      IBS   \n",
       "94       MXL      PJL      TSI   \n",
       "95       MXL      STU      CEU   \n",
       "96       MXL      STU      FIN   \n",
       "97       MXL      STU      GBR   \n",
       "98       MXL      STU      IBS   \n",
       "99       MXL      STU      TSI   \n",
       "100      PEL      BEB      CEU   \n",
       "101      PEL      BEB      FIN   \n",
       "102      PEL      BEB      GBR   \n",
       "103      PEL      BEB      IBS   \n",
       "104      PEL      BEB      TSI   \n",
       "105      PEL      CDX      CEU   \n",
       "106      PEL      CDX      FIN   \n",
       "107      PEL      CDX      GBR   \n",
       "108      PEL      CDX      IBS   \n",
       "109      PEL      CDX      TSI   \n",
       "110      PEL      CHB      CEU   \n",
       "111      PEL      CHB      FIN   \n",
       "112      PEL      CHB      GBR   \n",
       "113      PEL      CHB      IBS   \n",
       "114      PEL      CHB      TSI   \n",
       "115      PEL      CHS      CEU   \n",
       "116      PEL      CHS      FIN   \n",
       "117      PEL      CHS      GBR   \n",
       "118      PEL      CHS      IBS   \n",
       "119      PEL      CHS      TSI   \n",
       "120      PEL      GIH      CEU   \n",
       "121      PEL      GIH      FIN   \n",
       "122      PEL      GIH      GBR   \n",
       "123      PEL      GIH      IBS   \n",
       "124      PEL      GIH      TSI   \n",
       "125      PEL      ITU      CEU   \n",
       "126      PEL      ITU      FIN   \n",
       "127      PEL      ITU      GBR   \n",
       "128      PEL      ITU      IBS   \n",
       "129      PEL      ITU      TSI   \n",
       "130      PEL      JPT      CEU   \n",
       "131      PEL      JPT      FIN   \n",
       "132      PEL      JPT      GBR   \n",
       "133      PEL      JPT      IBS   \n",
       "134      PEL      JPT      TSI   \n",
       "135      PEL      KHV      CEU   \n",
       "136      PEL      KHV      FIN   \n",
       "137      PEL      KHV      GBR   \n",
       "138      PEL      KHV      IBS   \n",
       "139      PEL      KHV      TSI   \n",
       "140      PEL      PJL      CEU   \n",
       "141      PEL      PJL      FIN   \n",
       "142      PEL      PJL      GBR   \n",
       "143      PEL      PJL      IBS   \n",
       "144      PEL      PJL      TSI   \n",
       "145      PEL      STU      CEU   \n",
       "146      PEL      STU      FIN   \n",
       "147      PEL      STU      GBR   \n",
       "148      PEL      STU      IBS   \n",
       "149      PEL      STU      TSI   \n",
       "150      PUR      BEB      CEU   \n",
       "151      PUR      BEB      FIN   \n",
       "152      PUR      BEB      GBR   \n",
       "153      PUR      BEB      IBS   \n",
       "154      PUR      BEB      TSI   \n",
       "155      PUR      CDX      CEU   \n",
       "156      PUR      CDX      FIN   \n",
       "157      PUR      CDX      GBR   \n",
       "158      PUR      CDX      IBS   \n",
       "159      PUR      CDX      TSI   \n",
       "160      PUR      CHB      CEU   \n",
       "161      PUR      CHB      FIN   \n",
       "162      PUR      CHB      GBR   \n",
       "163      PUR      CHB      IBS   \n",
       "164      PUR      CHB      TSI   \n",
       "165      PUR      CHS      CEU   \n",
       "166      PUR      CHS      FIN   \n",
       "167      PUR      CHS      GBR   \n",
       "168      PUR      CHS      IBS   \n",
       "169      PUR      CHS      TSI   \n",
       "170      PUR      GIH      CEU   \n",
       "171      PUR      GIH      FIN   \n",
       "172      PUR      GIH      GBR   \n",
       "173      PUR      GIH      IBS   \n",
       "174      PUR      GIH      TSI   \n",
       "175      PUR      ITU      CEU   \n",
       "176      PUR      ITU      FIN   \n",
       "177      PUR      ITU      GBR   \n",
       "178      PUR      ITU      IBS   \n",
       "179      PUR      ITU      TSI   \n",
       "180      PUR      JPT      CEU   \n",
       "181      PUR      JPT      FIN   \n",
       "182      PUR      JPT      GBR   \n",
       "183      PUR      JPT      IBS   \n",
       "184      PUR      JPT      TSI   \n",
       "185      PUR      KHV      CEU   \n",
       "186      PUR      KHV      FIN   \n",
       "187      PUR      KHV      GBR   \n",
       "188      PUR      KHV      IBS   \n",
       "189      PUR      KHV      TSI   \n",
       "190      PUR      PJL      CEU   \n",
       "191      PUR      PJL      FIN   \n",
       "192      PUR      PJL      GBR   \n",
       "193      PUR      PJL      IBS   \n",
       "194      PUR      PJL      TSI   \n",
       "195      PUR      STU      CEU   \n",
       "196      PUR      STU      FIN   \n",
       "197      PUR      STU      GBR   \n",
       "198      PUR      STU      IBS   \n",
       "199      PUR      STU      TSI   \n",
       "\n",
       "     Focal 742kb Region $\\left( PBS_{A:B:C} \\right)$  \\\n",
       "0                                           0.006010   \n",
       "1                                           0.000000   \n",
       "2                                           0.007634   \n",
       "3                                           0.002293   \n",
       "4                                           0.005953   \n",
       "5                                           0.000000   \n",
       "6                                           0.000000   \n",
       "7                                           0.000000   \n",
       "8                                           0.000000   \n",
       "9                                           0.000000   \n",
       "10                                          0.000000   \n",
       "11                                          0.000000   \n",
       "12                                          0.000288   \n",
       "13                                          0.003791   \n",
       "14                                          0.002545   \n",
       "15                                          0.000000   \n",
       "16                                          0.000000   \n",
       "17                                          0.000000   \n",
       "18                                          0.003006   \n",
       "19                                          0.001817   \n",
       "20                                          0.010258   \n",
       "21                                          0.000995   \n",
       "22                                          0.011692   \n",
       "23                                          0.003962   \n",
       "24                                          0.008264   \n",
       "25                                          0.012877   \n",
       "26                                          0.000000   \n",
       "27                                          0.012636   \n",
       "28                                          0.004221   \n",
       "29                                          0.009036   \n",
       "30                                          0.000000   \n",
       "31                                          0.000000   \n",
       "32                                          0.000000   \n",
       "33                                          0.002176   \n",
       "34                                          0.001386   \n",
       "35                                          0.000000   \n",
       "36                                          0.000000   \n",
       "37                                          0.000000   \n",
       "38                                          0.003092   \n",
       "39                                          0.001809   \n",
       "40                                          0.014897   \n",
       "41                                          0.002713   \n",
       "42                                          0.015271   \n",
       "43                                          0.006846   \n",
       "44                                          0.008949   \n",
       "45                                          0.008643   \n",
       "46                                          0.000000   \n",
       "47                                          0.009792   \n",
       "48                                          0.003544   \n",
       "49                                          0.007058   \n",
       "50                                          0.073922   \n",
       "51                                          0.070940   \n",
       "52                                          0.078372   \n",
       "53                                          0.066875   \n",
       "54                                          0.068855   \n",
       "55                                          0.055320   \n",
       "56                                          0.052291   \n",
       "57                                          0.059716   \n",
       "58                                          0.053224   \n",
       "59                                          0.055809   \n",
       "60                                          0.065842   \n",
       "61                                          0.061304   \n",
       "62                                          0.071925   \n",
       "63                                          0.069272   \n",
       "64                                          0.066346   \n",
       "65                                          0.063583   \n",
       "66                                          0.060042   \n",
       "67                                          0.069558   \n",
       "68                                          0.068915   \n",
       "69                                          0.066046   \n",
       "70                                          0.087203   \n",
       "71                                          0.084596   \n",
       "72                                          0.091462   \n",
       "73                                          0.077576   \n",
       "74                                          0.080197   \n",
       "75                                          0.084820   \n",
       "76                                          0.077461   \n",
       "77                                          0.087405   \n",
       "78                                          0.072834   \n",
       "79                                          0.075968   \n",
       "80                                          0.063423   \n",
       "81                                          0.058515   \n",
       "82                                          0.068761   \n",
       "83                                          0.068404   \n",
       "84                                          0.065934   \n",
       "85                                          0.062689   \n",
       "86                                          0.058970   \n",
       "87                                          0.067149   \n",
       "88                                          0.065048   \n",
       "89                                          0.062084   \n",
       "90                                          0.095860   \n",
       "91                                          0.090333   \n",
       "92                                          0.099060   \n",
       "93                                          0.084479   \n",
       "94                                          0.084902   \n",
       "95                                          0.079544   \n",
       "96                                          0.075992   \n",
       "97                                          0.083518   \n",
       "98                                          0.071114   \n",
       "99                                          0.072948   \n",
       "100                                         0.039316   \n",
       "101                                         0.039025   \n",
       "102                                         0.042458   \n",
       "103                                         0.032087   \n",
       "104                                         0.035839   \n",
       "105                                         0.019981   \n",
       "106                                         0.019643   \n",
       "107                                         0.023069   \n",
       "108                                         0.017703   \n",
       "109                                         0.022059   \n",
       "110                                         0.022649   \n",
       "111                                         0.020801   \n",
       "112                                         0.027423   \n",
       "113                                         0.025897   \n",
       "114                                         0.024742   \n",
       "115                                         0.019627   \n",
       "116                                         0.018777   \n",
       "117                                         0.024294   \n",
       "118                                         0.024776   \n",
       "119                                         0.023679   \n",
       "120                                         0.052839   \n",
       "121                                         0.052923   \n",
       "122                                         0.055790   \n",
       "123                                         0.043030   \n",
       "124                                         0.047423   \n",
       "125                                         0.048203   \n",
       "126                                         0.043535   \n",
       "127                                         0.049480   \n",
       "128                                         0.036035   \n",
       "129                                         0.040941   \n",
       "130                                         0.016766   \n",
       "131                                         0.014548   \n",
       "132                                         0.020795   \n",
       "133                                         0.021564   \n",
       "134                                         0.020866   \n",
       "135                                         0.021261   \n",
       "136                                         0.020233   \n",
       "137                                         0.024413   \n",
       "138                                         0.023438   \n",
       "139                                         0.022245   \n",
       "140                                         0.055930   \n",
       "141                                         0.053094   \n",
       "142                                         0.057822   \n",
       "143                                         0.044367   \n",
       "144                                         0.046561   \n",
       "145                                         0.044166   \n",
       "146                                         0.043305   \n",
       "147                                         0.046833   \n",
       "148                                         0.035554   \n",
       "149                                         0.039160   \n",
       "150                                         0.001231   \n",
       "151                                         0.000000   \n",
       "152                                         0.002596   \n",
       "153                                         0.000000   \n",
       "154                                         0.002104   \n",
       "155                                         0.000000   \n",
       "156                                         0.000000   \n",
       "157                                         0.000000   \n",
       "158                                         0.000000   \n",
       "159                                         0.000000   \n",
       "160                                         0.000000   \n",
       "161                                         0.000000   \n",
       "162                                         0.000000   \n",
       "163                                         0.001048   \n",
       "164                                         0.001071   \n",
       "165                                         0.000000   \n",
       "166                                         0.000000   \n",
       "167                                         0.000000   \n",
       "168                                         0.000000   \n",
       "169                                         0.000000   \n",
       "170                                         0.006419   \n",
       "171                                         0.002998   \n",
       "172                                         0.007593   \n",
       "173                                         0.000000   \n",
       "174                                         0.005353   \n",
       "175                                         0.007360   \n",
       "176                                         0.000000   \n",
       "177                                         0.006859   \n",
       "178                                         0.000000   \n",
       "179                                         0.004447   \n",
       "180                                         0.000000   \n",
       "181                                         0.000000   \n",
       "182                                         0.000000   \n",
       "183                                         0.000000   \n",
       "184                                         0.000000   \n",
       "185                                         0.000000   \n",
       "186                                         0.000000   \n",
       "187                                         0.000000   \n",
       "188                                         0.000000   \n",
       "189                                         0.000000   \n",
       "190                                         0.009540   \n",
       "191                                         0.003198   \n",
       "192                                         0.009655   \n",
       "193                                         0.001149   \n",
       "194                                         0.004521   \n",
       "195                                         0.004478   \n",
       "196                                         0.000112   \n",
       "197                                         0.005368   \n",
       "198                                         0.000000   \n",
       "199                                         0.003822   \n",
       "\n",
       "     742kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                             0.004268    \n",
       "1                                             0.005621    \n",
       "2                                             0.004288    \n",
       "3                                             0.003760    \n",
       "4                                             0.004534    \n",
       "5                                             0.001491    \n",
       "6                                             0.003462    \n",
       "7                                             0.001566    \n",
       "8                                             0.001345    \n",
       "9                                             0.001596    \n",
       "10                                            0.001363    \n",
       "11                                            0.003355    \n",
       "12                                            0.001480    \n",
       "13                                            0.001231    \n",
       "14                                            0.001445    \n",
       "15                                            0.001409    \n",
       "16                                            0.003380    \n",
       "17                                            0.001521    \n",
       "18                                            0.001269    \n",
       "19                                            0.001494    \n",
       "20                                            0.006041    \n",
       "21                                            0.006906    \n",
       "22                                            0.006017    \n",
       "23                                            0.005306    \n",
       "24                                            0.006361    \n",
       "25                                            0.005126    \n",
       "26                                            0.006118    \n",
       "27                                            0.005112    \n",
       "28                                            0.004632    \n",
       "29                                            0.005634    \n",
       "30                                            0.001334    \n",
       "31                                            0.003355    \n",
       "32                                            0.001430    \n",
       "33                                            0.001210    \n",
       "34                                            0.001402    \n",
       "35                                            0.001487    \n",
       "36                                            0.003445    \n",
       "37                                            0.001593    \n",
       "38                                            0.001363    \n",
       "39                                            0.001583    \n",
       "40                                            0.006338    \n",
       "41                                            0.007069    \n",
       "42                                            0.006343    \n",
       "43                                            0.005573    \n",
       "44                                            0.006643    \n",
       "45                                            0.004942    \n",
       "46                                            0.005970    \n",
       "47                                            0.004923    \n",
       "48                                            0.004477    \n",
       "49                                            0.005440    \n",
       "50                                            0.017006    \n",
       "51                                            0.016964    \n",
       "52                                            0.016957    \n",
       "53                                            0.016979    \n",
       "54                                            0.017933    \n",
       "55                                            0.005987    \n",
       "56                                            0.007654    \n",
       "57                                            0.006011    \n",
       "58                                            0.006143    \n",
       "59                                            0.006463    \n",
       "60                                            0.005256    \n",
       "61                                            0.006950    \n",
       "62                                            0.005322    \n",
       "63                                            0.005388    \n",
       "64                                            0.005684    \n",
       "65                                            0.005435    \n",
       "66                                            0.007101    \n",
       "67                                            0.005510    \n",
       "68                                            0.005584    \n",
       "69                                            0.005877    \n",
       "70                                            0.020919    \n",
       "71                                            0.020081    \n",
       "72                                            0.020812    \n",
       "73                                            0.020713    \n",
       "74                                            0.021908    \n",
       "75                                            0.019040    \n",
       "76                                            0.018468    \n",
       "77                                            0.018940    \n",
       "78                                            0.019112    \n",
       "79                                            0.020306    \n",
       "80                                            0.005068    \n",
       "81                                            0.006821    \n",
       "82                                            0.005153    \n",
       "83                                            0.005252    \n",
       "84                                            0.005478    \n",
       "85                                            0.005937    \n",
       "86                                            0.007577    \n",
       "87                                            0.005976    \n",
       "88                                            0.006099    \n",
       "89                                            0.006397    \n",
       "90                                            0.021647    \n",
       "91                                            0.020649    \n",
       "92                                            0.021605    \n",
       "93                                            0.021463    \n",
       "94                                            0.022613    \n",
       "95                                            0.018787    \n",
       "96                                            0.018249    \n",
       "97                                            0.018657    \n",
       "98                                            0.018839    \n",
       "99                                            0.019992    \n",
       "100                                           0.060214    \n",
       "101                                           0.058218    \n",
       "102                                           0.060169    \n",
       "103                                           0.060870    \n",
       "104                                           0.061861    \n",
       "105                                           0.034397    \n",
       "106                                           0.035683    \n",
       "107                                           0.034424    \n",
       "108                                           0.035318    \n",
       "109                                           0.035712    \n",
       "110                                           0.032606    \n",
       "111                                           0.034113    \n",
       "112                                           0.032666    \n",
       "113                                           0.033438    \n",
       "114                                           0.033863    \n",
       "115                                           0.032833    \n",
       "116                                           0.034258    \n",
       "117                                           0.032904    \n",
       "118                                           0.033715    \n",
       "119                                           0.034115    \n",
       "120                                           0.066481    \n",
       "121                                           0.063654    \n",
       "122                                           0.066401    \n",
       "123                                           0.066962    \n",
       "124                                           0.068183    \n",
       "125                                           0.063681    \n",
       "126                                           0.061132    \n",
       "127                                           0.063592    \n",
       "128                                           0.064445    \n",
       "129                                           0.065663    \n",
       "130                                           0.031939    \n",
       "131                                           0.033598    \n",
       "132                                           0.032005    \n",
       "133                                           0.032852    \n",
       "134                                           0.033189    \n",
       "135                                           0.034631    \n",
       "136                                           0.035823    \n",
       "137                                           0.034660    \n",
       "138                                           0.035508    \n",
       "139                                           0.035920    \n",
       "140                                           0.067601    \n",
       "141                                           0.064611    \n",
       "142                                           0.067569    \n",
       "143                                           0.068086    \n",
       "144                                           0.069252    \n",
       "145                                           0.063216    \n",
       "146                                           0.060725    \n",
       "147                                           0.063113    \n",
       "148                                           0.063961    \n",
       "149                                           0.065149    \n",
       "150                                           0.002673    \n",
       "151                                           0.004606    \n",
       "152                                           0.002692    \n",
       "153                                           0.001883    \n",
       "154                                           0.002504    \n",
       "155                                           0.001362    \n",
       "156                                           0.004050    \n",
       "157                                           0.001431    \n",
       "158                                           0.000952    \n",
       "159                                           0.001193    \n",
       "160                                           0.001321    \n",
       "161                                           0.004164    \n",
       "162                                           0.001428    \n",
       "163                                           0.000935    \n",
       "164                                           0.001166    \n",
       "165                                           0.001349    \n",
       "166                                           0.004111    \n",
       "167                                           0.001455    \n",
       "168                                           0.000967    \n",
       "169                                           0.001189    \n",
       "170                                           0.003748    \n",
       "171                                           0.005263    \n",
       "172                                           0.003738    \n",
       "173                                           0.002739    \n",
       "174                                           0.003607    \n",
       "175                                           0.003084    \n",
       "176                                           0.004681    \n",
       "177                                           0.003071    \n",
       "178                                           0.002298    \n",
       "179                                           0.003098    \n",
       "180                                           0.001300    \n",
       "181                                           0.004187    \n",
       "182                                           0.001406    \n",
       "183                                           0.000931    \n",
       "184                                           0.001128    \n",
       "185                                           0.001350    \n",
       "186                                           0.004004    \n",
       "187                                           0.001445    \n",
       "188                                           0.000954    \n",
       "189                                           0.001180    \n",
       "190                                           0.003922    \n",
       "191                                           0.005299    \n",
       "192                                           0.003908    \n",
       "193                                           0.002850    \n",
       "194                                           0.003707    \n",
       "195                                           0.002975    \n",
       "196                                           0.004580    \n",
       "197                                           0.002945    \n",
       "198                                           0.002169    \n",
       "199                                           0.002952    \n",
       "\n",
       "     742kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                             0.006232       \n",
       "1                                             0.007057       \n",
       "2                                             0.006229       \n",
       "3                                             0.005823       \n",
       "4                                             0.006249       \n",
       "5                                             0.004624       \n",
       "6                                             0.007061       \n",
       "7                                             0.004953       \n",
       "8                                             0.004350       \n",
       "9                                             0.004769       \n",
       "10                                            0.004576       \n",
       "11                                            0.007149       \n",
       "12                                            0.004982       \n",
       "13                                            0.004295       \n",
       "14                                            0.004695       \n",
       "15                                            0.004641       \n",
       "16                                            0.007179       \n",
       "17                                            0.005029       \n",
       "18                                            0.004327       \n",
       "19                                            0.004737       \n",
       "20                                            0.007342       \n",
       "21                                            0.007777       \n",
       "22                                            0.007329       \n",
       "23                                            0.006877       \n",
       "24                                            0.007354       \n",
       "25                                            0.006790       \n",
       "26                                            0.007474       \n",
       "27                                            0.006870       \n",
       "28                                            0.006478       \n",
       "29                                            0.007010       \n",
       "30                                            0.004532       \n",
       "31                                            0.007255       \n",
       "32                                            0.004936       \n",
       "33                                            0.004353       \n",
       "34                                            0.004672       \n",
       "35                                            0.004641       \n",
       "36                                            0.007003       \n",
       "37                                            0.004960       \n",
       "38                                            0.004377       \n",
       "39                                            0.004716       \n",
       "40                                            0.007156       \n",
       "41                                            0.007561       \n",
       "42                                            0.007248       \n",
       "43                                            0.006799       \n",
       "44                                            0.007232       \n",
       "45                                            0.006804       \n",
       "46                                            0.007465       \n",
       "47                                            0.006836       \n",
       "48                                            0.006479       \n",
       "49                                            0.007002       \n",
       "50                                            0.014576       \n",
       "51                                            0.014552       \n",
       "52                                            0.014594       \n",
       "53                                            0.014481       \n",
       "54                                            0.014607       \n",
       "55                                            0.011866       \n",
       "56                                            0.013190       \n",
       "57                                            0.012065       \n",
       "58                                            0.012144       \n",
       "59                                            0.012273       \n",
       "60                                            0.011149       \n",
       "61                                            0.012709       \n",
       "62                                            0.011410       \n",
       "63                                            0.011437       \n",
       "64                                            0.011618       \n",
       "65                                            0.011446       \n",
       "66                                            0.012974       \n",
       "67                                            0.011677       \n",
       "68                                            0.011703       \n",
       "69                                            0.011834       \n",
       "70                                            0.015859       \n",
       "71                                            0.015681       \n",
       "72                                            0.015919       \n",
       "73                                            0.015691       \n",
       "74                                            0.015823       \n",
       "75                                            0.015499       \n",
       "76                                            0.015360       \n",
       "77                                            0.015562       \n",
       "78                                            0.015403       \n",
       "79                                            0.015586       \n",
       "80                                            0.011145       \n",
       "81                                            0.012868       \n",
       "82                                            0.011522       \n",
       "83                                            0.011504       \n",
       "84                                            0.011590       \n",
       "85                                            0.011682       \n",
       "86                                            0.013040       \n",
       "87                                            0.011859       \n",
       "88                                            0.011958       \n",
       "89                                            0.012092       \n",
       "90                                            0.015604       \n",
       "91                                            0.015376       \n",
       "92                                            0.015691       \n",
       "93                                            0.015463       \n",
       "94                                            0.015564       \n",
       "95                                            0.015486       \n",
       "96                                            0.015372       \n",
       "97                                            0.015496       \n",
       "98                                            0.015394       \n",
       "99                                            0.015591       \n",
       "100                                           0.034610       \n",
       "101                                           0.034190       \n",
       "102                                           0.034641       \n",
       "103                                           0.034536       \n",
       "104                                           0.034684       \n",
       "105                                           0.031146       \n",
       "106                                           0.031506       \n",
       "107                                           0.031349       \n",
       "108                                           0.031359       \n",
       "109                                           0.031498       \n",
       "110                                           0.030795       \n",
       "111                                           0.031287       \n",
       "112                                           0.030993       \n",
       "113                                           0.030974       \n",
       "114                                           0.031103       \n",
       "115                                           0.031045       \n",
       "116                                           0.031521       \n",
       "117                                           0.031231       \n",
       "118                                           0.031196       \n",
       "119                                           0.031309       \n",
       "120                                           0.036563       \n",
       "121                                           0.036016       \n",
       "122                                           0.036601       \n",
       "123                                           0.036440       \n",
       "124                                           0.036660       \n",
       "125                                           0.035872       \n",
       "126                                           0.035384       \n",
       "127                                           0.035911       \n",
       "128                                           0.035836       \n",
       "129                                           0.036047       \n",
       "130                                           0.030753       \n",
       "131                                           0.031336       \n",
       "132                                           0.031062       \n",
       "133                                           0.031044       \n",
       "134                                           0.031105       \n",
       "135                                           0.031097       \n",
       "136                                           0.031528       \n",
       "137                                           0.031266       \n",
       "138                                           0.031300       \n",
       "139                                           0.031401       \n",
       "140                                           0.036007       \n",
       "141                                           0.035405       \n",
       "142                                           0.036100       \n",
       "143                                           0.035949       \n",
       "144                                           0.036176       \n",
       "145                                           0.035986       \n",
       "146                                           0.035485       \n",
       "147                                           0.035958       \n",
       "148                                           0.035951       \n",
       "149                                           0.036187       \n",
       "150                                           0.004243       \n",
       "151                                           0.005818       \n",
       "152                                           0.004392       \n",
       "153                                           0.003477       \n",
       "154                                           0.004071       \n",
       "155                                           0.004149       \n",
       "156                                           0.007233       \n",
       "157                                           0.004500       \n",
       "158                                           0.003372       \n",
       "159                                           0.003797       \n",
       "160                                           0.004036       \n",
       "161                                           0.007313       \n",
       "162                                           0.004470       \n",
       "163                                           0.003314       \n",
       "164                                           0.003775       \n",
       "165                                           0.004107       \n",
       "166                                           0.007310       \n",
       "167                                           0.004532       \n",
       "168                                           0.003374       \n",
       "169                                           0.003844       \n",
       "170                                           0.004978       \n",
       "171                                           0.006134       \n",
       "172                                           0.005156       \n",
       "173                                           0.004188       \n",
       "174                                           0.004792       \n",
       "175                                           0.004571       \n",
       "176                                           0.005966       \n",
       "177                                           0.004765       \n",
       "178                                           0.003891       \n",
       "179                                           0.004575       \n",
       "180                                           0.004043       \n",
       "181                                           0.007409       \n",
       "182                                           0.004428       \n",
       "183                                           0.003321       \n",
       "184                                           0.003755       \n",
       "185                                           0.004061       \n",
       "186                                           0.007114       \n",
       "187                                           0.004450       \n",
       "188                                           0.003381       \n",
       "189                                           0.003735       \n",
       "190                                           0.004820       \n",
       "191                                           0.005786       \n",
       "192                                           0.005014       \n",
       "193                                           0.004095       \n",
       "194                                           0.004736       \n",
       "195                                           0.004586       \n",
       "196                                           0.005986       \n",
       "197                                           0.004717       \n",
       "198                                           0.003818       \n",
       "199                                           0.004528       \n",
       "\n",
       "     742kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                             0.000109    \n",
       "1                                             0.000123    \n",
       "2                                             0.000109    \n",
       "3                                             0.000102    \n",
       "4                                             0.000109    \n",
       "5                                             0.000081    \n",
       "6                                             0.000123    \n",
       "7                                             0.000087    \n",
       "8                                             0.000076    \n",
       "9                                             0.000083    \n",
       "10                                            0.000080    \n",
       "11                                            0.000125    \n",
       "12                                            0.000087    \n",
       "13                                            0.000075    \n",
       "14                                            0.000082    \n",
       "15                                            0.000081    \n",
       "16                                            0.000126    \n",
       "17                                            0.000088    \n",
       "18                                            0.000076    \n",
       "19                                            0.000083    \n",
       "20                                            0.000128    \n",
       "21                                            0.000136    \n",
       "22                                            0.000128    \n",
       "23                                            0.000120    \n",
       "24                                            0.000129    \n",
       "25                                            0.000119    \n",
       "26                                            0.000131    \n",
       "27                                            0.000120    \n",
       "28                                            0.000113    \n",
       "29                                            0.000123    \n",
       "30                                            0.000079    \n",
       "31                                            0.000127    \n",
       "32                                            0.000086    \n",
       "33                                            0.000076    \n",
       "34                                            0.000082    \n",
       "35                                            0.000081    \n",
       "36                                            0.000122    \n",
       "37                                            0.000087    \n",
       "38                                            0.000077    \n",
       "39                                            0.000082    \n",
       "40                                            0.000125    \n",
       "41                                            0.000132    \n",
       "42                                            0.000127    \n",
       "43                                            0.000119    \n",
       "44                                            0.000126    \n",
       "45                                            0.000119    \n",
       "46                                            0.000131    \n",
       "47                                            0.000120    \n",
       "48                                            0.000113    \n",
       "49                                            0.000122    \n",
       "50                                            0.000255    \n",
       "51                                            0.000254    \n",
       "52                                            0.000255    \n",
       "53                                            0.000253    \n",
       "54                                            0.000255    \n",
       "55                                            0.000207    \n",
       "56                                            0.000231    \n",
       "57                                            0.000211    \n",
       "58                                            0.000212    \n",
       "59                                            0.000215    \n",
       "60                                            0.000195    \n",
       "61                                            0.000222    \n",
       "62                                            0.000199    \n",
       "63                                            0.000200    \n",
       "64                                            0.000203    \n",
       "65                                            0.000200    \n",
       "66                                            0.000227    \n",
       "67                                            0.000204    \n",
       "68                                            0.000205    \n",
       "69                                            0.000207    \n",
       "70                                            0.000277    \n",
       "71                                            0.000274    \n",
       "72                                            0.000278    \n",
       "73                                            0.000274    \n",
       "74                                            0.000277    \n",
       "75                                            0.000271    \n",
       "76                                            0.000269    \n",
       "77                                            0.000272    \n",
       "78                                            0.000269    \n",
       "79                                            0.000272    \n",
       "80                                            0.000195    \n",
       "81                                            0.000225    \n",
       "82                                            0.000201    \n",
       "83                                            0.000201    \n",
       "84                                            0.000203    \n",
       "85                                            0.000204    \n",
       "86                                            0.000228    \n",
       "87                                            0.000207    \n",
       "88                                            0.000209    \n",
       "89                                            0.000211    \n",
       "90                                            0.000273    \n",
       "91                                            0.000269    \n",
       "92                                            0.000274    \n",
       "93                                            0.000270    \n",
       "94                                            0.000272    \n",
       "95                                            0.000271    \n",
       "96                                            0.000269    \n",
       "97                                            0.000271    \n",
       "98                                            0.000269    \n",
       "99                                            0.000273    \n",
       "100                                           0.000605    \n",
       "101                                           0.000598    \n",
       "102                                           0.000606    \n",
       "103                                           0.000604    \n",
       "104                                           0.000606    \n",
       "105                                           0.000544    \n",
       "106                                           0.000551    \n",
       "107                                           0.000548    \n",
       "108                                           0.000548    \n",
       "109                                           0.000551    \n",
       "110                                           0.000538    \n",
       "111                                           0.000547    \n",
       "112                                           0.000542    \n",
       "113                                           0.000541    \n",
       "114                                           0.000544    \n",
       "115                                           0.000543    \n",
       "116                                           0.000551    \n",
       "117                                           0.000546    \n",
       "118                                           0.000545    \n",
       "119                                           0.000547    \n",
       "120                                           0.000639    \n",
       "121                                           0.000630    \n",
       "122                                           0.000640    \n",
       "123                                           0.000637    \n",
       "124                                           0.000641    \n",
       "125                                           0.000627    \n",
       "126                                           0.000619    \n",
       "127                                           0.000628    \n",
       "128                                           0.000626    \n",
       "129                                           0.000630    \n",
       "130                                           0.000538    \n",
       "131                                           0.000548    \n",
       "132                                           0.000543    \n",
       "133                                           0.000543    \n",
       "134                                           0.000544    \n",
       "135                                           0.000544    \n",
       "136                                           0.000551    \n",
       "137                                           0.000547    \n",
       "138                                           0.000547    \n",
       "139                                           0.000549    \n",
       "140                                           0.000629    \n",
       "141                                           0.000619    \n",
       "142                                           0.000631    \n",
       "143                                           0.000628    \n",
       "144                                           0.000632    \n",
       "145                                           0.000629    \n",
       "146                                           0.000620    \n",
       "147                                           0.000629    \n",
       "148                                           0.000628    \n",
       "149                                           0.000633    \n",
       "150                                           0.000074    \n",
       "151                                           0.000102    \n",
       "152                                           0.000077    \n",
       "153                                           0.000061    \n",
       "154                                           0.000071    \n",
       "155                                           0.000073    \n",
       "156                                           0.000126    \n",
       "157                                           0.000079    \n",
       "158                                           0.000059    \n",
       "159                                           0.000066    \n",
       "160                                           0.000071    \n",
       "161                                           0.000128    \n",
       "162                                           0.000078    \n",
       "163                                           0.000058    \n",
       "164                                           0.000066    \n",
       "165                                           0.000072    \n",
       "166                                           0.000128    \n",
       "167                                           0.000079    \n",
       "168                                           0.000059    \n",
       "169                                           0.000067    \n",
       "170                                           0.000087    \n",
       "171                                           0.000107    \n",
       "172                                           0.000090    \n",
       "173                                           0.000073    \n",
       "174                                           0.000084    \n",
       "175                                           0.000080    \n",
       "176                                           0.000104    \n",
       "177                                           0.000083    \n",
       "178                                           0.000068    \n",
       "179                                           0.000080    \n",
       "180                                           0.000071    \n",
       "181                                           0.000130    \n",
       "182                                           0.000077    \n",
       "183                                           0.000058    \n",
       "184                                           0.000066    \n",
       "185                                           0.000071    \n",
       "186                                           0.000124    \n",
       "187                                           0.000078    \n",
       "188                                           0.000059    \n",
       "189                                           0.000065    \n",
       "190                                           0.000084    \n",
       "191                                           0.000101    \n",
       "192                                           0.000088    \n",
       "193                                           0.000072    \n",
       "194                                           0.000083    \n",
       "195                                           0.000080    \n",
       "196                                           0.000105    \n",
       "197                                           0.000082    \n",
       "198                                           0.000067    \n",
       "199                                           0.000079    \n",
       "\n",
       "     742kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                             0.000214              0.257562  \n",
       "1                                             0.000242              1.000000  \n",
       "2                                             0.000214              0.201344  \n",
       "3                                             0.000200              0.425298  \n",
       "4                                             0.000214              0.289337  \n",
       "5                                             0.000159              1.000000  \n",
       "6                                             0.000242              1.000000  \n",
       "7                                             0.000170              1.000000  \n",
       "8                                             0.000149              1.000000  \n",
       "9                                             0.000163              1.000000  \n",
       "10                                            0.000157              1.000000  \n",
       "11                                            0.000245              1.000000  \n",
       "12                                            0.000171              0.185151  \n",
       "13                                            0.000147              0.095325  \n",
       "14                                            0.000161              0.137489  \n",
       "15                                            0.000159              1.000000  \n",
       "16                                            0.000246              1.000000  \n",
       "17                                            0.000172              1.000000  \n",
       "18                                            0.000148              0.111213  \n",
       "19                                            0.000162              0.160403  \n",
       "20                                            0.000252              0.192789  \n",
       "21                                            0.000267              0.751299  \n",
       "22                                            0.000251              0.159487  \n",
       "23                                            0.000236              0.457684  \n",
       "24                                            0.000252              0.294531  \n",
       "25                                            0.000233              0.106936  \n",
       "26                                            0.000256              1.000000  \n",
       "27                                            0.000235              0.116407  \n",
       "28                                            0.000222              0.384662  \n",
       "29                                            0.000240              0.227009  \n",
       "30                                            0.000155              1.000000  \n",
       "31                                            0.000249              1.000000  \n",
       "32                                            0.000169              1.000000  \n",
       "33                                            0.000149              0.116712  \n",
       "34                                            0.000160              0.158265  \n",
       "35                                            0.000159              1.000000  \n",
       "36                                            0.000240              1.000000  \n",
       "37                                            0.000170              1.000000  \n",
       "38                                            0.000150              0.122823  \n",
       "39                                            0.000162              0.169569  \n",
       "40                                            0.000245              0.094714  \n",
       "41                                            0.000259              0.672166  \n",
       "42                                            0.000248              0.097159  \n",
       "43                                            0.000233              0.308585  \n",
       "44                                            0.000248              0.278032  \n",
       "45                                            0.000233              0.199817  \n",
       "46                                            0.000256              1.000000  \n",
       "47                                            0.000234              0.167430  \n",
       "48                                            0.000222              0.402994  \n",
       "49                                            0.000240              0.300336  \n",
       "50                                            0.000500              0.007333  \n",
       "51                                            0.000499              0.007333  \n",
       "52                                            0.000500              0.004888  \n",
       "53                                            0.000496              0.009777  \n",
       "54                                            0.000501              0.010694  \n",
       "55                                            0.000407              0.009777  \n",
       "56                                            0.000452              0.014971  \n",
       "57                                            0.000414              0.007333  \n",
       "58                                            0.000416              0.010082  \n",
       "59                                            0.000421              0.009777  \n",
       "60                                            0.000382              0.003972  \n",
       "61                                            0.000436              0.006416  \n",
       "62                                            0.000391              0.003972  \n",
       "63                                            0.000392              0.003972  \n",
       "64                                            0.000398              0.004583  \n",
       "65                                            0.000392              0.004583  \n",
       "66                                            0.000445              0.008555  \n",
       "67                                            0.000400              0.004583  \n",
       "68                                            0.000401              0.004277  \n",
       "69                                            0.000406              0.005500  \n",
       "70                                            0.000544              0.004277  \n",
       "71                                            0.000538              0.005805  \n",
       "72                                            0.000546              0.004277  \n",
       "73                                            0.000538              0.008249  \n",
       "74                                            0.000542              0.007333  \n",
       "75                                            0.000531              0.006111  \n",
       "76                                            0.000527              0.007638  \n",
       "77                                            0.000533              0.005194  \n",
       "78                                            0.000528              0.010082  \n",
       "79                                            0.000534              0.007638  \n",
       "80                                            0.000382              0.005194  \n",
       "81                                            0.000441              0.008555  \n",
       "82                                            0.000395              0.003972  \n",
       "83                                            0.000394              0.003361  \n",
       "84                                            0.000397              0.003972  \n",
       "85                                            0.000400              0.006111  \n",
       "86                                            0.000447              0.009777  \n",
       "87                                            0.000407              0.005194  \n",
       "88                                            0.000410              0.005805  \n",
       "89                                            0.000414              0.006722  \n",
       "90                                            0.000535              0.002750  \n",
       "91                                            0.000527              0.003972  \n",
       "92                                            0.000538              0.002139  \n",
       "93                                            0.000530              0.005500  \n",
       "94                                            0.000534              0.005805  \n",
       "95                                            0.000531              0.007333  \n",
       "96                                            0.000527              0.008249  \n",
       "97                                            0.000531              0.006111  \n",
       "98                                            0.000528              0.009777  \n",
       "99                                            0.000534              0.010694  \n",
       "100                                           0.001186              0.722884  \n",
       "101                                           0.001172              0.695692  \n",
       "102                                           0.001187              0.675527  \n",
       "103                                           0.001184              0.826459  \n",
       "104                                           0.001189              0.788268  \n",
       "105                                           0.001068              0.626948  \n",
       "106                                           0.001080              0.658723  \n",
       "107                                           0.001075              0.585396  \n",
       "108                                           0.001075              0.687137  \n",
       "109                                           0.001080              0.623587  \n",
       "110                                           0.001056              0.555454  \n",
       "111                                           0.001072              0.609838  \n",
       "112                                           0.001062              0.481515  \n",
       "113                                           0.001062              0.519707  \n",
       "114                                           0.001066              0.541399  \n",
       "115                                           0.001064              0.607088  \n",
       "116                                           0.001080              0.645585  \n",
       "117                                           0.001071              0.534067  \n",
       "118                                           0.001069              0.537427  \n",
       "119                                           0.001073              0.558509  \n",
       "120                                           0.001253              0.590284  \n",
       "121                                           0.001235              0.556370  \n",
       "122                                           0.001255              0.553315  \n",
       "123                                           0.001249              0.742744  \n",
       "124                                           0.001257              0.698136  \n",
       "125                                           0.001230              0.623587  \n",
       "126                                           0.001213              0.661167  \n",
       "127                                           0.001231              0.603727  \n",
       "128                                           0.001228              0.805988  \n",
       "129                                           0.001236              0.750687  \n",
       "130                                           0.001054              0.639474  \n",
       "131                                           0.001074              0.705469  \n",
       "132                                           0.001065              0.575619  \n",
       "133                                           0.001064              0.581118  \n",
       "134                                           0.001066              0.597006  \n",
       "135                                           0.001066              0.614421  \n",
       "136                                           0.001081              0.652612  \n",
       "137                                           0.001072              0.562481  \n",
       "138                                           0.001073              0.593645  \n",
       "139                                           0.001076              0.619310  \n",
       "140                                           0.001234              0.570425  \n",
       "141                                           0.001214              0.573480  \n",
       "142                                           0.001237              0.541094  \n",
       "143                                           0.001232              0.738772  \n",
       "144                                           0.001240              0.724412  \n",
       "145                                           0.001233              0.681943  \n",
       "146                                           0.001216              0.661167  \n",
       "147                                           0.001233              0.641002  \n",
       "148                                           0.001232              0.809044  \n",
       "149                                           0.001240              0.770241  \n",
       "150                                           0.000145              0.449435  \n",
       "151                                           0.000199              1.000000  \n",
       "152                                           0.000151              0.338527  \n",
       "153                                           0.000119              1.000000  \n",
       "154                                           0.000140              0.359609  \n",
       "155                                           0.000142              1.000000  \n",
       "156                                           0.000248              1.000000  \n",
       "157                                           0.000154              1.000000  \n",
       "158                                           0.000116              1.000000  \n",
       "159                                           0.000130              1.000000  \n",
       "160                                           0.000138              1.000000  \n",
       "161                                           0.000251              1.000000  \n",
       "162                                           0.000153              1.000000  \n",
       "163                                           0.000114              0.139627  \n",
       "164                                           0.000129              0.166208  \n",
       "165                                           0.000141              1.000000  \n",
       "166                                           0.000251              1.000000  \n",
       "167                                           0.000155              1.000000  \n",
       "168                                           0.000116              1.000000  \n",
       "169                                           0.000132              1.000000  \n",
       "170                                           0.000171              0.213566  \n",
       "171                                           0.000210              0.546593  \n",
       "172                                           0.000177              0.170180  \n",
       "173                                           0.000144              1.000000  \n",
       "174                                           0.000164              0.255118  \n",
       "175                                           0.000157              0.139322  \n",
       "176                                           0.000205              1.000000  \n",
       "177                                           0.000163              0.157959  \n",
       "178                                           0.000133              1.000000  \n",
       "179                                           0.000157              0.270394  \n",
       "180                                           0.000139              1.000000  \n",
       "181                                           0.000254              1.000000  \n",
       "182                                           0.000152              1.000000  \n",
       "183                                           0.000114              1.000000  \n",
       "184                                           0.000129              1.000000  \n",
       "185                                           0.000139              1.000000  \n",
       "186                                           0.000244              1.000000  \n",
       "187                                           0.000153              1.000000  \n",
       "188                                           0.000116              1.000000  \n",
       "189                                           0.000128              1.000000  \n",
       "190                                           0.000165              0.106936  \n",
       "191                                           0.000198              0.554537  \n",
       "192                                           0.000172              0.108463  \n",
       "193                                           0.000140              0.509624  \n",
       "194                                           0.000162              0.323862  \n",
       "195                                           0.000157              0.250535  \n",
       "196                                           0.000205              0.672777  \n",
       "197                                           0.000162              0.200428  \n",
       "198                                           0.000131              1.000000  \n",
       "199                                           0.000155              0.289948  "
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2da597ee",
   "metadata": {},
   "source": [
    "## S52"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "238b71a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/amr_asn_eur_pbs_per_region_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S52. $\\boldsymbol{PBS_{AMR:EAS/SAS:EUR}}$ values for the focal 72kb \\textit{MUC19} region.} \\newline The $PBS_{A:B:C}$ values observed in the focal 72kb region for all unique combinations of: AMR populations as the target population (\\textit{A}), EAS and SAS populations as one of the control populations (\\textit{B}), and EUR populations as the second control population (\\textit{C}). For each unique $PBS_{A:B:C}$ configuration, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{A:B:C}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the $PBS_{A:B:C}$ value is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant. Note that when the target population is MXL, $PBS_{MXL:B:C}$ is always significant, and is never significant any other AMR population is the target population.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/amr_asn_eur_pbs_per_region_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            ('0' if mat[i, j] == 0 else\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, 8))+r' & '\\\n",
    "        + (\n",
    "            r'$>0.999968$' if mat[i, 8] == 1 else\n",
    "            convert_sci_notation_to_latex(float(mat[i, 8])) if float(mat[i, 8]) < 0.001 else\n",
    "            '{:.3f}'.format(round(float(mat[i, 8]), 3)).rstrip('0').rstrip('.')\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1938c6fe",
   "metadata": {},
   "source": [
    "\\textbf{Table S52. $\\boldsymbol{PBS_{AMR:EAS/SAS:EUR}}$ values for the focal 72kb \\textit{MUC19} region.} \\newline The $PBS_{A:B:C}$ values observed in the focal 72kb region for all unique combinations of: AMR populations as the target population (\\textit{A}), EAS and SAS populations as one of the control populations (\\textit{B}), and EUR populations as the second control population (\\textit{C}). For each unique $PBS_{A:B:C}$ configuration, the mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the $PBS_{A:B:C}$  genomic background distribution used to compute the \\textit{P-value} are also reported. The \\textit{P-value} represents the proportion of non-overlapping 72kb windows of comparable effective sequence length where the $PBS_{A:B:C}$ value is greater than or equal to that observed at the focal 72kb region. A \\textit{P-value} less than 0.05 is considered statistically significant. Note that when the target population is MXL, $PBS_{MXL:B:C}$ is always significant, and is never significant any other AMR population is the target population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "e2cbc99f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$A$ Pop.</th>\n",
       "      <th>$B$ Pop.</th>\n",
       "      <th>$C$ Pop.</th>\n",
       "      <th>Focal 72kb Region $\\left( PBS_{A:B:C} \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\mu \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\sigma \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( SEM \\right)$</th>\n",
       "      <th>72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006200</td>\n",
       "      <td>0.010694</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007522</td>\n",
       "      <td>0.011968</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006258</td>\n",
       "      <td>0.010736</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005641</td>\n",
       "      <td>0.010115</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CLM</td>\n",
       "      <td>BEB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006430</td>\n",
       "      <td>0.010848</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004148</td>\n",
       "      <td>0.010036</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006781</td>\n",
       "      <td>0.013645</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000150</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004312</td>\n",
       "      <td>0.010417</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000115</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003848</td>\n",
       "      <td>0.009490</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CDX</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004210</td>\n",
       "      <td>0.010086</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.002491</td>\n",
       "      <td>0.003865</td>\n",
       "      <td>0.009727</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.247648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006560</td>\n",
       "      <td>0.013542</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.001481</td>\n",
       "      <td>0.004028</td>\n",
       "      <td>0.010126</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>0.272997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.009497</td>\n",
       "      <td>0.003557</td>\n",
       "      <td>0.009183</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>0.127596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003920</td>\n",
       "      <td>0.009781</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.001167</td>\n",
       "      <td>0.003950</td>\n",
       "      <td>0.009829</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.284582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006615</td>\n",
       "      <td>0.013535</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000561</td>\n",
       "      <td>0.004128</td>\n",
       "      <td>0.010274</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.299640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.008803</td>\n",
       "      <td>0.003652</td>\n",
       "      <td>0.009289</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.138645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CLM</td>\n",
       "      <td>CHS</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004016</td>\n",
       "      <td>0.009866</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.002446</td>\n",
       "      <td>0.007691</td>\n",
       "      <td>0.012046</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.521056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008560</td>\n",
       "      <td>0.012927</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000481</td>\n",
       "      <td>0.007729</td>\n",
       "      <td>0.012098</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.608435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.003510</td>\n",
       "      <td>0.006969</td>\n",
       "      <td>0.011324</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.445672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CLM</td>\n",
       "      <td>GIH</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.002923</td>\n",
       "      <td>0.007982</td>\n",
       "      <td>0.012136</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>0.000134</td>\n",
       "      <td>0.515531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.001326</td>\n",
       "      <td>0.007005</td>\n",
       "      <td>0.011549</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.527748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008031</td>\n",
       "      <td>0.012637</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007048</td>\n",
       "      <td>0.011657</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.001983</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>0.011005</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.475314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CLM</td>\n",
       "      <td>ITU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.002724</td>\n",
       "      <td>0.007466</td>\n",
       "      <td>0.011861</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.487026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003766</td>\n",
       "      <td>0.009506</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006511</td>\n",
       "      <td>0.013390</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>0.000147</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003937</td>\n",
       "      <td>0.009962</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.006979</td>\n",
       "      <td>0.003504</td>\n",
       "      <td>0.009140</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>0.156891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CLM</td>\n",
       "      <td>JPT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003817</td>\n",
       "      <td>0.009576</td>\n",
       "      <td>0.000054</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004067</td>\n",
       "      <td>0.009915</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006653</td>\n",
       "      <td>0.013479</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004238</td>\n",
       "      <td>0.010312</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.006029</td>\n",
       "      <td>0.003776</td>\n",
       "      <td>0.009407</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.183503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>CLM</td>\n",
       "      <td>KHV</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004150</td>\n",
       "      <td>0.009990</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.008293</td>\n",
       "      <td>0.007744</td>\n",
       "      <td>0.011648</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>0.320917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000577</td>\n",
       "      <td>0.008500</td>\n",
       "      <td>0.012435</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>0.646726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.006425</td>\n",
       "      <td>0.007824</td>\n",
       "      <td>0.011824</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.379285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.004662</td>\n",
       "      <td>0.007034</td>\n",
       "      <td>0.011046</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000122</td>\n",
       "      <td>0.411989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CLM</td>\n",
       "      <td>PJL</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.003408</td>\n",
       "      <td>0.008023</td>\n",
       "      <td>0.011836</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.510859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000385</td>\n",
       "      <td>0.006885</td>\n",
       "      <td>0.011537</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.562914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007930</td>\n",
       "      <td>0.012555</td>\n",
       "      <td>0.000071</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006911</td>\n",
       "      <td>0.011607</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.003495</td>\n",
       "      <td>0.006354</td>\n",
       "      <td>0.011018</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.405865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CLM</td>\n",
       "      <td>STU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.002485</td>\n",
       "      <td>0.007303</td>\n",
       "      <td>0.011814</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.488099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.087021</td>\n",
       "      <td>0.018218</td>\n",
       "      <td>0.023108</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.020898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.088450</td>\n",
       "      <td>0.018121</td>\n",
       "      <td>0.023044</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.019067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.088198</td>\n",
       "      <td>0.018175</td>\n",
       "      <td>0.023101</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0.019351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.076737</td>\n",
       "      <td>0.018197</td>\n",
       "      <td>0.022961</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.029421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>MXL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.074149</td>\n",
       "      <td>0.019077</td>\n",
       "      <td>0.023408</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>0.034535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.091409</td>\n",
       "      <td>0.009932</td>\n",
       "      <td>0.019563</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000215</td>\n",
       "      <td>0.009376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.092824</td>\n",
       "      <td>0.011646</td>\n",
       "      <td>0.021500</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.000237</td>\n",
       "      <td>0.011901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.092345</td>\n",
       "      <td>0.010005</td>\n",
       "      <td>0.019712</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>0.000217</td>\n",
       "      <td>0.008776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.080727</td>\n",
       "      <td>0.010120</td>\n",
       "      <td>0.019666</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000217</td>\n",
       "      <td>0.014111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.076255</td>\n",
       "      <td>0.010454</td>\n",
       "      <td>0.020131</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>0.017678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.127226</td>\n",
       "      <td>0.008982</td>\n",
       "      <td>0.018510</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.002241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.125368</td>\n",
       "      <td>0.010778</td>\n",
       "      <td>0.020604</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.003630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.130704</td>\n",
       "      <td>0.009066</td>\n",
       "      <td>0.018715</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000206</td>\n",
       "      <td>0.002304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.119850</td>\n",
       "      <td>0.009128</td>\n",
       "      <td>0.018648</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.003062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.106435</td>\n",
       "      <td>0.009486</td>\n",
       "      <td>0.019188</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0.005051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.126237</td>\n",
       "      <td>0.009222</td>\n",
       "      <td>0.018741</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000206</td>\n",
       "      <td>0.002589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.124784</td>\n",
       "      <td>0.010992</td>\n",
       "      <td>0.020845</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>0.000230</td>\n",
       "      <td>0.003914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.130119</td>\n",
       "      <td>0.009319</td>\n",
       "      <td>0.018959</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.002431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.119492</td>\n",
       "      <td>0.009380</td>\n",
       "      <td>0.018881</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.003157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>MXL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.106977</td>\n",
       "      <td>0.009722</td>\n",
       "      <td>0.019342</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.005430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.119690</td>\n",
       "      <td>0.021586</td>\n",
       "      <td>0.025493</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>0.009439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.121857</td>\n",
       "      <td>0.020856</td>\n",
       "      <td>0.025038</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>0.000276</td>\n",
       "      <td>0.008271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.122213</td>\n",
       "      <td>0.021522</td>\n",
       "      <td>0.025507</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>0.008839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.106372</td>\n",
       "      <td>0.021421</td>\n",
       "      <td>0.025233</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.014837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>MXL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.102799</td>\n",
       "      <td>0.022530</td>\n",
       "      <td>0.025710</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>0.000283</td>\n",
       "      <td>0.017741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.111729</td>\n",
       "      <td>0.020096</td>\n",
       "      <td>0.024803</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.011017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.112130</td>\n",
       "      <td>0.019571</td>\n",
       "      <td>0.024492</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.000270</td>\n",
       "      <td>0.010323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.113283</td>\n",
       "      <td>0.020017</td>\n",
       "      <td>0.024814</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.010228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.098005</td>\n",
       "      <td>0.020163</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.017646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>MXL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.095760</td>\n",
       "      <td>0.021251</td>\n",
       "      <td>0.025204</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.020266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.125634</td>\n",
       "      <td>0.008753</td>\n",
       "      <td>0.018275</td>\n",
       "      <td>0.000103</td>\n",
       "      <td>0.000201</td>\n",
       "      <td>0.001862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.124391</td>\n",
       "      <td>0.010604</td>\n",
       "      <td>0.020510</td>\n",
       "      <td>0.000115</td>\n",
       "      <td>0.000226</td>\n",
       "      <td>0.004072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.129711</td>\n",
       "      <td>0.008874</td>\n",
       "      <td>0.018668</td>\n",
       "      <td>0.000105</td>\n",
       "      <td>0.000206</td>\n",
       "      <td>0.002083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.119215</td>\n",
       "      <td>0.008950</td>\n",
       "      <td>0.018571</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>0.000205</td>\n",
       "      <td>0.002873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>MXL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.107683</td>\n",
       "      <td>0.009219</td>\n",
       "      <td>0.018941</td>\n",
       "      <td>0.000106</td>\n",
       "      <td>0.000209</td>\n",
       "      <td>0.004893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.100661</td>\n",
       "      <td>0.009704</td>\n",
       "      <td>0.019225</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>0.006819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.098704</td>\n",
       "      <td>0.011371</td>\n",
       "      <td>0.021173</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>0.000233</td>\n",
       "      <td>0.009754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.103333</td>\n",
       "      <td>0.009768</td>\n",
       "      <td>0.019326</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.005966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.095366</td>\n",
       "      <td>0.009877</td>\n",
       "      <td>0.019348</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.008081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>MXL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.084330</td>\n",
       "      <td>0.010222</td>\n",
       "      <td>0.019824</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>0.000218</td>\n",
       "      <td>0.012343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.147047</td>\n",
       "      <td>0.022055</td>\n",
       "      <td>0.025053</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>0.000276</td>\n",
       "      <td>0.004104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.150584</td>\n",
       "      <td>0.021184</td>\n",
       "      <td>0.024577</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>0.003125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.149667</td>\n",
       "      <td>0.022045</td>\n",
       "      <td>0.025131</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>0.003630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.129035</td>\n",
       "      <td>0.021903</td>\n",
       "      <td>0.024813</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.006314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>MXL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.124794</td>\n",
       "      <td>0.022984</td>\n",
       "      <td>0.025270</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.007513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.106629</td>\n",
       "      <td>0.019899</td>\n",
       "      <td>0.024849</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>0.000274</td>\n",
       "      <td>0.013448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.107419</td>\n",
       "      <td>0.019408</td>\n",
       "      <td>0.024503</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0.000270</td>\n",
       "      <td>0.012311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.108795</td>\n",
       "      <td>0.019799</td>\n",
       "      <td>0.024777</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>0.011459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.095357</td>\n",
       "      <td>0.019947</td>\n",
       "      <td>0.024714</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.018436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>MXL</td>\n",
       "      <td>STU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.091362</td>\n",
       "      <td>0.021003</td>\n",
       "      <td>0.025211</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.022918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.038576</td>\n",
       "      <td>0.059391</td>\n",
       "      <td>0.055058</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.000606</td>\n",
       "      <td>0.565566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.040104</td>\n",
       "      <td>0.057433</td>\n",
       "      <td>0.053972</td>\n",
       "      <td>0.000303</td>\n",
       "      <td>0.000594</td>\n",
       "      <td>0.534756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.038553</td>\n",
       "      <td>0.059354</td>\n",
       "      <td>0.055086</td>\n",
       "      <td>0.000310</td>\n",
       "      <td>0.000607</td>\n",
       "      <td>0.565598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.030040</td>\n",
       "      <td>0.060044</td>\n",
       "      <td>0.055003</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.000606</td>\n",
       "      <td>0.667624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>PEL</td>\n",
       "      <td>BEB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.031507</td>\n",
       "      <td>0.061048</td>\n",
       "      <td>0.055490</td>\n",
       "      <td>0.000312</td>\n",
       "      <td>0.000611</td>\n",
       "      <td>0.661058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.038193</td>\n",
       "      <td>0.037228</td>\n",
       "      <td>0.046961</td>\n",
       "      <td>0.000264</td>\n",
       "      <td>0.000517</td>\n",
       "      <td>0.353242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.039707</td>\n",
       "      <td>0.038186</td>\n",
       "      <td>0.047518</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>0.350401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.037929</td>\n",
       "      <td>0.037268</td>\n",
       "      <td>0.047137</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>0.000519</td>\n",
       "      <td>0.356399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.029258</td>\n",
       "      <td>0.038030</td>\n",
       "      <td>0.047302</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.000521</td>\n",
       "      <td>0.441095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CDX</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.028842</td>\n",
       "      <td>0.038409</td>\n",
       "      <td>0.047775</td>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.000526</td>\n",
       "      <td>0.447787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.059090</td>\n",
       "      <td>0.035392</td>\n",
       "      <td>0.045817</td>\n",
       "      <td>0.000257</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.207242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.057332</td>\n",
       "      <td>0.036553</td>\n",
       "      <td>0.046544</td>\n",
       "      <td>0.000262</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.220784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.061369</td>\n",
       "      <td>0.035439</td>\n",
       "      <td>0.045972</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>0.000506</td>\n",
       "      <td>0.196666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.053462</td>\n",
       "      <td>0.036116</td>\n",
       "      <td>0.046125</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.000508</td>\n",
       "      <td>0.242155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.044103</td>\n",
       "      <td>0.036520</td>\n",
       "      <td>0.046609</td>\n",
       "      <td>0.000262</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.302166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.059318</td>\n",
       "      <td>0.035729</td>\n",
       "      <td>0.046086</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.000508</td>\n",
       "      <td>0.209104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.057964</td>\n",
       "      <td>0.036830</td>\n",
       "      <td>0.046762</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.000515</td>\n",
       "      <td>0.220658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.062000</td>\n",
       "      <td>0.035801</td>\n",
       "      <td>0.046263</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.000509</td>\n",
       "      <td>0.197266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.054320</td>\n",
       "      <td>0.036491</td>\n",
       "      <td>0.046393</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.000511</td>\n",
       "      <td>0.240830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PEL</td>\n",
       "      <td>CHS</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.045861</td>\n",
       "      <td>0.036877</td>\n",
       "      <td>0.046832</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.000516</td>\n",
       "      <td>0.296420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.064684</td>\n",
       "      <td>0.065359</td>\n",
       "      <td>0.058749</td>\n",
       "      <td>0.000330</td>\n",
       "      <td>0.000647</td>\n",
       "      <td>0.381527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.066950</td>\n",
       "      <td>0.062623</td>\n",
       "      <td>0.057337</td>\n",
       "      <td>0.000322</td>\n",
       "      <td>0.000631</td>\n",
       "      <td>0.343803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.066007</td>\n",
       "      <td>0.065310</td>\n",
       "      <td>0.058802</td>\n",
       "      <td>0.000330</td>\n",
       "      <td>0.000648</td>\n",
       "      <td>0.370794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.053113</td>\n",
       "      <td>0.065849</td>\n",
       "      <td>0.058656</td>\n",
       "      <td>0.000330</td>\n",
       "      <td>0.000646</td>\n",
       "      <td>0.476924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PEL</td>\n",
       "      <td>GIH</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.053595</td>\n",
       "      <td>0.067094</td>\n",
       "      <td>0.059176</td>\n",
       "      <td>0.000332</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.484911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.058154</td>\n",
       "      <td>0.062763</td>\n",
       "      <td>0.057686</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.409622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.058655</td>\n",
       "      <td>0.060272</td>\n",
       "      <td>0.056420</td>\n",
       "      <td>0.000317</td>\n",
       "      <td>0.000621</td>\n",
       "      <td>0.387840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.058508</td>\n",
       "      <td>0.062693</td>\n",
       "      <td>0.057713</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000636</td>\n",
       "      <td>0.408675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.046177</td>\n",
       "      <td>0.063526</td>\n",
       "      <td>0.057701</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.521371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>PEL</td>\n",
       "      <td>ITU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.047988</td>\n",
       "      <td>0.064740</td>\n",
       "      <td>0.058215</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.513037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.058746</td>\n",
       "      <td>0.034801</td>\n",
       "      <td>0.045529</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>0.000501</td>\n",
       "      <td>0.204811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.057603</td>\n",
       "      <td>0.036074</td>\n",
       "      <td>0.046335</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.000510</td>\n",
       "      <td>0.215386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.061623</td>\n",
       "      <td>0.034865</td>\n",
       "      <td>0.045845</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>0.000505</td>\n",
       "      <td>0.192721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.054074</td>\n",
       "      <td>0.035611</td>\n",
       "      <td>0.046027</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>0.234642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PEL</td>\n",
       "      <td>JPT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.046599</td>\n",
       "      <td>0.035919</td>\n",
       "      <td>0.046387</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.000511</td>\n",
       "      <td>0.281173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.040479</td>\n",
       "      <td>0.037155</td>\n",
       "      <td>0.046829</td>\n",
       "      <td>0.000263</td>\n",
       "      <td>0.000516</td>\n",
       "      <td>0.335091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.038621</td>\n",
       "      <td>0.038050</td>\n",
       "      <td>0.047381</td>\n",
       "      <td>0.000266</td>\n",
       "      <td>0.000522</td>\n",
       "      <td>0.355325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.041951</td>\n",
       "      <td>0.037183</td>\n",
       "      <td>0.046952</td>\n",
       "      <td>0.000264</td>\n",
       "      <td>0.000517</td>\n",
       "      <td>0.325557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.036932</td>\n",
       "      <td>0.037929</td>\n",
       "      <td>0.047142</td>\n",
       "      <td>0.000265</td>\n",
       "      <td>0.000519</td>\n",
       "      <td>0.371362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>PEL</td>\n",
       "      <td>KHV</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.029951</td>\n",
       "      <td>0.038332</td>\n",
       "      <td>0.047602</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>0.438885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.085147</td>\n",
       "      <td>0.066402</td>\n",
       "      <td>0.058103</td>\n",
       "      <td>0.000326</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>0.263842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.088784</td>\n",
       "      <td>0.063493</td>\n",
       "      <td>0.056678</td>\n",
       "      <td>0.000318</td>\n",
       "      <td>0.000624</td>\n",
       "      <td>0.229497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.086568</td>\n",
       "      <td>0.066405</td>\n",
       "      <td>0.058229</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.256550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.068883</td>\n",
       "      <td>0.066916</td>\n",
       "      <td>0.058015</td>\n",
       "      <td>0.000326</td>\n",
       "      <td>0.000639</td>\n",
       "      <td>0.363943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>PEL</td>\n",
       "      <td>PJL</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.068698</td>\n",
       "      <td>0.068094</td>\n",
       "      <td>0.058591</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.000645</td>\n",
       "      <td>0.376570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.053461</td>\n",
       "      <td>0.062325</td>\n",
       "      <td>0.057623</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>0.445735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.054350</td>\n",
       "      <td>0.059895</td>\n",
       "      <td>0.056303</td>\n",
       "      <td>0.000316</td>\n",
       "      <td>0.000620</td>\n",
       "      <td>0.418682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.054427</td>\n",
       "      <td>0.062230</td>\n",
       "      <td>0.057578</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000634</td>\n",
       "      <td>0.438191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.043936</td>\n",
       "      <td>0.063063</td>\n",
       "      <td>0.057611</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000634</td>\n",
       "      <td>0.538418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>PEL</td>\n",
       "      <td>STU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.043996</td>\n",
       "      <td>0.064254</td>\n",
       "      <td>0.058133</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.000640</td>\n",
       "      <td>0.546310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004538</td>\n",
       "      <td>0.007951</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.000088</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006561</td>\n",
       "      <td>0.010292</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004581</td>\n",
       "      <td>0.008099</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003609</td>\n",
       "      <td>0.006799</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>PUR</td>\n",
       "      <td>BEB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004318</td>\n",
       "      <td>0.007659</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003905</td>\n",
       "      <td>0.008939</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007473</td>\n",
       "      <td>0.013753</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0.000151</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004030</td>\n",
       "      <td>0.009260</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003130</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CDX</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003498</td>\n",
       "      <td>0.008299</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.001894</td>\n",
       "      <td>0.003831</td>\n",
       "      <td>0.008930</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>0.285877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007506</td>\n",
       "      <td>0.013849</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.001622</td>\n",
       "      <td>0.003973</td>\n",
       "      <td>0.009221</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.295852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.005910</td>\n",
       "      <td>0.003045</td>\n",
       "      <td>0.007501</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.166014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHB</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003433</td>\n",
       "      <td>0.008248</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003862</td>\n",
       "      <td>0.008963</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007500</td>\n",
       "      <td>0.013837</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000152</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003998</td>\n",
       "      <td>0.009250</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.004057</td>\n",
       "      <td>0.003079</td>\n",
       "      <td>0.007536</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.204116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>PUR</td>\n",
       "      <td>CHS</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003470</td>\n",
       "      <td>0.008284</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0.005491</td>\n",
       "      <td>0.008919</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>0.576867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007115</td>\n",
       "      <td>0.010817</td>\n",
       "      <td>0.000061</td>\n",
       "      <td>0.000119</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005532</td>\n",
       "      <td>0.009093</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004404</td>\n",
       "      <td>0.007703</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>PUR</td>\n",
       "      <td>GIH</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005292</td>\n",
       "      <td>0.008577</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004969</td>\n",
       "      <td>0.008509</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006700</td>\n",
       "      <td>0.010554</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004999</td>\n",
       "      <td>0.008685</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.000096</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004060</td>\n",
       "      <td>0.007425</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>PUR</td>\n",
       "      <td>ITU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004924</td>\n",
       "      <td>0.008398</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003809</td>\n",
       "      <td>0.008892</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007537</td>\n",
       "      <td>0.013858</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003937</td>\n",
       "      <td>0.009164</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.004512</td>\n",
       "      <td>0.003040</td>\n",
       "      <td>0.007450</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.193131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>PUR</td>\n",
       "      <td>JPT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003376</td>\n",
       "      <td>0.008105</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003859</td>\n",
       "      <td>0.008866</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007369</td>\n",
       "      <td>0.013571</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003998</td>\n",
       "      <td>0.009192</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000101</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003087</td>\n",
       "      <td>0.007515</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>PUR</td>\n",
       "      <td>KHV</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003467</td>\n",
       "      <td>0.008194</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.008323</td>\n",
       "      <td>0.005426</td>\n",
       "      <td>0.008476</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.230412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.007623</td>\n",
       "      <td>0.006911</td>\n",
       "      <td>0.010112</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>0.315676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.007193</td>\n",
       "      <td>0.005482</td>\n",
       "      <td>0.008665</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.266557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.001703</td>\n",
       "      <td>0.004351</td>\n",
       "      <td>0.007353</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.454764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>PUR</td>\n",
       "      <td>PJL</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.002283</td>\n",
       "      <td>0.005226</td>\n",
       "      <td>0.008249</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>0.476482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004868</td>\n",
       "      <td>0.008455</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006639</td>\n",
       "      <td>0.010557</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004889</td>\n",
       "      <td>0.008620</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003942</td>\n",
       "      <td>0.007308</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>PUR</td>\n",
       "      <td>STU</td>\n",
       "      <td>TSI</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004795</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>0.000091</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    $A$ Pop. $B$ Pop. $C$ Pop.  \\\n",
       "0        CLM      BEB      CEU   \n",
       "1        CLM      BEB      FIN   \n",
       "2        CLM      BEB      GBR   \n",
       "3        CLM      BEB      IBS   \n",
       "4        CLM      BEB      TSI   \n",
       "5        CLM      CDX      CEU   \n",
       "6        CLM      CDX      FIN   \n",
       "7        CLM      CDX      GBR   \n",
       "8        CLM      CDX      IBS   \n",
       "9        CLM      CDX      TSI   \n",
       "10       CLM      CHB      CEU   \n",
       "11       CLM      CHB      FIN   \n",
       "12       CLM      CHB      GBR   \n",
       "13       CLM      CHB      IBS   \n",
       "14       CLM      CHB      TSI   \n",
       "15       CLM      CHS      CEU   \n",
       "16       CLM      CHS      FIN   \n",
       "17       CLM      CHS      GBR   \n",
       "18       CLM      CHS      IBS   \n",
       "19       CLM      CHS      TSI   \n",
       "20       CLM      GIH      CEU   \n",
       "21       CLM      GIH      FIN   \n",
       "22       CLM      GIH      GBR   \n",
       "23       CLM      GIH      IBS   \n",
       "24       CLM      GIH      TSI   \n",
       "25       CLM      ITU      CEU   \n",
       "26       CLM      ITU      FIN   \n",
       "27       CLM      ITU      GBR   \n",
       "28       CLM      ITU      IBS   \n",
       "29       CLM      ITU      TSI   \n",
       "30       CLM      JPT      CEU   \n",
       "31       CLM      JPT      FIN   \n",
       "32       CLM      JPT      GBR   \n",
       "33       CLM      JPT      IBS   \n",
       "34       CLM      JPT      TSI   \n",
       "35       CLM      KHV      CEU   \n",
       "36       CLM      KHV      FIN   \n",
       "37       CLM      KHV      GBR   \n",
       "38       CLM      KHV      IBS   \n",
       "39       CLM      KHV      TSI   \n",
       "40       CLM      PJL      CEU   \n",
       "41       CLM      PJL      FIN   \n",
       "42       CLM      PJL      GBR   \n",
       "43       CLM      PJL      IBS   \n",
       "44       CLM      PJL      TSI   \n",
       "45       CLM      STU      CEU   \n",
       "46       CLM      STU      FIN   \n",
       "47       CLM      STU      GBR   \n",
       "48       CLM      STU      IBS   \n",
       "49       CLM      STU      TSI   \n",
       "50       MXL      BEB      CEU   \n",
       "51       MXL      BEB      FIN   \n",
       "52       MXL      BEB      GBR   \n",
       "53       MXL      BEB      IBS   \n",
       "54       MXL      BEB      TSI   \n",
       "55       MXL      CDX      CEU   \n",
       "56       MXL      CDX      FIN   \n",
       "57       MXL      CDX      GBR   \n",
       "58       MXL      CDX      IBS   \n",
       "59       MXL      CDX      TSI   \n",
       "60       MXL      CHB      CEU   \n",
       "61       MXL      CHB      FIN   \n",
       "62       MXL      CHB      GBR   \n",
       "63       MXL      CHB      IBS   \n",
       "64       MXL      CHB      TSI   \n",
       "65       MXL      CHS      CEU   \n",
       "66       MXL      CHS      FIN   \n",
       "67       MXL      CHS      GBR   \n",
       "68       MXL      CHS      IBS   \n",
       "69       MXL      CHS      TSI   \n",
       "70       MXL      GIH      CEU   \n",
       "71       MXL      GIH      FIN   \n",
       "72       MXL      GIH      GBR   \n",
       "73       MXL      GIH      IBS   \n",
       "74       MXL      GIH      TSI   \n",
       "75       MXL      ITU      CEU   \n",
       "76       MXL      ITU      FIN   \n",
       "77       MXL      ITU      GBR   \n",
       "78       MXL      ITU      IBS   \n",
       "79       MXL      ITU      TSI   \n",
       "80       MXL      JPT      CEU   \n",
       "81       MXL      JPT      FIN   \n",
       "82       MXL      JPT      GBR   \n",
       "83       MXL      JPT      IBS   \n",
       "84       MXL      JPT      TSI   \n",
       "85       MXL      KHV      CEU   \n",
       "86       MXL      KHV      FIN   \n",
       "87       MXL      KHV      GBR   \n",
       "88       MXL      KHV      IBS   \n",
       "89       MXL      KHV      TSI   \n",
       "90       MXL      PJL      CEU   \n",
       "91       MXL      PJL      FIN   \n",
       "92       MXL      PJL      GBR   \n",
       "93       MXL      PJL      IBS   \n",
       "94       MXL      PJL      TSI   \n",
       "95       MXL      STU      CEU   \n",
       "96       MXL      STU      FIN   \n",
       "97       MXL      STU      GBR   \n",
       "98       MXL      STU      IBS   \n",
       "99       MXL      STU      TSI   \n",
       "100      PEL      BEB      CEU   \n",
       "101      PEL      BEB      FIN   \n",
       "102      PEL      BEB      GBR   \n",
       "103      PEL      BEB      IBS   \n",
       "104      PEL      BEB      TSI   \n",
       "105      PEL      CDX      CEU   \n",
       "106      PEL      CDX      FIN   \n",
       "107      PEL      CDX      GBR   \n",
       "108      PEL      CDX      IBS   \n",
       "109      PEL      CDX      TSI   \n",
       "110      PEL      CHB      CEU   \n",
       "111      PEL      CHB      FIN   \n",
       "112      PEL      CHB      GBR   \n",
       "113      PEL      CHB      IBS   \n",
       "114      PEL      CHB      TSI   \n",
       "115      PEL      CHS      CEU   \n",
       "116      PEL      CHS      FIN   \n",
       "117      PEL      CHS      GBR   \n",
       "118      PEL      CHS      IBS   \n",
       "119      PEL      CHS      TSI   \n",
       "120      PEL      GIH      CEU   \n",
       "121      PEL      GIH      FIN   \n",
       "122      PEL      GIH      GBR   \n",
       "123      PEL      GIH      IBS   \n",
       "124      PEL      GIH      TSI   \n",
       "125      PEL      ITU      CEU   \n",
       "126      PEL      ITU      FIN   \n",
       "127      PEL      ITU      GBR   \n",
       "128      PEL      ITU      IBS   \n",
       "129      PEL      ITU      TSI   \n",
       "130      PEL      JPT      CEU   \n",
       "131      PEL      JPT      FIN   \n",
       "132      PEL      JPT      GBR   \n",
       "133      PEL      JPT      IBS   \n",
       "134      PEL      JPT      TSI   \n",
       "135      PEL      KHV      CEU   \n",
       "136      PEL      KHV      FIN   \n",
       "137      PEL      KHV      GBR   \n",
       "138      PEL      KHV      IBS   \n",
       "139      PEL      KHV      TSI   \n",
       "140      PEL      PJL      CEU   \n",
       "141      PEL      PJL      FIN   \n",
       "142      PEL      PJL      GBR   \n",
       "143      PEL      PJL      IBS   \n",
       "144      PEL      PJL      TSI   \n",
       "145      PEL      STU      CEU   \n",
       "146      PEL      STU      FIN   \n",
       "147      PEL      STU      GBR   \n",
       "148      PEL      STU      IBS   \n",
       "149      PEL      STU      TSI   \n",
       "150      PUR      BEB      CEU   \n",
       "151      PUR      BEB      FIN   \n",
       "152      PUR      BEB      GBR   \n",
       "153      PUR      BEB      IBS   \n",
       "154      PUR      BEB      TSI   \n",
       "155      PUR      CDX      CEU   \n",
       "156      PUR      CDX      FIN   \n",
       "157      PUR      CDX      GBR   \n",
       "158      PUR      CDX      IBS   \n",
       "159      PUR      CDX      TSI   \n",
       "160      PUR      CHB      CEU   \n",
       "161      PUR      CHB      FIN   \n",
       "162      PUR      CHB      GBR   \n",
       "163      PUR      CHB      IBS   \n",
       "164      PUR      CHB      TSI   \n",
       "165      PUR      CHS      CEU   \n",
       "166      PUR      CHS      FIN   \n",
       "167      PUR      CHS      GBR   \n",
       "168      PUR      CHS      IBS   \n",
       "169      PUR      CHS      TSI   \n",
       "170      PUR      GIH      CEU   \n",
       "171      PUR      GIH      FIN   \n",
       "172      PUR      GIH      GBR   \n",
       "173      PUR      GIH      IBS   \n",
       "174      PUR      GIH      TSI   \n",
       "175      PUR      ITU      CEU   \n",
       "176      PUR      ITU      FIN   \n",
       "177      PUR      ITU      GBR   \n",
       "178      PUR      ITU      IBS   \n",
       "179      PUR      ITU      TSI   \n",
       "180      PUR      JPT      CEU   \n",
       "181      PUR      JPT      FIN   \n",
       "182      PUR      JPT      GBR   \n",
       "183      PUR      JPT      IBS   \n",
       "184      PUR      JPT      TSI   \n",
       "185      PUR      KHV      CEU   \n",
       "186      PUR      KHV      FIN   \n",
       "187      PUR      KHV      GBR   \n",
       "188      PUR      KHV      IBS   \n",
       "189      PUR      KHV      TSI   \n",
       "190      PUR      PJL      CEU   \n",
       "191      PUR      PJL      FIN   \n",
       "192      PUR      PJL      GBR   \n",
       "193      PUR      PJL      IBS   \n",
       "194      PUR      PJL      TSI   \n",
       "195      PUR      STU      CEU   \n",
       "196      PUR      STU      FIN   \n",
       "197      PUR      STU      GBR   \n",
       "198      PUR      STU      IBS   \n",
       "199      PUR      STU      TSI   \n",
       "\n",
       "     Focal 72kb Region $\\left( PBS_{A:B:C} \\right)$  \\\n",
       "0                                          0.000000   \n",
       "1                                          0.000000   \n",
       "2                                          0.000000   \n",
       "3                                          0.000000   \n",
       "4                                          0.000000   \n",
       "5                                          0.000000   \n",
       "6                                          0.000000   \n",
       "7                                          0.000000   \n",
       "8                                          0.000000   \n",
       "9                                          0.000000   \n",
       "10                                         0.002491   \n",
       "11                                         0.000000   \n",
       "12                                         0.001481   \n",
       "13                                         0.009497   \n",
       "14                                         0.000000   \n",
       "15                                         0.001167   \n",
       "16                                         0.000000   \n",
       "17                                         0.000561   \n",
       "18                                         0.008803   \n",
       "19                                         0.000000   \n",
       "20                                         0.002446   \n",
       "21                                         0.000000   \n",
       "22                                         0.000481   \n",
       "23                                         0.003510   \n",
       "24                                         0.002923   \n",
       "25                                         0.001326   \n",
       "26                                         0.000000   \n",
       "27                                         0.000000   \n",
       "28                                         0.001983   \n",
       "29                                         0.002724   \n",
       "30                                         0.000000   \n",
       "31                                         0.000000   \n",
       "32                                         0.000000   \n",
       "33                                         0.006979   \n",
       "34                                         0.000000   \n",
       "35                                         0.000000   \n",
       "36                                         0.000000   \n",
       "37                                         0.000000   \n",
       "38                                         0.006029   \n",
       "39                                         0.000000   \n",
       "40                                         0.008293   \n",
       "41                                         0.000577   \n",
       "42                                         0.006425   \n",
       "43                                         0.004662   \n",
       "44                                         0.003408   \n",
       "45                                         0.000385   \n",
       "46                                         0.000000   \n",
       "47                                         0.000000   \n",
       "48                                         0.003495   \n",
       "49                                         0.002485   \n",
       "50                                         0.087021   \n",
       "51                                         0.088450   \n",
       "52                                         0.088198   \n",
       "53                                         0.076737   \n",
       "54                                         0.074149   \n",
       "55                                         0.091409   \n",
       "56                                         0.092824   \n",
       "57                                         0.092345   \n",
       "58                                         0.080727   \n",
       "59                                         0.076255   \n",
       "60                                         0.127226   \n",
       "61                                         0.125368   \n",
       "62                                         0.130704   \n",
       "63                                         0.119850   \n",
       "64                                         0.106435   \n",
       "65                                         0.126237   \n",
       "66                                         0.124784   \n",
       "67                                         0.130119   \n",
       "68                                         0.119492   \n",
       "69                                         0.106977   \n",
       "70                                         0.119690   \n",
       "71                                         0.121857   \n",
       "72                                         0.122213   \n",
       "73                                         0.106372   \n",
       "74                                         0.102799   \n",
       "75                                         0.111729   \n",
       "76                                         0.112130   \n",
       "77                                         0.113283   \n",
       "78                                         0.098005   \n",
       "79                                         0.095760   \n",
       "80                                         0.125634   \n",
       "81                                         0.124391   \n",
       "82                                         0.129711   \n",
       "83                                         0.119215   \n",
       "84                                         0.107683   \n",
       "85                                         0.100661   \n",
       "86                                         0.098704   \n",
       "87                                         0.103333   \n",
       "88                                         0.095366   \n",
       "89                                         0.084330   \n",
       "90                                         0.147047   \n",
       "91                                         0.150584   \n",
       "92                                         0.149667   \n",
       "93                                         0.129035   \n",
       "94                                         0.124794   \n",
       "95                                         0.106629   \n",
       "96                                         0.107419   \n",
       "97                                         0.108795   \n",
       "98                                         0.095357   \n",
       "99                                         0.091362   \n",
       "100                                        0.038576   \n",
       "101                                        0.040104   \n",
       "102                                        0.038553   \n",
       "103                                        0.030040   \n",
       "104                                        0.031507   \n",
       "105                                        0.038193   \n",
       "106                                        0.039707   \n",
       "107                                        0.037929   \n",
       "108                                        0.029258   \n",
       "109                                        0.028842   \n",
       "110                                        0.059090   \n",
       "111                                        0.057332   \n",
       "112                                        0.061369   \n",
       "113                                        0.053462   \n",
       "114                                        0.044103   \n",
       "115                                        0.059318   \n",
       "116                                        0.057964   \n",
       "117                                        0.062000   \n",
       "118                                        0.054320   \n",
       "119                                        0.045861   \n",
       "120                                        0.064684   \n",
       "121                                        0.066950   \n",
       "122                                        0.066007   \n",
       "123                                        0.053113   \n",
       "124                                        0.053595   \n",
       "125                                        0.058154   \n",
       "126                                        0.058655   \n",
       "127                                        0.058508   \n",
       "128                                        0.046177   \n",
       "129                                        0.047988   \n",
       "130                                        0.058746   \n",
       "131                                        0.057603   \n",
       "132                                        0.061623   \n",
       "133                                        0.054074   \n",
       "134                                        0.046599   \n",
       "135                                        0.040479   \n",
       "136                                        0.038621   \n",
       "137                                        0.041951   \n",
       "138                                        0.036932   \n",
       "139                                        0.029951   \n",
       "140                                        0.085147   \n",
       "141                                        0.088784   \n",
       "142                                        0.086568   \n",
       "143                                        0.068883   \n",
       "144                                        0.068698   \n",
       "145                                        0.053461   \n",
       "146                                        0.054350   \n",
       "147                                        0.054427   \n",
       "148                                        0.043936   \n",
       "149                                        0.043996   \n",
       "150                                        0.000000   \n",
       "151                                        0.000000   \n",
       "152                                        0.000000   \n",
       "153                                        0.000000   \n",
       "154                                        0.000000   \n",
       "155                                        0.000000   \n",
       "156                                        0.000000   \n",
       "157                                        0.000000   \n",
       "158                                        0.000000   \n",
       "159                                        0.000000   \n",
       "160                                        0.001894   \n",
       "161                                        0.000000   \n",
       "162                                        0.001622   \n",
       "163                                        0.005910   \n",
       "164                                        0.000000   \n",
       "165                                        0.000000   \n",
       "166                                        0.000000   \n",
       "167                                        0.000000   \n",
       "168                                        0.004057   \n",
       "169                                        0.000000   \n",
       "170                                        0.000211   \n",
       "171                                        0.000000   \n",
       "172                                        0.000000   \n",
       "173                                        0.000000   \n",
       "174                                        0.000000   \n",
       "175                                        0.000000   \n",
       "176                                        0.000000   \n",
       "177                                        0.000000   \n",
       "178                                        0.000000   \n",
       "179                                        0.000000   \n",
       "180                                        0.000000   \n",
       "181                                        0.000000   \n",
       "182                                        0.000000   \n",
       "183                                        0.004512   \n",
       "184                                        0.000000   \n",
       "185                                        0.000000   \n",
       "186                                        0.000000   \n",
       "187                                        0.000000   \n",
       "188                                        0.000000   \n",
       "189                                        0.000000   \n",
       "190                                        0.008323   \n",
       "191                                        0.007623   \n",
       "192                                        0.007193   \n",
       "193                                        0.001703   \n",
       "194                                        0.002283   \n",
       "195                                        0.000000   \n",
       "196                                        0.000000   \n",
       "197                                        0.000000   \n",
       "198                                        0.000000   \n",
       "199                                        0.000000   \n",
       "\n",
       "     72kb Non-overlapping Windows $\\left( \\mu \\right)$  \\\n",
       "0                                             0.006200   \n",
       "1                                             0.007522   \n",
       "2                                             0.006258   \n",
       "3                                             0.005641   \n",
       "4                                             0.006430   \n",
       "5                                             0.004148   \n",
       "6                                             0.006781   \n",
       "7                                             0.004312   \n",
       "8                                             0.003848   \n",
       "9                                             0.004210   \n",
       "10                                            0.003865   \n",
       "11                                            0.006560   \n",
       "12                                            0.004028   \n",
       "13                                            0.003557   \n",
       "14                                            0.003920   \n",
       "15                                            0.003950   \n",
       "16                                            0.006615   \n",
       "17                                            0.004128   \n",
       "18                                            0.003652   \n",
       "19                                            0.004016   \n",
       "20                                            0.007691   \n",
       "21                                            0.008560   \n",
       "22                                            0.007729   \n",
       "23                                            0.006969   \n",
       "24                                            0.007982   \n",
       "25                                            0.007005   \n",
       "26                                            0.008031   \n",
       "27                                            0.007048   \n",
       "28                                            0.006496   \n",
       "29                                            0.007466   \n",
       "30                                            0.003766   \n",
       "31                                            0.006511   \n",
       "32                                            0.003937   \n",
       "33                                            0.003504   \n",
       "34                                            0.003817   \n",
       "35                                            0.004067   \n",
       "36                                            0.006653   \n",
       "37                                            0.004238   \n",
       "38                                            0.003776   \n",
       "39                                            0.004150   \n",
       "40                                            0.007744   \n",
       "41                                            0.008500   \n",
       "42                                            0.007824   \n",
       "43                                            0.007034   \n",
       "44                                            0.008023   \n",
       "45                                            0.006885   \n",
       "46                                            0.007930   \n",
       "47                                            0.006911   \n",
       "48                                            0.006354   \n",
       "49                                            0.007303   \n",
       "50                                            0.018218   \n",
       "51                                            0.018121   \n",
       "52                                            0.018175   \n",
       "53                                            0.018197   \n",
       "54                                            0.019077   \n",
       "55                                            0.009932   \n",
       "56                                            0.011646   \n",
       "57                                            0.010005   \n",
       "58                                            0.010120   \n",
       "59                                            0.010454   \n",
       "60                                            0.008982   \n",
       "61                                            0.010778   \n",
       "62                                            0.009066   \n",
       "63                                            0.009128   \n",
       "64                                            0.009486   \n",
       "65                                            0.009222   \n",
       "66                                            0.010992   \n",
       "67                                            0.009319   \n",
       "68                                            0.009380   \n",
       "69                                            0.009722   \n",
       "70                                            0.021586   \n",
       "71                                            0.020856   \n",
       "72                                            0.021522   \n",
       "73                                            0.021421   \n",
       "74                                            0.022530   \n",
       "75                                            0.020096   \n",
       "76                                            0.019571   \n",
       "77                                            0.020017   \n",
       "78                                            0.020163   \n",
       "79                                            0.021251   \n",
       "80                                            0.008753   \n",
       "81                                            0.010604   \n",
       "82                                            0.008874   \n",
       "83                                            0.008950   \n",
       "84                                            0.009219   \n",
       "85                                            0.009704   \n",
       "86                                            0.011371   \n",
       "87                                            0.009768   \n",
       "88                                            0.009877   \n",
       "89                                            0.010222   \n",
       "90                                            0.022055   \n",
       "91                                            0.021184   \n",
       "92                                            0.022045   \n",
       "93                                            0.021903   \n",
       "94                                            0.022984   \n",
       "95                                            0.019899   \n",
       "96                                            0.019408   \n",
       "97                                            0.019799   \n",
       "98                                            0.019947   \n",
       "99                                            0.021003   \n",
       "100                                           0.059391   \n",
       "101                                           0.057433   \n",
       "102                                           0.059354   \n",
       "103                                           0.060044   \n",
       "104                                           0.061048   \n",
       "105                                           0.037228   \n",
       "106                                           0.038186   \n",
       "107                                           0.037268   \n",
       "108                                           0.038030   \n",
       "109                                           0.038409   \n",
       "110                                           0.035392   \n",
       "111                                           0.036553   \n",
       "112                                           0.035439   \n",
       "113                                           0.036116   \n",
       "114                                           0.036520   \n",
       "115                                           0.035729   \n",
       "116                                           0.036830   \n",
       "117                                           0.035801   \n",
       "118                                           0.036491   \n",
       "119                                           0.036877   \n",
       "120                                           0.065359   \n",
       "121                                           0.062623   \n",
       "122                                           0.065310   \n",
       "123                                           0.065849   \n",
       "124                                           0.067094   \n",
       "125                                           0.062763   \n",
       "126                                           0.060272   \n",
       "127                                           0.062693   \n",
       "128                                           0.063526   \n",
       "129                                           0.064740   \n",
       "130                                           0.034801   \n",
       "131                                           0.036074   \n",
       "132                                           0.034865   \n",
       "133                                           0.035611   \n",
       "134                                           0.035919   \n",
       "135                                           0.037155   \n",
       "136                                           0.038050   \n",
       "137                                           0.037183   \n",
       "138                                           0.037929   \n",
       "139                                           0.038332   \n",
       "140                                           0.066402   \n",
       "141                                           0.063493   \n",
       "142                                           0.066405   \n",
       "143                                           0.066916   \n",
       "144                                           0.068094   \n",
       "145                                           0.062325   \n",
       "146                                           0.059895   \n",
       "147                                           0.062230   \n",
       "148                                           0.063063   \n",
       "149                                           0.064254   \n",
       "150                                           0.004538   \n",
       "151                                           0.006561   \n",
       "152                                           0.004581   \n",
       "153                                           0.003609   \n",
       "154                                           0.004318   \n",
       "155                                           0.003905   \n",
       "156                                           0.007473   \n",
       "157                                           0.004030   \n",
       "158                                           0.003130   \n",
       "159                                           0.003498   \n",
       "160                                           0.003831   \n",
       "161                                           0.007506   \n",
       "162                                           0.003973   \n",
       "163                                           0.003045   \n",
       "164                                           0.003433   \n",
       "165                                           0.003862   \n",
       "166                                           0.007500   \n",
       "167                                           0.003998   \n",
       "168                                           0.003079   \n",
       "169                                           0.003470   \n",
       "170                                           0.005491   \n",
       "171                                           0.007115   \n",
       "172                                           0.005532   \n",
       "173                                           0.004404   \n",
       "174                                           0.005292   \n",
       "175                                           0.004969   \n",
       "176                                           0.006700   \n",
       "177                                           0.004999   \n",
       "178                                           0.004060   \n",
       "179                                           0.004924   \n",
       "180                                           0.003809   \n",
       "181                                           0.007537   \n",
       "182                                           0.003937   \n",
       "183                                           0.003040   \n",
       "184                                           0.003376   \n",
       "185                                           0.003859   \n",
       "186                                           0.007369   \n",
       "187                                           0.003998   \n",
       "188                                           0.003087   \n",
       "189                                           0.003467   \n",
       "190                                           0.005426   \n",
       "191                                           0.006911   \n",
       "192                                           0.005482   \n",
       "193                                           0.004351   \n",
       "194                                           0.005226   \n",
       "195                                           0.004868   \n",
       "196                                           0.006639   \n",
       "197                                           0.004889   \n",
       "198                                           0.003942   \n",
       "199                                           0.004795   \n",
       "\n",
       "     72kb Non-overlapping Windows $\\left( \\sigma \\right)$  \\\n",
       "0                                             0.010694      \n",
       "1                                             0.011968      \n",
       "2                                             0.010736      \n",
       "3                                             0.010115      \n",
       "4                                             0.010848      \n",
       "5                                             0.010036      \n",
       "6                                             0.013645      \n",
       "7                                             0.010417      \n",
       "8                                             0.009490      \n",
       "9                                             0.010086      \n",
       "10                                            0.009727      \n",
       "11                                            0.013542      \n",
       "12                                            0.010126      \n",
       "13                                            0.009183      \n",
       "14                                            0.009781      \n",
       "15                                            0.009829      \n",
       "16                                            0.013535      \n",
       "17                                            0.010274      \n",
       "18                                            0.009289      \n",
       "19                                            0.009866      \n",
       "20                                            0.012046      \n",
       "21                                            0.012927      \n",
       "22                                            0.012098      \n",
       "23                                            0.011324      \n",
       "24                                            0.012136      \n",
       "25                                            0.011549      \n",
       "26                                            0.012637      \n",
       "27                                            0.011657      \n",
       "28                                            0.011005      \n",
       "29                                            0.011861      \n",
       "30                                            0.009506      \n",
       "31                                            0.013390      \n",
       "32                                            0.009962      \n",
       "33                                            0.009140      \n",
       "34                                            0.009576      \n",
       "35                                            0.009915      \n",
       "36                                            0.013479      \n",
       "37                                            0.010312      \n",
       "38                                            0.009407      \n",
       "39                                            0.009990      \n",
       "40                                            0.011648      \n",
       "41                                            0.012435      \n",
       "42                                            0.011824      \n",
       "43                                            0.011046      \n",
       "44                                            0.011836      \n",
       "45                                            0.011537      \n",
       "46                                            0.012555      \n",
       "47                                            0.011607      \n",
       "48                                            0.011018      \n",
       "49                                            0.011814      \n",
       "50                                            0.023108      \n",
       "51                                            0.023044      \n",
       "52                                            0.023101      \n",
       "53                                            0.022961      \n",
       "54                                            0.023408      \n",
       "55                                            0.019563      \n",
       "56                                            0.021500      \n",
       "57                                            0.019712      \n",
       "58                                            0.019666      \n",
       "59                                            0.020131      \n",
       "60                                            0.018510      \n",
       "61                                            0.020604      \n",
       "62                                            0.018715      \n",
       "63                                            0.018648      \n",
       "64                                            0.019188      \n",
       "65                                            0.018741      \n",
       "66                                            0.020845      \n",
       "67                                            0.018959      \n",
       "68                                            0.018881      \n",
       "69                                            0.019342      \n",
       "70                                            0.025493      \n",
       "71                                            0.025038      \n",
       "72                                            0.025507      \n",
       "73                                            0.025233      \n",
       "74                                            0.025710      \n",
       "75                                            0.024803      \n",
       "76                                            0.024492      \n",
       "77                                            0.024814      \n",
       "78                                            0.024691      \n",
       "79                                            0.025204      \n",
       "80                                            0.018275      \n",
       "81                                            0.020510      \n",
       "82                                            0.018668      \n",
       "83                                            0.018571      \n",
       "84                                            0.018941      \n",
       "85                                            0.019225      \n",
       "86                                            0.021173      \n",
       "87                                            0.019326      \n",
       "88                                            0.019348      \n",
       "89                                            0.019824      \n",
       "90                                            0.025053      \n",
       "91                                            0.024577      \n",
       "92                                            0.025131      \n",
       "93                                            0.024813      \n",
       "94                                            0.025270      \n",
       "95                                            0.024849      \n",
       "96                                            0.024503      \n",
       "97                                            0.024777      \n",
       "98                                            0.024714      \n",
       "99                                            0.025211      \n",
       "100                                           0.055058      \n",
       "101                                           0.053972      \n",
       "102                                           0.055086      \n",
       "103                                           0.055003      \n",
       "104                                           0.055490      \n",
       "105                                           0.046961      \n",
       "106                                           0.047518      \n",
       "107                                           0.047137      \n",
       "108                                           0.047302      \n",
       "109                                           0.047775      \n",
       "110                                           0.045817      \n",
       "111                                           0.046544      \n",
       "112                                           0.045972      \n",
       "113                                           0.046125      \n",
       "114                                           0.046609      \n",
       "115                                           0.046086      \n",
       "116                                           0.046762      \n",
       "117                                           0.046263      \n",
       "118                                           0.046393      \n",
       "119                                           0.046832      \n",
       "120                                           0.058749      \n",
       "121                                           0.057337      \n",
       "122                                           0.058802      \n",
       "123                                           0.058656      \n",
       "124                                           0.059176      \n",
       "125                                           0.057686      \n",
       "126                                           0.056420      \n",
       "127                                           0.057713      \n",
       "128                                           0.057701      \n",
       "129                                           0.058215      \n",
       "130                                           0.045529      \n",
       "131                                           0.046335      \n",
       "132                                           0.045845      \n",
       "133                                           0.046027      \n",
       "134                                           0.046387      \n",
       "135                                           0.046829      \n",
       "136                                           0.047381      \n",
       "137                                           0.046952      \n",
       "138                                           0.047142      \n",
       "139                                           0.047602      \n",
       "140                                           0.058103      \n",
       "141                                           0.056678      \n",
       "142                                           0.058229      \n",
       "143                                           0.058015      \n",
       "144                                           0.058591      \n",
       "145                                           0.057623      \n",
       "146                                           0.056303      \n",
       "147                                           0.057578      \n",
       "148                                           0.057611      \n",
       "149                                           0.058133      \n",
       "150                                           0.007951      \n",
       "151                                           0.010292      \n",
       "152                                           0.008099      \n",
       "153                                           0.006799      \n",
       "154                                           0.007659      \n",
       "155                                           0.008939      \n",
       "156                                           0.013753      \n",
       "157                                           0.009260      \n",
       "158                                           0.007570      \n",
       "159                                           0.008299      \n",
       "160                                           0.008930      \n",
       "161                                           0.013849      \n",
       "162                                           0.009221      \n",
       "163                                           0.007501      \n",
       "164                                           0.008248      \n",
       "165                                           0.008963      \n",
       "166                                           0.013837      \n",
       "167                                           0.009250      \n",
       "168                                           0.007536      \n",
       "169                                           0.008284      \n",
       "170                                           0.008919      \n",
       "171                                           0.010817      \n",
       "172                                           0.009093      \n",
       "173                                           0.007703      \n",
       "174                                           0.008577      \n",
       "175                                           0.008509      \n",
       "176                                           0.010554      \n",
       "177                                           0.008685      \n",
       "178                                           0.007425      \n",
       "179                                           0.008398      \n",
       "180                                           0.008892      \n",
       "181                                           0.013858      \n",
       "182                                           0.009164      \n",
       "183                                           0.007450      \n",
       "184                                           0.008105      \n",
       "185                                           0.008866      \n",
       "186                                           0.013571      \n",
       "187                                           0.009192      \n",
       "188                                           0.007515      \n",
       "189                                           0.008194      \n",
       "190                                           0.008476      \n",
       "191                                           0.010112      \n",
       "192                                           0.008665      \n",
       "193                                           0.007353      \n",
       "194                                           0.008249      \n",
       "195                                           0.008455      \n",
       "196                                           0.010557      \n",
       "197                                           0.008620      \n",
       "198                                           0.007308      \n",
       "199                                           0.008264      \n",
       "\n",
       "     72kb Non-overlapping Windows $\\left( SEM \\right)$  \\\n",
       "0                                             0.000060   \n",
       "1                                             0.000067   \n",
       "2                                             0.000060   \n",
       "3                                             0.000057   \n",
       "4                                             0.000061   \n",
       "5                                             0.000056   \n",
       "6                                             0.000077   \n",
       "7                                             0.000059   \n",
       "8                                             0.000053   \n",
       "9                                             0.000057   \n",
       "10                                            0.000055   \n",
       "11                                            0.000076   \n",
       "12                                            0.000057   \n",
       "13                                            0.000052   \n",
       "14                                            0.000055   \n",
       "15                                            0.000055   \n",
       "16                                            0.000076   \n",
       "17                                            0.000058   \n",
       "18                                            0.000052   \n",
       "19                                            0.000055   \n",
       "20                                            0.000068   \n",
       "21                                            0.000073   \n",
       "22                                            0.000068   \n",
       "23                                            0.000064   \n",
       "24                                            0.000068   \n",
       "25                                            0.000065   \n",
       "26                                            0.000071   \n",
       "27                                            0.000065   \n",
       "28                                            0.000062   \n",
       "29                                            0.000067   \n",
       "30                                            0.000053   \n",
       "31                                            0.000075   \n",
       "32                                            0.000056   \n",
       "33                                            0.000051   \n",
       "34                                            0.000054   \n",
       "35                                            0.000056   \n",
       "36                                            0.000076   \n",
       "37                                            0.000058   \n",
       "38                                            0.000053   \n",
       "39                                            0.000056   \n",
       "40                                            0.000065   \n",
       "41                                            0.000070   \n",
       "42                                            0.000066   \n",
       "43                                            0.000062   \n",
       "44                                            0.000066   \n",
       "45                                            0.000065   \n",
       "46                                            0.000071   \n",
       "47                                            0.000065   \n",
       "48                                            0.000062   \n",
       "49                                            0.000066   \n",
       "50                                            0.000130   \n",
       "51                                            0.000129   \n",
       "52                                            0.000130   \n",
       "53                                            0.000129   \n",
       "54                                            0.000132   \n",
       "55                                            0.000110   \n",
       "56                                            0.000121   \n",
       "57                                            0.000111   \n",
       "58                                            0.000110   \n",
       "59                                            0.000113   \n",
       "60                                            0.000104   \n",
       "61                                            0.000116   \n",
       "62                                            0.000105   \n",
       "63                                            0.000105   \n",
       "64                                            0.000108   \n",
       "65                                            0.000105   \n",
       "66                                            0.000117   \n",
       "67                                            0.000107   \n",
       "68                                            0.000106   \n",
       "69                                            0.000109   \n",
       "70                                            0.000143   \n",
       "71                                            0.000141   \n",
       "72                                            0.000143   \n",
       "73                                            0.000142   \n",
       "74                                            0.000144   \n",
       "75                                            0.000139   \n",
       "76                                            0.000138   \n",
       "77                                            0.000139   \n",
       "78                                            0.000139   \n",
       "79                                            0.000142   \n",
       "80                                            0.000103   \n",
       "81                                            0.000115   \n",
       "82                                            0.000105   \n",
       "83                                            0.000104   \n",
       "84                                            0.000106   \n",
       "85                                            0.000108   \n",
       "86                                            0.000119   \n",
       "87                                            0.000109   \n",
       "88                                            0.000109   \n",
       "89                                            0.000111   \n",
       "90                                            0.000141   \n",
       "91                                            0.000138   \n",
       "92                                            0.000141   \n",
       "93                                            0.000139   \n",
       "94                                            0.000142   \n",
       "95                                            0.000140   \n",
       "96                                            0.000138   \n",
       "97                                            0.000139   \n",
       "98                                            0.000139   \n",
       "99                                            0.000142   \n",
       "100                                           0.000309   \n",
       "101                                           0.000303   \n",
       "102                                           0.000310   \n",
       "103                                           0.000309   \n",
       "104                                           0.000312   \n",
       "105                                           0.000264   \n",
       "106                                           0.000267   \n",
       "107                                           0.000265   \n",
       "108                                           0.000266   \n",
       "109                                           0.000268   \n",
       "110                                           0.000257   \n",
       "111                                           0.000262   \n",
       "112                                           0.000258   \n",
       "113                                           0.000259   \n",
       "114                                           0.000262   \n",
       "115                                           0.000259   \n",
       "116                                           0.000263   \n",
       "117                                           0.000260   \n",
       "118                                           0.000261   \n",
       "119                                           0.000263   \n",
       "120                                           0.000330   \n",
       "121                                           0.000322   \n",
       "122                                           0.000330   \n",
       "123                                           0.000330   \n",
       "124                                           0.000332   \n",
       "125                                           0.000324   \n",
       "126                                           0.000317   \n",
       "127                                           0.000324   \n",
       "128                                           0.000324   \n",
       "129                                           0.000327   \n",
       "130                                           0.000256   \n",
       "131                                           0.000260   \n",
       "132                                           0.000258   \n",
       "133                                           0.000259   \n",
       "134                                           0.000261   \n",
       "135                                           0.000263   \n",
       "136                                           0.000266   \n",
       "137                                           0.000264   \n",
       "138                                           0.000265   \n",
       "139                                           0.000267   \n",
       "140                                           0.000326   \n",
       "141                                           0.000318   \n",
       "142                                           0.000327   \n",
       "143                                           0.000326   \n",
       "144                                           0.000329   \n",
       "145                                           0.000324   \n",
       "146                                           0.000316   \n",
       "147                                           0.000324   \n",
       "148                                           0.000324   \n",
       "149                                           0.000327   \n",
       "150                                           0.000045   \n",
       "151                                           0.000058   \n",
       "152                                           0.000046   \n",
       "153                                           0.000038   \n",
       "154                                           0.000043   \n",
       "155                                           0.000050   \n",
       "156                                           0.000077   \n",
       "157                                           0.000052   \n",
       "158                                           0.000043   \n",
       "159                                           0.000047   \n",
       "160                                           0.000050   \n",
       "161                                           0.000078   \n",
       "162                                           0.000052   \n",
       "163                                           0.000042   \n",
       "164                                           0.000046   \n",
       "165                                           0.000050   \n",
       "166                                           0.000078   \n",
       "167                                           0.000052   \n",
       "168                                           0.000042   \n",
       "169                                           0.000047   \n",
       "170                                           0.000050   \n",
       "171                                           0.000061   \n",
       "172                                           0.000051   \n",
       "173                                           0.000043   \n",
       "174                                           0.000048   \n",
       "175                                           0.000048   \n",
       "176                                           0.000059   \n",
       "177                                           0.000049   \n",
       "178                                           0.000042   \n",
       "179                                           0.000047   \n",
       "180                                           0.000050   \n",
       "181                                           0.000078   \n",
       "182                                           0.000051   \n",
       "183                                           0.000042   \n",
       "184                                           0.000046   \n",
       "185                                           0.000050   \n",
       "186                                           0.000076   \n",
       "187                                           0.000052   \n",
       "188                                           0.000042   \n",
       "189                                           0.000046   \n",
       "190                                           0.000048   \n",
       "191                                           0.000057   \n",
       "192                                           0.000049   \n",
       "193                                           0.000041   \n",
       "194                                           0.000046   \n",
       "195                                           0.000048   \n",
       "196                                           0.000059   \n",
       "197                                           0.000048   \n",
       "198                                           0.000041   \n",
       "199                                           0.000046   \n",
       "\n",
       "     72kb Non-overlapping Windows $\\left( \\pm CI_{95\\%} \\right)$  $P-value$  \n",
       "0                                             0.000118             1.000000  \n",
       "1                                             0.000132             1.000000  \n",
       "2                                             0.000118             1.000000  \n",
       "3                                             0.000111             1.000000  \n",
       "4                                             0.000119             1.000000  \n",
       "5                                             0.000111             1.000000  \n",
       "6                                             0.000150             1.000000  \n",
       "7                                             0.000115             1.000000  \n",
       "8                                             0.000105             1.000000  \n",
       "9                                             0.000111             1.000000  \n",
       "10                                            0.000107             0.247648  \n",
       "11                                            0.000149             1.000000  \n",
       "12                                            0.000112             0.272997  \n",
       "13                                            0.000101             0.127596  \n",
       "14                                            0.000108             1.000000  \n",
       "15                                            0.000108             0.284582  \n",
       "16                                            0.000149             1.000000  \n",
       "17                                            0.000113             0.299640  \n",
       "18                                            0.000102             0.138645  \n",
       "19                                            0.000109             1.000000  \n",
       "20                                            0.000133             0.521056  \n",
       "21                                            0.000142             1.000000  \n",
       "22                                            0.000133             0.608435  \n",
       "23                                            0.000125             0.445672  \n",
       "24                                            0.000134             0.515531  \n",
       "25                                            0.000127             0.527748  \n",
       "26                                            0.000139             1.000000  \n",
       "27                                            0.000128             1.000000  \n",
       "28                                            0.000121             0.475314  \n",
       "29                                            0.000131             0.487026  \n",
       "30                                            0.000105             1.000000  \n",
       "31                                            0.000147             1.000000  \n",
       "32                                            0.000110             1.000000  \n",
       "33                                            0.000101             0.156891  \n",
       "34                                            0.000105             1.000000  \n",
       "35                                            0.000109             1.000000  \n",
       "36                                            0.000148             1.000000  \n",
       "37                                            0.000114             1.000000  \n",
       "38                                            0.000104             0.183503  \n",
       "39                                            0.000110             1.000000  \n",
       "40                                            0.000128             0.320917  \n",
       "41                                            0.000137             0.646726  \n",
       "42                                            0.000130             0.379285  \n",
       "43                                            0.000122             0.411989  \n",
       "44                                            0.000130             0.510859  \n",
       "45                                            0.000127             0.562914  \n",
       "46                                            0.000138             1.000000  \n",
       "47                                            0.000128             1.000000  \n",
       "48                                            0.000121             0.405865  \n",
       "49                                            0.000130             0.488099  \n",
       "50                                            0.000254             0.020898  \n",
       "51                                            0.000254             0.019067  \n",
       "52                                            0.000254             0.019351  \n",
       "53                                            0.000253             0.029421  \n",
       "54                                            0.000258             0.034535  \n",
       "55                                            0.000215             0.009376  \n",
       "56                                            0.000237             0.011901  \n",
       "57                                            0.000217             0.008776  \n",
       "58                                            0.000217             0.014111  \n",
       "59                                            0.000222             0.017678  \n",
       "60                                            0.000204             0.002241  \n",
       "61                                            0.000227             0.003630  \n",
       "62                                            0.000206             0.002304  \n",
       "63                                            0.000205             0.003062  \n",
       "64                                            0.000211             0.005051  \n",
       "65                                            0.000206             0.002589  \n",
       "66                                            0.000230             0.003914  \n",
       "67                                            0.000209             0.002431  \n",
       "68                                            0.000208             0.003157  \n",
       "69                                            0.000213             0.005430  \n",
       "70                                            0.000281             0.009439  \n",
       "71                                            0.000276             0.008271  \n",
       "72                                            0.000281             0.008839  \n",
       "73                                            0.000278             0.014837  \n",
       "74                                            0.000283             0.017741  \n",
       "75                                            0.000273             0.011017  \n",
       "76                                            0.000270             0.010323  \n",
       "77                                            0.000273             0.010228  \n",
       "78                                            0.000272             0.017646  \n",
       "79                                            0.000278             0.020266  \n",
       "80                                            0.000201             0.001862  \n",
       "81                                            0.000226             0.004072  \n",
       "82                                            0.000206             0.002083  \n",
       "83                                            0.000205             0.002873  \n",
       "84                                            0.000209             0.004893  \n",
       "85                                            0.000212             0.006819  \n",
       "86                                            0.000233             0.009754  \n",
       "87                                            0.000213             0.005966  \n",
       "88                                            0.000213             0.008081  \n",
       "89                                            0.000218             0.012343  \n",
       "90                                            0.000276             0.004104  \n",
       "91                                            0.000271             0.003125  \n",
       "92                                            0.000277             0.003630  \n",
       "93                                            0.000273             0.006314  \n",
       "94                                            0.000278             0.007513  \n",
       "95                                            0.000274             0.013448  \n",
       "96                                            0.000270             0.012311  \n",
       "97                                            0.000273             0.011459  \n",
       "98                                            0.000272             0.018436  \n",
       "99                                            0.000278             0.022918  \n",
       "100                                           0.000606             0.565566  \n",
       "101                                           0.000594             0.534756  \n",
       "102                                           0.000607             0.565598  \n",
       "103                                           0.000606             0.667624  \n",
       "104                                           0.000611             0.661058  \n",
       "105                                           0.000517             0.353242  \n",
       "106                                           0.000523             0.350401  \n",
       "107                                           0.000519             0.356399  \n",
       "108                                           0.000521             0.441095  \n",
       "109                                           0.000526             0.447787  \n",
       "110                                           0.000505             0.207242  \n",
       "111                                           0.000513             0.220784  \n",
       "112                                           0.000506             0.196666  \n",
       "113                                           0.000508             0.242155  \n",
       "114                                           0.000513             0.302166  \n",
       "115                                           0.000508             0.209104  \n",
       "116                                           0.000515             0.220658  \n",
       "117                                           0.000509             0.197266  \n",
       "118                                           0.000511             0.240830  \n",
       "119                                           0.000516             0.296420  \n",
       "120                                           0.000647             0.381527  \n",
       "121                                           0.000631             0.343803  \n",
       "122                                           0.000648             0.370794  \n",
       "123                                           0.000646             0.476924  \n",
       "124                                           0.000652             0.484911  \n",
       "125                                           0.000635             0.409622  \n",
       "126                                           0.000621             0.387840  \n",
       "127                                           0.000636             0.408675  \n",
       "128                                           0.000635             0.521371  \n",
       "129                                           0.000641             0.513037  \n",
       "130                                           0.000501             0.204811  \n",
       "131                                           0.000510             0.215386  \n",
       "132                                           0.000505             0.192721  \n",
       "133                                           0.000507             0.234642  \n",
       "134                                           0.000511             0.281173  \n",
       "135                                           0.000516             0.335091  \n",
       "136                                           0.000522             0.355325  \n",
       "137                                           0.000517             0.325557  \n",
       "138                                           0.000519             0.371362  \n",
       "139                                           0.000524             0.438885  \n",
       "140                                           0.000640             0.263842  \n",
       "141                                           0.000624             0.229497  \n",
       "142                                           0.000641             0.256550  \n",
       "143                                           0.000639             0.363943  \n",
       "144                                           0.000645             0.376570  \n",
       "145                                           0.000635             0.445735  \n",
       "146                                           0.000620             0.418682  \n",
       "147                                           0.000634             0.438191  \n",
       "148                                           0.000634             0.538418  \n",
       "149                                           0.000640             0.546310  \n",
       "150                                           0.000088             1.000000  \n",
       "151                                           0.000113             1.000000  \n",
       "152                                           0.000089             1.000000  \n",
       "153                                           0.000075             1.000000  \n",
       "154                                           0.000084             1.000000  \n",
       "155                                           0.000098             1.000000  \n",
       "156                                           0.000151             1.000000  \n",
       "157                                           0.000102             1.000000  \n",
       "158                                           0.000083             1.000000  \n",
       "159                                           0.000091             1.000000  \n",
       "160                                           0.000098             0.285877  \n",
       "161                                           0.000153             1.000000  \n",
       "162                                           0.000102             0.295852  \n",
       "163                                           0.000083             0.166014  \n",
       "164                                           0.000091             1.000000  \n",
       "165                                           0.000099             1.000000  \n",
       "166                                           0.000152             1.000000  \n",
       "167                                           0.000102             1.000000  \n",
       "168                                           0.000083             0.204116  \n",
       "169                                           0.000091             1.000000  \n",
       "170                                           0.000098             0.576867  \n",
       "171                                           0.000119             1.000000  \n",
       "172                                           0.000100             1.000000  \n",
       "173                                           0.000085             1.000000  \n",
       "174                                           0.000094             1.000000  \n",
       "175                                           0.000094             1.000000  \n",
       "176                                           0.000116             1.000000  \n",
       "177                                           0.000096             1.000000  \n",
       "178                                           0.000082             1.000000  \n",
       "179                                           0.000092             1.000000  \n",
       "180                                           0.000098             1.000000  \n",
       "181                                           0.000153             1.000000  \n",
       "182                                           0.000101             1.000000  \n",
       "183                                           0.000082             0.193131  \n",
       "184                                           0.000089             1.000000  \n",
       "185                                           0.000098             1.000000  \n",
       "186                                           0.000149             1.000000  \n",
       "187                                           0.000101             1.000000  \n",
       "188                                           0.000083             1.000000  \n",
       "189                                           0.000090             1.000000  \n",
       "190                                           0.000093             0.230412  \n",
       "191                                           0.000111             0.315676  \n",
       "192                                           0.000095             0.266557  \n",
       "193                                           0.000081             0.454764  \n",
       "194                                           0.000091             0.476482  \n",
       "195                                           0.000093             1.000000  \n",
       "196                                           0.000116             1.000000  \n",
       "197                                           0.000095             1.000000  \n",
       "198                                           0.000080             1.000000  \n",
       "199                                           0.000091             1.000000  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8806db26",
   "metadata": {},
   "source": [
    "## S53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "431b6449",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/pbs_empirical_significance_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S53. Summary of per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values for the focal 742kb \\textit{MUC19} region.} \\newline The total number of SNPs, the number of SNPs with a $PBS_{MXL:CHB:CEU}$ value greater than the $99.95^{th}$ genome-wide percentile, and those identified as significant based on different statistical criteria: uncorrected \\textit{P-values} less than 0.05, Bonferroni adjusted \\textit{P-values} less than 0.05, and Benjamini-Hochberg adjusted \\textit{P-values} less than 0.01. We report the results for all SNPs, as well as stratified by SNP set (see Methods for a description of these SNP sets).},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/pbs_empirical_significance_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(mat.shape[1]))\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90e2c32a",
   "metadata": {},
   "source": [
    "\\textbf{Table S53. Summary of per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values for the focal 742kb \\textit{MUC19} region.} \\newline The total number of SNPs, the number of SNPs with a $PBS_{MXL:CHB:CEU}$ value greater than the $99.95^{th}$ genome-wide percentile, and those identified as significant based on different statistical criteria: uncorrected \\textit{P-values} less than 0.05, Bonferroni adjusted \\textit{P-values} less than 0.05, and Benjamini-Hochberg adjusted \\textit{P-values} less than 0.01. We report the results for all SNPs, as well as stratified by SNP set (see Methods for a description of these SNP sets)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "14e98f3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP Set</th>\n",
       "      <th>Total SNPs</th>\n",
       "      <th>$PBS &gt; 99.95^{th}$ Genome-Wide Percentile</th>\n",
       "      <th>Uncorrected $P-values &lt; 0.05$</th>\n",
       "      <th>Bonferroni Adjusted $P-values &lt; 0.05$</th>\n",
       "      <th>Benjamini-Hochberg Adjusted $P-values &lt; 0.01$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All SNPs</td>\n",
       "      <td>6144</td>\n",
       "      <td>417</td>\n",
       "      <td>1111</td>\n",
       "      <td>0</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>217</td>\n",
       "      <td>208</td>\n",
       "      <td>217</td>\n",
       "      <td>0</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Denisovan-specific SNPs</td>\n",
       "      <td>135</td>\n",
       "      <td>135</td>\n",
       "      <td>135</td>\n",
       "      <td>0</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neanderthal-specific SNPs</td>\n",
       "      <td>80</td>\n",
       "      <td>72</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Human-specific SNPs</td>\n",
       "      <td>5420</td>\n",
       "      <td>141</td>\n",
       "      <td>677</td>\n",
       "      <td>0</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Shared Hominin SNPs</td>\n",
       "      <td>507</td>\n",
       "      <td>68</td>\n",
       "      <td>217</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     SNP Set  Total SNPs  \\\n",
       "0                   All SNPs        6144   \n",
       "1               Archaic SNPs         217   \n",
       "2    Denisovan-specific SNPs         135   \n",
       "3  Neanderthal-specific SNPs          80   \n",
       "4        Human-specific SNPs        5420   \n",
       "5        Shared Hominin SNPs         507   \n",
       "\n",
       "   $PBS > 99.95^{th}$ Genome-Wide Percentile  Uncorrected $P-values < 0.05$  \\\n",
       "0                                        417                           1111   \n",
       "1                                        208                            217   \n",
       "2                                        135                            135   \n",
       "3                                         72                             80   \n",
       "4                                        141                            677   \n",
       "5                                         68                            217   \n",
       "\n",
       "   Bonferroni Adjusted $P-values < 0.05$  \\\n",
       "0                                      0   \n",
       "1                                      0   \n",
       "2                                      0   \n",
       "3                                      0   \n",
       "4                                      0   \n",
       "5                                      0   \n",
       "\n",
       "   Benjamini-Hochberg Adjusted $P-values < 0.01$  \n",
       "0                                            485  \n",
       "1                                            217  \n",
       "2                                            135  \n",
       "3                                             80  \n",
       "4                                            200  \n",
       "5                                             68  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8198837d",
   "metadata": {},
   "source": [
    "## S54"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "a7a8f32e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/pbs_empirical_outlier_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 5, 8, 11, 14]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 5, 8, 11, 14]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S54. Summary of per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ outliers for the focal 742kb \\textit{MUC19} region.} \\newline The total number of SNPs, the number of outlier SNPs with a $PBS_{MXL:CHB}$ value greater than the $99.95^{th}$ genome-wide percentile, the mean $PBS_{MXL:CHB}$ value among these outliers, and the associated 95\\% confidence intervals ($\\pm CI_{95\\%}$) for the mean $PBS_{MXL:CHB}$ value. Additionally, the mean percentile rank of outlier SNPs and the associated 95\\% confidence intervals ($\\pm CI_{95\\%}$) for the mean percentile rank is reported. We report the results for all SNPs, as well as stratified by SNP set (see Methods for a description of these SNP sets).},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/pbs_empirical_outlier_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01d412b2",
   "metadata": {},
   "source": [
    "\\textbf{Table S54. Summary of per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ outliers for the focal 742kb \\textit{MUC19} region.} \\newline The total number of SNPs, the number of outlier SNPs with a $PBS_{MXL:CHB}$ value greater than the $99.95^{th}$ genome-wide percentile, the mean $PBS_{MXL:CHB}$ value among these outliers, and the associated 95\\% confidence intervals ($\\pm CI_{95\\%}$) for the mean $PBS_{MXL:CHB}$ value. Additionally, the mean percentile rank of outlier SNPs and the associated 95\\% confidence intervals ($\\pm CI_{95\\%}$) for the mean percentile rank is reported. We report the results for all SNPs, as well as stratified by SNP set (see Methods for a description of these SNP sets)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "5453cc0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP Set</th>\n",
       "      <th>Total SNPs</th>\n",
       "      <th>$PBS &gt; 99.95^{th}$ Genome-Wide Percentile</th>\n",
       "      <th>$PBS \\; \\left( \\mu \\right)$</th>\n",
       "      <th>$PBS \\; \\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>Percentile Rank $\\left( \\mu \\right)$</th>\n",
       "      <th>Percentile Rank $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All SNPs</td>\n",
       "      <td>6144</td>\n",
       "      <td>417</td>\n",
       "      <td>0.301969</td>\n",
       "      <td>0.003209</td>\n",
       "      <td>99.975275</td>\n",
       "      <td>0.001031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>217</td>\n",
       "      <td>208</td>\n",
       "      <td>0.293409</td>\n",
       "      <td>0.003612</td>\n",
       "      <td>99.972262</td>\n",
       "      <td>0.001309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Denisovan-specific SNPs</td>\n",
       "      <td>135</td>\n",
       "      <td>135</td>\n",
       "      <td>0.279338</td>\n",
       "      <td>0.000319</td>\n",
       "      <td>99.966727</td>\n",
       "      <td>0.000229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neanderthal-specific SNPs</td>\n",
       "      <td>80</td>\n",
       "      <td>72</td>\n",
       "      <td>0.320095</td>\n",
       "      <td>0.007143</td>\n",
       "      <td>99.982795</td>\n",
       "      <td>0.002248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Human-specific SNPs</td>\n",
       "      <td>5420</td>\n",
       "      <td>141</td>\n",
       "      <td>0.313188</td>\n",
       "      <td>0.006865</td>\n",
       "      <td>99.978522</td>\n",
       "      <td>0.001937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Shared Hominin SNPs</td>\n",
       "      <td>507</td>\n",
       "      <td>68</td>\n",
       "      <td>0.304890</td>\n",
       "      <td>0.006364</td>\n",
       "      <td>99.977758</td>\n",
       "      <td>0.002286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     SNP Set  Total SNPs  \\\n",
       "0                   All SNPs        6144   \n",
       "1               Archaic SNPs         217   \n",
       "2    Denisovan-specific SNPs         135   \n",
       "3  Neanderthal-specific SNPs          80   \n",
       "4        Human-specific SNPs        5420   \n",
       "5        Shared Hominin SNPs         507   \n",
       "\n",
       "   $PBS > 99.95^{th}$ Genome-Wide Percentile  $PBS \\; \\left( \\mu \\right)$  \\\n",
       "0                                        417                     0.301969   \n",
       "1                                        208                     0.293409   \n",
       "2                                        135                     0.279338   \n",
       "3                                         72                     0.320095   \n",
       "4                                        141                     0.313188   \n",
       "5                                         68                     0.304890   \n",
       "\n",
       "   $PBS \\; \\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                               0.003209   \n",
       "1                               0.003612   \n",
       "2                               0.000319   \n",
       "3                               0.007143   \n",
       "4                               0.006865   \n",
       "5                               0.006364   \n",
       "\n",
       "   Percentile Rank $\\left( \\mu \\right)$  \\\n",
       "0                             99.975275   \n",
       "1                             99.972262   \n",
       "2                             99.966727   \n",
       "3                             99.982795   \n",
       "4                             99.978522   \n",
       "5                             99.977758   \n",
       "\n",
       "   Percentile Rank $\\left( \\pm CI_{95\\%} \\right)$  \n",
       "0                                        0.001031  \n",
       "1                                        0.001309  \n",
       "2                                        0.000229  \n",
       "3                                        0.002248  \n",
       "4                                        0.001937  \n",
       "5                                        0.002286  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 2, 5, 8, 11, 14]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94f97f63",
   "metadata": {},
   "source": [
    "## S55"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "041efb75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/pbs_empirical_bhp_significance_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 5, 8, 11, 14]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 5, 8, 11, 14]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S55. Summary of significant per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values for the focal 742kb \\textit{MUC19} region.} \\newline The total number of SNPs, the number of significant SNPs (i.e., SNPs with a Benjamini-Hochberg adjusted \\textit{P-values} less than 0.01), the mean $PBS_{MXL:CHB}$ value among these significant SNPs, and the associated 95\\% confidence intervals ($\\pm CI_{95\\%}$) for the mean $PBS_{MXL:CHB}$ value. Additionally, the mean percentile rank of significant SNPs and the associated 95\\% confidence intervals ($\\pm CI_{95\\%}$) for the mean percentile rank is reported. We report the results for all SNPs, as well as stratified by SNP set (see Methods for a description of these SNP sets).},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/pbs_empirical_bhp_significance_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf0a867c",
   "metadata": {},
   "source": [
    "\\textbf{Table S55. Summary of significant per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values for the focal 742kb \\textit{MUC19} region.} \\newline The total number of SNPs, the number of significant SNPs (i.e., SNPs with a Benjamini-Hochberg adjusted \\textit{P-values} less than 0.01), the mean $PBS_{MXL:CHB}$ value among these significant SNPs, and the associated 95\\% confidence intervals ($\\pm CI_{95\\%}$) for the mean $PBS_{MXL:CHB}$ value. Additionally, the mean percentile rank of significant SNPs and the associated 95\\% confidence intervals ($\\pm CI_{95\\%}$) for the mean percentile rank is reported. We report the results for all SNPs, as well as stratified by SNP set (see Methods for a description of these SNP sets)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "2b32b327",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP Set</th>\n",
       "      <th>Total SNPs</th>\n",
       "      <th>Benjamini-Hochberg Adjusted $P-values &lt; 0.01$</th>\n",
       "      <th>$PBS \\; \\left( \\mu \\right)$</th>\n",
       "      <th>$PBS \\; \\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "      <th>Percentile Rank $\\left( \\mu \\right)$</th>\n",
       "      <th>Percentile Rank $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All SNPs</td>\n",
       "      <td>6144</td>\n",
       "      <td>485</td>\n",
       "      <td>0.293157</td>\n",
       "      <td>0.003382</td>\n",
       "      <td>99.968599</td>\n",
       "      <td>0.001737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "      <td>0.291069</td>\n",
       "      <td>0.003776</td>\n",
       "      <td>99.970305</td>\n",
       "      <td>0.001780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Denisovan-specific SNPs</td>\n",
       "      <td>135</td>\n",
       "      <td>135</td>\n",
       "      <td>0.279338</td>\n",
       "      <td>0.000319</td>\n",
       "      <td>99.966727</td>\n",
       "      <td>0.000229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neanderthal-specific SNPs</td>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>0.311784</td>\n",
       "      <td>0.008503</td>\n",
       "      <td>99.977022</td>\n",
       "      <td>0.004372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Human-specific SNPs</td>\n",
       "      <td>5420</td>\n",
       "      <td>200</td>\n",
       "      <td>0.291433</td>\n",
       "      <td>0.006750</td>\n",
       "      <td>99.963633</td>\n",
       "      <td>0.003535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Shared Hominin SNPs</td>\n",
       "      <td>507</td>\n",
       "      <td>68</td>\n",
       "      <td>0.304890</td>\n",
       "      <td>0.006364</td>\n",
       "      <td>99.977758</td>\n",
       "      <td>0.002286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     SNP Set  Total SNPs  \\\n",
       "0                   All SNPs        6144   \n",
       "1               Archaic SNPs         217   \n",
       "2    Denisovan-specific SNPs         135   \n",
       "3  Neanderthal-specific SNPs          80   \n",
       "4        Human-specific SNPs        5420   \n",
       "5        Shared Hominin SNPs         507   \n",
       "\n",
       "   Benjamini-Hochberg Adjusted $P-values < 0.01$  $PBS \\; \\left( \\mu \\right)$  \\\n",
       "0                                            485                     0.293157   \n",
       "1                                            217                     0.291069   \n",
       "2                                            135                     0.279338   \n",
       "3                                             80                     0.311784   \n",
       "4                                            200                     0.291433   \n",
       "5                                             68                     0.304890   \n",
       "\n",
       "   $PBS \\; \\left( \\pm CI_{95\\%} \\right)$  \\\n",
       "0                               0.003382   \n",
       "1                               0.003776   \n",
       "2                               0.000319   \n",
       "3                               0.008503   \n",
       "4                               0.006750   \n",
       "5                               0.006364   \n",
       "\n",
       "   Percentile Rank $\\left( \\mu \\right)$  \\\n",
       "0                             99.968599   \n",
       "1                             99.970305   \n",
       "2                             99.966727   \n",
       "3                             99.977022   \n",
       "4                             99.963633   \n",
       "5                             99.977758   \n",
       "\n",
       "   Percentile Rank $\\left( \\pm CI_{95\\%} \\right)$  \n",
       "0                                        0.001737  \n",
       "1                                        0.001780  \n",
       "2                                        0.000229  \n",
       "3                                        0.004372  \n",
       "4                                        0.003535  \n",
       "5                                        0.002286  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 2, 5, 8, 11, 14]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "472da5f4",
   "metadata": {},
   "source": [
    "## S56"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "96034fea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('../meta_data/slim_demographic_info.txt', sep='\\t')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S56. Demographic model from Medina-Mu\\~{n}oz et al., 2023 with archaic introgression.} \\newline Demographic parameters used for \\textit{PBS} simulations.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/slim_demographic_info.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(1, 2))+r' & '\\\n",
    "        + r' & '.join(f'{mat[i, j]}' for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beb7025c",
   "metadata": {},
   "source": [
    "\\textbf{Table S56. Demographic model from Medina-Mu\\~{n}oz et al., 2023 with archaic introgression.} \\newline Demographic parameters used for \\textit{PBS} simulations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "2aadab0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parameter</th>\n",
       "      <th>Value</th>\n",
       "      <th>Description</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Population size</td>\n",
       "      <td>1.358000e+04</td>\n",
       "      <td>Ancestral Pop. size</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Population size</td>\n",
       "      <td>1.324900e+04</td>\n",
       "      <td>Archaic Pop. size</td>\n",
       "      <td>Jacobs et al., 2019; Malaspinas et al., 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Population size</td>\n",
       "      <td>2.714200e+04</td>\n",
       "      <td>AMH Pop. size</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Population size</td>\n",
       "      <td>8.260000e+02</td>\n",
       "      <td>NEA Pop. size</td>\n",
       "      <td>Jacobs et al., 2019; Malaspinas et al., 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Population size</td>\n",
       "      <td>5.083000e+03</td>\n",
       "      <td>DEN Pop. size</td>\n",
       "      <td>Jacobs et al., 2019; Malaspinas et al., 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Population size</td>\n",
       "      <td>1.835000e+03</td>\n",
       "      <td>OOA Pop. size</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Population size</td>\n",
       "      <td>2.714200e+04</td>\n",
       "      <td>AFR Pop. size</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Population size</td>\n",
       "      <td>2.761000e+03</td>\n",
       "      <td>EUR Pop. size</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Population size</td>\n",
       "      <td>1.955000e+03</td>\n",
       "      <td>EAS Pop. size</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Population size</td>\n",
       "      <td>1.313000e+03</td>\n",
       "      <td>MXB Pop. size</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Population size</td>\n",
       "      <td>2.000000e+04</td>\n",
       "      <td>MXL Pop. size</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Growth rate per generation</td>\n",
       "      <td>1.282727e-03</td>\n",
       "      <td>EUR Pop. growth rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Growth rate per generation</td>\n",
       "      <td>1.170195e-03</td>\n",
       "      <td>EAS Pop. growth rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Growth rate per generation</td>\n",
       "      <td>3.014781e-03</td>\n",
       "      <td>MXB Pop. growth rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Migration rate per generation</td>\n",
       "      <td>1.585950e-04</td>\n",
       "      <td>AFR-OOA migration rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Migration rate per generation</td>\n",
       "      <td>2.501606e-05</td>\n",
       "      <td>AFR-EUR migration rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Migration rate per generation</td>\n",
       "      <td>3.294832e-06</td>\n",
       "      <td>AFR-EAS migration rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Migration rate per generation</td>\n",
       "      <td>6.624007e-05</td>\n",
       "      <td>EUR-EAS migration rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Migration rate per generation</td>\n",
       "      <td>5.030000e-02</td>\n",
       "      <td>MXB-MXL migration rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Migration rate per generation</td>\n",
       "      <td>5.760000e-02</td>\n",
       "      <td>EUR-MXL migration rate</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Admixture percentage</td>\n",
       "      <td>5.000000e-02</td>\n",
       "      <td>NEA admixture into OOA</td>\n",
       "      <td>Green et al., 2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Admixture percentage</td>\n",
       "      <td>5.000000e-02</td>\n",
       "      <td>DEN admixture into EAS</td>\n",
       "      <td>Skoglund and Jakobsson 2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Admixture percentage</td>\n",
       "      <td>1.117800e-01</td>\n",
       "      <td>AFR admixture into MXL</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Admixture percentage</td>\n",
       "      <td>7.422000e-03</td>\n",
       "      <td>EAS admixture into MXL</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>2.022500e+04</td>\n",
       "      <td>Ancestral-Archaic split</td>\n",
       "      <td>Jacobs et al., 2019; Malaspinas et al., 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.667100e+04</td>\n",
       "      <td>Ancestral-AMH split</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.509000e+04</td>\n",
       "      <td>Archaic-DEN split</td>\n",
       "      <td>Jacobs et al., 2019; Malaspinas et al., 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.509000e+04</td>\n",
       "      <td>Archaic-NEA split</td>\n",
       "      <td>Jacobs et al., 2019; Malaspinas et al., 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>3.046000e+03</td>\n",
       "      <td>AMH-OoA split</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>3.046000e+03</td>\n",
       "      <td>AMH-AFR split</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>2.069000e+03</td>\n",
       "      <td>NEA admixture into OOA</td>\n",
       "      <td>Green et al., 2010; Pr\\\"{u}fer et al., 2014; P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.764000e+03</td>\n",
       "      <td>OOA-EUR split</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.764000e+03</td>\n",
       "      <td>OOA-EAS split</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.552000e+03</td>\n",
       "      <td>DEN admixture into EAS</td>\n",
       "      <td>Reich et al., 2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.123000e+03</td>\n",
       "      <td>EAS-MXB split</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.600000e+01</td>\n",
       "      <td>MXL divergence from EUR and MXB</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Epoch time (g. a.)</td>\n",
       "      <td>1.300000e+01</td>\n",
       "      <td>AFR and EAS admixture into MXL</td>\n",
       "      <td>Medina-Mu\\~{n}oz et al., 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Mutation rate</td>\n",
       "      <td>1.315407e-08</td>\n",
       "      <td>Per-base per-generation mutation rate</td>\n",
       "      <td>Hubisz and Siepel 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Uniform recombination rate</td>\n",
       "      <td>5.999665e-09</td>\n",
       "      <td>Per-base per-generation recombination rate</td>\n",
       "      <td>Hinch et al., 2011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>MXL sample size</td>\n",
       "      <td>6.400000e+01</td>\n",
       "      <td>MXL sample size</td>\n",
       "      <td>Based on the number of individuals in MXL from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>EAS sample size</td>\n",
       "      <td>1.030000e+02</td>\n",
       "      <td>EAS sample size</td>\n",
       "      <td>Based on the number of individuals in CHB from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>EUR sample size</td>\n",
       "      <td>9.900000e+01</td>\n",
       "      <td>EUR sample size</td>\n",
       "      <td>Based on the number of individuals in CEU from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Generation time</td>\n",
       "      <td>2.900000e+01</td>\n",
       "      <td>Number of years per generation</td>\n",
       "      <td>Fenner 2005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Parameter         Value  \\\n",
       "0                 Population size  1.358000e+04   \n",
       "1                 Population size  1.324900e+04   \n",
       "2                 Population size  2.714200e+04   \n",
       "3                 Population size  8.260000e+02   \n",
       "4                 Population size  5.083000e+03   \n",
       "5                 Population size  1.835000e+03   \n",
       "6                 Population size  2.714200e+04   \n",
       "7                 Population size  2.761000e+03   \n",
       "8                 Population size  1.955000e+03   \n",
       "9                 Population size  1.313000e+03   \n",
       "10                Population size  2.000000e+04   \n",
       "11     Growth rate per generation  1.282727e-03   \n",
       "12     Growth rate per generation  1.170195e-03   \n",
       "13     Growth rate per generation  3.014781e-03   \n",
       "14  Migration rate per generation  1.585950e-04   \n",
       "15  Migration rate per generation  2.501606e-05   \n",
       "16  Migration rate per generation  3.294832e-06   \n",
       "17  Migration rate per generation  6.624007e-05   \n",
       "18  Migration rate per generation  5.030000e-02   \n",
       "19  Migration rate per generation  5.760000e-02   \n",
       "20           Admixture percentage  5.000000e-02   \n",
       "21           Admixture percentage  5.000000e-02   \n",
       "22           Admixture percentage  1.117800e-01   \n",
       "23           Admixture percentage  7.422000e-03   \n",
       "24             Epoch time (g. a.)  2.022500e+04   \n",
       "25             Epoch time (g. a.)  1.667100e+04   \n",
       "26             Epoch time (g. a.)  1.509000e+04   \n",
       "27             Epoch time (g. a.)  1.509000e+04   \n",
       "28             Epoch time (g. a.)  3.046000e+03   \n",
       "29             Epoch time (g. a.)  3.046000e+03   \n",
       "30             Epoch time (g. a.)  2.069000e+03   \n",
       "31             Epoch time (g. a.)  1.764000e+03   \n",
       "32             Epoch time (g. a.)  1.764000e+03   \n",
       "33             Epoch time (g. a.)  1.552000e+03   \n",
       "34             Epoch time (g. a.)  1.123000e+03   \n",
       "35             Epoch time (g. a.)  1.600000e+01   \n",
       "36             Epoch time (g. a.)  1.300000e+01   \n",
       "37                  Mutation rate  1.315407e-08   \n",
       "38     Uniform recombination rate  5.999665e-09   \n",
       "39                MXL sample size  6.400000e+01   \n",
       "40                EAS sample size  1.030000e+02   \n",
       "41                EUR sample size  9.900000e+01   \n",
       "42                Generation time  2.900000e+01   \n",
       "\n",
       "                                   Description  \\\n",
       "0                          Ancestral Pop. size   \n",
       "1                            Archaic Pop. size   \n",
       "2                                AMH Pop. size   \n",
       "3                                NEA Pop. size   \n",
       "4                                DEN Pop. size   \n",
       "5                                OOA Pop. size   \n",
       "6                                AFR Pop. size   \n",
       "7                                EUR Pop. size   \n",
       "8                                EAS Pop. size   \n",
       "9                                MXB Pop. size   \n",
       "10                               MXL Pop. size   \n",
       "11                        EUR Pop. growth rate   \n",
       "12                        EAS Pop. growth rate   \n",
       "13                        MXB Pop. growth rate   \n",
       "14                      AFR-OOA migration rate   \n",
       "15                      AFR-EUR migration rate   \n",
       "16                      AFR-EAS migration rate   \n",
       "17                      EUR-EAS migration rate   \n",
       "18                      MXB-MXL migration rate   \n",
       "19                      EUR-MXL migration rate   \n",
       "20                      NEA admixture into OOA   \n",
       "21                      DEN admixture into EAS   \n",
       "22                      AFR admixture into MXL   \n",
       "23                      EAS admixture into MXL   \n",
       "24                     Ancestral-Archaic split   \n",
       "25                         Ancestral-AMH split   \n",
       "26                           Archaic-DEN split   \n",
       "27                           Archaic-NEA split   \n",
       "28                               AMH-OoA split   \n",
       "29                               AMH-AFR split   \n",
       "30                      NEA admixture into OOA   \n",
       "31                               OOA-EUR split   \n",
       "32                               OOA-EAS split   \n",
       "33                      DEN admixture into EAS   \n",
       "34                               EAS-MXB split   \n",
       "35             MXL divergence from EUR and MXB   \n",
       "36              AFR and EAS admixture into MXL   \n",
       "37       Per-base per-generation mutation rate   \n",
       "38  Per-base per-generation recombination rate   \n",
       "39                             MXL sample size   \n",
       "40                             EAS sample size   \n",
       "41                             EUR sample size   \n",
       "42              Number of years per generation   \n",
       "\n",
       "                                            Reference  \n",
       "0                       Medina-Mu\\~{n}oz et al., 2023  \n",
       "1        Jacobs et al., 2019; Malaspinas et al., 2016  \n",
       "2                       Medina-Mu\\~{n}oz et al., 2023  \n",
       "3        Jacobs et al., 2019; Malaspinas et al., 2016  \n",
       "4        Jacobs et al., 2019; Malaspinas et al., 2016  \n",
       "5                       Medina-Mu\\~{n}oz et al., 2023  \n",
       "6                       Medina-Mu\\~{n}oz et al., 2023  \n",
       "7                       Medina-Mu\\~{n}oz et al., 2023  \n",
       "8                       Medina-Mu\\~{n}oz et al., 2023  \n",
       "9                       Medina-Mu\\~{n}oz et al., 2023  \n",
       "10                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "11                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "12                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "13                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "14                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "15                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "16                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "17                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "18                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "19                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "20                                 Green et al., 2010  \n",
       "21                        Skoglund and Jakobsson 2011  \n",
       "22                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "23                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "24       Jacobs et al., 2019; Malaspinas et al., 2016  \n",
       "25                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "26       Jacobs et al., 2019; Malaspinas et al., 2016  \n",
       "27       Jacobs et al., 2019; Malaspinas et al., 2016  \n",
       "28                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "29                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "30  Green et al., 2010; Pr\\\"{u}fer et al., 2014; P...  \n",
       "31                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "32                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "33                                 Reich et al., 2011  \n",
       "34                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "35                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "36                      Medina-Mu\\~{n}oz et al., 2023  \n",
       "37                             Hubisz and Siepel 2020  \n",
       "38                                 Hinch et al., 2011  \n",
       "39  Based on the number of individuals in MXL from...  \n",
       "40  Based on the number of individuals in CHB from...  \n",
       "41  Based on the number of individuals in CEU from...  \n",
       "42                                        Fenner 2005  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8790302",
   "metadata": {},
   "source": [
    "## S57"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "a7de0cb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/slim_per_snp_geq_freq_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S57. Summary of SNPs segregating at high frequency in MXL across 10,000 simulated replicates.} \\newline The total number of SNPs, the number of SNPs segregating a frequency greater than or equal to 30\\%, and the associated \\textit{P-value} based on 10,000 replicates per simulation model. The \\textit{P-value} represents the proportion of SNPs across all 10,000 simulations that are segregating at a frequency greater than or equal to 30\\%, where a \\textit{P-value} less than 0.05 is considered statistically significant. We report the results for all SNPs, as well as archaic SNPs.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/slim_per_snp_geq_freq_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(2))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            (convert_sci_notation_to_latex(mat[i, j]) if abs(mat[i, j]) < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(2, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48862c1d",
   "metadata": {},
   "source": [
    "\\textbf{Table S57. Summary of SNPs segregating at high frequency in MXL across 10,000 simulated replicates.} \\newline The total number of SNPs, the number of SNPs segregating a frequency greater than or equal to 30\\%, and the associated \\textit{P-value} based on 10,000 replicates per simulation model. The \\textit{P-value} represents the proportion of SNPs across all 10,000 simulations that are segregating at a frequency greater than or equal to 30\\%, where a \\textit{P-value} less than 0.05 is considered statistically significant. We report the results for all SNPs, as well as archaic SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "2f049e7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Selection Model</th>\n",
       "      <th>SNP Set</th>\n",
       "      <th>Total SNPs</th>\n",
       "      <th>MXL SNPs $\\geq 30\\%$</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>55296617</td>\n",
       "      <td>7572328</td>\n",
       "      <td>0.136940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>55296617</td>\n",
       "      <td>47581</td>\n",
       "      <td>0.000860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Negative</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>53988190</td>\n",
       "      <td>6910329</td>\n",
       "      <td>0.127997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>53988190</td>\n",
       "      <td>65682</td>\n",
       "      <td>0.001217</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Selection Model       SNP Set  Total SNPs  MXL SNPs $\\geq 30\\%$  $P-value$\n",
       "0         Neutral      All SNPs    55296617               7572328   0.136940\n",
       "1         Neutral  Archaic SNPs    55296617                 47581   0.000860\n",
       "2        Negative      All SNPs    53988190               6910329   0.127997\n",
       "3        Negative  Archaic SNPs    53988190                 65682   0.001217"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3226640b",
   "metadata": {},
   "source": [
    "## S58"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "5e1c48e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/slim_per_snp_pbs_significance_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S58. Summary of significant per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values across 10,000 simulated replicates for all 417 outlier SNPs.} \\newline The number of  $PBS_{MXL:CHB:CEU}$ values among the 417 SNPs with a per-SNP \\textit{PBS} value greater than the empirical $99.95^{th}$ genome-wide percentile, and those identified as significant based on 10,000 replicate simulations per model and different statistical criteria: uncorrected \\textit{P-values} less than 0.05, Bonferroni adjusted \\textit{P-values} less than 0.05, and Benjamini-Hochberg adjusted \\textit{P-values} less than 0.01. We report the results for all SNPs, as well as archaic SNPs.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/slim_per_snp_pbs_significance_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(mat.shape[1]))\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9802c013",
   "metadata": {},
   "source": [
    "\\textbf{Table S58. Summary of significant per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values across 10,000 simulated replicates for all 417 outlier SNPs.} \\newline The number of  $PBS_{MXL:CHB:CEU}$ values among the 417 SNPs with a per-SNP \\textit{PBS} value greater than the empirical $99.95^{th}$ genome-wide percentile, and those identified as significant based on 10,000 replicate simulations per model and different statistical criteria: uncorrected \\textit{P-values} less than 0.05, Bonferroni adjusted \\textit{P-values} less than 0.05, and Benjamini-Hochberg adjusted \\textit{P-values} less than 0.01. We report the results for all SNPs, as well as archaic SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "0560156f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Selection Model</th>\n",
       "      <th>SNP Set</th>\n",
       "      <th>Number of $PBS_{MXL:CHB:CEU}$ Values</th>\n",
       "      <th>Uncorrected $P-value &lt; 0.05$</th>\n",
       "      <th>Bonferroni Adj. $P-value &lt; 0.05$</th>\n",
       "      <th>Benjamini-Hochberg Adj. $P-value &lt; 0.01$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Negative</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>417</td>\n",
       "      <td>417</td>\n",
       "      <td>184</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>208</td>\n",
       "      <td>208</td>\n",
       "      <td>208</td>\n",
       "      <td>208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>417</td>\n",
       "      <td>417</td>\n",
       "      <td>97</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>208</td>\n",
       "      <td>208</td>\n",
       "      <td>208</td>\n",
       "      <td>208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Selection Model       SNP Set  Number of $PBS_{MXL:CHB:CEU}$ Values  \\\n",
       "0        Negative      All SNPs                                   417   \n",
       "1        Negative  Archaic SNPs                                   208   \n",
       "2         Neutral      All SNPs                                   417   \n",
       "3         Neutral  Archaic SNPs                                   208   \n",
       "\n",
       "   Uncorrected $P-value < 0.05$  Bonferroni Adj. $P-value < 0.05$  \\\n",
       "0                           417                               184   \n",
       "1                           208                               208   \n",
       "2                           417                                97   \n",
       "3                           208                               208   \n",
       "\n",
       "   Benjamini-Hochberg Adj. $P-value < 0.01$  \n",
       "0                                       417  \n",
       "1                                       208  \n",
       "2                                       417  \n",
       "3                                       208  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e8a90bc",
   "metadata": {},
   "source": [
    "## S59"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "70c3c393",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/slim_per_snp_unique_pbs_significance_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S59. Summary of significant per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values across 10,000 simulated replicates for the 57 unique outlier SNPs.} \\newline The number of unique $PBS_{MXL:CHB:CEU}$ values among the 57 unique per-SNP \\textit{PBS} value greater than the empirical $99.95^{th}$ genome-wide percentile, and those identified as significant based on 10,000 replicate simulations per model and different statistical criteria: uncorrected \\textit{P-values} less than 0.05, Bonferroni adjusted \\textit{P-values} less than 0.05, and Benjamini-Hochberg adjusted \\textit{P-values} less than 0.01. We report the results for all SNPs, as well as archaic SNPs.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/slim_per_snp_pbs_unique_significance_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(mat.shape[1]))\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0507a75",
   "metadata": {},
   "source": [
    "\\textbf{Table S59. Summary of significant per-SNP $\\boldsymbol{PBS_{MXL:CHB:CEU}}$ values across 10,000 simulated replicates for the 57 unique outlier SNPs.} \\newline The number of unique $PBS_{MXL:CHB:CEU}$ values among the 57 unique per-SNP \\textit{PBS} value greater than the empirical $99.95^{th}$ genome-wide percentile, and those identified as significant based on 10,000 replicate simulations per model and different statistical criteria: uncorrected \\textit{P-values} less than 0.05, Bonferroni adjusted \\textit{P-values} less than 0.05, and Benjamini-Hochberg adjusted \\textit{P-values} less than 0.01. We report the results for all SNPs, as well as archaic SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "81bdc10f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Selection Model</th>\n",
       "      <th>SNP Set</th>\n",
       "      <th>Unique $PBS_{MXL:CHB:CEU}$ Values</th>\n",
       "      <th>Uncorrected $P-value &lt; 0.05$</th>\n",
       "      <th>Bonferroni Adj. $P-value &lt; 0.05$</th>\n",
       "      <th>Benjamini-Hochberg Adj. $P-value &lt; 0.01$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Negative</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Selection Model       SNP Set  Unique $PBS_{MXL:CHB:CEU}$ Values  \\\n",
       "0        Negative      All SNPs                                 57   \n",
       "1        Negative  Archaic SNPs                                 20   \n",
       "2         Neutral      All SNPs                                 57   \n",
       "3         Neutral  Archaic SNPs                                 20   \n",
       "\n",
       "   Uncorrected $P-value < 0.05$  Bonferroni Adj. $P-value < 0.05$  \\\n",
       "0                            57                                57   \n",
       "1                            20                                20   \n",
       "2                            57                                57   \n",
       "3                            20                                20   \n",
       "\n",
       "   Benjamini-Hochberg Adj. $P-value < 0.01$  \n",
       "0                                        57  \n",
       "1                                        20  \n",
       "2                                        57  \n",
       "3                                        20  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "015d625d",
   "metadata": {},
   "source": [
    "## S60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "21b9a933",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/slim_per_region_geq_freq_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 3, 5, 6]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 3, 5, 6]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S60. Comparison of the number of SNPs segregating at high frequency in MXL for the observed focal regions and across 10,000 simulated replicates.} \\newline The number of SNPs segregating a frequency greater than or equal to 30\\% observed in MXL, along with the number of simulated replicates where the number SNPs segregating a frequency greater than or equal to 30\\% is greater than or equal to what is observed in the empirical data. The \\textit{P-value} represents the proportion of the 10,000 simulation replicates where the simulated value is greater than or equal to what is observed in the empirical data, where a \\textit{P-value} less than 0.05 is considered statistically significant. We report the results for both the focal 742kb and 72kb \\textit{MUC19} regions, for all SNPs, as well as archaic SNPs.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/slim_per_region_geq_freq_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(5))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            r'$<0.0001$' if mat[i, j][0] == '<' else\n",
    "            (convert_sci_notation_to_latex(float(mat[i, j])) if abs(float(mat[i, j])) < 0.0001 else\n",
    "            '{:.4f}'.format(round(float(mat[i, j]), 4)).rstrip('0').rstrip('.'))\n",
    "            for j in range(5, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0e26551",
   "metadata": {},
   "source": [
    "\\textbf{Table S60. Comparison of the number of SNPs segregating at high frequency in MXL for the observed focal regions and across 10,000 simulated replicates.} \\newline The number of SNPs segregating a frequency greater than or equal to 30\\% observed in MXL, along with the number of simulated replicates where the number SNPs segregating a frequency greater than or equal to 30\\% is greater than or equal to what is observed in the empirical data. The \\textit{P-value} represents the proportion of the 10,000 simulation replicates where the simulated value is greater than or equal to what is observed in the empirical data, where a \\textit{P-value} less than 0.05 is considered statistically significant. We report the results for both the focal 742kb and 72kb \\textit{MUC19} regions, for all SNPs, as well as archaic SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "2731d69f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Selection Model</th>\n",
       "      <th>SNP Set</th>\n",
       "      <th>Region</th>\n",
       "      <th>MXL AAF $\\geq 30\\%$ (Observed)</th>\n",
       "      <th>Total Replicates</th>\n",
       "      <th>Simulated Replicates $\\geq$ Observed</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>742kb</td>\n",
       "      <td>1311</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>742kb</td>\n",
       "      <td>208</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>72kb</td>\n",
       "      <td>300</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>72kb</td>\n",
       "      <td>140</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Negative</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>742kb</td>\n",
       "      <td>1311</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>742kb</td>\n",
       "      <td>208</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Negative</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>72kb</td>\n",
       "      <td>300</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>72kb</td>\n",
       "      <td>140</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Selection Model       SNP Set Region  MXL AAF $\\geq 30\\%$ (Observed)  \\\n",
       "0         Neutral      All SNPs  742kb                            1311   \n",
       "1         Neutral  Archaic SNPs  742kb                             208   \n",
       "2         Neutral      All SNPs   72kb                             300   \n",
       "3         Neutral  Archaic SNPs   72kb                             140   \n",
       "4        Negative      All SNPs  742kb                            1311   \n",
       "5        Negative  Archaic SNPs  742kb                             208   \n",
       "6        Negative      All SNPs   72kb                             300   \n",
       "7        Negative  Archaic SNPs   72kb                             140   \n",
       "\n",
       "   Total Replicates  Simulated Replicates $\\geq$ Observed $P-value$  \n",
       "0             10000                                     0   <0.0001  \n",
       "1             10000                                     0   <0.0001  \n",
       "2             10000                                     0   <0.0001  \n",
       "3             10000                                     0   <0.0001  \n",
       "4             10000                                     0   <0.0001  \n",
       "5             10000                                     0   <0.0001  \n",
       "6             10000                                     0   <0.0001  \n",
       "7             10000                                     0   <0.0001  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43162df7",
   "metadata": {},
   "source": [
    "## S61"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "d5e2b41a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/slim_per_region_pbs_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 2, 3, 5, 6]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 2, 3, 5, 6]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S61. Comparison of per-region \\textit{PBS} values for the observed focal regions and across 10,000 simulated replicates.} \\newline The per-region \\textit{PBS} values observed in MXL, along with the number of simulated replicates where the per-region \\textit{PBS} value is greater than or equal to what is observed in the empirical data. The \\textit{P-value} represents the proportion of the 10,000 simulation replicates where the simulated value is greater than or equal to what is observed in the empirical data, where a \\textit{P-value} less than 0.05 is considered statistically significant. We report the results for both the focal 742kb and 72kb \\textit{MUC19} regions, for all SNPs, as well as archaic SNPs.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/slim_per_region_pbs_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(3, 5))+r' & '\\\n",
    "        + r' & '.join(\n",
    "            r'$<0.0001$' if mat[i, j][0] == '<' else\n",
    "            (convert_sci_notation_to_latex(float(mat[i, j])) if abs(float(mat[i, j])) < 0.0001 else\n",
    "            '{:.4f}'.format(round(float(mat[i, j]), 4)).rstrip('0').rstrip('.'))\n",
    "            for j in range(5, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61ae458a",
   "metadata": {},
   "source": [
    "\\textbf{Table S61. Comparison of per-region \\textit{PBS} values for the observed focal regions and across 10,000 simulated replicates.} \\newline The per-region \\textit{PBS} values observed in MXL, along with the number of simulated replicates where the per-region \\textit{PBS} value is greater than or equal to what is observed in the empirical data. The \\textit{P-value} represents the proportion of the 10,000 simulation replicates where the simulated value is greater than or equal to what is observed in the empirical data, where a \\textit{P-value} less than 0.05 is considered statistically significant. We report the results for both the focal 742kb and 72kb \\textit{MUC19} regions, for all SNPs, as well as archaic SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "668d079b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Selection Model</th>\n",
       "      <th>SNP Set</th>\n",
       "      <th>Region</th>\n",
       "      <th>$PBS$ (Observed)</th>\n",
       "      <th>Total Replicates</th>\n",
       "      <th>Simulated Replicates $\\geq$ Observed</th>\n",
       "      <th>$P-value$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>742kb</td>\n",
       "      <td>0.065842</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>742kb</td>\n",
       "      <td>0.291188</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>72kb</td>\n",
       "      <td>0.127226</td>\n",
       "      <td>10000</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Neutral</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>72kb</td>\n",
       "      <td>0.279597</td>\n",
       "      <td>10000</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Negative</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>742kb</td>\n",
       "      <td>0.065842</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>742kb</td>\n",
       "      <td>0.291188</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Negative</td>\n",
       "      <td>All SNPs</td>\n",
       "      <td>72kb</td>\n",
       "      <td>0.127226</td>\n",
       "      <td>10000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Negative</td>\n",
       "      <td>Archaic SNPs</td>\n",
       "      <td>72kb</td>\n",
       "      <td>0.279597</td>\n",
       "      <td>10000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Selection Model       SNP Set Region  $PBS$ (Observed)  Total Replicates  \\\n",
       "0         Neutral      All SNPs  742kb          0.065842             10000   \n",
       "1         Neutral  Archaic SNPs  742kb          0.291188             10000   \n",
       "2         Neutral      All SNPs   72kb          0.127226             10000   \n",
       "3         Neutral  Archaic SNPs   72kb          0.279597             10000   \n",
       "4        Negative      All SNPs  742kb          0.065842             10000   \n",
       "5        Negative  Archaic SNPs  742kb          0.291188             10000   \n",
       "6        Negative      All SNPs   72kb          0.127226             10000   \n",
       "7        Negative  Archaic SNPs   72kb          0.279597             10000   \n",
       "\n",
       "   Simulated Replicates $\\geq$ Observed $P-value$  \n",
       "0                                     0   <0.0001  \n",
       "1                                     0   <0.0001  \n",
       "2                                     3    0.0003  \n",
       "3                                     6    0.0006  \n",
       "4                                     0   <0.0001  \n",
       "5                                     0   <0.0001  \n",
       "6                                     1    0.0001  \n",
       "7                                     1    0.0001  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6b1488a",
   "metadata": {},
   "source": [
    "## S62"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "dd2abf96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/slim_pbs_power_summary.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S62. Power to detect positive selection in MXL using per-region \\textit{PBS} values.} \\newline The power to detect positive selection for different strengths of positive selection: strong ($s = 0.1$), weak ($s = 0.01$), and extremely weak ($s = 0.0015$). The power represents the proportion of the 1,000 simulated replicates with per-region \\textit{PBS} values greater than the upper 5\\% quantile from its respective neutral distribution of per-region \\textit{PBS} values consisting of 10,000 replicates. For each selection coefficient we report the results for both the simulated 742kb and 72kb \\textit{MUC19} regions, for all SNPs, as well as archaic SNPs.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/slim_pbs_power_summary.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(\n",
    "            '{:.4f}'.format(round(mat[i, j], 4)).rstrip('0').rstrip('.')\n",
    "        for j in range(mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c997e28",
   "metadata": {},
   "source": [
    "\\textbf{Table S62. Power to detect positive selection in MXL using per-region \\textit{PBS} values.} \\newline The power to detect positive selection for different strengths of positive selection: strong ($s = 0.1$), weak ($s = 0.01$), and extremely weak ($s = 0.0015$). The power represents the proportion of the 1,000 simulated replicates with per-region \\textit{PBS} values greater than the upper 5\\% quantile from its respective neutral distribution of per-region \\textit{PBS} values consisting of 10,000 replicates. For each selection coefficient we report the results for both the simulated 742kb and 72kb \\textit{MUC19} regions, for all SNPs, as well as archaic SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "cc8ee4d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>$s$</th>\n",
       "      <th>All SNPs (72kb)</th>\n",
       "      <th>Archaic SNPs (72kb)</th>\n",
       "      <th>All SNPs (742kb)</th>\n",
       "      <th>Archaic SNPs (742kb)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.1000</td>\n",
       "      <td>0.986</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.982</td>\n",
       "      <td>0.981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0100</td>\n",
       "      <td>0.873</td>\n",
       "      <td>0.935</td>\n",
       "      <td>0.794</td>\n",
       "      <td>0.857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.096</td>\n",
       "      <td>0.193</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      $s$  All SNPs (72kb)  Archaic SNPs (72kb)  All SNPs (742kb)  \\\n",
       "0  0.1000            0.986                0.984             0.982   \n",
       "1  0.0100            0.873                0.935             0.794   \n",
       "2  0.0015            0.096                0.193             0.081   \n",
       "\n",
       "   Archaic SNPs (742kb)  \n",
       "0                 0.981  \n",
       "1                 0.857  \n",
       "2                 0.126  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e98a516",
   "metadata": {},
   "source": [
    "## S63"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "78506d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_ihs_critical_scores_proportions_742kb.csv.gz')\n",
    "df = df[(df['SNP Set'] == 'Archaic SNPs') & (df['Super Population'] != 'AFR')]\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 3, 4, 5, 6, 7, 8, 9]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 3, 4, 5, 6, 7, 8, 9]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S63. Integrated Haplotype Scores (\\textit{iHS}) for archaic SNPs in the focal 742kb \\textit{MUC19} region among 1000 Genomes Project non-African populations.} \\newline The normalized $\\mid iHS \\mid$ scores for archaic SNPs observed at the focal 742kb region for each non-African population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each population, the normalized \\textit{iHS} scores were calculated for archaic SNPs with a minor allele frequency $> 5\\%$. For each population, the table reports the total number of \\textit{iHS} scores observed for archaic SNPs, the number of extreme \\textit{iHS} scores for archaic SNPs (i.e., $\\mid iHS \\mid > 2$), the observed proportion of extreme \\textit{iHS} scores for archaic SNPs (calculated as the number of extreme \\textit{iHS} scores normalized by the total number of \\textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \\textit{iHS} scores for archaic SNPs determined from the genomic background distribution (i.e., non-overlapping 742kb windows with more than 10 archaic SNPs), the total number of 742kb windows in the genomic background distribution, the number of 742kb windows where the observed proportion of extreme \\textit{iHS} scores for archaic SNPs is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \\textit{iHS} scores for archaic SNPs (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \\textit{iHS} scores). An observed proportion of extreme \\textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection. Note that a \"---\" denotes that the population did not meet the minimum number of archaic SNPs threshold for the \\textit{iHS} analysis.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_ihs_critical_scores_proportions_archaic_snps_742kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            '0' if mat[i, j] == 0 else\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3da54a9d",
   "metadata": {},
   "source": [
    "\\textbf{Table S63. Integrated Haplotype Scores (\\textit{iHS}) for archaic SNPs in the focal 742kb \\textit{MUC19} region among 1000 Genomes Project non-African populations.} \\newline The normalized $\\mid iHS \\mid$ scores for archaic SNPs observed at the focal 742kb region for each non-African population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each population, the normalized \\textit{iHS} scores were calculated for archaic SNPs with a minor allele frequency $> 5\\%$. For each population, the table reports the total number of \\textit{iHS} scores observed for archaic SNPs, the number of extreme \\textit{iHS} scores for archaic SNPs (i.e., $\\mid iHS \\mid > 2$), the observed proportion of extreme \\textit{iHS} scores for archaic SNPs (calculated as the number of extreme \\textit{iHS} scores normalized by the total number of \\textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \\textit{iHS} scores for archaic SNPs determined from the genomic background distribution (i.e., non-overlapping 742kb windows with more than 10 archaic SNPs), the total number of 742kb windows in the genomic background distribution, the number of 742kb windows where the observed proportion of extreme \\textit{iHS} scores for archaic SNPs is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \\textit{iHS} scores for archaic SNPs (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \\textit{iHS} scores). An observed proportion of extreme \\textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection. Note that a \"---\" denotes that the population did not meet the minimum number of archaic SNPs threshold for the \\textit{iHS} analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "d1c0714f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Focal 742kb Region (Total SNPs)</th>\n",
       "      <th>Focal 742kb Region (SNPs with $\\mid iHS \\mid &gt; 2$)</th>\n",
       "      <th>Focal 742kb Region (Prop. of SNPs with $\\mid iHS \\mid &gt; 2$)</th>\n",
       "      <th>742kb Non-overlapping Windows ($99^{th}$ Percentile)</th>\n",
       "      <th>742kb Non-overlapping Windows (Total SNPs $&gt; 10$)</th>\n",
       "      <th>Focal 742kb Region $&gt;$ 742kb Non-overlapping Windows</th>\n",
       "      <th>Focal 742kb Region (Percentile Rank)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>217</td>\n",
       "      <td>208</td>\n",
       "      <td>0.958525</td>\n",
       "      <td>0.733320</td>\n",
       "      <td>1250.0</td>\n",
       "      <td>1248.0</td>\n",
       "      <td>99.840000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>217</td>\n",
       "      <td>66</td>\n",
       "      <td>0.304147</td>\n",
       "      <td>0.686908</td>\n",
       "      <td>1119.0</td>\n",
       "      <td>1063.0</td>\n",
       "      <td>94.995532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>217</td>\n",
       "      <td>6</td>\n",
       "      <td>0.027650</td>\n",
       "      <td>0.520430</td>\n",
       "      <td>1276.0</td>\n",
       "      <td>1032.0</td>\n",
       "      <td>80.877743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>217</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.562500</td>\n",
       "      <td>1148.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>197</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.514489</td>\n",
       "      <td>1263.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>214</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>0.494474</td>\n",
       "      <td>1264.0</td>\n",
       "      <td>908.0</td>\n",
       "      <td>71.835443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>184</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.544545</td>\n",
       "      <td>1269.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>156</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.493746</td>\n",
       "      <td>1312.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>214</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.650556</td>\n",
       "      <td>1288.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>171</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.458070</td>\n",
       "      <td>1213.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>191</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.459089</td>\n",
       "      <td>1198.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>152</td>\n",
       "      <td>1</td>\n",
       "      <td>0.006579</td>\n",
       "      <td>0.563448</td>\n",
       "      <td>1254.0</td>\n",
       "      <td>932.0</td>\n",
       "      <td>74.322169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.429064</td>\n",
       "      <td>1281.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0.058824</td>\n",
       "      <td>0.587201</td>\n",
       "      <td>1232.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>85.795455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.497959</td>\n",
       "      <td>1321.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Focal 742kb Region (Total SNPs)  \\\n",
       "1               AMR        MXL                              217   \n",
       "5               AMR        PEL                              217   \n",
       "9               AMR        CLM                              217   \n",
       "13              AMR        PUR                              217   \n",
       "17              SAS        BEB                              197   \n",
       "21              SAS        STU                              214   \n",
       "25              SAS        ITU                              184   \n",
       "29              SAS        PJL                              156   \n",
       "33              SAS        GIH                              214   \n",
       "37              EAS        CHB                                0   \n",
       "41              EAS        KHV                              171   \n",
       "45              EAS        CHS                                0   \n",
       "49              EAS        JPT                                2   \n",
       "53              EAS        CDX                              191   \n",
       "57              EUR        TSI                              152   \n",
       "61              EUR        CEU                               17   \n",
       "65              EUR        IBS                               17   \n",
       "69              EUR        GBR                                0   \n",
       "73              EUR        FIN                               17   \n",
       "\n",
       "    Focal 742kb Region (SNPs with $\\mid iHS \\mid > 2$)  \\\n",
       "1                                                 208    \n",
       "5                                                  66    \n",
       "9                                                   6    \n",
       "13                                                  0    \n",
       "17                                                  0    \n",
       "21                                                  1    \n",
       "25                                                  0    \n",
       "29                                                  0    \n",
       "33                                                  0    \n",
       "37                                                  0    \n",
       "41                                                  0    \n",
       "45                                                  0    \n",
       "49                                                  0    \n",
       "53                                                  0    \n",
       "57                                                  1    \n",
       "61                                                  0    \n",
       "65                                                  1    \n",
       "69                                                  0    \n",
       "73                                                  0    \n",
       "\n",
       "    Focal 742kb Region (Prop. of SNPs with $\\mid iHS \\mid > 2$)  \\\n",
       "1                                            0.958525             \n",
       "5                                            0.304147             \n",
       "9                                            0.027650             \n",
       "13                                           0.000000             \n",
       "17                                           0.000000             \n",
       "21                                           0.004673             \n",
       "25                                           0.000000             \n",
       "29                                           0.000000             \n",
       "33                                           0.000000             \n",
       "37                                                NaN             \n",
       "41                                           0.000000             \n",
       "45                                                NaN             \n",
       "49                                           0.000000             \n",
       "53                                           0.000000             \n",
       "57                                           0.006579             \n",
       "61                                           0.000000             \n",
       "65                                           0.058824             \n",
       "69                                                NaN             \n",
       "73                                           0.000000             \n",
       "\n",
       "    742kb Non-overlapping Windows ($99^{th}$ Percentile)  \\\n",
       "1                                            0.733320      \n",
       "5                                            0.686908      \n",
       "9                                            0.520430      \n",
       "13                                           0.562500      \n",
       "17                                           0.514489      \n",
       "21                                           0.494474      \n",
       "25                                           0.544545      \n",
       "29                                           0.493746      \n",
       "33                                           0.650556      \n",
       "37                                                NaN      \n",
       "41                                           0.458070      \n",
       "45                                                NaN      \n",
       "49                                                NaN      \n",
       "53                                           0.459089      \n",
       "57                                           0.563448      \n",
       "61                                           0.429064      \n",
       "65                                           0.587201      \n",
       "69                                                NaN      \n",
       "73                                           0.497959      \n",
       "\n",
       "    742kb Non-overlapping Windows (Total SNPs $> 10$)  \\\n",
       "1                                              1250.0   \n",
       "5                                              1119.0   \n",
       "9                                              1276.0   \n",
       "13                                             1148.0   \n",
       "17                                             1263.0   \n",
       "21                                             1264.0   \n",
       "25                                             1269.0   \n",
       "29                                             1312.0   \n",
       "33                                             1288.0   \n",
       "37                                                NaN   \n",
       "41                                             1213.0   \n",
       "45                                                NaN   \n",
       "49                                                NaN   \n",
       "53                                             1198.0   \n",
       "57                                             1254.0   \n",
       "61                                             1281.0   \n",
       "65                                             1232.0   \n",
       "69                                                NaN   \n",
       "73                                             1321.0   \n",
       "\n",
       "    Focal 742kb Region $>$ 742kb Non-overlapping Windows  \\\n",
       "1                                              1248.0      \n",
       "5                                              1063.0      \n",
       "9                                              1032.0      \n",
       "13                                                0.0      \n",
       "17                                                0.0      \n",
       "21                                              908.0      \n",
       "25                                                0.0      \n",
       "29                                                0.0      \n",
       "33                                                0.0      \n",
       "37                                                NaN      \n",
       "41                                                0.0      \n",
       "45                                                NaN      \n",
       "49                                                NaN      \n",
       "53                                                0.0      \n",
       "57                                              932.0      \n",
       "61                                                0.0      \n",
       "65                                             1057.0      \n",
       "69                                                NaN      \n",
       "73                                                0.0      \n",
       "\n",
       "    Focal 742kb Region (Percentile Rank)  \n",
       "1                              99.840000  \n",
       "5                              94.995532  \n",
       "9                              80.877743  \n",
       "13                              0.000000  \n",
       "17                              0.000000  \n",
       "21                             71.835443  \n",
       "25                              0.000000  \n",
       "29                              0.000000  \n",
       "33                              0.000000  \n",
       "37                                   NaN  \n",
       "41                              0.000000  \n",
       "45                                   NaN  \n",
       "49                                   NaN  \n",
       "53                              0.000000  \n",
       "57                             74.322169  \n",
       "61                              0.000000  \n",
       "65                             85.795455  \n",
       "69                                   NaN  \n",
       "73                              0.000000  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 3, 4, 5, 6, 7, 8, 9]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0adb53b9",
   "metadata": {},
   "source": [
    "## S64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "0d109f09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/tgp_ihs_critical_scores_proportions_72kb.csv.gz')\n",
    "df = df[(df['SNP Set'] == 'Archaic SNPs') & (df['Super Population'] != 'AFR')]\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, [0, 1, 3, 4, 5, 6, 7, 8, 9]]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[[0, 1, 3, 4, 5, 6, 7, 8, 9]]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S64. Integrated Haplotype Scores (\\textit{iHS}) for archaic SNPs in the focal 72kb \\textit{MUC19} region among 1000 Genomes Project non-African populations.} \\newline The normalized $\\mid iHS \\mid$ scores for archaic SNPs observed at the focal 72kb region for each non-African population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each population, the normalized \\textit{iHS} scores were calculated for archaic SNPs with a minor allele frequency $> 5\\%$. For each population, the table reports the total number of \\textit{iHS} scores observed for archaic SNPs, the number of extreme \\textit{iHS} scores for archaic SNPs (i.e., $\\mid iHS \\mid > 2$), the observed proportion of extreme \\textit{iHS} scores for archaic SNPs (calculated as the number of extreme \\textit{iHS} scores normalized by the total number of \\textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \\textit{iHS} scores for archaic SNPs determined from the genomic background distribution (i.e., non-overlapping 72kb windows with more than 10 archaic SNPs), the total number of 72kb windows in the genomic background distribution, the number of 72kb windows where the observed proportion of extreme \\textit{iHS} scores for archaic SNPs is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \\textit{iHS} scores for archaic SNPs (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \\textit{iHS} scores). An observed proportion of extreme \\textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection. Note that a \"---\" denotes that the population did not meet the minimum number of archaic SNPs threshold for the \\textit{iHS} analysis.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/tgp_ihs_critical_scores_proportions_archaic_snps_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(3))+r' & '\\\n",
    "        + r' & '.join((\n",
    "            '---' if np.isnan(mat[i, j]) else\n",
    "            '0' if mat[i, j] == 0 else\n",
    "            convert_sci_notation_to_latex(mat[i, j]) if mat[i, j] < 0.001 else\n",
    "            '{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.'))\n",
    "            for j in range(3, mat.shape[1])\n",
    "        )\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "609d6bf6",
   "metadata": {},
   "source": [
    "\\textbf{Table S64. Integrated Haplotype Scores (\\textit{iHS}) for archaic SNPs in the focal 72kb \\textit{MUC19} region among 1000 Genomes Project non-African populations.} \\newline The normalized $\\mid iHS \\mid$ scores for archaic SNPs observed at the focal 72kb region for each non-African population stratified by super population: Admixed Americans (AMR), South Asians (SAS), East Asians (EAS), and Europeans (EUR). For each population, the normalized \\textit{iHS} scores were calculated for archaic SNPs with a minor allele frequency $> 5\\%$. For each population, the table reports the total number of \\textit{iHS} scores observed for archaic SNPs, the number of extreme \\textit{iHS} scores for archaic SNPs (i.e., $\\mid iHS \\mid > 2$), the observed proportion of extreme \\textit{iHS} scores for archaic SNPs (calculated as the number of extreme \\textit{iHS} scores normalized by the total number of \\textit{iHS} scores), the $99^{th}$ percentile of the proportion of extreme \\textit{iHS} scores for archaic SNPs determined from the genomic background distribution (i.e., non-overlapping 72kb windows with more than 10 archaic SNPs), the total number of 72kb windows in the genomic background distribution, the number of 72kb windows where the observed proportion of extreme \\textit{iHS} scores for archaic SNPs is greater than the genomic background distribution, and the percentile rank of the observed proportion of extreme \\textit{iHS} scores for archaic SNPs (i.e., the percentage of the genomic background distribution that is less than the observed proportion of extreme \\textit{iHS} scores). An observed proportion of extreme \\textit{iHS} scores greater than the $99^{th}$ percentile of the genomic background distribution is considered evidence of positive selection. Note that a \"---\" denotes that the population did not meet the minimum number of archaic SNPs threshold for the \\textit{iHS} analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "cfa18d2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Super Population</th>\n",
       "      <th>Population</th>\n",
       "      <th>Focal 72kb Region (Total SNPs)</th>\n",
       "      <th>Focal 72kb Region (SNPs with $\\mid iHS \\mid &gt; 2$)</th>\n",
       "      <th>Focal 72kb Region (Prop. of SNPs with $\\mid iHS \\mid &gt; 2$)</th>\n",
       "      <th>72kb Non-overlapping Windows ($99^{th}$ Percentile)</th>\n",
       "      <th>72kb Non-overlapping Windows (Total SNPs $&gt; 10$)</th>\n",
       "      <th>Focal 72kb Region $&gt;$ 72kb Non-overlapping Windows</th>\n",
       "      <th>Focal 72kb Region (Percentile Rank)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMR</td>\n",
       "      <td>MXL</td>\n",
       "      <td>140</td>\n",
       "      <td>135</td>\n",
       "      <td>0.964286</td>\n",
       "      <td>0.937500</td>\n",
       "      <td>2699.0</td>\n",
       "      <td>2677.0</td>\n",
       "      <td>99.184883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PEL</td>\n",
       "      <td>140</td>\n",
       "      <td>58</td>\n",
       "      <td>0.414286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2332.0</td>\n",
       "      <td>2225.0</td>\n",
       "      <td>95.411664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AMR</td>\n",
       "      <td>CLM</td>\n",
       "      <td>140</td>\n",
       "      <td>6</td>\n",
       "      <td>0.042857</td>\n",
       "      <td>0.823529</td>\n",
       "      <td>2630.0</td>\n",
       "      <td>2339.0</td>\n",
       "      <td>88.935361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AMR</td>\n",
       "      <td>PUR</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.877850</td>\n",
       "      <td>2344.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SAS</td>\n",
       "      <td>BEB</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.731524</td>\n",
       "      <td>3048.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SAS</td>\n",
       "      <td>STU</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.733333</td>\n",
       "      <td>2901.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SAS</td>\n",
       "      <td>ITU</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.733344</td>\n",
       "      <td>2900.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SAS</td>\n",
       "      <td>PJL</td>\n",
       "      <td>139</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.801667</td>\n",
       "      <td>2952.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SAS</td>\n",
       "      <td>GIH</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.848132</td>\n",
       "      <td>2883.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHB</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>EAS</td>\n",
       "      <td>KHV</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.686412</td>\n",
       "      <td>3248.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CHS</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>EAS</td>\n",
       "      <td>JPT</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>EAS</td>\n",
       "      <td>CDX</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.798966</td>\n",
       "      <td>3116.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>EUR</td>\n",
       "      <td>TSI</td>\n",
       "      <td>139</td>\n",
       "      <td>1</td>\n",
       "      <td>0.007194</td>\n",
       "      <td>0.766199</td>\n",
       "      <td>2668.0</td>\n",
       "      <td>2352.0</td>\n",
       "      <td>88.155922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>EUR</td>\n",
       "      <td>CEU</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>EUR</td>\n",
       "      <td>IBS</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>EUR</td>\n",
       "      <td>GBR</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>EUR</td>\n",
       "      <td>FIN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Super Population Population  Focal 72kb Region (Total SNPs)  \\\n",
       "1               AMR        MXL                             140   \n",
       "5               AMR        PEL                             140   \n",
       "9               AMR        CLM                             140   \n",
       "13              AMR        PUR                             140   \n",
       "17              SAS        BEB                             140   \n",
       "21              SAS        STU                             140   \n",
       "25              SAS        ITU                             140   \n",
       "29              SAS        PJL                             139   \n",
       "33              SAS        GIH                             140   \n",
       "37              EAS        CHB                               0   \n",
       "41              EAS        KHV                             140   \n",
       "45              EAS        CHS                               0   \n",
       "49              EAS        JPT                               2   \n",
       "53              EAS        CDX                             140   \n",
       "57              EUR        TSI                             139   \n",
       "61              EUR        CEU                               0   \n",
       "65              EUR        IBS                               0   \n",
       "69              EUR        GBR                               0   \n",
       "73              EUR        FIN                               0   \n",
       "\n",
       "    Focal 72kb Region (SNPs with $\\mid iHS \\mid > 2$)  \\\n",
       "1                                                 135   \n",
       "5                                                  58   \n",
       "9                                                   6   \n",
       "13                                                  0   \n",
       "17                                                  0   \n",
       "21                                                  0   \n",
       "25                                                  0   \n",
       "29                                                  0   \n",
       "33                                                  0   \n",
       "37                                                  0   \n",
       "41                                                  0   \n",
       "45                                                  0   \n",
       "49                                                  0   \n",
       "53                                                  0   \n",
       "57                                                  1   \n",
       "61                                                  0   \n",
       "65                                                  0   \n",
       "69                                                  0   \n",
       "73                                                  0   \n",
       "\n",
       "    Focal 72kb Region (Prop. of SNPs with $\\mid iHS \\mid > 2$)  \\\n",
       "1                                            0.964286            \n",
       "5                                            0.414286            \n",
       "9                                            0.042857            \n",
       "13                                           0.000000            \n",
       "17                                           0.000000            \n",
       "21                                           0.000000            \n",
       "25                                           0.000000            \n",
       "29                                           0.000000            \n",
       "33                                           0.000000            \n",
       "37                                                NaN            \n",
       "41                                           0.000000            \n",
       "45                                                NaN            \n",
       "49                                           0.000000            \n",
       "53                                           0.000000            \n",
       "57                                           0.007194            \n",
       "61                                                NaN            \n",
       "65                                                NaN            \n",
       "69                                                NaN            \n",
       "73                                                NaN            \n",
       "\n",
       "    72kb Non-overlapping Windows ($99^{th}$ Percentile)  \\\n",
       "1                                            0.937500     \n",
       "5                                            1.000000     \n",
       "9                                            0.823529     \n",
       "13                                           0.877850     \n",
       "17                                           0.731524     \n",
       "21                                           0.733333     \n",
       "25                                           0.733344     \n",
       "29                                           0.801667     \n",
       "33                                           0.848132     \n",
       "37                                                NaN     \n",
       "41                                           0.686412     \n",
       "45                                                NaN     \n",
       "49                                                NaN     \n",
       "53                                           0.798966     \n",
       "57                                           0.766199     \n",
       "61                                                NaN     \n",
       "65                                                NaN     \n",
       "69                                                NaN     \n",
       "73                                                NaN     \n",
       "\n",
       "    72kb Non-overlapping Windows (Total SNPs $> 10$)  \\\n",
       "1                                             2699.0   \n",
       "5                                             2332.0   \n",
       "9                                             2630.0   \n",
       "13                                            2344.0   \n",
       "17                                            3048.0   \n",
       "21                                            2901.0   \n",
       "25                                            2900.0   \n",
       "29                                            2952.0   \n",
       "33                                            2883.0   \n",
       "37                                               NaN   \n",
       "41                                            3248.0   \n",
       "45                                               NaN   \n",
       "49                                               NaN   \n",
       "53                                            3116.0   \n",
       "57                                            2668.0   \n",
       "61                                               NaN   \n",
       "65                                               NaN   \n",
       "69                                               NaN   \n",
       "73                                               NaN   \n",
       "\n",
       "    Focal 72kb Region $>$ 72kb Non-overlapping Windows  \\\n",
       "1                                              2677.0    \n",
       "5                                              2225.0    \n",
       "9                                              2339.0    \n",
       "13                                                0.0    \n",
       "17                                                0.0    \n",
       "21                                                0.0    \n",
       "25                                                0.0    \n",
       "29                                                0.0    \n",
       "33                                                0.0    \n",
       "37                                                NaN    \n",
       "41                                                0.0    \n",
       "45                                                NaN    \n",
       "49                                                NaN    \n",
       "53                                                0.0    \n",
       "57                                             2352.0    \n",
       "61                                                NaN    \n",
       "65                                                NaN    \n",
       "69                                                NaN    \n",
       "73                                                NaN    \n",
       "\n",
       "    Focal 72kb Region (Percentile Rank)  \n",
       "1                             99.184883  \n",
       "5                             95.411664  \n",
       "9                             88.935361  \n",
       "13                             0.000000  \n",
       "17                             0.000000  \n",
       "21                             0.000000  \n",
       "25                             0.000000  \n",
       "29                             0.000000  \n",
       "33                             0.000000  \n",
       "37                                  NaN  \n",
       "41                             0.000000  \n",
       "45                                  NaN  \n",
       "49                                  NaN  \n",
       "53                             0.000000  \n",
       "57                            88.155922  \n",
       "61                                  NaN  \n",
       "65                                  NaN  \n",
       "69                                  NaN  \n",
       "73                                  NaN  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, [0, 1, 3, 4, 5, 6, 7, 8, 9]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dd84956",
   "metadata": {},
   "source": [
    "## S65"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "c7a927c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/chagyrskaya_nean_manual_phasing_info.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, 1:]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[1:]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S65. Read-based phasing information for the Chagyrskaya Neanderthal.} \\newline The four heterozygous sites that could not be resolved with statistical phasing, preceding and proceeding heterozygous sites and their distance from the heterozygous site to resolve, and the results from manual read-based phasing for the Chagyrskaya Neanderthal at the focal 72kb \\textit{MUC19} region.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/chagyrskaya_nean_manual_phasing_info.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(mat.shape[1]))\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76627e14",
   "metadata": {},
   "source": [
    "\\textbf{Table S65. Read-based phasing information for the Chagyrskaya Neanderthal.} \\newline The four heterozygous sites that could not be resolved with statistical phasing, preceding and proceeding heterozygous sites and their distance from the heterozygous site to resolve, and the results from manual read-based phasing for the Chagyrskaya Neanderthal at the focal 72kb \\textit{MUC19} region."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "c70afb0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Late Neanderthal</th>\n",
       "      <th>Het. Pos. to Resolve</th>\n",
       "      <th>Preceding Het. Pos.</th>\n",
       "      <th>Dist. to Preceding Het. Pos.</th>\n",
       "      <th>Proceeding Het. Pos.</th>\n",
       "      <th>Dist. to Proceeding Het. Pos.</th>\n",
       "      <th>Is Resvolable?</th>\n",
       "      <th>Phasing Info.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>40767590</td>\n",
       "      <td>40761946</td>\n",
       "      <td>5644</td>\n",
       "      <td>40768905</td>\n",
       "      <td>1315</td>\n",
       "      <td>False</td>\n",
       "      <td>Distant Flanking Het. Sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>40821807</td>\n",
       "      <td>40821795</td>\n",
       "      <td>12</td>\n",
       "      <td>40821847</td>\n",
       "      <td>40</td>\n",
       "      <td>True</td>\n",
       "      <td>G-C-C and A-T-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>40826155</td>\n",
       "      <td>40826138</td>\n",
       "      <td>17</td>\n",
       "      <td>40826201</td>\n",
       "      <td>46</td>\n",
       "      <td>True</td>\n",
       "      <td>T-A-G and G-G-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chagyrskaya Nean.</td>\n",
       "      <td>40828121</td>\n",
       "      <td>40828019</td>\n",
       "      <td>102</td>\n",
       "      <td>40828306</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>Not Enough Overlapping Reads</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Late Neanderthal  Het. Pos. to Resolve  Preceding Het. Pos.  \\\n",
       "0  Chagyrskaya Nean.              40767590             40761946   \n",
       "1  Chagyrskaya Nean.              40821807             40821795   \n",
       "2  Chagyrskaya Nean.              40826155             40826138   \n",
       "3  Chagyrskaya Nean.              40828121             40828019   \n",
       "\n",
       "   Dist. to Preceding Het. Pos.  Proceeding Het. Pos.  \\\n",
       "0                          5644              40768905   \n",
       "1                            12              40821847   \n",
       "2                            17              40826201   \n",
       "3                           102              40828306   \n",
       "\n",
       "   Dist. to Proceeding Het. Pos.  Is Resvolable?                 Phasing Info.  \n",
       "0                           1315           False   Distant Flanking Het. Sites  \n",
       "1                             40            True               G-C-C and A-T-A  \n",
       "2                             46            True               T-A-G and G-G-A  \n",
       "3                            185           False  Not Enough Overlapping Reads  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "935e9a19",
   "metadata": {},
   "source": [
    "## S66"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "4c349512",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/vindija_nean_manual_phasing_info.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()[:, 1:]\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values[1:]\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S66. Read-based phasing information for the Vindija Neanderthal.} \\newline The five heterozygous sites that could not be resolved with statistical phasing, preceding and proceeding heterozygous sites and their distance from the heterozygous site to resolve, and the results from manual read-based phasing for the Vindija Neanderthal at the focal 72kb \\textit{MUC19} region. Note that position 40829306 is the last heterozygous site in the focal 72kb \\textit{MUC19} region for the Vindija Neanderthal, and thus, the proceeding heterozygous site information is denoted by a \"-1\".},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/vindija_nean_manual_phasing_info.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(mat.shape[1]))\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e2ffab4",
   "metadata": {},
   "source": [
    "\\textbf{Table S66. Read-based phasing information for the Vindija Neanderthal.} \\newline The five heterozygous sites that could not be resolved with statistical phasing, preceding and proceeding heterozygous sites and their distance from the heterozygous site to resolve, and the results from manual read-based phasing for the Vindija Neanderthal at the focal 72kb \\textit{MUC19} region. Note that position 40829306 is the last heterozygous site in the focal 72kb \\textit{MUC19} region for the Vindija Neanderthal, and thus, the proceeding heterozygous site information is denoted by a \"-1\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "584771f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Late Neanderthal</th>\n",
       "      <th>Het. Pos. to Resolve</th>\n",
       "      <th>Preceding Het. Pos.</th>\n",
       "      <th>Dist. to Preceding Het. Pos.</th>\n",
       "      <th>Proceeding Het. Pos.</th>\n",
       "      <th>Dist. to Proceeding Het. Pos.</th>\n",
       "      <th>Is Resvolable?</th>\n",
       "      <th>Phasing Info.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>40765370</td>\n",
       "      <td>40761946</td>\n",
       "      <td>3424</td>\n",
       "      <td>40768905</td>\n",
       "      <td>3535</td>\n",
       "      <td>False</td>\n",
       "      <td>Distant Flanking Het. Sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>40783344</td>\n",
       "      <td>40775127</td>\n",
       "      <td>8217</td>\n",
       "      <td>40784418</td>\n",
       "      <td>1074</td>\n",
       "      <td>False</td>\n",
       "      <td>Distant Flanking Het. Sites</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>40821807</td>\n",
       "      <td>40821795</td>\n",
       "      <td>12</td>\n",
       "      <td>40821847</td>\n",
       "      <td>40</td>\n",
       "      <td>True</td>\n",
       "      <td>G-C-C and A-T-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>40824215</td>\n",
       "      <td>40824154</td>\n",
       "      <td>61</td>\n",
       "      <td>40824266</td>\n",
       "      <td>51</td>\n",
       "      <td>True</td>\n",
       "      <td>G-A and A-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vindija Nean.</td>\n",
       "      <td>40829306</td>\n",
       "      <td>40828306</td>\n",
       "      <td>1000</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>Distant Flanking Het. Sites</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Late Neanderthal  Het. Pos. to Resolve  Preceding Het. Pos.  \\\n",
       "0    Vindija Nean.              40765370             40761946   \n",
       "1    Vindija Nean.              40783344             40775127   \n",
       "2    Vindija Nean.              40821807             40821795   \n",
       "3    Vindija Nean.              40824215             40824154   \n",
       "4    Vindija Nean.              40829306             40828306   \n",
       "\n",
       "   Dist. to Preceding Het. Pos.  Proceeding Het. Pos.  \\\n",
       "0                          3424              40768905   \n",
       "1                          8217              40784418   \n",
       "2                            12              40821847   \n",
       "3                            61              40824266   \n",
       "4                          1000                    -1   \n",
       "\n",
       "   Dist. to Proceeding Het. Pos.  Is Resvolable?                Phasing Info.  \n",
       "0                           3535           False  Distant Flanking Het. Sites  \n",
       "1                           1074           False  Distant Flanking Het. Sites  \n",
       "2                             40            True              G-C-C and A-T-A  \n",
       "3                             51            True                  G-A and A-C  \n",
       "4                             -1           False  Distant Flanking Het. Sites  "
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1ce45ac",
   "metadata": {},
   "source": [
    "## S67"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "a5e335f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe.\n",
    "df = pd.read_csv('./dataframes/287_denisovan_haps_v_archaics_pairwise_diffs_72kb.csv.gz')\n",
    "# Convert the dataframe to a numpy matrix.\n",
    "mat = df.to_numpy()\n",
    "# Intialize the columns.\n",
    "cols = df.columns.values\n",
    "# Intialize a table header.\n",
    "header = r'''\\begin{longtblr}\n",
    "{\n",
    "colspec={XXXXX},\n",
    "hlines={solid, 1pt},\n",
    "vlines={solid, 1pt},\n",
    "cells={halign=c, valign=m},\n",
    "width=\\linewidth,\n",
    "rowhead=1,\n",
    "row{odd}={Grey},\n",
    "row{1}={Brown, fg=white, halign=c, valign=m},\n",
    "caption={\\textbf{Table S67. Patterns of pairwise differences at the focal 72kb \\textit{MUC19} region between the \\textit{Denisovan-like} haplotypes in 1000 Genomes Project and the archaic individuals} \\newline The mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the number of pairwise distance distribution for the focal 72kb \\textit{MUC19} region between the 287 \\textit{Denisovan-like} haplotypes in 1000 Genomes Project and the four high-coverage archaic individuals. For the Denisovan and Altai Neanderthal, we compute the number of pairwise differences between a modern human haplotype and the two unphased archaic chromosomes, and for the Chagyrskaya and Vindija Neanderthals, we compute the number of pairwise differences between a modern human haplotype and a phased archaic haplotype.},\n",
    "}'''+'\\n'\\\n",
    "+ r' & '.join(f'{col}' for col in cols)+r' \\\\'+'\\n'\n",
    "# Open the table.\n",
    "table = open('./supp_tables/287_denisovan_haps_v_archaics_pairwise_diffs_72kb.tex', 'w')\n",
    "# Write the header.\n",
    "table.write(header)\n",
    "# For every row...\n",
    "for i in range(mat.shape[0]):\n",
    "    # Write the results.\n",
    "    table.write(\n",
    "        r' & '.join(f'{mat[i, j]}' for j in range(1))+r' & '\\\n",
    "        + r' & '.join('{:.3f}'.format(round(mat[i, j], 3)).rstrip('0').rstrip('.') for j in range(1, mat.shape[1]))\\\n",
    "        + r' \\\\'+'\\n'\n",
    "    )\n",
    "# Write the footer.\n",
    "table.write(r'\\end{longtblr}'+'\\n')\n",
    "# Close the table file.\n",
    "table.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed7d263",
   "metadata": {},
   "source": [
    "\\textbf{Table S67. Patterns of pairwise differences at the focal 72kb \\textit{MUC19} region between the \\textit{Denisovan-like} haplotypes in 1000 Genomes Project and the archaic individuals} \\newline The mean ($\\mu$), standard deviation ($\\sigma$), standard error of the mean (\\textit{SEM}), and 95\\% confidence intervals ($\\pm CI_{95\\%}$) of the number of pairwise distance distribution for the focal 72kb \\textit{MUC19} region between the 287 \\textit{Denisovan-like} haplotypes in 1000 Genomes Project and the four high-coverage archaic individuals. For the Denisovan and Altai Neanderthal, we compute the number of pairwise differences between a modern human haplotype and the two unphased archaic chromosomes, and for the Chagyrskaya and Vindija Neanderthals, we compute the number of pairwise differences between a modern human haplotype and a phased archaic haplotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "96677c22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Archaic</th>\n",
       "      <th>Focal 72kb Region Pairwise Diffs. $\\left( \\mu\\right)$</th>\n",
       "      <th>Focal 72kb Region Pairwise Diffs. $\\left( \\sigma \\right)$</th>\n",
       "      <th>Focal 72kb Region Pairwise Diffs. $\\left( SEM \\right)$</th>\n",
       "      <th>Focal 72kb Region Pairwise Diffs. $\\left( \\pm CI_{95\\%} \\right)$</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Denisovan</td>\n",
       "      <td>49.010453</td>\n",
       "      <td>1.993866</td>\n",
       "      <td>0.117900</td>\n",
       "      <td>0.232061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Altai Nean.</td>\n",
       "      <td>179.837979</td>\n",
       "      <td>2.503110</td>\n",
       "      <td>0.148012</td>\n",
       "      <td>0.291331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chagyrskaya Nean. Hap. 1</td>\n",
       "      <td>170.337979</td>\n",
       "      <td>2.456745</td>\n",
       "      <td>0.145270</td>\n",
       "      <td>0.285935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chagyrskaya Nean. Hap. 2</td>\n",
       "      <td>7.547038</td>\n",
       "      <td>2.840177</td>\n",
       "      <td>0.167943</td>\n",
       "      <td>0.330561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vindija Nean. Hap. 1</td>\n",
       "      <td>170.898955</td>\n",
       "      <td>2.178506</td>\n",
       "      <td>0.128818</td>\n",
       "      <td>0.253551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Vindija Nean. Hap. 2</td>\n",
       "      <td>6.108014</td>\n",
       "      <td>2.611721</td>\n",
       "      <td>0.154434</td>\n",
       "      <td>0.303972</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Archaic  \\\n",
       "0                 Denisovan   \n",
       "1               Altai Nean.   \n",
       "2  Chagyrskaya Nean. Hap. 1   \n",
       "3  Chagyrskaya Nean. Hap. 2   \n",
       "4      Vindija Nean. Hap. 1   \n",
       "5      Vindija Nean. Hap. 2   \n",
       "\n",
       "   Focal 72kb Region Pairwise Diffs. $\\left( \\mu\\right)$  \\\n",
       "0                                          49.010453       \n",
       "1                                         179.837979       \n",
       "2                                         170.337979       \n",
       "3                                           7.547038       \n",
       "4                                         170.898955       \n",
       "5                                           6.108014       \n",
       "\n",
       "   Focal 72kb Region Pairwise Diffs. $\\left( \\sigma \\right)$  \\\n",
       "0                                           1.993866           \n",
       "1                                           2.503110           \n",
       "2                                           2.456745           \n",
       "3                                           2.840177           \n",
       "4                                           2.178506           \n",
       "5                                           2.611721           \n",
       "\n",
       "   Focal 72kb Region Pairwise Diffs. $\\left( SEM \\right)$  \\\n",
       "0                                           0.117900        \n",
       "1                                           0.148012        \n",
       "2                                           0.145270        \n",
       "3                                           0.167943        \n",
       "4                                           0.128818        \n",
       "5                                           0.154434        \n",
       "\n",
       "   Focal 72kb Region Pairwise Diffs. $\\left( \\pm CI_{95\\%} \\right)$  \n",
       "0                                           0.232061                 \n",
       "1                                           0.291331                 \n",
       "2                                           0.285935                 \n",
       "3                                           0.330561                 \n",
       "4                                           0.253551                 \n",
       "5                                           0.303972                 "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anc_der_intro_env",
   "language": "python",
   "name": "anc_der_intro_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
